// @baidu/wuji-san/lib/wuji-text-image/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _san = require("san");

var _wujiGrid = _interopRequireDefault(require("../wuji-grid"));

var _wujiImage = _interopRequireDefault(require("../wuji-image"));

var _wujiTitle = _interopRequireDefault(require("../wuji-title"));

var _wujiWidget = _interopRequireDefault(require("../wuji-widget"));

var _pr2rem = _interopRequireDefault(require("../utils/pr2rem"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 'w-text-image';

var WujiTextImage = /*#__PURE__*/function (_Component) {
  _inherits(WujiTextImage, _Component);

  var _super = _createSuper(WujiTextImage);

  function WujiTextImage() {
    _classCallCheck(this, WujiTextImage);

    return _super.apply(this, arguments);
  }

  _createClass(WujiTextImage, [{
    key: "initData",
    value: function initData() {
      return {
        title: '',
        titleProps: {},
        labelProps: {},
        imageProps: {},
        margin: 51,
        gutter: 24,
        spacing: true,
        userSelect: true,
        justify: 'flex-start',
        hairline: true,
        reverse: false
      };
    } // 安全距离M_W_X005

  }, {
    key: "safeSpace",
    value: function safeSpace(gutter, reverse) {
      if (reverse) {
        return;
      }

      return "margin-right: ".concat((0, _pr2rem["default"])(24 - gutter), "rem");
    }
  }, {
    key: "handleClick",
    value: function handleClick(event) {
      this.fire('click', {
        event: event,
        data: this.data.get()
      });
    }
  }, {
    key: "dislike",
    value: function dislike(event) {
      this.fire('dislike', {
        event: event,
        data: this.data.get()
      });
    }
  }]);

  return WujiTextImage;
}(_san.Component);

exports["default"] = WujiTextImage;

_defineProperty(WujiTextImage, "template",
/* html */
"\n        <div class=\"{{wrapClass}}\" on-click=\"handleClick\">\n            <wuji-row gutter=\"{{gutter}}\" wrap=\"{{false}}\" margin=\"{{spacing ? margin : 0}}\">\n                <wuji-col\n                    span=\"{{hasImage ? 8 : 12}}\"\n                    class=\"".concat(prefixCls, "-left ").concat(prefixCls, "-{{justify}}\"\n                    style=\"{{safeSpace(gutter, reverse)}}\"\n                >\n                    <wuji-title\n                        s-bind=\"titleProps\"\n                        text=\"{{title || titleProps.text}}\"\n                        label=\"{{label}}\"\n                        labelProps=\"{{labelProps}}\"\n                        spacing=\"{{false}}\"\n                        size=\"{{titleSize}}\"\n                    />\n                    <wuji-widget\n                        items=\"{{widgets}}\"\n                        dislike=\"{{dislike}}\"\n                        on-dislike=\"dislike\"\n                    />\n                </wuji-col>\n                <wuji-col s-if=\"hasImage\" class=\"").concat(prefixCls, "-right\" span=\"4\">\n                    <wuji-img\n                        s-bind=\"imageProps\"\n                        src=\"{{image}}\"\n                        user-select=\"{{userSelect}}\"\n                    >\n                        <slot name=\"imageMask\" slot=\"mask\"/>\n                    </wuji-img>\n                </wuji-col>\n            </wuji-row>\n            <div s-if=\"hairline\" class=\"").concat(prefixCls, "-hairline\"/>\n        </div>\n    "));

_defineProperty(WujiTextImage, "components", {
  'wuji-row': _wujiGrid["default"].Row,
  'wuji-col': _wujiGrid["default"].Col,
  'wuji-img': _wujiImage["default"],
  'wuji-title': _wujiTitle["default"],
  'wuji-widget': _wujiWidget["default"]
});

_defineProperty(WujiTextImage, "dataTypes", {
  title: _san.DataTypes.string,
  label: _san.DataTypes.string,
  labelProps: _san.DataTypes.object,
  image: _san.DataTypes.string,
  imageProps: _san.DataTypes.object,
  dislike: _san.DataTypes.bool,
  widgets: _san.DataTypes.arrayOf(_san.DataTypes.object),
  margin: _san.DataTypes.number,
  gutter: _san.DataTypes.number,
  spacing: _san.DataTypes.bool,
  reverse: _san.DataTypes.bool,
  hairline: _san.DataTypes.bool,
  justify: _san.DataTypes.oneOf(['center', 'flex-start', 'flex-end', 'space-between', 'space-around'])
});

_defineProperty(WujiTextImage, "computed", {
  wrapClass: function wrapClass() {
    var reverse = this.data.get('reverse');
    return [prefixCls, reverse && "".concat(prefixCls, "-reverse")].filter(function (c) {
      return c;
    });
  },
  hasImage: function hasImage() {
    var image = this.data.get('image');
    var imageProps = this.data.get('imageProps');
    return imageProps && imageProps.src || image;
  }
});
