// @baidu/wuji-uikit/src/font/fontSize.js
/**
 * @file 字号编码 fontSize.js
 *
 * DO NOT CHANGE, THIS IS GENERATED BY SCRIPT!
 */

module.exports.Android_FontSize = {
    T_X01: 17,
    T_X03: 15,
    T_X04: 14,
    T_X06: 12,
    T_X08: 10,
    T_X09: 9,
    T_X10: 31,
    T_X13: 10,
    T_X14: 30,
    T_X02: 16,
    T_X05: 13,
    T_X07: 11,
    T_X11: 18,
    T_X12: 20,
    T_X15: 18,
    T_X17: 26,
    T_X18: 13,
    T_X19: 11,
    T_X20: 12,
    T_X21: 23,
    T_X22: 14,
    T_X23: 19,
    T_X24: 20,
    T_X25: 17,
    T_X26: 21,
    T_X27: 17,
    T_X28: 19,
    T_X29: 19,
    T_X31: 21,
    T_X32: 18,
    T_X33: 14,
    T_X34: 12,
    T_X35: 13,
    T_X36: 11,
    T_X37: 14,
    T_X38: 15,
    T_X39: 15,
    T_X40: 3,
};

module.exports.Ios_FontSize = {
    T_X01: 19,
    T_X03: 17,
    T_X04: 16,
    T_X06: 14,
    T_X08: 12,
    T_X09: 10,
    T_X10: 24,
    T_X13: 11,
    T_X14: 35,
    T_X02: 18,
    T_X05: 15,
    T_X07: 13,
    T_X11: 21,
    T_X12: 23,
    T_X15: 20,
    T_X17: 30,
    T_X18: 14,
    T_X19: 12,
    T_X20: 13,
    T_X21: 27,
    T_X22: 17,
    T_X23: 21,
    T_X24: 25,
    T_X25: 20,
    T_X26: 24,
    T_X27: 18,
    T_X28: 22,
    T_X29: 23,
    T_X31: 24,
    T_X32: 21,
    T_X33: 14,
    T_X34: 12,
    T_X35: 14,
    T_X36: 12,
    T_X37: 16,
    T_X38: 16,
    T_X39: 17,
    T_X40: 4,
};

module.exports.H5_FontSize = {
    T_X01: 57,
    T_X03: 51,
    T_X04: 48,
    T_X06: 42,
    T_X08: 36,
    T_X09: 30,
    T_X10: 72,
    T_X13: 33,
    T_X14: 105,
    T_X02: 54,
    T_X05: 45,
    T_X07: 39,
    T_X11: 63,
    T_X12: 69,
    T_X15: 60,
    T_X17: 90,
    T_X18: 42,
    T_X19: 36,
    T_X20: 39,
    T_X21: 81,
    T_X22: 51,
    T_X23: 63,
    T_X24: 75,
    T_X25: 60,
    T_X26: 72,
    T_X27: 54,
    T_X28: 66,
    T_X29: 69,
    T_X31: 72,
    T_X32: 63,
    T_X33: 42,
    T_X34: 36,
    T_X35: 42,
    T_X36: 36,
    T_X37: 48,
    T_X38: 48,
    T_X39: 51,
    T_X40: 12,
};
