// @baidu/wuji-uikit/src/spacing/spacing.js
/**
 * @file 间距编码 spacing.js
 *
 * DO NOT CHANGE, THIS IS GENERATED BY SCRIPT!
 */

module.exports.Android_Spacing = {
    M_H_X001: 2,
    M_H_X002: 5,
    M_H_X003: 8,
    M_H_X004: 10,
    M_H_X005: 13,
    M_H_X006: 15,
    M_H_X007: 18,
    M_H_X008: 26,
    M_H_X009: 31,
    M_H_X010: 50,
    M_H_X011: 9,
    M_H_X012: 4,
    M_H_X013: 30,
    M_H_X014: 14,
    M_H_X015: 1,
    M_H_X016: 7,
    M_H_X017: 22,
    M_H_X018: 12,
    M_H_X019: 8,
    M_H_X020: 7,
    M_H_X021: 11,
    M_H_X022: 6,
    M_H_X023: 9,
    M_H_X024: 33,
    M_H_X025: 3,
    M_H_X026: 5,
    M_H_X027: 7,
    M_H_X028: 3,
    M_H_X029: 3,
    M_H_X030: 2,
    M_H_X031: 2,
    M_H_X032: 16,
    M_H_X033: 6,
    M_H_X034: 4,
    M_H_X035: 0,
    M_H_X036: 21,
    M_H_X037: 5,
    M_H_X038: 10,
    M_H_X039: 9,
    M_H_X040: 6,
    M_H_X041: 24,
    M_H_X042: 28,
    M_H_X043: 35,
    M_H_X044: 6,
    M_H_X045: 11,
    M_H_X046: 14,
    M_H_X047: 3,
    M_H_X048: 11,
    M_H_X049: 48,
    M_H_X050: 39,
    M_H_X051: 18,
    M_H_X052: 13,
    M_H_X053: 12,
    M_H_X054: 20,
    M_H_X055: 17,
    M_H_X056: 27,
    M_W_X001: 2,
    M_W_X002: 4,
    M_W_X003: 5,
    M_W_X004: 6,
    M_W_X005: 7,
    M_W_X006: 8,
    M_W_X007: 9,
    M_W_X008: 10,
    M_W_X009: 13,
    M_W_X010: 15,
    M_W_X011: 3,
    M_W_X016: 17,
    M_W_X017: 11,
    M_W_X018: 2,
    M_W_X019: 26,
    M_W_X021: 1,
    M_W_X022: 12,
    M_W_X023: 3,
    M_W_X024: 14,
    M_W_X025: 8,
    M_W_X026: 5,
    M_W_X027: 13,
    M_W_X028: 5,
    M_W_X029: 9,
    M_W_X030: 3,
    M_W_X031: 21,
    M_W_X032: 35,
    M_W_X033: 1,
    M_W_X034: 9,
    M_W_X035: 18,
    M_W_X036: 9,
    M_W_X037: 52,
    M_W_X038: 12,
    M_W_X039: 24,
    M_W_X040: 12,
    M_W_X041: 8,
    M_W_X042: 19,
    M_W_X043: 16,
    M_W_X044: 70,
    M_W_X045: 39,
    M_W_X046: 18,
    M_W_X047: 44,
    M_W_X048: 23,
    M_T_X001: 5,
    M_T_X002: 4,
    M_T_X003: 3,
    M_T_X004: 8,
    M_T_X005: 10,
    M_T_X006: 10,
    M_T_X007: 9,
    M_T_X008: 2,
    M_T_X009: 13,
    M_T_X010: 15,
    M_T_X011: 6,
    M_T_X012: 7,
    M_T_X013: 3,
};

module.exports.Ios_Spacing = {
    M_H_X001: 3,
    M_H_X002: 6,
    M_H_X003: 9,
    M_H_X004: 12,
    M_H_X005: 15,
    M_H_X006: 18,
    M_H_X007: 21,
    M_H_X008: 30,
    M_H_X009: 36,
    M_H_X010: 58,
    M_H_X011: 10,
    M_H_X012: 5,
    M_H_X013: 35,
    M_H_X014: 17,
    M_H_X015: 1,
    M_H_X016: 8,
    M_H_X017: 24,
    M_H_X018: 15,
    M_H_X019: 12,
    M_H_X020: 9,
    M_H_X021: 15,
    M_H_X022: 15,
    M_H_X023: 12,
    M_H_X024: 38,
    M_H_X025: 12,
    M_H_X026: 9,
    M_H_X027: 12,
    M_H_X028: 4,
    M_H_X029: 3,
    M_H_X030: 1,
    M_H_X031: 2,
    M_H_X032: 18,
    M_H_X033: 7,
    M_H_X034: 6,
    M_H_X035: 1,
    M_H_X036: 24,
    M_H_X037: 6,
    M_H_X038: 12,
    M_H_X039: 10,
    M_H_X040: 7,
    M_H_X041: 27,
    M_H_X042: 33,
    M_H_X043: 40,
    M_H_X044: 10,
    M_H_X045: 13,
    M_H_X046: 16,
    M_H_X047: 4,
    M_H_X048: 13,
    M_H_X049: 56,
    M_H_X050: 45,
    M_H_X051: 21,
    M_H_X052: 15,
    M_H_X053: 14,
    M_H_X054: 60,
    M_H_X055: 20,
    M_H_X056: 32,
    M_W_X001: 3,
    M_W_X002: 5,
    M_W_X003: 6,
    M_W_X004: 7,
    M_W_X005: 8,
    M_W_X006: 9,
    M_W_X007: 10,
    M_W_X008: 12,
    M_W_X009: 15,
    M_W_X010: 17,
    M_W_X011: 4,
    M_W_X016: 20,
    M_W_X017: 13,
    M_W_X018: 2,
    M_W_X019: 30,
    M_W_X021: 1,
    M_W_X022: 15,
    M_W_X023: 3,
    M_W_X024: 16,
    M_W_X025: 10,
    M_W_X026: 5,
    M_W_X027: 15,
    M_W_X028: 6,
    M_W_X029: 10,
    M_W_X030: 3,
    M_W_X031: 24,
    M_W_X032: 40,
    M_W_X033: 6,
    M_W_X034: 3,
    M_W_X035: 21,
    M_W_X036: 11,
    M_W_X037: 60,
    M_W_X038: 14,
    M_W_X039: 27,
    M_W_X040: 12,
    M_W_X041: 8,
    M_W_X042: 22,
    M_W_X043: 18,
    M_W_X044: 80,
    M_W_X045: 45,
    M_W_X046: 21,
    M_W_X047: 50,
    M_W_X048: 26,
    M_T_X001: 6,
    M_T_X002: 8,
    M_T_X003: 4,
    M_T_X004: 9,
    M_T_X005: 11,
    M_T_X006: 12,
    M_T_X007: 10,
    M_T_X008: 2,
    M_T_X009: 15,
    M_T_X010: 18,
    M_T_X011: 7,
    M_T_X012: 8,
    M_T_X013: 4,
};

module.exports.H5_Spacing = {
    M_H_X001: 9,
    M_H_X002: 18,
    M_H_X003: 27,
    M_H_X004: 36,
    M_H_X005: 45,
    M_H_X006: 54,
    M_H_X007: 63,
    M_H_X008: 90,
    M_H_X009: 108,
    M_H_X010: 174,
    M_H_X011: 30,
    M_H_X012: 15,
    M_H_X013: 105,
    M_H_X014: 51,
    M_H_X015: 3,
    M_H_X016: 24,
    M_H_X017: 72,
    M_H_X018: 45,
    M_H_X019: 36,
    M_H_X020: 27,
    M_H_X021: 45,
    M_H_X022: 45,
    M_H_X023: 36,
    M_H_X024: 114,
    M_H_X025: 36,
    M_H_X026: 27,
    M_H_X027: 36,
    M_H_X028: 12,
    M_H_X029: 9,
    M_H_X030: 3,
    M_H_X031: 6,
    M_H_X032: 54,
    M_H_X033: 21,
    M_H_X034: 18,
    M_H_X035: 3,
    M_H_X036: 72,
    M_H_X037: 18,
    M_H_X038: 36,
    M_H_X039: 30,
    M_H_X040: 21,
    M_H_X041: 81,
    M_H_X042: 99,
    M_H_X043: 120,
    M_H_X044: 30,
    M_H_X045: 39,
    M_H_X046: 48,
    M_H_X047: 12,
    M_H_X048: 39,
    M_H_X049: 168,
    M_H_X050: 135,
    M_H_X051: 63,
    M_H_X052: 45,
    M_H_X053: 42,
    M_H_X054: 17,
    M_H_X055: 60,
    M_H_X056: 96,
    M_W_X001: 9,
    M_W_X002: 15,
    M_W_X003: 18,
    M_W_X004: 21,
    M_W_X005: 24,
    M_W_X006: 27,
    M_W_X007: 30,
    M_W_X008: 36,
    M_W_X009: 45,
    M_W_X010: 51,
    M_W_X011: 12,
    M_W_X016: 60,
    M_W_X017: 39,
    M_W_X018: 6,
    M_W_X019: 90,
    M_W_X021: 3,
    M_W_X022: 45,
    M_W_X023: 9,
    M_W_X024: 48,
    M_W_X025: 30,
    M_W_X026: 15,
    M_W_X027: 45,
    M_W_X028: 18,
    M_W_X029: 30,
    M_W_X030: 9,
    M_W_X031: 72,
    M_W_X032: 120,
    M_W_X033: 18,
    M_W_X034: 9,
    M_W_X035: 63,
    M_W_X036: 33,
    M_W_X037: 180,
    M_W_X038: 42,
    M_W_X039: 81,
    M_W_X040: 36,
    M_W_X041: 24,
    M_W_X042: 66,
    M_W_X043: 54,
    M_W_X044: 240,
    M_W_X045: 135,
    M_W_X046: 63,
    M_W_X047: 150,
    M_W_X048: 78,
    M_T_X001: 18,
    M_T_X002: 24,
    M_T_X003: 12,
    M_T_X004: 27,
    M_T_X005: 33,
    M_T_X006: 36,
    M_T_X007: 30,
    M_T_X008: 6,
    M_T_X009: 45,
    M_T_X010: 54,
    M_T_X011: 21,
    M_T_X012: 24,
    M_T_X013: 12,
};
