// @baidu/auc-refute-rumor/index.ts
/**
 * @file index.ts
 * @author gaohong01
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucRefuteRumorData} from './types';
import {WujiAvatar, WujiIcon} from '@baidu/wuji-san';
import '@baidu/wuji-icons/svg/arrow-goto-o-48.svg';
import {jump} from '@baidu/auc-base/jump';
import {componentLog} from '@baidu/auc-log';

const titlePrefixMap = {
    rumour4: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/2141838041.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/3362262595.png'
    },
    rumour5: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/2878891855.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/266937643.png'
    },
    rumour6: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/431536991.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/2158612619.png'
    },
    rumour7: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/1107054048.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/3436525879.png'
    },
    rumour8: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/2532855551.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/3184130363.png'
    },
    rumour9: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/618173167.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/33055047.png'
    },
    rumour10: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/753650406.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/3390687274.png'
    },
    rumour11: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/1393075198.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/1212768251.png'
    },
    rumour12: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/3512218671.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/1965651531.png'
    },
    rumour13: {
        day: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/3963100575.png',
        night: 'https://b.bdstatic.com/searchbox/image/gcp/20221226/847842459.png'
    }
}
export class AucRefuteRumor extends AucComponent {
    static template:string = /* html */ `
    <div class="{{styles.aucRefuteRumorWrapper}} {{isBigEvent ? styles.bigEvent : ''}}" on-click="comClick">
        <div class="{{styles.title}}">{{title}}</div>
        <div class="{{styles.userInfo}}" on-click="avatarClick">
            <wuji-avatar
                src="{{auimage}}"
                size="72"
                vip="{{vflag + ''}}"
                class="{{styles.avatar}}"
            ></wuji-avatar>
            <div class="{{styles.auname}}">{{auname || source}}</div>
            <div class="{{styles.point}}">·</div>
            <div class="{{styles.auabs}}">{{auabs}}</div>
        </div>
        <div class="{{styles.desc}}">{{description}}</div>
        <div class="{{styles.toDetail}}">
            查看详情
            <wuji-icon class="{{styles.arrow}}" type="arrow-goto-o-48" size="36"></wuji-icon>
        </div>
        <div class="{{styles.titlePrefix}}" style="background-image:url({{titlePrefixPic}})"></div>
    </div>
    `;
    
    static components:object = {
        'wuji-avatar': WujiAvatar,
        'wuji-icon': WujiIcon
    };
    static computed = {
        titlePrefixPic() {
            const type = this.data.get('title_prefix');
            const theme = this.data.get('pageMode') === 'night' ? 'night' : 'day';
            return titlePrefixMap[type] ? titlePrefixMap[type][theme] : '';
        }
    };
    initData(): AucRefuteRumorData {
        return {
            styles,
            logBaseType: 'refute_rumor_',
            logExtraParams: {}
        };
    }

    attached(): void {
        this.registerScrollInView().then(() => {
            componentLog(this.data.get('logBaseType'), 'show', this.data.get('logExtraParams'));
        });
    }

    comClick(e) {
        e.preventDefault();
        e.stopPropagation();
        componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
        jump({
            cmd: this.data.get('cmd'),
            link: this.data.get('link')
        });
    }

    avatarClick(e) {
        e.preventDefault();
        e.stopPropagation();
        componentLog(this.data.get('logBaseType'), 'avatar_click', this.data.get('logExtraParams'));
        jump({
            cmd: this.data.get('auhomepagecmd'),
            link: this.data.get('auhomepageurl')
        });
    }
}

export default AucRefuteRumor;
