// @baidu/feed-baselog/src/index.devtools.ts
/**
 * @file baselog devtools cdp
 * @author okaychen<chenglong13@baidu.com>
 */

import BaseLog from './index';
import merge from 'lodash.merge';
import ThunderH5 from './lib/thunder.h5';
import {TBASEURL, UBASEURL} from './lib/const';
import {ConfigType, ArgsType} from './lib/interface';

// @ts-ignore
const $devtools = global.$devtools;
const isDevtools = typeof $devtools !== 'undefined' && $devtools.sendCommand && $devtools.registerEvent;

if (isDevtools) {
    $devtools.baselogMonitor = [];
    /* istanbul ignore next */
    $devtools.registerEvent('baselog.getData', () => {
        return $devtools.baselogMonitor;
    });
}

class $BaseLog<U> extends BaseLog {
    constructor(InitArgs?: U) {
        super(InitArgs);
    }

    sendThunder<T>(params: T, callback?: Function, url: string = TBASEURL): void {
        super.sendThunder(params, callback, url);
        const thunder = new ThunderH5(this.baseParams);
        const thunderURL = thunder.assembleUrl(url, this.baseParams, params);
        const tFinnalParams = <T>merge({}, this.baseParams, params);

        this.$handleParams('thunder', tFinnalParams, url, thunderURL);
    }

    sendUbc<T, U>(ubcList: ConfigType<T, U>, callback?: Function, url: string = UBASEURL): void {
        super.sendUbc(ubcList, callback);
        /* istanbul ignore next */
        const uFinnalParams = merge({}, this.finalParams?.params, ubcList);

        let ubcName = 'ubc';
        // 如果打点ID是13587，则为thunder迁移ubc点位
        if (+ubcList.ubcId === 13587) {
            ubcName = 'thunder to ubc';
        }

        this.$handleParams(ubcName, uFinnalParams, url);
    }

    // devtools.getData 参数
    $handleParams(ubcName?: String, args?: ArgsType, url?: String, requestURL?: String): void {
        /* istanbul ignore next */
        if (!isDevtools) {
            return;
        }
        
        let $params = {
            args,
            path: url,
            method: 'get',
            tid: args?.tid,
            ubcid: args?.ubcId,
            name: `${ubcName}-${args?.ubcId || args?.tid}`,
            request: {
                'Request Method': 'get',
                'Request URL': requestURL
            }
        }
            
        $devtools.baselogMonitor.push($params);
    }
};

export default $BaseLog;
