// @baidu/auc-loader/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import {Component} from 'san';
import styles from './style.module.less';
import {WujiIcon} from '@baidu/wuji-san';
import '@baidu/wuji-icons/svg/d20-arrow-down-o-48.svg';
import {AucLoaderData, LoaderState} from './types';

export {
    LoaderState
};

export default class AucLoader extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{loaderClass}}" on-click="clickLoader">
            <div class="{{styles.loaderContainer}} {{stateClass}}">
                <template s-if="state === LoaderState.DEFAULT">
                    <slot name="default">&nbsp;&nbsp;查看更多<wuji-icon class="{{styles.defaultIcon}}" type="d20-arrow-down-o-48" size="36"></wuji-icon></slot>
                </template>
                <template s-if="state === LoaderState.LOADING">
                    <slot name="loading">
                        <div class="{{styles.loading}}">
                            <div class="{{styles.iconWrap}}"><div class="{{styles.icon}}"/></div>
                            <div class="{{styles.text}}">加载中...</div>
                        </div>
                    </slot>
                </template>
                <template s-if="state === LoaderState.END">
                    <slot name="end">已经到底啦</slot>
                </template>
                <template s-if="state === LoaderState.FAIL">
                    <slot name="error">加载失败，点击重试</slot>
                </template>
            </div>
        </div>
    `;

    static computed = {
        stateClass() {
            const state:number = this.data.get('state');
            const styles:object = this.data.get('styles');
            const className:string = `state${state}`;
            return styles[className];
        },
        loaderClass() {
            return [
                styles.aucLoader,
                this.data.get('isDoubleRow') ? styles.doubleRowLoader : '',
                this.data.get('isColumnBackColor') ? styles.isColumnBackColor : ''
            ]
        }
    };

    static components = {
        'wuji-icon': WujiIcon
    };

    initData():AucLoaderData {
        return {
            state: LoaderState.DEFAULT,
            styles,
            LoaderState
        };
    }

    clickLoader(e) {
        const state:number = this.data.get('state');
        if ([LoaderState.FAIL, LoaderState.DEFAULT].includes(state)) {
            this.fire('click-loader', e);
        }
    }
}
