// @baidu/ubc-report-sdk/src/Command.js
/**
 * @file 指令类
 * @author wangbing11(wangbing11@baidu.com)
 * // command命令的调用方法：ubc(command, [...fields], [fieldsObject])
 */

// command 标识符的正则解析 [trackerName.][pluginName:]methodName
const commandRegex = /^(?:(\w+)\.)?(?:(\w+)\.)?(\w+)$/;

/**
 * 初始化一个命令
 *
 * @class Command
 * @param {Array} args 用户输入的或者内部生成的不正规的指令参数
 *
 */
/* eslint-disable */
export default function Command(args) {
/* eslint-enable */
    let r = commandRegex.exec(args[0]);
    if (r != null && r.length === 4) {
        // Tracker名字默认为t0
        this.trackerName = r[1] || 't0';

        // pluginName
        this.pluginName = r[2] || '';

        // methodName
        this.methodName = r[3];

        this.fields = [].slice.call(args, 1);

        this.isCreateCommand = 'init' === this.methodName;

        this.isConfigCommand = 'config' === this.methodName;

        this.isEventCommand = 'event' === this.methodName || 'pv' === this.methodName;

        this.isTimingCommand = 'timing' === this.methodName;

        // 必须要有一个方法名
        if (!this.methodName) {
            throw 'UBC SDK methodName can not be empty!';
        }

        // 如果要创建一个新的跟踪器，必须要要有一个ID，ID为数字或者数字组成的字符
        if (this.isCreateCommand) {
            if (!this.fields[0]) {
                throw 'UBC SDK serverId can not be empty!';
            }
        }
    }
    else {
        throw 'UBC SDK error inputs';
    }
}
