// @baidu/nano/es/loader/index.js
/**
 * @file 组件 loader 进度加载
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';

const prefixCls = 's-loader';
const stateType = [0, 1, 2, 3];

export default class Loader extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="${prefixCls}">
            <div class="${prefixCls}-container state-{{state}}">
                <div class="${prefixCls}-ing">
                    <slot name="spinner">
                        <div class="icon-wrap"><div class="icon"/></div>
                    </slot>
                    <div class="text">{{textIng}}</div>
                </div>
                <div class="${prefixCls}-end">{{textEnd}}</div>
                <div class="${prefixCls}-err" on-click="handleErrorClick">{{textErr}}</div>
            </div>
        </div>
    `;

    static dataTypes = {
        state: DataTypes.oneOf(stateType)
    };

    initData() {
        return {
            state: 0,
            textIng: '正在加载...',
            textEnd: '已经到底啦',
            textErr: '加载失败 点击重新加载'
        };
    }

    handleErrorClick(e) {
        this.fire('error-click', e);
    }
}

Loader.stateType = stateType;
