// @baidu/auc-operation-position/components/opetration-beancurd-cube/item.ts
/**
 * @file index.ts
 * @author gaohong01
 */
import {Component} from 'san';
import styles from './style.module.less';
import {BeancurdCubeItemData} from './types';

export class beancurdCubeItem extends Component {
    static template: string = /* html */ `
        <div class="{{itemClass}}" style="background-image:url({{itemIcon}})" on-click="handleClick"></div>
     `;
    static components: object = {
    };

    static computed: object = {
        itemClass() {
            let item = this.data.get('item');
            return [
                styles.beancurdCubeItem,
                item.isBig ? styles.beancurdCubeItemBig : styles.beancurdCubeItemSmall,
                item.needMr ? styles.beancurdCubeItemMr : '',
                item.needMb ? styles.beancurdCubeItemMb : ''
            ];
        },
        itemIcon() {
            let divIcon = this.data.get('item.divIcon');
            const pageMode = this.data.get('pageMode');
            return divIcon ? typeof divIcon === 'string' ? divIcon : divIcon[pageMode === 'dark' ? 'day' : pageMode] : '';
        }
    };

    initData(): BeancurdCubeItemData {
        return {
            styles,
            item: {},
            pageMode: 'day'
        };
    }

    inited(): void {
    }

    attached(): void {
    }

    handleClick(): void {
        this.fire('bcclick', this.data.get('index'));
    }
}

export default beancurdCubeItem;
