// @baidu/ug-invoke-app/src/config/app/marketMap.ts
/**
 * @file 安卓应用商店调起协议
 * @author zhangsiyuan(zhangsiyuan@baidu.com)
 */
import {androidStoreRules} from '../rules/androidStoreRules';
import {getHostInfo} from '../../utils/detect/getHostInfo';

export const marketMap = {
    mimarket: (pkgName: string) => `mimarket://details?id=${pkgName}&back=true`,
    appmarket: (pkgName: string) => `appmarket://details?id=${pkgName}&back=true`,
    oppomarket: (pkgName: string) => `oppomarket://details?packagename=${pkgName}`,
    vivomarket: (pkgName: string) => `vivomarket://details?id=${pkgName}`,
    samsungapps: (pkgName: string) => `samsungapps://ProductDetail/${pkgName}?back=1`,
    market: (pkgName: string) => `market://details?id=${pkgName}`,
};

export type MarketName = keyof typeof marketMap;

/**
 * 获得当前机型 调用安卓市场协议的scheme
 *
 */
export function getMarketScheme(pkgName: string) {
    const {browser, brand} = getHostInfo();
    let marketScheme = `market://details?id=${pkgName}`;

    for (const [browserRule, brandRule, schemeName] of androidStoreRules) {
        if (
            (browserRule === '*' || browser === browserRule)
            && (brandRule === '*' || brand === brandRule)
        ) {
            marketScheme = marketMap[schemeName](pkgName);
            break;
        }
    }

    // 贴吧特殊case
    // 贴吧专用协议,贴吧不加后面参数生效不了
    if (browser === 'tieba') {
        marketScheme += '&enable_tieba_native_open=1';
    }

    return marketScheme;
}
