// @baidu/wuji-san/es/wuji-more/index.js
/**
 * @file 组件 WujiMore 无极-查看更多
 * @author yanghongxue <yanghongxue@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';
import '@baidu/wuji-icons/svg/cardbody-slidemore-o-124.svg';

const prefixCls = 'w-more';

export default class WujiMore extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap" on-click="handleClick">
            <div class="${prefixCls}-content">
                <div class="${prefixCls}-icon">
                    <wuji-icon size="auto" type="cardbody-slidemore-o-124"/>
                </div>
                <div class="${prefixCls}-text">{{text}}</div>
            </div>
        </div>
    `;

    static components = {
        'wuji-icon': Icon,
    };

    static dataTypes = {
        url: DataTypes.string,
        text: DataTypes.string,
        jumpFn: DataTypes.func,
    };

    initData() {
        return {
            text: '查看更多',
        };
    }

    redirect(url) {
        if (!url) {
            return;
        }

        window.location.href = url;
    }

    handleClick() {
        const {url, jumpFn} = this.data.get();
        let flag = true;

        if (jumpFn) {
            let ret = jumpFn();

            if (ret && typeof ret.then === 'function') {
                flag = false;
                ret.then(() => {
                    this.redirect(url);
                });
            }
        }

        flag && this.redirect(url);
    }
}
