// @baidu/wuji-san/es/wuji-quick-evaluate/index.js
/**
 * @file 组件 WujiQuickEvaluate 无极-敏捷评估
 * @author baozhixin <baozhixin@baidu.com>
 */

import {Component, DataTypes} from 'san';
import Button from './button';
import Star from './star';

export default class WujiQuickEvaluate extends Component {
    static template = /* html */ `
        <template>
            <type-button
                s-if="type === 'button'"
                init-title="{{initTitle}}"
                succ-title="{{succTitle}}"
                items="{{items}}"
                completed="{{completed}}"
                env="{{env}}"
                spacing="{{spacing}}"
                on-submit="handleSubmit"
            />
            <type-star
                s-if="type === 'star'"
                init-title="{{initTitle}}"
                succ-title="{{succTitle}}"
                star-desc="{{starDesc}}"
                spacing="{{spacing}}"
                on-click="handleClick"
            />
        </template>
    `;

    static components = {
        'type-button': Button,
        'type-star': Star,
    };

    static dataTypes = {
        type: DataTypes.oneOf([
            'button',
            'star',
        ]),
    };

    initData() {
        return {
            type: 'star',
            initTitle: '',
            succTitle: '',
            starDesc: [],
            items: [],
            completed: false,
            spacing: true,
        };
    }

    handleClick(data) {
        this.fire('click', data);
    }

    handleSubmit(data) {
        this.fire('submit', data);
    }
}

export {Button, Star};
