// @baidu/wuji-san/es/wuji-quick-evaluate/star.js
/**
 * @file 组件 WujiQuickEvaluate 无极-敏捷评估-星级评分
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/star.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';
import '@baidu/wuji-icons/svg/score-star-f-96.svg';
import '@baidu/wuji-icons/svg/score-star-o-96.svg';

const prefixCls = 'w-qes';
// eslint-disable-next-line
const sparkHtml = '<div class="spark"><svg class="line" xmlns="http://www.w3.org/2000/svg"><rect width="100%" height="100%"></rect></svg></div>';

export default class WujiQuickEvaluate extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap {{spacing ? '${prefixCls}-spacing' : ''}}">
            <div s-if="title" class="${prefixCls}-title">{{title | raw}}</div>
            <div class="${prefixCls}-score {{animate ? 'animate' : ''}}">
                <wuji-icon
                    s-for="item, index in itemList trackBy item"
                    class="${prefixCls}-score-item ipos{{index + 1}} {{score > index ? 'active' : ''}}"
                    type="{{score > index ? activeIcon : defaultIcon}}"
                    size="{{96}}"
                    on-click="handleSelect($event, index)"
                />
                <div s-if="score > 3" class="${prefixCls}-fireworks">
                    <div class="fire-bind for{{score}}">{{html1 | raw}}</div>
                    <div s-if="score === 5">{{html2 | raw}}</div>
                </div>
            </div>
            <div
                s-if="stext"
                class="${prefixCls}-stext"
            >{{stext}}</div>
        </div>
    `;

    static components = {
        'wuji-icon': Icon,
    };

    static dataTypes = {
        animate: DataTypes.bool,
        initTitle: DataTypes.string,
        succTitle: DataTypes.string,
        starDesc: DataTypes.arrayOf(DataTypes.string),
        score: DataTypes.number,
        defaultIcon: DataTypes.string,
        activeIcon: DataTypes.string,
        spacing: DataTypes.bool,
    };

    static computed = {
        title() {
            const data = this.data;
            const initTitle = data.get('initTitle');
            const succTitle = data.get('succTitle');
            const score = data.get('score');
            return (score > 0 && succTitle) || initTitle;
        },
        stext() {
            const data = this.data;
            const starDesc = data.get('starDesc');
            const score = data.get('score');
            return starDesc[score] || '';
        },
        itemList() {
            return new Array(5).fill('');
        },
        html1() {
            return sparkHtml.repeat(5);
        },
        html2() {
            const html = [];
            for (let index = 1; index < 9; index++) {
                const wraper = [
                    `<div class="fire fire${index}">`,
                    '</div>',
                ];
                const sparks = sparkHtml.repeat(8);
                html.push(wraper.join(sparks));
            }
            return html.join('');
        },
    };

    initData() {
        return {
            animate: false,
            initTitle: '',
            succTitle: '',
            starDesc: [],
            score: 0,
            defaultIcon: 'score-star-o-96',
            activeIcon: 'score-star-f-96',
            spacing: true,
        };
    }

    // 点击评分：1、设置分数；2、重置动效；
    handleSelect({event}, index) {
        event.stopPropagation();
        this.data.set('score', index + 1);
        this.data.set('animate', false);
        this.fire('click', {event, score: index + 1});

        setTimeout(() => {
            this.data.set('animate', true);
        }, 20);
    }
}
