// @baidu/mo-opinion/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var moActions = require('@baidu/mo-actions');
var Base = require('@baidu/mo-base');
var wujiSan = require('@baidu/wuji-san');
var xbox = require('@baidu/xbox');
var styles = require('./style.module.less');
require('@baidu/wuji-icons/svg/defaultpic-logo-87.svg');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var Base__default = /*#__PURE__*/_interopDefaultLegacy(Base);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author lvkunjie <lvkunjie@baidu.com>
 */
// @san/component
var Opinion = /** @class */ (function (_super) {
    __extends(Opinion, _super);
    function Opinion() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Opinion.prototype.initData = function () {
        return {
            styles: styles__default["default"],
        };
    };
    Opinion.prototype.gifOnLoad = function () {
        this.eventDispatch('gifload', {
            status: 'success',
        });
        // 图片加载完成后删掉占位图
        var iconContainerRef = this.ref('gifContainer');
        if (iconContainerRef) {
            iconContainerRef.style.display = 'none';
        }
    };
    Opinion.prototype.onLoad = function (val, isGif) {
        this.eventDispatch('load', {
            status: 'success',
        });
        // 图片加载完成后删掉占位图
        var iconContainerRef = this.ref("iconContainer".concat(isGif ? 'Gif' : '', "-").concat(val));
        if (iconContainerRef) {
            iconContainerRef.style.display = 'none';
        }
    };
    Opinion.prototype.handelOpinionClick = function (index, list) {
        this.eventDispatch(moActions.CLICK_OPINION_ITEM, {
            index: index,
            list: list,
        });
    };
    Opinion.template = "\n        <div\n            class=\"{{styles.memeWrapper}}\"\n        >\n            <ul\n                s-if=\"list && list.length\"\n                class=\"{{styles.list}} {{selectedIndex > -1 ? styles.hasSelected : ''}}\"\n            >\n                <li\n                    s-for=\"item, index in list trackBy index\"\n                    class=\"{{[\n                        styles.item,\n                        selectedIndex === index ? styles.selected : '',\n                    ]}}\"\n                >\n                    <div\n                        class=\"{{styles.itemGroup}}\"\n                        on-click=\"handelOpinionClick(index, list)\"\n                    >\n\n                        <!-- <div\n                            s-if=\"gifAnimation\"\n                            class=\"{{[\n                                styles.gifAnimation,\n                                styles.defaultImg\n                            ]}}\"\n                        > -->\n                        <div\n                            class=\"{{[\n                                gifAnimation ? styles.show : styles.hidden,\n                                styles.gifAnimation,\n                                styles.defaultImg\n                            ]}}\"\n                        >\n                            <div\n                                s-ref=\"{{'iconContainerGif-' + index}}\"\n                                class=\"{{styles.iconContainer}}\"\n                            >\n                                <wuji-icon\n                                    class=\"{{styles.logo}}\"\n                                    type=\"defaultpic-logo-87\"\n                                    size=\"70\"\n                                />\n                            </div>\n                            <img\n                                src=\"{{item.gifurl || item.pngurl}}\"\n                                class=\"{{[\n                                    styles.itemImg,\n                                    styles.imgCover\n                                ]}}\"\n                                on-load=\"onLoad(index, true)\"\n                            />\n                        </div>\n                        <div\n                            class=\"{{[\n                                styles.defaultImg,\n                                gifAnimation ? styles.hidden : styles.show\n                            ]}}\"\n                        >\n                        <!-- <div\n                            s-else\n                            class=\"{{[\n                                styles.defaultImg\n                            ]}}\"\n                        > -->\n                            <div\n                                s-ref=\"{{'iconContainer-' + index}}\"\n                                class=\"{{styles.iconContainer}}\"\n                            >\n                                <wuji-icon\n                                    class=\"{{styles.logo}}\"\n                                    type=\"defaultpic-logo-87\"\n                                    size=\"70\"\n                                />\n                            </div>\n                            <img\n                                src=\"{{item.pngurl}}\"\n                                class=\"{{[\n                                    styles.itemImg,\n                                    styles.imgCover\n                                ]}}\"\n                                on-load=\"onLoad(index, false)\"\n                            />\n                        </div>\n                        <div\n                            s-if=\"selectedIndex >= 0\"\n                            class=\"{{styles.itemCount}}\"\n                        >{{(item.like_num > 0 ? item.like_num : 0) | numberFormat}}</div>\n                        <div\n                            s-else\n                            class=\"{{styles.itemName}}\"\n                        >{{item.name}}</div>\n                    </div>\n                </li>\n            </ul>\n            <div\n                s-if=\"list && list[selectedIndex] && list[selectedIndex].gifurl\"\n                class=\"{{animateIndex >= 0 ? styles.popup : styles.hidden}}\"\n            >\n                <div\n                    class=\"{{[\n                        styles.bgWrapper,\n                        styles.defaultImg\n                    ]}}\"\n                >\n\n                    <div\n                        s-ref=\"gifContainer\"\n                        class=\"{{styles.gifContainer}}\"\n                    >\n                        <wuji-icon\n                            class=\"{{styles.logo}}\"\n                            type=\"defaultpic-logo-87\"\n                            size=\"87\"\n                        />\n                    </div>\n                    <img\n                        src=\"{{list[selectedIndex].gifurl}}\"\n                        class=\"{{[\n                            styles.popupimg,\n                            styles.imgCover\n                        ]}}\"\n                        on-load=\"gifOnLoad\"\n                    />\n                    <span\n                        class=\"{{styles.popuptxt}}\"\n                    ></span>\n                </div>\n            </div>\n        </div>\n    ";
    // eslint-disable-next-line @typescript-eslint/member-ordering
    Opinion.components = {
        'wuji-icon': wujiSan.WujiIcon,
    };
    Opinion.filters = {
        // 以高精度模式格式化数字
        numberFormat: function (num) {
            return xbox.numberFormat(num);
        },
    };
    return Opinion;
}(Base__default["default"]));

exports["default"] = Opinion;
