// @baidu/mo-opinion/dist/index.js
/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author lvkunjie <lvkunjie@baidu.com>
 */
import { CLICK_OPINION_ITEM, } from '@baidu/mo-actions';
import Base from '@baidu/mo-base';
import { WujiIcon } from '@baidu/wuji-san';
import { numberFormat } from '@baidu/xbox';
import styles from './style.module.less';
import '@baidu/wuji-icons/svg/defaultpic-logo-87.svg';
export * from './interface';
// @san/component
export default class Opinion extends Base {
    static template = /* html */ `
        <div
            class="{{styles.memeWrapper}}"
        >
            <ul
                s-if="list && list.length"
                class="{{styles.list}} {{selectedIndex > -1 ? styles.hasSelected : ''}}"
            >
                <li
                    s-for="item, index in list trackBy index"
                    class="{{[
                        styles.item,
                        selectedIndex === index ? styles.selected : '',
                    ]}}"
                >
                    <div
                        class="{{styles.itemGroup}}"
                        on-click="handelOpinionClick(index, list)"
                    >

                        <!-- <div
                            s-if="gifAnimation"
                            class="{{[
                                styles.gifAnimation,
                                styles.defaultImg
                            ]}}"
                        > -->
                        <div
                            class="{{[
                                gifAnimation ? styles.show : styles.hidden,
                                styles.gifAnimation,
                                styles.defaultImg
                            ]}}"
                        >
                            <div
                                s-ref="{{'iconContainerGif-' + index}}"
                                class="{{styles.iconContainer}}"
                            >
                                <wuji-icon
                                    class="{{styles.logo}}"
                                    type="defaultpic-logo-87"
                                    size="70"
                                />
                            </div>
                            <img
                                src="{{item.gifurl || item.pngurl}}"
                                class="{{[
                                    styles.itemImg,
                                    styles.imgCover
                                ]}}"
                                on-load="onLoad(index, true)"
                            />
                        </div>
                        <div
                            class="{{[
                                styles.defaultImg,
                                gifAnimation ? styles.hidden : styles.show
                            ]}}"
                        >
                        <!-- <div
                            s-else
                            class="{{[
                                styles.defaultImg
                            ]}}"
                        > -->
                            <div
                                s-ref="{{'iconContainer-' + index}}"
                                class="{{styles.iconContainer}}"
                            >
                                <wuji-icon
                                    class="{{styles.logo}}"
                                    type="defaultpic-logo-87"
                                    size="70"
                                />
                            </div>
                            <img
                                src="{{item.pngurl}}"
                                class="{{[
                                    styles.itemImg,
                                    styles.imgCover
                                ]}}"
                                on-load="onLoad(index, false)"
                            />
                        </div>
                        <div
                            s-if="selectedIndex >= 0"
                            class="{{styles.itemCount}}"
                        >{{(item.like_num > 0 ? item.like_num : 0) | numberFormat}}</div>
                        <div
                            s-else
                            class="{{styles.itemName}}"
                        >{{item.name}}</div>
                    </div>
                </li>
            </ul>
            <div
                s-if="list && list[selectedIndex] && list[selectedIndex].gifurl"
                class="{{animateIndex >= 0 ? styles.popup : styles.hidden}}"
            >
                <div
                    class="{{[
                        styles.bgWrapper,
                        styles.defaultImg
                    ]}}"
                >

                    <div
                        s-ref="gifContainer"
                        class="{{styles.gifContainer}}"
                    >
                        <wuji-icon
                            class="{{styles.logo}}"
                            type="defaultpic-logo-87"
                            size="87"
                        />
                    </div>
                    <img
                        src="{{list[selectedIndex].gifurl}}"
                        class="{{[
                            styles.popupimg,
                            styles.imgCover
                        ]}}"
                        on-load="gifOnLoad"
                    />
                    <span
                        class="{{styles.popuptxt}}"
                    ></span>
                </div>
            </div>
        </div>
    `;
    // eslint-disable-next-line @typescript-eslint/member-ordering
    static components = {
        'wuji-icon': WujiIcon,
    };
    static filters = {
        // 以高精度模式格式化数字
        numberFormat(num) {
            return numberFormat(num);
        },
    };
    initData() {
        return {
            styles,
        };
    }
    gifOnLoad() {
        this.eventDispatch('gifload', {
            status: 'success',
        });
        // 图片加载完成后删掉占位图
        const iconContainerRef = this.ref('gifContainer');
        if (iconContainerRef) {
            iconContainerRef.style.display = 'none';
        }
    }
    onLoad(val, isGif) {
        this.eventDispatch('load', {
            status: 'success',
        });
        // 图片加载完成后删掉占位图
        const iconContainerRef = this.ref(`iconContainer${isGif ? 'Gif' : ''}-${val}`);
        if (iconContainerRef) {
            iconContainerRef.style.display = 'none';
        }
    }
    handelOpinionClick(index, list) {
        this.eventDispatch(CLICK_OPINION_ITEM, {
            index,
            list,
        });
    }
}
