// @baidu/xbox/na/o2o.js
/**
 * 调起o2o框架，支持轻应用（直达号）调起
 * @memberOf Bdbox.client
 * @name o2o
 * @param  {string} url 调起的url
 * @param  {object} opt 调起的参数
 * @param  {string} schemeHead
 * @return {object}     Bdbox
 *
 * @author wangyongqing01
 * @version $Id: o2o.js 289534 2016-04-07 06:00:51Z yupeng07 $
 *
 * @example
 * ```js
 * //例如：调起o2o打开百度首页
 * Box.o2o('http://m.baidu.com', {color:'0000ff'});
 * //调起o2o框架，打开首页，并且指定顶框黑条颜色改变成0000ff(16进制颜色值)
 * Box.o2o('http://m.baidu.com', {color:'0000ff'});
 * //调起直达号
 * Bdbox.client.o2o('http://m.baidu.com', {appid:xxxx})
 * ```
 */

import invoke, {invokeP} from './invoke';
import {isPlainObject} from '../is';
import each from '../each';
import {isIOS, isAndroid} from '../browser/detect';
import {command} from './android';

const encode = encodeURIComponent;
/* eslint-disable fecs-camelcase */
let min_v = '';
/* eslint-disable fecs-camelcase */

export default function o2o(url, opt, schemeHead) {
    if (isIOS()) {
        if (isPlainObject(url)) {
            opt = url;
            url = url.url;
            delete opt.url;
        }
        let obj = {
            openurl: encode(url),
            minver: '5.3.0.0',
            isla: 0,
            opentype: 1,
            append: 0,
            rbtnstyle: 2
        };
        if (isPlainObject(opt)) {
            let alias = {
                color: 'barcolor'
            };
            each(opt, function o2o(value, key) {
                key = alias[key] || key;
                obj[key] = value;
            });
        }

        if (obj.appid) {
            obj.isla = 1;
        }
        invokeP('easybrowse', obj, schemeHead);
        return;
    }
    if (isPlainObject(url)) {
        opt = url;
        url = url.url;
        delete opt.url;
    }
    let intent = ['S.bdsb_light_start_url=' + encode(url)];

    if (isPlainObject(opt)) {
        // android intent中不能传递不认参数，不然会报错,因为他们只是透传 intent,没有处理，所以要把多于参数去掉。
        // 2016.04.07
        /* eslint-disable fecs-camelcase */
        min_v = opt.min_v;
        delete opt.min_v;
        /* eslint-disable fecs-camelcase */
        const alias = {
            color: 'i.extra_actionbar_color_id',
            appid: 'S.bdsb_wallet_appid'
        };
        each(opt, function (value, key) {
            if (key === 'color') {
                value = parseInt(('0x' + value), 16) | 0xff000000;
            }

            key = alias[key] || key;
            intent.push(key + '=' + value);
        });
    }
    intent = intent.join(';');
    let params = {
        intent: 'intent:#Intent;' + intent + ';end',
        /* eslint-disable fecs-camelcase */
        min_v: (min_v !== '' ? min_v : '16783629'),
        /* eslint-disable fecs-camelcase */
        mode: '0'
    };
    // android需要统计各种O2O打开时长，扩展了几个O2O 框架，需要传component参数来区分，以及版本信息，参数不能向下兼容。
    // 当参数存在component时，不设置class,因为端需要做统计使用的class不一样，为了统一以后都只传 component参数就OK。为了兼容之前的代码如果为空时class保持不变。
    // 2016.04.07

    if (!opt || !opt.component || opt.component === '') {
        params.class = 'com.baidu.searchbox.wallet.WalletServiceActivity';
    }

    command(params.intent, params.mode, params.class, params.min_v);
}
