// @baidu/wuji-san/es/wuji-smallapp-container/components/Layout.js
/**
 * @file Layout
 * @author baozhixin@baidu.com
 */

import {Component} from 'san';
import Image from '../../wuji-image';
import Title from '../../wuji-title';
import Widget from '../../wuji-widget';

class Layout extends Component {
    static template = /* html */ `
        <div on-click="handleClick">
            <slot name="header"/>
            <div class="img-txt-wrapper">
                <div s-if="hasImage" class="img-left">
                    <wuji-image
                        s-bind="imageProps"
                        s-ref="image"
                        src="{{image}}"
                        user-select="{{userSelect}}"
                    >
                        <slot name="imageMask" slot="mask"/>
                    </wuji-image>
                </div>
                <div class="txt-right">
                    <slot name="title">
                        <wuji-title
                            s-if="title || titleProps.text"
                            s-bind="{{titleProps}}"
                            text="{{title || titleProps.text}}"
                            label="{{label}}"
                            labelProps="{{labelProps}}"
                            spacing="{{false}}"
                            size="{{titleSize}}"
                        />
                    </slot>
                    <slot name="detail"/>
                </div>
            </div>
            <slot name="footer" test="test"/>
            <wuji-widget
                s-if="widgets && widgets.length > 0"
                items="{{widgets}}"
                spacing
                dislike="{{dislike}}"
                has-margin-top="{{false}}"
                on-dislike="dislike"
            />
        </div>
    `;

    static components = {
        'wuji-image': Image,
        'wuji-title': Title,
        'wuji-widget': Widget,
    };

    static computed = {
        hasImage() {
            const image = this.data.get('image');
            const imageProps = this.data.get('imageProps');
            return (imageProps && imageProps.src) || image;
        },
    };

    initData() {
        return {
            title: '',
            dislike: false,
            imageProps: {},
            userSelect: true,
        };
    }

    handleClick(event) {
        this.fire('click', {event, data: this.data.get()});
    }
}

export default Layout;
