// @baidu/wuji-san/es/wuji-smallapp-container/components/Progress.js
/**
 * @file Progress
 * @author baozhixin@baidu.com
 */

import {Component} from 'san';
import pr2rem from '../../utils/pr2rem';
import isNumeric from '../../utils/is-numeric';

class Progress extends Component {
    static template = /* html */ `
        <div class="progressContainer" style="color:{{color}}">
            <div class="progress" style="{{getProgressStyle(width)}}">
                <svg class="track" style="{{getTrackStyle(trackColor)}}">
                    <rect width="100%" height="100%"/>
                </svg>
                <svg class="line" style="{{format(percentage)}}">
                    <rect width="100%" height="100%"/>
                </svg>
            </div>
            <div s-if="text" class="text">{{text}}</div>
        </div>
    `;

    getProgressStyle(width) {
        if (isNumeric(width)) {
            return {
                width: `${pr2rem(width)}rem`,
            };
        }
    }

    getTrackStyle(color) {
        if (color) {
            return {
                color,
                opacity: 1,
            };
        }
    }

    format(val) {
        let num = (val + '').replace('%', '');
        if (!isNaN(num) && num >= 0 && num <= 100) {
            return {
                width: `${num}%`,
            };
        }
    }
}

export default Progress;
