// san-ssr/dist/target-js/compilers/component-compiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentClassCompiler = void 0;
const lodash_1 = require("lodash");
const component_1 = require("../../models/component");
const lang_1 = require("../../utils/lang");
const js_emitter_1 = require("../js-emitter");
class ComponentClassCompiler {
    constructor(emitter = new js_emitter_1.JSEmitter()) {
        this.emitter = emitter;
    }
    /**
     * 编译一个 ComponentClass 的 proto。例如输入组件类：
     *
     * class Foo { computed = { author: () => 'harttle' } }
     *
     * 输出字符串：
     *
     * computed: {
     *     author: () => 'harttle'
     * },
     */
    compile(componentInfo) {
        const { emitter } = this;
        for (const key of Object.getOwnPropertyNames(componentInfo.proto)) {
            const member = componentInfo.proto[key];
            if (component_1.COMPONENT_RESERVED_MEMBERS.has(key) || !member)
                continue;
            emitter.nextLine(key + ': ');
            if (typeof member === 'function')
                this.emitMethod(member, emitter);
            else if (member instanceof Array)
                this.emitArray(member, emitter);
            else
                this.emitObjectLiteral(member, emitter);
        }
    }
    emitObjectLiteral(obj, emitter) {
        emitter.feedLine('{');
        const members = Object.getOwnPropertyNames(obj).filter(key => lodash_1.isFunction(obj[key]));
        for (const itemKey of members) {
            const item = obj[itemKey];
            emitter.writeIndentedLines(itemKey + ':' + lang_1.functionString(item) + ',');
        }
        emitter.writeLine('},');
    }
    emitArray(arr, emitter) {
        emitter.feedLine('[');
        emitter.writeIndentedLines(arr.map(item => lodash_1.isFunction(item)
            ? lang_1.functionString(item)
            : String(item)).join(',\n'));
        emitter.writeLine('],');
    }
    emitMethod(fn, emitter) {
        const funcString = lang_1.functionString(fn);
        emitter.feedLine(funcString + ',');
    }
}
exports.ComponentClassCompiler = ComponentClassCompiler;
