// boxjs/dist/thirdParty/bdbox/client/nativeShare.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /* eslint-disable */


  var version_compare = require('../utils/version_compare');

  var DOC = document;
  /**
   * 默认分享
   * @memberOf Bdbox.client
   * @function
   * @name nativeShare
   * @version $Id: nativeShare.js 304176 2016-08-09 05:47:18Z baozhixin $
   * @param  {object} opts 默认对象
   * @example
   * Bdbox.client.nativeShare({
   *     linkUrl:'',
   *     title: '',
   *     content: '',
   *     wechatIcon
   * });
   */

  function nativeShare(opts) {
    var defOpts = {
      wechatIcon: location.protocol + '//m.baidu.com/static/search/logo300.png',
      type: 'url',
      mediaType: 'all',
      linkUrl: location.href,
      title: DOC.title,
      success: 'console.log',
      error: 'console.log'
    };
    $.each(opts || {}, function (value, key) {
      defOpts[key] = value;
    });

    if (defOpts.image) {
      defOpts.imageUrl = defOpts.image;
    }

    if (defOpts.iconUrl && !defOpts.imageUrl && defOpts.type === 'url') {
      //如果类型是url，iconUrl存在，但是imageUrl不存在，那么就赋值
      defOpts.imageUrl = defOpts.iconUrl;
    }

    var BoxShareData = {};
    ['success', 'error'].forEach(function (v, i) {
      var name = v;

      if ($.isFunction(defOpts[v])) {
        name = '_xSHARE_SUCCESS_' + $.getId();
        window[name] = defOpts[v];
      }

      BoxShareData[v + 'callback'] = name;
      delete defOpts[v];
    });
    BoxShareData.options = defOpts;

    if (BoxShareData.options.imageUrl && $.$isAndroid() && version_compare($.version, '6.5') < 0) {
      //android有了imageUrl就是图片分享，但是去掉，微博就是截图分享，orz
      delete BoxShareData.options.imageUrl;
    }

    window.BoxShareData = BoxShareData; //添加微信分享默认图片

    var wechatIcon = defOpts.wechatIcon;

    if (/micromessenger\//i.test(navigator.userAgent) && wechatIcon && wechatIcon.length > 20) {
      var $div = DOC.createElement('div');
      $div.id = 'wa-generalevent-wx-logo';
      $div.style.display = 'none';
      $div.innerHTML = '<img src="' + wechatIcon + '"/>';
      var firstChild = DOC.body.firstChild;

      if (firstChild) {
        DOC.body.insertBefore($div, firstChild);
      } else {
        DOC.body.appendChild($div);
      }
    }
  }

  return nativeShare;
}();
