// boxjs/dist/thirdParty/bdbox/client/service_lightapp.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /* eslint-disable */

  /**
   * 调起轻应用
   * @memberOf Bdbox.client
   * @name lightapp
   */


  var getVersion = require('../utils/getVersion'),
      version_compare = require('../utils/version_compare'),
      A_NAMESPACE = 'Bdbox_android_utils',
      ios_invokeApp = require('../ios/invokeApp'),
      android_invokeApp = require('../android/invokeApp');

  var self = function (options) {
    if (version_compare(getVersion(), '5.3') < 0) {
      return;
    }

    if (!(options && typeof options === 'object' && options.url && options.appid)) {
      return;
    }

    var url = options.url,
        appid = options.appid;

    if ($.$isIOS()) {
      ios_invokeApp('easybrowse', {
        opentype: 1,
        isla: 1,
        appid: appid,
        openurl: encodeURIComponent(url),
        minver: '5.3.0.0'
      });
    } else {
      var intent = 'intent:#Intent;S.bdsb_light_start_url=' + encodeURIComponent(url) + ';S.bdsb_wallet_appid=' + appid + ';end',
          opt = {
        intent: intent,
        'class': 'com.baidu.searchbox.wallet.WalletServiceActivity',
        min_v: '16783629',
        mode: 0
      };
      android_invokeApp(A_NAMESPACE, 'command', [JSON.stringify(opt)]);
    }
  };

  return self;
}();
