// @baidu/wuji-san/es/wuji-progress/index.js
/**
 * @file 组件 WujiProgress 无极-进度条
 * @author raowenjuan <raowenjuan@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import pr2rem from '../utils/pr2rem';
import isNumeric from '../utils/is-numeric';

const prefixCls = 'wuji-progress';

export default class WujiProgress extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap {{color}}">
            <div class="progress" style="{{getProgressStyle(width)}}">
                <svg class="track" style="{{getTrackStyle(trackColor)}}">
                    <rect width="100%" height="100%"/>
                </svg>
                <svg class="line" style="{{format(percentage)}}">
                    <rect width="100%" height="100%"/>
                </svg>
            </div>
            <div s-if="text" class="text">{{text}}</div>
        </div>
    `;
    static computed = {
        color() {
            const percentage = this.data.get('percentage');
            let num = (percentage + '').replace('%', '');
            if (num <= 20) {
                return 'green';
            }
            else if (num <= 40) {
                return 'yellow';
            }
            else if (num <= 60) {
                return 'orange';
            }
            return 'red';
        },
    };

    getProgressStyle(width) {
        if (isNumeric(width)) {
            return {
                width: `${pr2rem(width)}rem`,
            };
        }
    }

    getTrackStyle(color) {
        if (color) {
            return {
                color,
                opacity: 1,
            };
        }
    }

    format(percentage) {
        let num = (percentage + '').replace('%', '');
        if (!isNaN(num) && num >= 0 && num <= 100) {
            return {
                width: `${num}%`,
            };
        }
    }

    static dataTypes = {
        foo: DataTypes.any,
    };

    initData() {
        return {
            // 进度条宽度
            width: 120,
            // 进度条背景色颜色
            trackColor: '#eeeeee',
            // 进度条颜色和字体颜色
            color: 'red',
            // 进度
            percentage: '50%',
            // 进度条右侧文字
            text: '',
        };
    }
}
