// @baidu/nano/es/switch/index.js
/**
 * @file 组件 switch 开关
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';

const prefixCls = 's-switch';

export default class Switch extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{wrapClass}}" style="{{wrapStyle}}" on-click="handleClick">
            <div class="${prefixCls}-node"/>
        </div>
    `;

    static dataTypes = {
        color: DataTypes.string,
        checked: DataTypes.bool,
        disabled: DataTypes.bool,
        disableAutoChecked: DataTypes.bool
    };

    static computed = {
        wrapClass() {
            const checked = this.data.get('checked');
            const disabled = this.data.get('disabled');
            return [
                prefixCls,
                checked && `${prefixCls}-checked`,
                disabled && `${prefixCls}-disabled`
            ].filter(c => c);
        },
        wrapStyle() {
            const checked = this.data.get('checked');
            const color = this.data.get('color');
            return checked && color ? {
                'background-color': color
            } : {};
        }
    };

    initData() {
        return {
            color: '',
            checked: true,
            disabled: false,
            disableAutoChecked: false
        };
    }

    handleClick() {
        const {checked, disabled, disableAutoChecked} = this.data.get();
        if (disabled) {
            return;
        }
        !disableAutoChecked && this.data.set('checked', !checked);
        this.fire('change', !checked);
    }
}
