// @baidu/ug-invoke-app/src/api/createWxTag.ts
/**
 * @file 创建微信开放标签
 * @author zhuzhibo(zhuzhibo@baidu.com)
 */
import {supportWxTag} from '../utils/detect/supportWxTag';
import {Status} from '../config/Status';
import {AppConfig, appConfigMap, AppConfigName} from '../config/app/appConfigMap';
import {copyText} from '../utils/copy/copyText';

interface Event {
    status: number;
    msg: string;
    e?: any;
}

interface WxInfo {
    appId: string;
    nonceStr: string;
    signature: string;
    timestamp: number;
    url: string;
}

export interface WxTagConfig {
    targetDom: HTMLElement;
    wxInfo: WxInfo;
    appName?: AppConfigName;
    scheme?: string;
    token?: string;
    toYYB?: boolean;
    debug?: boolean;
    onClick?: () => void;
    onSucceed?: (e: Event) => void;
    onFailed?: (e: Event) => void;
}

function _openYYB(pkgName: string, scheme: string): void {
    const YYBUrl = `https://a.app.qq.com/o/simple.jsp?pkgname=${pkgName}`;
    location.href = `${YYBUrl}&android_schema=${encodeURIComponent(scheme)}`;
}

function insertDom(targetDom: HTMLElement, scheme: string, appid: string): void {
    const wxTag = document.createElement('wx-open-launch-app');
    wxTag.setAttribute('id', 'ug-wx-btn');
    wxTag.setAttribute('appid', appid);
    wxTag.setAttribute('extinfo', scheme);
    wxTag.setAttribute('style', 'position: absolute; left: 0; right: 0; top: 0; bottom: 0;');
    const wxScript = document.createElement('script');
    wxScript.setAttribute('type', 'text/wxtag-template');
    const wxBtn = document.createElement('div');
    wxBtn.setAttribute('style', 'width: 100%; height: 100%; position: absolute; bottom: 0px;');
    wxScript.appendChild(wxBtn);
    wxTag.appendChild(wxScript);
    targetDom.appendChild(wxTag);
}

function execCreateWxTag(options: WxTagConfig, appConfig: AppConfig) {
    options.scheme = options.scheme || appConfig.scheme;
    // 创建标签插入DOM中
    insertDom(options.targetDom, options.scheme, appConfig.wxAppid as string);
    // 创建监听事件
    document.addEventListener('WeixinOpenTagsError', e => {
        if (options.onFailed) {
            options.onFailed({
                status: Status.NOT_SUPPORT_WX_TAG,
                msg: '不支持微信开放标签',
                e,
            });
        }
    });
    const btn: HTMLElement = document.querySelector('#ug-wx-btn') as HTMLElement;
    btn.addEventListener('click', async () => {
        // 复制口令
        if (options.token) {
            await copyText(options.token);
        }
        if (options.onClick) {
            options.onClick();
        }
    });
    btn.addEventListener('launch', () => {
        if (options.onSucceed) {
            options.onSucceed({
                status: Status.INVOKE_SUCCESS,
                msg: '调起成功',
            });
        }
    });
    btn.addEventListener('error', e => {
        if (options.onFailed) {
            if (options.toYYB) {
                _openYYB(appConfig.pkgName, options.scheme as string);
            }
            options.onFailed({
                status: Status.INVOKE_FAIL,
                msg: '调起失败',
                e,
            });
        }
    });
    // 注入数据
    /* global wx */
    // @ts-ignore
    wx.config({
        debug: options.debug ?? false,
        appId: options.wxInfo.appId,
        timestamp: options.wxInfo.timestamp,
        nonceStr: options.wxInfo.nonceStr,
        signature: options.wxInfo.signature,
        jsApiList: [
            'onMenuShareWeibo',
            'onMenuShareTimeline', // 分享到朋友圈
            'onMenuShareAppMessage', // 分享给好友
            'onMenuShareQZone', // 分享到QQ空间
            'onMenuShareQQ', // 分享到QQ好友
            'updateAppMessageShareData', // 新版分享给好友
            'updateTimelineShareData', // 新版分享朋友圈
        ],
        openTagList: ['wx-open-launch-app'],
    });
}

export function createWxTag(options: WxTagConfig): void {
    options.appName = options.appName || 'baiduboxapp';
    const appConfig: AppConfig = appConfigMap[options.appName];
    if (!supportWxTag(options, appConfig)) {
        if (options.onFailed) {
            options.onFailed({
                status: Status.NOT_SUPPORT_WX_TAG,
                msg: '不支持微信开放标签',
            });
        }
        return;
    }
    /* global wx */
    // @ts-ignore
    if (typeof wx !== 'object') {
        const scriptTag = document.createElement('script');
        scriptTag.setAttribute('src', '//res.wx.qq.com/open/js/jweixin-1.6.0.js');
        scriptTag.setAttribute('type', 'text/javascript');
        scriptTag.onload = () => {
            execCreateWxTag(options, appConfig);
        };
        document.getElementsByTagName('head')[0].appendChild(scriptTag);
    }
    else {
        console.log('[ug-invoke-app] 检测到微信JS-SDK已被引入，请确保SDK版本大于等于1.6.0');
        execCreateWxTag(options, appConfig);
    }
}
