// @baidu/hd-core/dist/index.amd.js
define(['exports', '@baidu/xbox'], (function (exports, xbox) { 'use strict';

    /**
     * Copyright (C) 2021 ~ present baidu.com
     */
    var env = {
        ua: xbox.UA,
        isIOS: xbox.isIOS(),
        isSSR: xbox.isSSR(),
    };

    /**
     * Copyright (C) 2021 ~ present baidu.com
     */
    /**
     * @description 获取数组中多个不重复的随机数
     * @param {Array} arr 数组
     * @param {Number} num 随机数个数
     * @returns {Array} 随机数组
     */
    function getRandoms(arr, num) {
        var _a;
        if (!Array.isArray(arr)) {
            throw new TypeError('参数不是数组类型');
        }
        var result = [];
        var length = arr.length;
        for (var i = 0; i < num; i++) {
            var randomIndex = Math.floor(Math.random() * (length - i - 1));
            var cur = arr[randomIndex];
            result.push(cur);
            /* eslint-disable no-param-reassign */
            _a = [
                arr[length - 1 - i],
                arr[randomIndex],
            ], arr[randomIndex] = _a[0], arr[length - 1 - i] = _a[1];
            /* eslint-enable no-param-reassign */
        }
        return result;
    }
    function encodeHtml(source) {
        source = String(source).replace(/</g, '&lt;').replace(/>/g, '&gt;');
        return source;
    }

    exports.encodeHtml = encodeHtml;
    exports.env = env;
    exports.getRandoms = getRandoms;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
