// @baidu/ug-invoke-app/src/utils/detect/os.ts
/**
 * @file 环境判断
 * @author zhuzhibo <zhuzhibo@baidu.com>
 */

const ua = navigator.userAgent;

function detectIpad() {
    let isMobile = RegExp(/iPad/i).test(navigator.userAgent);

    if (!isMobile) {
        const isMac = RegExp(/Macintosh/i).test(navigator.userAgent);

        if (isMac && navigator.maxTouchPoints && navigator.maxTouchPoints > 2) {
            isMobile = true;
        }
    }
    return isMobile;
}

export const IS_IPAD = detectIpad();

export const IS_IOS = /(iPhone|iPod|iPad|Macintosh)/i.test(ua) || IS_IPAD;
export const IS_ANDROID = /(Android)/i.test(ua);

function getIOSVersion() {
    let reg = /OS ((\d+_?){2,3})\s/i;
    if (IS_IOS) {
        let osv = reg.exec(navigator.userAgent);
        // @ts-ignore
        if (osv?.length > 0) {
            // @ts-ignore
            return osv[0]
                .replace('OS', '')
                .replace('os', '')
                .replace(/\s+/g, '')
                .replace(/_/g, '.');
        }
    }
    return '';
}

export const IOS_VERSION = Number(getIOSVersion().split('.')[0]);
