// @baidu/auc-qianfan-components/comp/header.js
/**
 * @file TestEs
 * @author wuji-dsl
 *
 * THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
 */

import {Component} from 'san';
import {WujiTopicHeader} from '@baidu/wuji-san';
// import {WujiTopicHeader} from '@baidu/auc-base/node_modules/@baidu/wuji-san';

class qfTopicHeader extends Component {
    static template = /* html */ `
        <div wnid="wn0" class="header {{wn1.noHeadDesc ? 'headerNodesc' : ''}}">
            <wuji-topic-header wnid="wn1" s-bind="wn1"></wuji-topic-header>
        </div>
    `;

    static components = {
        'wuji-topic-header': WujiTopicHeader
    };

    initData() {
        return {
            wn1: {
                title: '',
                imgType: 0,
                imgUrl: '',
                description: '',
                hotType: 0,
                authorNameList: [],
                authorImgList: [],
                authorNum: 0,
                modifyTime: 0,
                headImgLogo: '',
                headImgColor: ''
            }
        };
    }
}

export default qfTopicHeader;
