// @baidu/wuji-san/es/wuji-text-image/index.js
/**
 * @file 组件 WujiTextImage 无极-左文右图
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Grid from '../wuji-grid';
import Image from '../wuji-image';
import Title from '../wuji-title';
import Widget from '../wuji-widget';
import pr2rem from '../utils/pr2rem';

const prefixCls = 'w-text-image';

export default class WujiTextImage extends Component {
    static template = /* html */ `
        <div class="{{wrapClass}}" on-click="handleClick">
            <wuji-row gutter="{{gutter}}" wrap="{{false}}" margin="{{spacing ? margin : 0}}">
                <wuji-col
                    span="{{hasImage ? 8 : 12}}"
                    class="${prefixCls}-left ${prefixCls}-{{justify}}"
                    style="{{safeSpace(gutter, reverse)}}"
                >
                    <wuji-title
                        s-bind="titleProps"
                        text="{{title || titleProps.text}}"
                        label="{{label}}"
                        labelProps="{{labelProps}}"
                        spacing="{{false}}"
                        size="{{titleSize}}"
                    />
                    <wuji-widget
                        items="{{widgets}}"
                        dislike="{{dislike}}"
                        on-dislike="dislike"
                    />
                </wuji-col>
                <wuji-col s-if="hasImage" class="${prefixCls}-right" span="4">
                    <wuji-img
                        s-bind="imageProps"
                        src="{{image}}"
                        user-select="{{userSelect}}"
                    >
                        <slot name="imageMask" slot="mask"/>
                    </wuji-img>
                </wuji-col>
            </wuji-row>
            <div s-if="hairline" class="${prefixCls}-hairline"/>
        </div>
    `;

    static components = {
        'wuji-row': Grid.Row,
        'wuji-col': Grid.Col,
        'wuji-img': Image,
        'wuji-title': Title,
        'wuji-widget': Widget,
    };

    static dataTypes = {
        title: DataTypes.string,
        label: DataTypes.string,
        labelProps: DataTypes.object,
        image: DataTypes.string,
        imageProps: DataTypes.object,
        dislike: DataTypes.bool,
        widgets: DataTypes.arrayOf(DataTypes.object),
        margin: DataTypes.number,
        gutter: DataTypes.number,
        spacing: DataTypes.bool,
        reverse: DataTypes.bool,
        hairline: DataTypes.bool,
        justify: DataTypes.oneOf([
            'center',
            'flex-start',
            'flex-end',
            'space-between',
            'space-around',
        ]),
    };

    static computed = {
        wrapClass() {
            const reverse = this.data.get('reverse');
            return [
                prefixCls,
                reverse && `${prefixCls}-reverse`,
            ].filter(c => c);
        },
        hasImage() {
            const image = this.data.get('image');
            const imageProps = this.data.get('imageProps');
            return (imageProps && imageProps.src) || image;
        },
    };

    initData() {
        return {
            title: '',
            titleProps: {},
            labelProps: {},
            imageProps: {},
            margin: 51,
            gutter: 24,
            spacing: true,
            userSelect: true,
            justify: 'flex-start',
            hairline: true,
            reverse: false,
        };
    }

    // 安全距离M_W_X005
    safeSpace(gutter, reverse) {
        if (reverse) {
            return;
        }
        return `margin-right: ${pr2rem(24 - gutter)}rem`;
    }

    handleClick(event) {
        this.fire('click', {event, data: this.data.get()});
    }

    dislike(event) {
        this.fire('dislike', {event, data: this.data.get()});
    }
}
