// @baidu/nano/es/grid/col.js
/**
 * @file 组件 grid 网格布局
 * @author baozhixin <baozhixin@baidu.com>
 */

import {Component, DataTypes} from 'san';
import pr2rem from '../utils/pr2rem';

const prefixCls = 's-col';

export default class Col extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{colClass}}" style="{{colStyle}}">
            <slot/>
        </div>
    `;

    static dataTypes = {
        span: DataTypes.oneOfType([DataTypes.string, DataTypes.number]),
        offset: DataTypes.oneOfType([DataTypes.string, DataTypes.number])
    };

    static computed = {
        colClass() {
            const data = this.data;
            const span = Math.floor(data.get('span'));
            const offset = Math.floor(data.get('offset'));
            return [
                prefixCls,
                (span > 0 && span <= 12) && `${prefixCls}-${span}`,
                (offset > 0 && offset < 12) && `offset-${offset}`
            ].filter(c => c);
        },

        colStyle() {
            const gutter = this.data.get('gutter');
            const padding = `${pr2rem(gutter / 2)}rem`;
            if (gutter > 0) {
                return {
                    'padding-left': padding,
                    'padding-right': padding
                };
            }
        }
    };

    inited() {
        const gutter = this.parentComponent.data.get('gutter') || 0;
        this.data.set('gutter', gutter);
    }
}
