// @baidu/nano/es/grid/row.js
/**
 * @file 组件 grid 网格布局
 * @author baozhixin <baozhixin@baidu.com>
 */

import {Component, DataTypes} from 'san';
// import classNames from 'classnames';
import pr2rem from '../utils/pr2rem';

const prefixCls = 's-row';

export default class Row extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{rowClass}}" style="{{rowStyle}}">
            <slot/>
        </div>
    `;

    static dataTypes = {
        align: DataTypes.string,
        justify: DataTypes.string,
        margin: DataTypes.oneOfType([DataTypes.string, DataTypes.number]),
        gutter: DataTypes.oneOfType([DataTypes.string, DataTypes.number]),
        type: DataTypes.oneOf(['flex', 'float']),
        wrap: DataTypes.bool
    };

    static computed = {
        rowClass() {
            const data = this.data;
            const flex = 'flex' === data.get('type');
            const type = data.get('type');
            const align = data.get('align');
            const justify = data.get('justify');
            const wrap = data.get('wrap');

            return [
                prefixCls,
                `${prefixCls}-${type}`,
                flex && align && `${prefixCls}-align-${align}`,
                flex && justify && `${prefixCls}-justify-${justify}`,
                wrap && `${prefixCls}-wrap`
            ].filter(c => c);
        },

        rowStyle() {
            const data = this.data;
            const baseMargin = data.get('margin');
            const gutter = data.get('gutter');
            const margin = pr2rem(baseMargin - gutter / 2);
            const val = margin > 0 ? `${margin}rem` : 0;
            return {
                'margin-left': val,
                'margin-right': val
            };
        }
    };

    initData() {
        return {
            margin: 51,
            gutter: 24,
            type: 'flex',
            wrap: true
        };
    }
}
