// @baidu/xbox-native/view/closeCurrentWindow.js
/**
 * @file Created on Thu Nov 15 2018
 * @module view
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {close} from './index';

register(
    'currentWindow/close',
    {
        v: 11,
        hasCB: true,
        versionRange: ['9.1'],
        androidliteVersionRange: ['2.2.2'],
        iosinfoVersionRange: ['1.1'],
        filter() {
            return {
                module: 'browser',
                action: 'closeWindow'
            };
        }
    },
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        ios: {
            supportFrame: ['search'],
            versionRange: ['8.2'],
            filter() {
                return {
                    module: 'apppage',
                    action: 'closeSimpleBrowseView',
                    extData: {
                        minver: '8.2.0.0'
                    }
                };
            }
        },
        android: {
            hasCB: false,
            versionRange: ['8.2'],
            filter() {
                return {
                    module: 'Bdbox_android_utils',
                    action: 'closeWindow'
                };
            }
        }
    }
);

/**
 * 关闭当前窗口
 * http://bdbox.baidu.com/jssdk/api/viewclose.html#closecurrentwindow
 * @function closeCurrentWindow
 * @param {Object} params 端能力参数，参数参看 http://ft.baidu.com/#/scheme/utils_showtoast
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    return close(getParams('currentWindow', params));
};
