// @baidu/xbox-native/view/easybrowse.js
/**
 * @file Created on Thu Nov 15 2018
 * @module view
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
import register from '../register';
import {encode, getParams, isPlainObject} from '../utils';
import stringify from '../stringify';

import each from '@baidu/xbox/each';
import {open} from './index';

import {isMatrix} from '../detect';
import invoke from '../invoke';
import json2query from '@baidu/xbox/json2query';

function easybrowseParams(data, type) {
    if (type === 'hybrid') {
        if (data.backup && data.backup.url) {
            data.backup.url = encode(data.backup.url);
        }
    } else {
        data.url = encode(data.url);
    }
    data.style = stringify(data.style, true);
    for (let key in data) {
        if (isPlainObject(data[key])) {
            data[key] = stringify(data[key]);
        }
    }
    return data;
}
register(
    'easybrowse/open',
    {
        v: 1,
        hasCB: true,
        versionRange: ['8.2'],
        iosinfoVersionRange: ['1.1'],
        iosliteVersionRange: ['1.1'],
        iosmissionVersionRange: ['1.0'],
        androidmissionVersionRange: ['1.0'],
        androidliteVersionRange: ['2.2.2'],
        androidinfoVersionRange: ['1.0'],
        filter(data) {
            if (data.type === 'h5') {
                delete data.type;
            }
            return {
                data: easybrowseParams(data, 'h5')
            };
        }
    },
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        ios: {
            noParams: true,
            hasCB: false,
            versionRange: ['5.3'],
            filter(module, action, data) {
                let obj = {
                    openurl: encode(data.url),
                    minver: '5.3.0.0',
                    isla: 0,
                    opentype: 1,
                    append: 0,
                    rbtnstyle: 2
                };
                let alias = {};

                if (data.type === 'h5') {
                    delete data.type;
                }
                each(data, (value, key) => {
                    if (key !== 'url' && key !== 'style') {
                        key = alias[key] || key;
                        obj[key] = value;
                    }
                });
                let toolids;
                if (isPlainObject(data.style)) {
                    alias.color = 'barcolor';
                    if (data.style.toolbaricons && data.style.toolbaricons.tids) {
                        toolids = stringify({
                            toolids: data.style.toolbaricons.tids
                        });
                    }
                    data.toolbar = toolids;

                    each(data.style, (value, key) => {
                        key = alias[key] || key;
                        obj[key] = value;
                    });

                    delete data.style;
                    delete data.url;
                }
                if (obj.appid) {
                    obj.isla = 1;
                }

                return {
                    module: 'easybrowse',
                    action: '',
                    data: obj,
                    noParams: true,
                    extData: {
                        minver: '5.3.0.0'
                    }
                };
            }
        },
        android: {
            hasCB: false,
            versionRange: ['5.3'],
            filter(module, action, data) {
                let intent = ['S.bdsb_light_start_url=' + encode(data.url)];
                const map = {
                    comments: 'component=com.baidu.searchbox/.home.feed.FeedCommentActivity',
                    video: 'component=com.baidu.searchbox/.home.feed.ShortVideoDetailActivity',
                    tags: 'component=com.baidu.searchbox/.xsearch.UserSubscribeCenterActivity',
                    h5: 'component=com.baidu.searchbox/.lightbrowser.LightBrowserActivity',
                    lockscreen: 'component=com.baidu.searchbox/.lockscreen.LockScreenContentDetail'
                };
                intent.push(map[data.type]);

                let minV = data['min_v'];
                delete data['min_v'];
                let toolids = stringify(
                    {
                        toolids: data.style.toolbaricons.tids
                    },
                    true
                );
                intent.push('S.toolbaricons=' + toolids);

                if (data.menumode) {
                    intent.push('S.menumode=' + data.menumode);
                }
                if (data.context) {
                    intent.push('S.context=' + data.context);
                }

                intent = intent.join(';');
                let params = {
                    intent: 'intent:#Intent;' + intent + ';end',
                    min_v: minV && minV !== '' ? minV : '16783629',
                    mode: '0'
                };

                return {
                    module: 'Bdbox_android_utils',
                    action: 'command',
                    data: stringify(params)
                };
            }
        }
    }
);

register(
    'easybrowse/hybrid',
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        ios: {
            hasCB: false,
            versionRange: ['5.3'],
            filter(module, action, data) {
                let obj = {
                    openurl: encode(data.url),
                    minver: '5.3.0.0',
                    isla: 0,
                    opentype: 1,
                    append: 0,
                    rbtnstyle: 2,
                    type: 'Hybrid'
                };
                let alias = {};
                delete data.type;
                each(data, (value, key) => {
                    if (key !== 'url' && key !== 'style') {
                        key = alias[key] || key;
                        obj[key] = value;
                    }
                });
                let toolids;
                if (isPlainObject(data.style)) {
                    alias.color = 'barcolor';
                    if (data.style.toolbaricons && data.style.toolbaricons.tids) {
                        toolids = stringify({
                            toolids: data.style.toolbaricons.tids
                        });
                    }
                    data.toolbar = toolids;
                    delete data.style;
                    delete data.url;
                    each(data.style, (value, key) => {
                        key = alias[key] || key;
                        obj[key] = value;
                    });
                }
                if (obj.appid) {
                    obj.isla = 1;
                }
                return {
                    module: 'common',
                    action: 'easybrowse',
                    data: obj,
                    extData: {
                        minver: '5.3.0.0'
                    }
                };
            }
        },
        android: {
            hasCB: false,
            versionRange: ['5.3'],
            filter(module, action, data) {
                let backup = {
                    url: encode(data.backup.url)
                };
                let intent = [
                    'S.backup=' + JSON.stringify(backup),
                    'component=com.baidu.searchbox/.home.feed.FeedDetailActivity'
                ];
                let minV = data.min_v;
                delete data.min_v;
                let toolids = stringify({
                    toolids: data.style.toolbaricons.tids
                });
                intent.push('S.toolbaricons=' + toolids);

                if (data.menumode) {
                    intent.push('S.menumode=' + data.menumode);
                }

                intent = intent.join(';');
                let params = {
                    intent: 'intent:#Intent;' + intent + ';end',
                    min_v: minV && minV !== '' ? minV : '16783629',
                    mode: '0'
                };

                return {
                    module: 'Bdbox_android_utils',
                    action: 'command',
                    data: stringify(params)
                };
            }
        }
    },
    {
        v: 1,
        hasCB: true,
        versionRange: ['8.2'],
        iosinfoVersionRange: ['1.3.1.10'],
        androidliteVersionRange: ['2.2.0.10'],
        androidinfoVersionRange: ['1.1'],
        iosmissionVersionRange: ['1.0'],
        androidmissionVersionRange: ['1.0'],
        filter(data) {
            return {
                data: easybrowseParams(data, 'hybrid')
            };
        }
    }
);
/**
 * 打开h5或者hybrid框架
 * @function easybrowse
 * @param {Object} params 端能力参数，参数参看 http://ft.baidu.com/#/scheme/utils_showtoast
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    if (isMatrix('baiduboxvision')) {
        const martixParams = Object.assign({}, params, params.data);
        delete martixParams.data;
        params.name && delete martixParams.name;
        params.type && delete martixParams.type;
        return invoke(`baiduboxapp://v1/easybrowse/${params.type === 'hybrid' ? 'hybrid' : 'open'}?${json2query(martixParams, true)}`);
    }
    return open(getParams('easybrowse', params, {type: 'h5'}));
};
