// @baidu/xbox-native/view/swangame.js
/**
 * @file Created on Thu Nov 15 2018
 * @module view
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
import register from '../register';
import {noEmptyQueryToJson, getParams} from '../utils';
import {open} from './index';
import stringify from '../stringify';


register('swan/open', {
    v: 22,
    versionRange: ['10.13.5'],
    schemeNoV: true,
    filter(data) {
        /* istanbul ignore if */
        if (stringify(data) === '{}') {
            return {
                module: 'swangame',
                action: 'open',
                data: {}
            };
        }

        // data.url中可能包含域名：
        // 1、域名中可能有 smartapps.cn ，并且后面紧跟 webmapp，需要去掉 webmapp，再摘出path+query
        // 2、域名中可能有 smartapps.cn，但没有紧跟 webmapp，那么域名后面的就是需要的path+query
        // 3、域名中可能没有 smartapps.cn，就是普通三方域名，需要摘出path+query
        //
        // data.url中可能没有域名：
        // 1、将前面的杠去掉，其他就是需要的path+query

        let url = '';
        let host = data.url.match(/^http(s)?:\/\/(.*?)\//);

        if (host) {
            if (data.url.match('smartapps.cn/webmapp/')) {
                url = data.url.split('smartapps.cn/webmapp/')[1];
            } else {
                url = data.url.split(host[0])[1];
            }
        } else {
            url = data.url.substring(0, 1) === '/' ? data.url.substring(1) : data.url;
        }

        let path = url.split('?')[0];
        let params = {};
        if (url.indexOf('?') > -1) {
            params = noEmptyQueryToJson(data.url);
        } else {
            params = data.params || {};
        }
        let baiduboxapp = {
            from: params.from || data.from || '',
            ext: params.ext || data.ext || {}
        };
        let result = {
            _baiduboxapp: stringify(baiduboxapp, true)
        };

        let appId = data.appId || data.appKey;
        if (appId) {
            appId = appId.replace(/\w-/g, item => {
                return item.substring(0, 1).toUpperCase();
            });
        }

        if (path && path.length > 0) {
            for (let key in params) {
                params[key] = encodeURIComponent(params[key]);
            }
            Object.assign(result, params);
        }

        return {
            module: 'swangame/' + appId,
            action: path,
            data: result
        };
    }
});

/**
 * @function swangame
 * @param {Object} [params={}] - 参数
 */
export default (params = {}) => open(getParams('swangame', params));
