// @baidu/xbox-concern/concern/invoke.js
/**
* @file: 端能力调用
* @author: xulinfeng@baidu.com
* @date:   2019-11-27 15:38:11
*/
import getBoxVersion from '@baidu/xbox/get-box-version';
import versionCompare from '@baidu/xbox/version_compare';
import {isBox, isIOS, isAndroid} from '@baidu/xbox/browser/detect';
import {isEmptyObject} from '@baidu/xbox/is';
import toJSON from '@baidu/xbox/query2json';
import invoke from '@baidu/xbox/na/ios';
import {prompt} from '@baidu/xbox/na/android';
import boxx from './boxx';

let boxxConfig = {
    emit: 'event.emit', // 发送广播
    webStorage: 'easyBrowser.webStorage', // 存储
    open: 'easyBrowser.open', // 轻浏览打开
    showToast: 'commonUI.showToast', // 端toast
    loginInvoke: 'account.login' // 登录
};
let boxxInvoke = (search, params) => new Promise((resolve, reject) => { // boxx 调用处理
    let basicSearch = params => boxx.call(boxxConfig[`${search}`], params);
    if (basicSearch && !isEmptyObject(basicSearch)) {
        basicSearch({
            ...params,
            success: res => {
                resolve(res);
            },
            fail: res => {
                reject(res);
            }
        });
    } else {
        reject(`${search} not exit`);
    }
});

/**
 * 关注数据广播 发送
 * @function sendBroadcast
 */
export const sendBroadcast = data => {
    boxxInvoke('emit', {
        action: 'com.baidu.channel.foundation.followchanged',
        data: JSON.stringify({data: [data]})
    });
    boxxInvoke('emit', {
        action: 'com.baidu.channel.subscribe',
        data: JSON.stringify({
            sign: `${data.type}_${data.third_id}`,
            relation: data.is_follow
        })
    });
};

/**
 * 存储数据
 * @function setStorage
 */
export const setStorage = (key, data, callback = () => {}) => {
    boxxInvoke('webStorage', {
        key,
        data: JSON.stringify(data),
        action: 'set'
    }).then(res => {
        callback(res);
    }).catch(err => {
        callback(err);
    });
};

/**
 * 端toast
 * @function clientToast
 */
export const clientToast = message => {
    boxxInvoke('showToast', {
        type: '1',
        message
    });
};
/**
 * 登录
 * @function login
 */
export const login = (callback = () => {}) =>  {
    if ((isBox() && versionCompare(getBoxVersion(), '5.5') >= 0)) {
        boxxInvoke('loginInvoke', {
            loginType: 'sms',
            showThirdLogin: '1',
            loginSource: 'webpage_card',
            loginSubSource: 'webpage_card',
            tpl: 'webpage_card',
            normalizeAccount: '0'
        }).then(res => {
            if (+res.status === 0) {
                location.reload(1);
            }
            callback(res);
        }).catch(err => {
            callback(err);
        });
    } else {
        // 5.5以下版本跳转至baidu passport web页登录
        window.location.href = 'http://wappass.baidu.com/passport/login?u=' + encodeURIComponent(window.location.href);
    }
};

/*
 * 框内cmd打开，框外location跳转
 * 无url则跳转我的关注，7.6以下以及框外执行location跳转
 * @function toLandingPage
 * @param url  {String}  - 跳转url
 * @param isToast {Boolean}
 */
export function openPage(url) {
    if (url) {
        let params = isIOS() ? {newbrowser: '1'} : {};
        isBox() ? boxxInvoke('open', {
            url,
            ...params
        }) : (window.location.href = url);
        // isBox() ? o2o(url, params) : (window.location.href = url);
        return;
    }
    if (isBox() && versionCompare(getBoxVersion(), '7.6.0') >= 0) {
        if (isIOS()) {
            /* eslint-disable max-len */
            let urlSchame = 'baiduboxapp://apppage?action=openPage&params=%7b%22pageid%22%3a%22mysubscription%22%2c%22url%22%3a%22https%3a%2f%2fmbd.baidu.com%2fwebpage%3faction%3dicard%26type%3dsubscribe%22%2c%22titile%22%3a%22%22%7d';
            /* eslint-enable max-len */
            invoke(urlSchame); // todo 端能力js-native有了再迁移
        } else if (isAndroid()) {
            let cmd = {
                mode: '7',
                commands: [{
                    'mode': '6',
                    /* eslint-disable max-len */
                    'intent': 'intent:#Intent;S.rn_search_box_key=6;S.rn_bundle_id=box.rnplugin.myattention;S.rn_component_name=MyAttention;end',
                    /* eslint-enable max-len */
                    'class': 'com.baidu.searchbox.reactnative.RNSearchBoxMainActivity',
                    /* eslint-disable fecs-camelcase */
                    'min_v': '25167488'
                    /* eslint-enable fecs-camelcase */
                }, {
                    'mode': '0',
                    /* eslint-disable max-len */
                    'intent': 'intent:#Intent;S.user_sub_center_load_url=/webpage?action=icard&type=subscribe;B.launch_center=true;B.user_sub_center_search_enable=false;B.create_menu_key=false;end',
                    /* eslint-enable max-len */
                    'class': 'com.baidu.searchbox.xsearch.UserSubscribeCenterActivity',
                    /* eslint-disable fecs-camelcase */
                    'dyna_url_key': 'user_sub_center_load_url',
                    'http_style': true,
                    'min_v': '24381184'
                }],
                min_v: '24381184'
                /* eslint-enable fecs-camelcase */
            };
            // todo js-native 支持再迁移
            prompt('Bdbox_android_utils', 'command', [JSON.stringify(cmd)]);
        }
        return;
    }
    let suburl = 'https://mbd.baidu.com/webpage?action=icard&type=subscribe';
    let urlJSON = toJSON(window.location.href);
    if (urlJSON.channel && urlJSON.channel === 'wise_home') {
        suburl += '&channel=wise_home';
    }
    window.location.href = suburl;
}
