// @baidu/wise-invoke-app/src/index.js
/**
 * @file index
 * @author xiongyang(xiongyang@baidu.com)
 */

import {IvkApp} from './invoke/index';
import {getCommonParams, setCommonParams, getInvokeInfo, execInvoke, invokeApp} from '@baidu/mcp-sdk';
import {ubcLogger} from '@baidu/ubc-report-sdk';

/**
 * IvkAppCreator
 *
 * @param {string} pos 调起策略位
 * @param {Array} opt 调起参数（mcp返回）
 * @param {string} logId 打点日志id
 * @param {number} index 要判断的物料索引，默认设为第一位
 * @param {Object} config 设置的common公参数
 * @return {Object} mcp实例
 */
export class IvkAppCreator {
    create({pos, opt, logId, index = 0}) {
        return new IvkApp({pos, opt, logId, index});
    }
    getCommonParams() {
        return getCommonParams();
    }
    setCommonParams(config = {}, reset = false) {
        return setCommonParams(config, reset);
    }
    getInvokeInfo(config = {}) {
        return getInvokeInfo(config);
    }
    execInvoke({sucCb, failCb} = {}, params = {}) {
         // 执行调起逻辑
        return execInvoke(params.pos, params.invokeData, params.logId).then(res => {
            if (sucCb && sucCb instanceof Function) {
                sucCb(res);
            }
        }).catch(err => {
            if (failCb && failCb instanceof Function) {
                failCb(err);
            }
        });
    }
    ubcEvent(...args) {
        return ubcLogger(...args);
    }
    invokeApp(...args) {
        return invokeApp(...args);
    }
}
