// @baidu/xbox/index.js
/**
 * 一般不要这样用，单独引用哦
 * @file Created on Thu Nov 15 2018
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
import guid from './guid';
import escapeHtml from './escape-html';
import dateFormat from './date-format';
import typeError from './type-error';
import ready from './browser/ready';
import storage, {Storage} from './storage';
import getGlobal from './browser/get-global';
import strictURIEncode from './strict-uri-encode';
import decodeURIComponent from './decode-uri-component';
import broadcast, {EventEmitter} from './emitter';
import Monitor from './monitor/monitor';
import json2query from './json2query';
import query2json from './query2json';
import parseUrl from './parse-url';
import each from './each';
import versionCompare from './version_compare';
import substr, {str2Array} from './substr';
import visibilityChange from './visibilitychange';
import {clone} from './clone';
import * as delegate from './delegate';
import packageInfo from './package.json';

const version = packageInfo.version;

export {
    version
};

// na
export {default as o2o} from './na/o2o';
export {default as invoke, invokeP, isMaybeBox, xRegister} from './na/invoke';

export {
    each,
    typeError,
    parseUrl,
    query2json,
    json2query,
    Monitor,
    guid,
    escapeHtml,
    dateFormat,
    ready,
    storage,
    Storage,
    substr,
    str2Array,
    getGlobal,
    strictURIEncode,
    decodeURIComponent,
    broadcast,
    EventEmitter,
    versionCompare,
    versionCompare as version_compare,
    visibilityChange,
    clone,
    delegate
};

export {
    isJSON,
    isString,
    isArray,
    isBoolean,
    isBlank,
    isEmpty,
    isFunction,
    isNil,
    isNumber,
    isPlainObject,
    isUndefined,
    isWindow
} from './is';
export {
    default as getType,
    isType
} from './type';

export {toArray, toBoolean, toString, toNumber} from './to';

export {default as isMatrix} from './is-matrix';
export {default as getMatrixSchema} from './get-matrix-schema';
export {default as isBox} from './is-box';
export {isAndroid, isIOS, getOSName, getOSVersion} from './os';
export {default as getBoxName} from './get-box-name';
export {default as getBoxVersion} from './get-box-version';
export {default as extend} from './extend';
export {default as emitter} from './emitter';
export {default as decodeHtml} from './decode-html';
export {default as byteLen} from './byte-len';

export {default as stringify} from './stringify';

export {default as kebab2camel} from './kebab2camel';
