// @baidu/auc-timeline/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import {Component} from 'san';
import styles from './style.module.less';
import {AucTimelineData} from './types';
import AucPicTextSm from '@baidu/auc-pic-text-sm';
import AucKol from '@baidu/auc-kol';
import {componentLog} from '@baidu/auc-log';

export class AucTimeline extends Component {
    static template:string = /* html */ `
    <div class="{{styles.aucTimelineWrapper}} {{isFirst ? styles.isFirst : ''}} {{isBigEvent ? styles.isBigEvent : ''}}">
        <div class="{{styles.aucTimelineList}}">
            <div s-for="timeline in timelineList" class="{{styles.aucTimelineItem}}">
            <div class="{{styles.line}}"></div>
            <div class="{{styles.time}}">
                <span class="{{styles.circle}}"></span>{{timeline.time}}
            </div>
                <template s-if="timeline.type === 'guardian'">
                    <div class="{{styles.kolWrapper}}">
                        <auc-kol
                            s-bind="timeline"
                            on-click-kol="clickItem(timeline)"
                            on-show="showItem(timeline)"
                        ></auc-kol>
                    </div>
                </template>
                <template s-else>
                    <div class="{{styles.description}}"></div>
                        <div class="{{styles.picTextSmWrapper}}">
                        <auc-pic-text-sm
                            class="{{styles.picTextSm}}"
                            title="{{timeline.title}}"
                            imageProps="{{timeline.imageProps}}"
                            labelProps="{{timeline.labelProps}}"
                            on-click-item="clickItem(timeline)"
                            on-show="showItem(timeline)"
                        ></auc-pic-text-sm>
                    </div>                
                </template>
            </div>
        </div>
    </div>
    `;
    static components:object = {
        'auc-pic-text-sm': AucPicTextSm,
        'auc-kol': AucKol
    };

    initData(): AucTimelineData {
        return {
            styles,
            timelineList: [],
            time: '',
            isFirst: false,
            logBaseType: 'timeline_',
        };
    }

    attached(): void {
    }

    clickItem(timeline) {
        this.fire('jump', timeline.jumpProps);
        componentLog(this.data.get('logBaseType'), 'click', timeline.logExtraParams);
    }

    showItem(timeline) {
        componentLog(this.data.get('logBaseType'), 'show', timeline.logExtraParams);
    }
}

export default AucTimeline;
