// @baidu/xbox-native/es5/storage/get.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _index = require("./index");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module storage
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
function thunk(res) {
  var result = {};

  try {
    result = JSON.parse((0, _utils.decode)(res.data));
  } catch (e) {
    result = (0, _utils.decode)(res.data);
  }

  return {
    status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
    message: res.errmsg,
    data: {
      data: result
    }
  };
}

(0, _register.default)('storage/get', {
  v: 16,
  hasCB: true,
  versionRange: ['10.0'],
  filter: function filter() {
    return {
      module: 'dataStorage',
      action: 'getStorage'
    };
  },
  thunk: function thunk(res) {
    var result = {};

    try {
      result = JSON.parse((0, _utils.decode)(res.data.data));
    } catch (e) {
      result = (0, _utils.decode)(res.data.data);
    }

    if (res.data) {
      return {
        status: res.status,
        message: res.message,
        data: {
          data: result
        }
      };
    } else {
      return res;
    }
  }
}, {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      if (data.domain) {
        delete data.domain;
      }

      data.action = 'get';
      return {
        module: 'utils',
        action: 'webStorage',
        data: data
      };
    },
    thunk: thunk
  },
  android: {
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      if (data.domain) {
        delete data.domain;
      }

      data.action = 'get';
      return {
        module: 'Bdbox_android_utils',
        action: 'webStorage',
        data: data
      };
    },
    thunk: thunk
  }
});
/**
 * 详细文档：http://bdbox.baidu.com/jssdk/api/cache.html#setstorage
 * @function get
 */

var _default = _index.get;
exports.default = _default;
