// @baidu/xbox-native/es5/storage/remove.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _index = require("./index");

var _stringify = _interopRequireDefault(require("../stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module storage
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
function thunk(res) {
  return {
    status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
    message: res.errmsg
  };
}

(0, _register.default)('storage/delete', {
  v: 16,
  hasCB: true,
  versionRange: ['10.0'],
  filter: function filter(data) {
    var d = data;

    if (d.data) {
      delete d.data;
    }

    return {
      module: 'dataStorage',
      action: 'removeStorage',
      data: {
        params: (0, _stringify.default)(data, true)
      }
    };
  }
}, {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  ios: {
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      data.action = 'delete';
      return {
        module: 'utils',
        action: 'webStorage',
        data: data
      };
    },
    thunk: thunk
  },
  android: {
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      data.action = 'delete';
      return {
        module: 'Bdbox_android_utils',
        action: 'webStorage',
        data: data
      };
    },
    thunk: thunk
  }
});
/**
 * strore/delete
 * @function remove
 */

var _default = _index.del;
exports.default = _default;
