// @baidu/xbox-native/es5/storage/set.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module storage
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
function thunk(res) {
  return {
    status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
    message: res.errmsg
  };
}

(0, _register.default)('storage/set', {
  v: 16,
  hasCB: true,
  versionRange: ['10.0'],
  iosinfoVersionRange: ['1.3.1.10'],
  androidliteVersionRange: ['3.3'],
  iosmissionVersionRange: ['1.0'],
  androidmissionVersionRange: ['1.0'],
  filter: function filter(data) {
    return {
      module: 'dataStorage',
      action: 'setStorage'
    };
  }
}, {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      if (data.trustedDomains) {
        delete data.trustedDomains;
      }

      data.action = 'set';
      return {
        module: 'utils',
        action: 'webStorage',
        data: data
      };
    },
    thunk: thunk
  },
  android: {
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      if (data.trustedDomains) {
        delete data.trustedDomains;
      }

      data.action = 'set';
      return {
        module: 'Bdbox_android_utils',
        action: 'webStorage',
        data: data
      };
    },
    thunk: thunk
  }
});
/**
 * 详细文档：http://bdbox.baidu.com/jssdk/api/cache.html#setstorage
 * @function set
 */

var _default = _index.set;
exports.default = _default;
