// @baidu/wuji-san/es/index.js
/**
 * @file 全部组件列表
 * @author wuji
 *
 * THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
 */

export const version = '1.0.2';
export {default as WujiAvatar} from './wuji-avatar';
export {default as WujiButton} from './wuji-button';
export {default as WujiGrid} from './wuji-grid';
export {default as WujiHeader} from './wuji-header';
export {default as WujiIcon} from './wuji-icon';
export {default as WujiImage} from './wuji-image';
export {default as WujiImageContent} from './wuji-image-content';
export {default as WujiImageText} from './wuji-image-text';
export {default as WujiInput} from './wuji-input';
export {default as WujiLabel} from './wuji-label';
export {default as WujiLink} from './wuji-link';
export {default as WujiLiveStatus} from './wuji-live-status';
export {default as WujiMore} from './wuji-more';
export {default as WujiMultiTag} from './wuji-multi-tag';
export {default as WujiPicker} from './wuji-picker';
export {default as WujiPoi} from './wuji-poi';
export {default as WujiPopup} from './wuji-popup';
export {default as WujiPrice} from './wuji-price';
export {default as WujiProgress} from './wuji-progress';
export {default as WujiQuickEvaluate} from './wuji-quick-evaluate';
export {default as WujiRank} from './wuji-rank';
export {default as WujiRate} from './wuji-rate';
export {default as WujiScroll} from './wuji-scroll';
export {default as WujiSelect} from './wuji-select';
export {default as WujiSingleHead} from './wuji-single-head';
export {default as WujiSmallappContainer} from './wuji-smallapp-container';
export {default as WujiSource} from './wuji-source';
export {default as WujiStext} from './wuji-stext';
export {default as WujiSubsidiaryAvatar} from './wuji-subsidiary-avatar';
export {default as WujiTableCovid} from './wuji-table-covid';
export {default as WujiTextImage} from './wuji-text-image';
export {default as WujiThreeImages} from './wuji-three-images';
export {default as WujiTitle} from './wuji-title';
export {default as WujiTopicHeader} from './wuji-topic-header';
export {default as WujiWidget} from './wuji-widget';
export const componentNameMap = {
    Avatar: 'wuji-avatar',
    Button: 'wuji-button',
    Grid: 'wuji-grid',
    Header: 'wuji-header',
    Icon: 'wuji-icon',
    Image: 'wuji-image',
    ImageContent: 'wuji-image-content',
    ImageText: 'wuji-image-text',
    Input: 'wuji-input',
    Label: 'wuji-label',
    Link: 'wuji-link',
    LiveStatus: 'wuji-live-status',
    More: 'wuji-more',
    MultiTag: 'wuji-multi-tag',
    Picker: 'wuji-picker',
    Poi: 'wuji-poi',
    Popup: 'wuji-popup',
    Price: 'wuji-price',
    Progress: 'wuji-progress',
    QuickEvaluate: 'wuji-quick-evaluate',
    Rank: 'wuji-rank',
    Rate: 'wuji-rate',
    Scroll: 'wuji-scroll',
    Select: 'wuji-select',
    SingleHead: 'wuji-single-head',
    SmallappContainer: 'wuji-smallapp-container',
    Source: 'wuji-source',
    Stext: 'wuji-stext',
    SubsidiaryAvatar: 'wuji-subsidiary-avatar',
    TableCovid: 'wuji-table-covid',
    TextImage: 'wuji-text-image',
    ThreeImages: 'wuji-three-images',
    Title: 'wuji-title',
    TopicHeader: 'wuji-topic-header',
    Widget: 'wuji-widget',
};
