// @baidu/wuji-san/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "WujiAvatar", {
  enumerable: true,
  get: function get() {
    return _wujiAvatar["default"];
  }
});
Object.defineProperty(exports, "WujiButton", {
  enumerable: true,
  get: function get() {
    return _wujiButton["default"];
  }
});
Object.defineProperty(exports, "WujiGrid", {
  enumerable: true,
  get: function get() {
    return _wujiGrid["default"];
  }
});
Object.defineProperty(exports, "WujiHeader", {
  enumerable: true,
  get: function get() {
    return _wujiHeader["default"];
  }
});
Object.defineProperty(exports, "WujiIcon", {
  enumerable: true,
  get: function get() {
    return _wujiIcon["default"];
  }
});
Object.defineProperty(exports, "WujiImage", {
  enumerable: true,
  get: function get() {
    return _wujiImage["default"];
  }
});
Object.defineProperty(exports, "WujiImageContent", {
  enumerable: true,
  get: function get() {
    return _wujiImageContent["default"];
  }
});
Object.defineProperty(exports, "WujiImageText", {
  enumerable: true,
  get: function get() {
    return _wujiImageText["default"];
  }
});
Object.defineProperty(exports, "WujiInput", {
  enumerable: true,
  get: function get() {
    return _wujiInput["default"];
  }
});
Object.defineProperty(exports, "WujiLabel", {
  enumerable: true,
  get: function get() {
    return _wujiLabel["default"];
  }
});
Object.defineProperty(exports, "WujiLink", {
  enumerable: true,
  get: function get() {
    return _wujiLink["default"];
  }
});
Object.defineProperty(exports, "WujiLiveStatus", {
  enumerable: true,
  get: function get() {
    return _wujiLiveStatus["default"];
  }
});
Object.defineProperty(exports, "WujiMore", {
  enumerable: true,
  get: function get() {
    return _wujiMore["default"];
  }
});
Object.defineProperty(exports, "WujiMultiTag", {
  enumerable: true,
  get: function get() {
    return _wujiMultiTag["default"];
  }
});
Object.defineProperty(exports, "WujiPicker", {
  enumerable: true,
  get: function get() {
    return _wujiPicker["default"];
  }
});
Object.defineProperty(exports, "WujiPoi", {
  enumerable: true,
  get: function get() {
    return _wujiPoi["default"];
  }
});
Object.defineProperty(exports, "WujiPopup", {
  enumerable: true,
  get: function get() {
    return _wujiPopup["default"];
  }
});
Object.defineProperty(exports, "WujiPrice", {
  enumerable: true,
  get: function get() {
    return _wujiPrice["default"];
  }
});
Object.defineProperty(exports, "WujiProgress", {
  enumerable: true,
  get: function get() {
    return _wujiProgress["default"];
  }
});
Object.defineProperty(exports, "WujiQuickEvaluate", {
  enumerable: true,
  get: function get() {
    return _wujiQuickEvaluate["default"];
  }
});
Object.defineProperty(exports, "WujiRank", {
  enumerable: true,
  get: function get() {
    return _wujiRank["default"];
  }
});
Object.defineProperty(exports, "WujiRate", {
  enumerable: true,
  get: function get() {
    return _wujiRate["default"];
  }
});
Object.defineProperty(exports, "WujiScroll", {
  enumerable: true,
  get: function get() {
    return _wujiScroll["default"];
  }
});
Object.defineProperty(exports, "WujiSelect", {
  enumerable: true,
  get: function get() {
    return _wujiSelect["default"];
  }
});
Object.defineProperty(exports, "WujiSingleHead", {
  enumerable: true,
  get: function get() {
    return _wujiSingleHead["default"];
  }
});
Object.defineProperty(exports, "WujiSmallappContainer", {
  enumerable: true,
  get: function get() {
    return _wujiSmallappContainer["default"];
  }
});
Object.defineProperty(exports, "WujiSource", {
  enumerable: true,
  get: function get() {
    return _wujiSource["default"];
  }
});
Object.defineProperty(exports, "WujiStext", {
  enumerable: true,
  get: function get() {
    return _wujiStext["default"];
  }
});
Object.defineProperty(exports, "WujiSubsidiaryAvatar", {
  enumerable: true,
  get: function get() {
    return _wujiSubsidiaryAvatar["default"];
  }
});
Object.defineProperty(exports, "WujiTableCovid", {
  enumerable: true,
  get: function get() {
    return _wujiTableCovid["default"];
  }
});
Object.defineProperty(exports, "WujiTextImage", {
  enumerable: true,
  get: function get() {
    return _wujiTextImage["default"];
  }
});
Object.defineProperty(exports, "WujiThreeImages", {
  enumerable: true,
  get: function get() {
    return _wujiThreeImages["default"];
  }
});
Object.defineProperty(exports, "WujiTitle", {
  enumerable: true,
  get: function get() {
    return _wujiTitle["default"];
  }
});
Object.defineProperty(exports, "WujiTopicHeader", {
  enumerable: true,
  get: function get() {
    return _wujiTopicHeader["default"];
  }
});
Object.defineProperty(exports, "WujiWidget", {
  enumerable: true,
  get: function get() {
    return _wujiWidget["default"];
  }
});
exports.componentNameMap = exports.version = void 0;

var _wujiAvatar = _interopRequireDefault(require("./wuji-avatar"));

var _wujiButton = _interopRequireDefault(require("./wuji-button"));

var _wujiGrid = _interopRequireDefault(require("./wuji-grid"));

var _wujiHeader = _interopRequireDefault(require("./wuji-header"));

var _wujiIcon = _interopRequireDefault(require("./wuji-icon"));

var _wujiImage = _interopRequireDefault(require("./wuji-image"));

var _wujiImageContent = _interopRequireDefault(require("./wuji-image-content"));

var _wujiImageText = _interopRequireDefault(require("./wuji-image-text"));

var _wujiInput = _interopRequireDefault(require("./wuji-input"));

var _wujiLabel = _interopRequireDefault(require("./wuji-label"));

var _wujiLink = _interopRequireDefault(require("./wuji-link"));

var _wujiLiveStatus = _interopRequireDefault(require("./wuji-live-status"));

var _wujiMore = _interopRequireDefault(require("./wuji-more"));

var _wujiMultiTag = _interopRequireDefault(require("./wuji-multi-tag"));

var _wujiPicker = _interopRequireDefault(require("./wuji-picker"));

var _wujiPoi = _interopRequireDefault(require("./wuji-poi"));

var _wujiPopup = _interopRequireDefault(require("./wuji-popup"));

var _wujiPrice = _interopRequireDefault(require("./wuji-price"));

var _wujiProgress = _interopRequireDefault(require("./wuji-progress"));

var _wujiQuickEvaluate = _interopRequireDefault(require("./wuji-quick-evaluate"));

var _wujiRank = _interopRequireDefault(require("./wuji-rank"));

var _wujiRate = _interopRequireDefault(require("./wuji-rate"));

var _wujiScroll = _interopRequireDefault(require("./wuji-scroll"));

var _wujiSelect = _interopRequireDefault(require("./wuji-select"));

var _wujiSingleHead = _interopRequireDefault(require("./wuji-single-head"));

var _wujiSmallappContainer = _interopRequireDefault(require("./wuji-smallapp-container"));

var _wujiSource = _interopRequireDefault(require("./wuji-source"));

var _wujiStext = _interopRequireDefault(require("./wuji-stext"));

var _wujiSubsidiaryAvatar = _interopRequireDefault(require("./wuji-subsidiary-avatar"));

var _wujiTableCovid = _interopRequireDefault(require("./wuji-table-covid"));

var _wujiTextImage = _interopRequireDefault(require("./wuji-text-image"));

var _wujiThreeImages = _interopRequireDefault(require("./wuji-three-images"));

var _wujiTitle = _interopRequireDefault(require("./wuji-title"));

var _wujiTopicHeader = _interopRequireDefault(require("./wuji-topic-header"));

var _wujiWidget = _interopRequireDefault(require("./wuji-widget"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * @file 全部组件列表
 * @author wuji
 *
 * THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
 */
var version = '1.0.2';
exports.version = version;
var componentNameMap = {
  Avatar: 'wuji-avatar',
  Button: 'wuji-button',
  Grid: 'wuji-grid',
  Header: 'wuji-header',
  Icon: 'wuji-icon',
  Image: 'wuji-image',
  ImageContent: 'wuji-image-content',
  ImageText: 'wuji-image-text',
  Input: 'wuji-input',
  Label: 'wuji-label',
  Link: 'wuji-link',
  LiveStatus: 'wuji-live-status',
  More: 'wuji-more',
  MultiTag: 'wuji-multi-tag',
  Picker: 'wuji-picker',
  Poi: 'wuji-poi',
  Popup: 'wuji-popup',
  Price: 'wuji-price',
  Progress: 'wuji-progress',
  QuickEvaluate: 'wuji-quick-evaluate',
  Rank: 'wuji-rank',
  Rate: 'wuji-rate',
  Scroll: 'wuji-scroll',
  Select: 'wuji-select',
  SingleHead: 'wuji-single-head',
  SmallappContainer: 'wuji-smallapp-container',
  Source: 'wuji-source',
  Stext: 'wuji-stext',
  SubsidiaryAvatar: 'wuji-subsidiary-avatar',
  TableCovid: 'wuji-table-covid',
  TextImage: 'wuji-text-image',
  ThreeImages: 'wuji-three-images',
  Title: 'wuji-title',
  TopicHeader: 'wuji-topic-header',
  Widget: 'wuji-widget'
};
exports.componentNameMap = componentNameMap;
