// @baidu/wuji-san/es/wuji-input/index.js
/**
 * @file 组件 WujiInput 无极-输入框
 * placeholder文本过长省略号显示时，使用input的placeholder在focus时省略号丢失了，所以用span显示placeholder
 * @author yangmei03 <yangmei03@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';
import '@baidu/wuji-icons/svg/d20-inputbox-delete-f-48.svg';

const prefixCls = 'w-input';

export default class WujiInput extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap">
            <label class="${prefixCls}-label">
                <span s-if="labelText"
                    class="${prefixCls}-text">
                    {{labelText}}
                </span>
                <div class="${prefixCls}-content">
                    <span s-if="showPlaceholder" class="${prefixCls}-placeholder">
                        {{placeholder}}
                    </span>
                    <input
                        s-ref="wuji-input"
                        class="${prefixCls}-input"
                        value="{=inputValue=}"
                        maxlength="{{maxlength}}"
                        autocomplete="off"
                        on-blur="handleEvent($event, 'blur')"
                        on-focus="handleEvent($event, 'focus')"
                        on-input="handleEvent($event, 'input')"
                        on-click="triggerFocus"
                    />
                </div>
                <span s-if="showClear"
                    class="${prefixCls}-clear"
                    on-click="handleEvent($event, 'clear')"
                >
                    <wuji-icon type="d20-inputbox-delete-f-48"/>
                </span>
                <span s-if="unit"
                    class="${prefixCls}-unit">
                    {{unit}}
                </span>
                <span s-if="icon"
                    class="${prefixCls}-icon"
                    on-click="handleEvent($event, 'icon')"
                >
                    <wuji-icon type="{{icon}}"/>
                </span>
            </label>
            <div s-if="error" class="${prefixCls}-error">{{error}}</div>
        </div>
    `;

    static components = {
        'wuji-icon': Icon,
    };

    static dataTypes = {
        inputValue: DataTypes.string,
        unit: DataTypes.string,
        icon: DataTypes.string,
        error: DataTypes.string,
        labelText: DataTypes.string,
        showClear: DataTypes.bool,
        showPlaceholder: DataTypes.bool,
        maxlength: DataTypes.number,
        placeholder: DataTypes.string,
    };

    static computed = {
        showClear() {
            return this.data.get('inputValue.length') > 0;
        },
    };

    initData() {
        return {
            inputValue: '',
            icon: '',
            error: '',
            showClear: false,
            showPlaceholder: true,
            maxlength: 20,
            labelText: '',
            placeholder: '',
        };
    }

    triggerFocus() {
        const $input = this.ref('wuji-input');
        $input.focus();
    }

    // 事件类型:  input, focus, blur, clear, icon
    handleEvent(e, eventType) {
        e.preventDefault();
        e.stopPropagation();

        const value = this.data.get('inputValue').trim();

        if (eventType === 'clear') {
            this.data.set('inputValue', '');
            this.triggerFocus();
        }

        if (eventType === 'input') {
            this.data.set('showPlaceholder', false);
        }

        // 输入文字的长度
        const ingValue = e.data ? value + e.data : value;
        if (!ingValue.length) {
            this.data.set('showPlaceholder', true);
        }

        this.fire(eventType, {event: e, inputValue: value});
    }
}
