// @baidu/auc-titlebar/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import {Component} from 'san';
import styles from './style.module.less';
import {AucTitlebarData} from './types';
import {WujiIcon} from '@baidu/wuji-san';
import '@baidu/wuji-icons/svg/d20-titlebar-more-o-66.svg';

export class AucTitleBar extends Component {
    static template:string = /* html */ `
    <div class="{{styles.aucTitleBar}}" style="{{titleBarStyle}}" id="title-bar">
        <slot></slot>
        <div class="{{styles.aucTitleText}}">
            {{title}}
        </div>
        <div class="{{styles.icon}}" on-click="clickSet" role=text aria-label="更多">
            <wuji-icon type="d20-titlebar-more-o-66" size="66"></wuji-icon>
        </div>
    </div>
    `;
    static components:object = {
        'wuji-icon': WujiIcon
    };

    static computed:object = {
        titleBarStyle() {
            const top = this.data.get('top');
            const background = this.data.get('isBigEvent') ? `background-color:${this.data.get('bigEventColor')};color:#fff` : '';
            return `top: ${top};${background}`;
        }
    }

    initData(): AucTitlebarData {
        return {
            styles,
            title: '',
            top: false
        };
    }
    attached(): void {
    }

    clickSet() {
        this.fire('click-set', '');
    }
}

export default AucTitleBar;
