// @baidu/wuji-san/lib/wuji-popup/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _san = require("san");

var _wujiIcon = _interopRequireDefault(require("../wuji-icon"));

var _list = _interopRequireDefault(require("./list"));

require("@baidu/wuji-icons/svg/popupmenu-close-o-48.svg");

require("@baidu/wuji-icons/svg/arrow-back-o-48.svg");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 'w-popup';

var WujiPopup = /*#__PURE__*/function (_Component) {
  _inherits(WujiPopup, _Component);

  var _super = _createSuper(WujiPopup);

  function WujiPopup() {
    _classCallCheck(this, WujiPopup);

    return _super.apply(this, arguments);
  }

  _createClass(WujiPopup, [{
    key: "initData",
    value: function initData() {
      return {
        value: '',
        content: ''
      };
    }
  }, {
    key: "attached",
    value: function attached() {
      var _this = this;

      this.watch('isOpen', function (value) {
        if (value) {
          _this.nextTick(function () {
            var windowHeight = document.documentElement.offsetHeight;
            var boxHeight = windowHeight * 0.7;

            var title = _this.el.querySelector(".".concat(prefixCls, "-title"));

            var titleHeight = title ? title.offsetHeight : 0;

            _this.data.set('contentHeight', boxHeight - titleHeight);
          });
        }
      });
    }
  }, {
    key: "close",
    value: function close(event) {
      this.data.set('isOpen', false);
      this.fire('close', {
        event: event
      });
    }
  }, {
    key: "handleSelect",
    value: function handleSelect(_ref) {
      var event = _ref.event,
          value = _ref.value;
      this.data.set('value', value);
      this.fire('select', {
        event: event,
        value: value
      });
    }
  }, {
    key: "handleButtonClick",
    value: function handleButtonClick(_ref2) {
      var event = _ref2.event;
      this.close(event);
    }
  }]);

  return WujiPopup;
}(_san.Component);

exports["default"] = WujiPopup;

_defineProperty(WujiPopup, "template",
/* html */
"\n        <div class=\"".concat(prefixCls, "-wrap\">\n            <div class=\"").concat(prefixCls, "-mask\" on-click=\"close\" s-if=\"isOpen\"/>\n            <div class=\"").concat(prefixCls, "-box ").concat(prefixCls, "-{{type}}\" s-if=\"isOpen\">\n                <div s-if=\"showToolbar\" class=\"").concat(prefixCls, "-title\">\n                    <div class=\"").concat(prefixCls, "-left\" on-click=\"close\">\n                        <wuji-icon\n                            s-if=\"type === 'answer'\"\n                            type=\"arrow-back-o-48\"\n                        />\n                        <div s-if=\"type === 'calendar'\">\u53D6\u6D88</div>\n                    </div>\n                    <div class=\"").concat(prefixCls, "-title-text\">{{title}}</div>\n                    <div class=\"").concat(prefixCls, "-right\" on-click=\"close\">\n                        <div s-if=\"type === 'calendar'\">\u5B8C\u6210</div>\n                        <wuji-icon s-else type=\"popupmenu-close-o-48\"/>\n                    </div>\n                </div>\n                <div class=\"").concat(prefixCls, "-content\" style=\"max-height: {{contentHeight}}px\">\n                    <span s-if=\"type === 'text'\">{{content}}</span>\n                    <img\n                        s-else-if=\"type === 'image'\"\n                        src=\"{{imageUrl}}\"\n                        class=\"image-box\"\n                    />\n                    <wuji-list\n                        s-else\n                        options=\"{{options}}\"\n                        value=\"{{value}}\"\n                        type=\"{{type}}\"\n                        buttonText=\"{{buttonText}}\"\n                        on-select=\"handleSelect\"\n                        on-button-click=\"handleButtonClick\"\n                    />\n                </div>\n            </div>\n        </div>\n    "));

_defineProperty(WujiPopup, "components", {
  'wuji-icon': _wujiIcon["default"],
  'wuji-list': _list["default"]
});

_defineProperty(WujiPopup, "dataTypes", {
  title: _san.DataTypes.string,
  type: _san.DataTypes.string
});
