// @baidu/mo-actions/dist/index.amd.js
define(['exports'], (function (exports) { 'use strict';

    /**
     * action: 音频初始化
     */
    var AUDIO_INITED = 'audio:inited';
    /**
     * action: 播放过程中断网
     */
    var AUDIO_PLAYING_OFFLINE = 'audio:playing:offline';
    /**
     * action: 播放过程链接网络
     */
    var AUDIO_PLAYING_ONLINE = 'audio:playing:online';
    /**
     * action: 音频播放状态更新
     */
    var AUDIO_PLAY_STATE = 'audio:play:state';
    /**
     * action: 播放过程链接网络
     */
    var CLICK_AUDIO_LINK = 'click:audio:link';

    /**
     * action: 点击跳转更多链接
     */
    var CLICK_MORE_LINK = 'click:more:link';
    /**
     * action: 单条表态点击事件
     */
    var CLICK_OPINION_ITEM = 'click:opinion:item';

    /**
     * Copyright (C) 2021 ~ present baidu.com
     *
     * 将全部派发事件 action 的名字定义为常量放在这个文件中
     * 需要注释说明派发事件的 action 及 payload 含义
     *
     * 如果 payload 包含多个字段，尽量说明每个字段的含义
     * 建议使用 payload.event 传递事件对象
     *
     * 当 payload 主要是透传所有传递给组件的属性以及组件自身处理的属性时
     * 可以使用 const payload = this.data.get()
     * 注释中可以用 payload.* 表示组件当前 action 会透传所有属性（包括计算属性）
     */
    /**
     * action: 推荐语点击事件
     */
    var CLICK_PREORDER = 'click:preorder';
    /**
     * action: 图片封面点击事件\
     * payload.transData: 透传transData
     */
    var CLICK_VIDEOCOVER = 'click:videoCover';
    /**
     * action: 推荐语展现\
     * payload.*: 透传所有属性
     */
    var SHOW_PREORDER = 'show:preorder';
    /**
     * action: 段落点击事件\
     * payload.index: 推荐语在全文中的TTS阅读进度, 主要用于阅读段落切换\
     * payload.event: 推荐语内容-text, 转载自作者-name
     */
    var CLICK_PARAGRAPH = 'click:paragraph';
    /**
     * action: 段落展现\
     * payload.*: 透传所有属性
     */
    var SHOW_PARAGRAPH = 'show:paragraph';
    /**
     * action: 预览图片\
     * payload: 点击事件
     */
    var PREVIEW_IMAGE = 'preview:image';
    /**
     * action: 点击 AI 图片
     */
    var CLICK_IMAGE_AI = 'click:image:ai';
    /**
     * action: 图片 onload 事件
     */
    var LOAD_IMAGE = 'load:image';
    /**
     * action: 转发的第三方资源的点击事件\
     * payload.xxx: 组件运行时所有数据this.data.get()
     */
    var CLICK_THIRDORIGINAL = 'click:thirdOriginal';
    /**
     * action: 点击 AI 图片
     */
    var CLICK_CAPSULE_YOUJIA = 'click:capsule:youjia';
    /**
     * action: 正文大图小程序点击事件\
     * payload.*: 透传所有属性，主要包含：\
     * payload.appType: 0-小程序 1-小游戏\
     * payload.appKey: 小程序或小游戏key\
     * payload.url: 小程序或小游戏首页路径\
     * payload.midurl: 小程序或小游戏链接\
     * payload.nid: 文章id
     */
    var CLICK_MINI = 'click:mini';
    /**
     * action: 正文大图小程序展现事件\
     * payload.*: 透传所有属性，主要包含：\
     * payload.nid: 文章id
     */
    var SHOW_MINI = 'show:mini';
    /**
     * action: 标题点击事件\
     * payload.index: TTS阅读进度, 标题默认 -1
     */
    var CLICK_TITLE = 'click:title';
    /**
     * action: 视频点击事件
     */
    var CLICK_VIDEO = 'click:video';
    /**
     * action: 视频播放事件
     */
    var PLAY_VIDEO = 'play:video';
    /**
     * action: 视频播放状态变化事件\
     * payload.idx: 索引位置\
     * payload.state: 播放状态
     */
    var CHANGESTATE_VIDEO = 'changeState:video';
    /**
     * action: 宫格图点击图片\
     * payload.index: 点击的图片索引\
     * payload.itemInfo: 点击的图片信息
     */
    var CLICK_IMAGEGRID_ITEM = 'click:imageGrid:item';
    /**
     * action: 宫格图图片加载失败\
     * payload.index: 失败的图片索引\
     * payload.itemInfo: 失败的图片信息
     */
    var RENDER_IMAGEGRID_ERROR = 'render:imageGrid:error';
    /**
     * action: 宫格图点击更多\
     * payload.index: 点击的图片索引
     */
    var CLICK_IMAGEGRID_MORE = 'click:imageGrid:more';
    /**
     * action: 点击定位\
     * payload.extraData: 透传的对象\
     * payload.styleType: 样式类型
     */
    var CLICK_LOCATION_ITEM = 'click:location:item';
    /**
     * action: 可滑动的图文组件，点击列表中某一个\
     * payload.index: 点击的组件列表的索引\
     * payload.itemData: 点击的组件列表的信息
     */
    var CLICK_SLIDABLEIMGTEXT_ITEM = 'click:slidableImgtext:item';
    /**
     * action: 可滑动的图文组件，图片加载失败\
     * payload.index: 点击的组件列表的索引\
     * payload.itemData: 点击的组件列表的信息
     */
    var RENDER_SLIDABLEIMGTEXT_IMGERROR = 'render:slidableImgtext:imgerror';
    /**
     * action: 可滑动的图文组件，点击图片\
     * payload.index: 点击的组件列表的索引\
     * payload.itemData: 点击的组件列表的信息
     */
    var CLICK_SLIDABLEIMGTEXT_IMG = 'click:slidableImgtext:img';
    /**
     * action: 点击 PK
     */
    var CLICK_PK = 'click:pk';
    /**
     * action: 富文本标题点击事件\
     * payload: 组件运行时所有数据this.data.get()
     */
    var CLICK_RICHTEXT = 'click:richText';
    /**
     * action: 富文本标题点击话题事件\
     * payload: 该话题的全部数据
     */
    var CLICK_RICHTEXT_TOPIC = 'click:richText:topic';
    /**
     * action: 富文本标题点击用户事件\
     * payload: 该用户的全部数据
     */
    var CLICK_RICHTEXT_AT = 'click:richText:at';
    /**
     * action: 富文本标题点击普通链接事件\
     * payload: 该普通链的全部数据
     */
    var CLICK_RICHTEXT_LINK = 'click:richText:link';
    /**
     * action: 富文本标题点击视频链接事件\
     * payload: 该视频链接的全部数据
     */
    var CLICK_RICHTEXT_VIDEO = 'click:richText:video';
    /**
     * action: 富文本标题点击投票链接事件\
     * payload: 该投票链的全部数据
     */
    var CLICK_RICHTEXT_VOTE = 'click:richText:Vote';
    /**
     * action: 富文本标题点击图片链接事件\
     * payload: 该图片链接的全部数据
     */
    var CLICK_RICHTEXT_IMAGE = 'click:richText:Image';
    /**
     * action: 全民投票点击事件\
     * payload: 组件运行时所有数据this.data.get()
     */
    var CLICK_VOTE = 'click:vote';
    /**
     * action: 全民投票展现事件\
     * payload: 组件运行时所有数据this.data.get()
     */
    var SHOW_VOTE = 'show:vote';
    /**
     * action: 点击 footer left内容
     */
    var CLICK_FOOTER_LEFT = 'click:footer:left';
    /**
     * action: 点击 footer ext内容
     */
    var CLICK_FOOTER_EXT = 'click:footer:ext';
    /**
     * action: 点击 footer right内容
     */
    var CLICK_FOOTER_RIGHT = 'click:footer:right';
    /**
     * action: 点击关注按钮
     */
    var CLICK_FOLLOW_BUTTON = 'click:follow:button';
    /**
     * action: 关注锁点击事件
     */
    var CLICK_FOLLOW_LOCK = 'click:follow:lock';
    /**
     * action: 关注锁展现事件
     */
    var SHOW_FOLLOW_LOCK = 'show:follow:lock';
    /**
     * action: 广告卡点击事件
     */
    var CLICK_ADS_CARD = 'click:ads:card';
    /**
     * action: 广告卡展现事件\
     * payload: 组件运行时所有数据this.data.get()
     */
    var SHOW_ADS_CARD = 'show:ads:card';
    /**
     * action: 点击作者信息区域的 Badge（徽标）
     */
    var CLICK_AUTHOR_BADGE = 'click:author:badge';
    /**
     * action: 点击作者信息区域
     */
    var CLICK_AUTHOR_INFO = 'click:author:info';
    /**
     * action: 导航栏使用粘性布局时随页面滚动吸顶状态的变化事件
     */
    var CHANGE_NAVBAR_STICKY = 'change:navbar:sticky';
    /**
     * action: 正文小说卡卡片点击事件
     */
    var CLICK_NOVEL_CARD = 'click:novel:card';
    /**
     * action: 正文小说卡按钮点击事件
     */
    var CLICK_NOVEL_BUTTON = 'click:novel:button';
    /**
     * action: 正文小说卡展现事件\
     * payload: 正文小说卡组件全部数据
     */
    var SHOW_NOVEL = 'show:novel';
    /**
     * action: TTS按钮点击事件
     */
    var CLICK_TTS_BUTTON = 'click:tts:button';
    /**
     * action: 导航栏菜单按钮点击事件
     */
    var CLICK_NAVBAR_MENU = 'click:navbar:menu';
    /**
     * action: 点赞点击事件
     */
    var CLICK_PRAISE = 'click:praise';
    /**
     * action: 单条弹幕的点击事件
     */
    var CLICK_DANMAKU_H_ITEM = 'click:danmakuHItem';

    exports.AUDIO_INITED = AUDIO_INITED;
    exports.AUDIO_PLAYING_OFFLINE = AUDIO_PLAYING_OFFLINE;
    exports.AUDIO_PLAYING_ONLINE = AUDIO_PLAYING_ONLINE;
    exports.AUDIO_PLAY_STATE = AUDIO_PLAY_STATE;
    exports.CHANGESTATE_VIDEO = CHANGESTATE_VIDEO;
    exports.CHANGE_NAVBAR_STICKY = CHANGE_NAVBAR_STICKY;
    exports.CLICK_ADS_CARD = CLICK_ADS_CARD;
    exports.CLICK_AUDIO_LINK = CLICK_AUDIO_LINK;
    exports.CLICK_AUTHOR_BADGE = CLICK_AUTHOR_BADGE;
    exports.CLICK_AUTHOR_INFO = CLICK_AUTHOR_INFO;
    exports.CLICK_CAPSULE_YOUJIA = CLICK_CAPSULE_YOUJIA;
    exports.CLICK_DANMAKU_H_ITEM = CLICK_DANMAKU_H_ITEM;
    exports.CLICK_FOLLOW_BUTTON = CLICK_FOLLOW_BUTTON;
    exports.CLICK_FOLLOW_LOCK = CLICK_FOLLOW_LOCK;
    exports.CLICK_FOOTER_EXT = CLICK_FOOTER_EXT;
    exports.CLICK_FOOTER_LEFT = CLICK_FOOTER_LEFT;
    exports.CLICK_FOOTER_RIGHT = CLICK_FOOTER_RIGHT;
    exports.CLICK_IMAGEGRID_ITEM = CLICK_IMAGEGRID_ITEM;
    exports.CLICK_IMAGEGRID_MORE = CLICK_IMAGEGRID_MORE;
    exports.CLICK_IMAGE_AI = CLICK_IMAGE_AI;
    exports.CLICK_LOCATION_ITEM = CLICK_LOCATION_ITEM;
    exports.CLICK_MINI = CLICK_MINI;
    exports.CLICK_MORE_LINK = CLICK_MORE_LINK;
    exports.CLICK_NAVBAR_MENU = CLICK_NAVBAR_MENU;
    exports.CLICK_NOVEL_BUTTON = CLICK_NOVEL_BUTTON;
    exports.CLICK_NOVEL_CARD = CLICK_NOVEL_CARD;
    exports.CLICK_OPINION_ITEM = CLICK_OPINION_ITEM;
    exports.CLICK_PARAGRAPH = CLICK_PARAGRAPH;
    exports.CLICK_PK = CLICK_PK;
    exports.CLICK_PRAISE = CLICK_PRAISE;
    exports.CLICK_PREORDER = CLICK_PREORDER;
    exports.CLICK_RICHTEXT = CLICK_RICHTEXT;
    exports.CLICK_RICHTEXT_AT = CLICK_RICHTEXT_AT;
    exports.CLICK_RICHTEXT_IMAGE = CLICK_RICHTEXT_IMAGE;
    exports.CLICK_RICHTEXT_LINK = CLICK_RICHTEXT_LINK;
    exports.CLICK_RICHTEXT_TOPIC = CLICK_RICHTEXT_TOPIC;
    exports.CLICK_RICHTEXT_VIDEO = CLICK_RICHTEXT_VIDEO;
    exports.CLICK_RICHTEXT_VOTE = CLICK_RICHTEXT_VOTE;
    exports.CLICK_SLIDABLEIMGTEXT_IMG = CLICK_SLIDABLEIMGTEXT_IMG;
    exports.CLICK_SLIDABLEIMGTEXT_ITEM = CLICK_SLIDABLEIMGTEXT_ITEM;
    exports.CLICK_THIRDORIGINAL = CLICK_THIRDORIGINAL;
    exports.CLICK_TITLE = CLICK_TITLE;
    exports.CLICK_TTS_BUTTON = CLICK_TTS_BUTTON;
    exports.CLICK_VIDEO = CLICK_VIDEO;
    exports.CLICK_VIDEOCOVER = CLICK_VIDEOCOVER;
    exports.CLICK_VOTE = CLICK_VOTE;
    exports.LOAD_IMAGE = LOAD_IMAGE;
    exports.PLAY_VIDEO = PLAY_VIDEO;
    exports.PREVIEW_IMAGE = PREVIEW_IMAGE;
    exports.RENDER_IMAGEGRID_ERROR = RENDER_IMAGEGRID_ERROR;
    exports.RENDER_SLIDABLEIMGTEXT_IMGERROR = RENDER_SLIDABLEIMGTEXT_IMGERROR;
    exports.SHOW_ADS_CARD = SHOW_ADS_CARD;
    exports.SHOW_FOLLOW_LOCK = SHOW_FOLLOW_LOCK;
    exports.SHOW_MINI = SHOW_MINI;
    exports.SHOW_NOVEL = SHOW_NOVEL;
    exports.SHOW_PARAGRAPH = SHOW_PARAGRAPH;
    exports.SHOW_PREORDER = SHOW_PREORDER;
    exports.SHOW_VOTE = SHOW_VOTE;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
