// @baidu/auc-star-content/index.ts
import {Component} from 'san';
import styles from './style.module.less';
import {aucStarContentTypes} from './types';
import aucStarTitle from '@baidu/auc-star-title';
import aucStarImageArea from '@baidu/auc-star-imagearea';

export class aucStarContent extends Component {
    static template:string = `
    <div class="{{styles.AucStarContent}}"
        on-touchstart="setActive($event)"
        on-touchmove="deleteActive($event)"
        on-touchend="deleteActive($event)">
        <auc-star-title
            title="{{title}}"
            titleLink="{{titleLink}}"
            class="{{styles.aucStarContentTitle}}"
            on-jump-detail="jumpDetail"
            on-back-shoubai="backShoubai"
        ></auc-star-title>
        <auc-star-image-area
            s-if="images.length !== 0"
            images="{{images}}"
            class="{{styles.aucStarContentImage}}"
            type="{{type}}"
            nid="{{nid}}"
            on-out-jump="outJump"
            on-jump-detail="jumpDetail"
        ></auc-star-image-area>
    </div>
    `;
    static components:object = {
        'auc-star-title': aucStarTitle,
        'auc-star-image-area': aucStarImageArea
    };

    initData(): aucStarContentTypes {
        return {
            styles,
            title: '',
            liketype: '', //数据来源
            type: '', // 模板类型
            images: [], // 图片集合
            time: '',
            titleLink: {}, // 头部跳转信息 包括话题，@等
            source: '', // 来源信息
            ts: '', // 时间
            nid: '', // nid
            rid: '', // rid
            description: '', // 描述信息
            cmd: {
                mode: 2,
                url: '',
                intent: ''
            }
        };
    }
    static computed = {

    };
    inited(): void {

    }
    attached(): void {

    }
    jumpDetail(): void {
        this.fire('jump-detail');
    }
    backShoubai(): void {
        this.fire('back-shoubai');
    }
    outJump():void {
        this.fire('out-jump');
    }
    setActive(event) {
        if (/titleLink/.test(event.target.className)){
            return;
        }
        this.fire('set-active');
    }
    deleteActive(event) {
        if (/titleLink/.test(event.target.className)){
            return;
        }
        this.fire('delete-active');
    }
}

export default aucStarContent;
