// @baidu/xbox-native/net/imageCache.js
/**
 * @file Created on Thu Nov 15 2018
 * @module net
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {download} from './index';
function thunk(res) {
    // errno为1是成功，0是失败
    return {
        status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
        message: res.errmsg,
        data: res.data
    };
}
register('image/cache', {
    androidlite: {
        versionRange: ['2.2.2']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidinfo: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        supportFrame: ['h5', 'hybrid', 'video', 'tags'],
        filter() {
            return {
                module: 'net',
                action: 'imageCache',
                extData: {
                    minver: '7.3.0.0'
                }
            };
        },
        thunk
    },
    android: {
        supportFrame: ['hybrid', 'tags'],
        filter() {
            return {
                module: 'Bdbox_android_net',
                action: 'imageCache'
            };
        },
        thunk
    }
});
/**
 * @function imageCache
 * @param {object} [data={}] - 参数
 */
export default (data = {}) => download(getParams('image-cache', data));
