// @baidu/xbox-native/net/request.js
/**
 * @file net requrest
 * @module net
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {decode} from '../utils';
import {request} from './index';

register('net/request', {
    androidlite: {
        versionRange: ['2.2.2'],
        supportFrame: ['tags']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidinfo: {
        versionRange: ['1.0'],
        supportFrame: ['tags']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        versionRange: ['7.3'],
        supportFrame: ['hybrid', 'h5', 'tags', 'video', 'search'],
        filter() {
            return {
                module: 'net',
                action: 'request'
            };
        },
        thunk(res) {
            let data;
            if (res && res.data && res.errno === '1') {
                data = JSON.parse(decode(res.data));
            } else {
                data = res.data;
            }
            return {
                status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
                message: res.message,
                data: data
            };
        }
    },
    android: {
        supportFrame: ['hybrid', 'tags'],
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'Bdbox_android_net',
                action: 'request'
            };
        },
        thunk(res) {
            return {
                status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
                message: res.errmsg,
                data: JSON.parse(res.data)
            };
        }
    }
});
/**
 * net/requst
 * @function request
 */
export default request;
