// @baidu/wuji-san/es/wuji-subsidiary-avatar/index.js
/**
 * @file 带头像的附属信息区块
 * @author xiewenbo01@baidu.com
 */
import './style/index.less';
import {Component, DataTypes} from 'san';
import WujiAvatar from '../wuji-avatar';

const prefixCls = 'w-topic-author-info';

export default class WujiHeaderAuthorInfo extends Component {
    static template = /* html */ `
        <div class="${prefixCls}">
            <wuji-avatar
                s-for="item in authorAvatars" src="{{item}}"
                class="${prefixCls}-avatar"
                border-color="{{borderColor}}"
                border-width="{{borderWidth}}"
                size="{{photoSize}}"
            >
            </wuji-avatar>
            <div class="{{descCls}}">{{authorDesc}}</div>
        </div>
    `;

    static components = {
        'wuji-avatar': WujiAvatar,
    };

    static computed = {
        // 默认显示三个头像
        authorAvatars() {
            const maxNum = this.data.get('maxNum');
            const avatarList = this.data.get('authorImgList') || [];

            return avatarList.slice(0, maxNum);
        },
        descCls() {
            const descColor = this.data.get('descColor') || 'FC2';
            return `${prefixCls}-name-${descColor.toLocaleLowerCase()}`;
        },
    };

    static trimWhitespace = 'all';

    static dataTypes = {
        // 头像尺寸
        photoSize: DataTypes.oneOf([
            48,
            66,
            72,
            84,
            105,
            126,
            156,
            170,
            192,
            219,
            252,
        ]),
        // 描述文字的颜色
        descColor: DataTypes.oneOf([
            'FC2',
            'FC6',
        ]),
        // 展示几个头像的叠加
        avatarsNum: DataTypes.number,
        // 头像数组
        authorImgList: DataTypes.array,
        // 描边宽度
        borderWidth: DataTypes.oneOf([
            1,
            3,
        ]),
        // 描边颜色
        borderColor: DataTypes.oneOf([
            'FC18',
            'FC107',
        ]),
        // 描述信息
        authorDesc: DataTypes.string,
    };

    initData() {
        return {
            borderColor: 'FC18',
            borderWidth: 3,
            descColor: 'FC2',
            photoSize: 48,
            maxNum: 3,
            authorImgList: [],
            authorDesc: '',
        };
    }
}
