// @baidu/wuji-san/lib/wuji-image-text/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _san = require("san");

var _wujiGrid = _interopRequireDefault(require("../wuji-grid"));

var _wujiImage = _interopRequireDefault(require("../wuji-image"));

var _wujiTitle = _interopRequireDefault(require("../wuji-title"));

var _wujiWidget = _interopRequireDefault(require("../wuji-widget"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 'w-image-text';

var WujiImageText = /*#__PURE__*/function (_Component) {
  _inherits(WujiImageText, _Component);

  var _super = _createSuper(WujiImageText);

  function WujiImageText() {
    _classCallCheck(this, WujiImageText);

    return _super.apply(this, arguments);
  }

  _createClass(WujiImageText, [{
    key: "initData",
    value: function initData() {
      return {
        title: '',
        price: '',
        discount: '123元',
        dislike: false,
        imageProps: {},
        margin: 51,
        spacing: true,
        userSelect: true,
        justify: 'flex-start',
        hairline: false,
        imgColSpan: 3
      };
    }
  }, {
    key: "handleClick",
    value: function handleClick(event) {
      this.fire('click', {
        event: event,
        data: this.data.get()
      });
    }
  }, {
    key: "dislike",
    value: function dislike(event) {
      this.fire('dislike', {
        event: event,
        data: this.data.get()
      });
    }
  }]);

  return WujiImageText;
}(_san.Component);

exports["default"] = WujiImageText;

_defineProperty(WujiImageText, "template",
/* html */
"\n        <div class=\"{{wrapClass}}\" on-click=\"handleClick\">\n            <slot name=\"header\"/>\n            <wuji-row gutter=\"{{gutter}}\" wrap=\"{{false}}\" margin=\"{{spacing ? margin : 0}}\">\n                <wuji-col s-if=\"hasImage\" class=\"".concat(prefixCls, "-left\" span=\"{{imgColSpan}}\">\n                    <wuji-img s-bind=\"imageProps\" src=\"{{image}}\" user-select=\"{{userSelect}}\" s-ref=\"image\">\n                        <slot s-if=\"hasMaskSlot\" name=\"imageMask\" slot=\"{{hasMaskSlot ? 'mask': ''}}\"></slot>\n                    </wuji-img>\n                </wuji-col>\n                <wuji-col\n                    span=\"{{hasImage ? 12 - imgColSpan : 12}}\"\n                    class=\"").concat(prefixCls, "-right ").concat(prefixCls, "-{{justify}}\"\n                >\n                    <slot name=\"title\">\n                        <wuji-title\n                            s-if=\"title || titleProps.text\"\n                            s-bind=\"{{titleProps}}\"\n                            text=\"{{title || titleProps.text}}\"\n                            label=\"{{label}}\"\n                            labelProps=\"{{labelProps}}\"\n                            spacing=\"{{false}}\"\n                            size=\"{{titleSize}}\"\n                        />\n                    </slot>\n                    <slot name=\"detail\">\n                        <div s-if=\"discount || price\" class=\"").concat(prefixCls, "-price-wrap\">\n                            <span class=\"").concat(prefixCls, "-price\">{{discount || price}}</span>\n                            <span s-if=\"discount\" class=\"").concat(prefixCls, "-price-origin\">{{price}}</span>\n                        </div>\n                    </slot>\n                </wuji-col>\n            </wuji-row>\n            <slot name=\"footer\"/>\n            <wuji-widget\n                s-if=\"widgets && widgets.length > 0\"\n                items=\"{{widgets}}\"\n                spacing\n                dislike=\"{{dislike}}\"\n                has-margin-top=\"{{false}}\"\n                on-dislike=\"dislike\"\n            />\n            <div s-if=\"hairline\" class=\"").concat(prefixCls, "-hairline\"/>\n        </div>\n    "));

_defineProperty(WujiImageText, "components", {
  'wuji-row': _wujiGrid["default"].Row,
  'wuji-col': _wujiGrid["default"].Col,
  'wuji-img': _wujiImage["default"],
  'wuji-title': _wujiTitle["default"],
  'wuji-widget': _wujiWidget["default"]
});

_defineProperty(WujiImageText, "dataTypes", {
  title: _san.DataTypes.string,
  label: _san.DataTypes.string,
  price: _san.DataTypes.string,
  discount: _san.DataTypes.string,
  image: _san.DataTypes.string,
  imageProps: _san.DataTypes.object,
  dislike: _san.DataTypes.bool,
  gutter: _san.DataTypes.number,
  margin: _san.DataTypes.number,
  spacing: _san.DataTypes.bool,
  hairline: _san.DataTypes.bool,
  justify: _san.DataTypes.oneOf(['center', 'flex-start', 'flex-end', 'space-between', 'space-around'])
});

_defineProperty(WujiImageText, "computed", {
  wrapClass: function wrapClass() {
    return [prefixCls].filter(function (c) {
      return c;
    });
  },
  hasImage: function hasImage() {
    var image = this.data.get('image');
    var imageProps = this.data.get('imageProps');
    return imageProps && imageProps.src || image;
  }
});
