// @baidu/boxx/dist/imageSearchDescription.js
/**
 * boxx.js 
 * v3.0.69
 */

(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
    typeof define === 'function' && define.amd ? define(factory) :
    (global = global || self, global.boxx = factory());
}(this, (function () { 'use strict';

    /**
     * @file  多模视觉预置描述表
     * @author zhuxingyu01
     * @date 2021/06/25
     */
    var android = [
        // http://es.baidu-int.com/#/apidetail/1576
        {
            name: 'imageSearch.setPerformanceData',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/setPerformanceData',
            authority: 'imageSearch',
            args: [{
                name: 'data',
                value: 'Object='
            }],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1579
        {
            name: 'imageSearch.showFullImage',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/showFullImage',
            authority: 'imageSearch',
            args: [{
                name: 'url',
                value: 'string'
            }],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1582
        {
            name: 'imageSearch.executeEvent',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/executeEvent',
            authority: 'imageSearch',
            args: [
                {
                    name: 'eventName',
                    value: 'string'
                },
                {
                    name: 'content',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1583
        {
            name: 'imageSearch.studyFriendIntegral',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/studyFriendIntegral',
            authority: 'imageSearch',
            args: [
                {
                    name: 'text',
                    value: 'string='
                },
                {
                    name: 'id',
                    value: 'string='
                },
                {
                    name: 'url',
                    value: 'string='
                },
                {
                    name: 'imgurl_bg',
                    value: 'string='
                },
                {
                    name: 'imgurl_btn',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1630
        {
            name: 'imageSearch.verifyFace',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/verifyFace',
            authority: 'imageSearch',
            args: [{
                name: 'successCallBack',
                value: 'string'
            }],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1721
        {
            name: 'imageSearch.widgetControl',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/widgetControl',
            authority: 'imageSearch',
            args: [
                {
                    name: 'from',
                    value: 'string'
                },
                {
                    name: 'type',
                    value: 'number'
                }
            ],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        }
    ];

    var ios = [
        // http://es.baidu-int.com/#/apidetail/1576
        {
            name: 'imageSearch.setPerformanceData',
            scheme: 'baiduboxapp',
            authority: 'imagesearch',
            path: '/setPerformanceData',
            invoke: 'iframe.bdboxURL',
            args: [{
                name: 'data',
                value: 'Object='
            }],
            env: {
                main: {
                    imagesearch: '12.20'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1579
        {
            name: 'imageSearch.showFullImage',
            scheme: 'baiduboxapp',
            authority: 'imagesearch',
            path: '/showFullImage',
            invoke: 'iframe.bdboxURL',
            args: [{
                name: 'url',
                value: 'string'
            }],
            env: {
                main: {
                    imagesearch: '12.20'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1582
        {
            name: 'imageSearch.executeEvent',
            scheme: 'baiduboxapp',
            authority: 'imagesearch',
            path: '/executeEvent',
            invoke: 'iframe.bdboxURL',
            args: [
                {
                    name: 'eventName',
                    value: 'string'
                },
                {
                    name: 'content',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    imagesearch: '12.20'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1583
        {
            name: 'imageSearch.studyFriendIntegral',
            scheme: 'baiduboxapp',
            authority: 'imagesearch',
            path: '/studyFriendIntegral',
            invoke: 'iframe.bdboxURL',
            args: [
                {
                    name: 'id',
                    value: 'string='
                },
                {
                    name: 'url',
                    value: 'string='
                },
                {
                    name: 'text',
                    value: 'string='
                },
                {
                    name: 'imgurl_bg',
                    value: 'string='
                },
                {
                    name: 'imgurl_btn',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    imagesearch: '12.20'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1584
        {
            name: 'imageSearch.imageTextSearch',
            scheme: 'baiduboxapp',
            authority: 'imagesearch',
            path: '/imageTextSearch',
            invoke: 'iframe.bdboxURL',
            args: [
                {
                    name: 'guessWord',
                    value: 'string='
                },
                {
                    name: 'relatedRequestUrl',
                    value: 'string='
                },
                {
                    name: 'searchResultUrl',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    imagesearch: '12.20'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1630
        {
            name: 'imageSearch.verifyFace',
            scheme: 'baiduboxapp',
            invoke: 'iframe.bdboxURL',
            path: '/verifyFace',
            authority: 'imageSearch',
            args: [{
                // 此处是传一个字符串函数名，fe提前注册事件，调用此处将事件名传给端上，端上在验证成功后调用
                name: 'successCallBack',
                value: 'string'
            }],
            env: {
                main: {
                    imagesearch: '12.23'
                }
            }
        }
    ];

    var imageSearchDescription = /*#__PURE__*/Object.freeze({
        __proto__: null,
        android: android,
        ios: ios
    });

    /**
     * @file  预置描述表
     * @author chenqi14
     * @date 2019/09/26
     */

    var android$1 = [
        // 手百通用的数据跨域获取
        // doc: http://cp01-invoker-3-mgt161657.epc.baidu.com:8007/invokers?id=56f3941329c7a9e44394f90b
        {
            name: 'easyBrowser.webStorage',
            invoke: 'method.args-json+callback',
            method: 'Bdbox_android_utils.webStorage',
            args: [
                {name: 'action', value: 'string'},
                {name: 'key', value: 'string'},
                {name: 'data', value: 'string='}
            ],
            env: {
                main: {
                    search: '7.3',
                    light: '7.3'
                },
                lite: {
                    search: '2.6.0.0',
                    light: '2.6.0.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            name: 'easyBrowser.openHybrid',
            // 打开 feed hybrid
            args: [
                {name: 'type', value: 'string='},
                {name: 'url', value: 'string='},
                {
                    name: 'backup', value: {
                        type: {
                            url: 'string',
                            version: 'string'
                        }
                    }
                },
                {name: 'tpl_id', value: 'string='},
                {name: 'append', value: 'string='},
                {name: 'tplpath', value: 'string='},
                {name: 'context', value: 'Object'},
                {name: 'style', value: 'Object='},
                {name: 'newbrowser', value: 'string='},
                {name: 'forbidautorotate', value: 'string='},
                {name: 'slog', value: 'Object='},
                {name: 'videoinfo', value: 'string='},
                {name: 'jssids', value: 'string='},
                {name: 'launchlogin', value: 'string='},
                {name: 'stay', value: 'string='},
                {name: 'next', value: 'string='},
                {name: 'commentInfo', value: 'Object='},
                {name: 'pagefrom', value: 'string='},
                {name: 'isSearchVideo', value: 'string='},
                {name: 'ch_url', value: 'string='},
                {name: 'first_image', value: 'string='}
            ],
            path: '/easybrowse/hybrid',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'v1',
            invoke: 'method.url',
            scheme: 'baiduboxapp',
            env: {
                main: {light: '8.2'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // data 和 header 字段，为了友好，对开发者要求的是对象。但低版本客户端要求的是字符串，boxx 对齐进行了兼容处理
            name: 'easyBrowser.request',
            invoke: 'method.args-json+callback',
            method: 'Bdbox_android_net.request',
            args: [
                {name: 'header', value: 'string='},
                {name: 'pu', value: 'string'},
                {name: 'method', value: 'string'},
                {name: 'data', value: 'string='},
                {name: 'url', value: 'string'},
                {name: 'cache', value: 'string'},
                {name: 'refresh', value: 'string'}
            ],
            env: {
                main: {light: '7.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // 网络 imageCache
        // baiduboxapp://v28/net/imageCache?params=$params&callback=$callback
        {
            // data 和 header 字段，为了友好，对开发者要求的是对象。但低版本客户端要求的是字符串，boxx 对齐进行了兼容处理
            name: 'easyBrowser.imageCache',
            invoke: 'method.args-json+callback',
            method: 'Bdbox_android_net.request',
            authority: 'v28',
            path: '/net/imageCache',
            args: [
                {name: 'list', value: 'Array'}
            ],
            env: {
                main: {light: '10.8'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 轻框架 通用性能打点接口
            // baiduboxapp://v20/utils/onPerformanceFlowEvent?params=$params&callback=$callback
            name: 'easyBrowser.performanceFlowLog',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'v20',
            path: '/utils/onPerformanceFlowEvent',
            args: [
                {name: 'actionId', value: 'string'},
                {name: 'operation', value: 'string'},
                {name: 'value', value: 'string'},
                {name: 'frameSource', value: 'string='},
                {name: 'businessType', value: 'string='}
            ],
            env: {
                main: {light: '10.4'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 轻浏览框架showmenu
            // baiduboxapp://v9/easybrowse/showmenu?callback=_bdbox_js_526&upgrade=0
            name: 'easyBrowser.showMenu',
            scheme: 'baiduboxapp',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'v9',
            path: '/easybrowse/showmenu',
            args: [
                {name: 'menuItem', value: 'Object='}
            ],
            query: {
                upgrade: 0
            },
            env: {
                main: {light: '9.0'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // todo 安卓的 query 存在的意义是？给谁鉴权用？
            // 打开轻浏览框架
            // 'baiduboxapp://v1/easybrowse/open?url=http://cp01-renzhonghua.epc.baidu.com:8888/html/fjm/newfeed/demo.html?
            //  frame=h5&style={'toolbaricons':{'tids':['4','1','2','3']}}&newbrowser=1&callback=_bdbox_js_403&upgrade=0
            //  &oauthType=feed'
            name: 'easyBrowser.open',
            scheme: 'baiduboxapp',
            invoke: 'method.url',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'v1',
            path: '/easybrowse/open',
            query: {
                upgrade: 0,
                oauthType: 'feed'
            },
            args: [
                {name: 'url', value: 'string'},
                {name: 'ch_url', value: 'string='},
                {name: 'type', value: 'string='},
                {name: 'append', value: 'string='},
                {name: 'context', value: 'Object='},
                {name: 'style', value: 'Object='},
                {name: 'newbrowser', value: 'string='},
                {name: 'forbidautorotate', value: 'string='},
                {name: 'fullscreen', value: 'string='},
                {name: 'slog', value: 'Object='},
                {name: 'videoinfo', value: 'string='},
                {name: 'jssids', value: 'string='},
                {name: 'launchlogin', value: 'string='},
                {name: 'stay', value: 'string='},
                {name: 'next', value: 'string='},
                {name: 'commentInfo', value: 'Object='},
                {name: 'pagefrom', value: 'string='},
                {name: 'isSearchVideo', value: 'string='}
            ],
            env: {
                main: {
                    search: '7.3',
                    light: '7.3',
                    imagesearch: '13.0'
                },
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 进入落地页或者个人中心无网，展示大WiFi页面
            // baiduboxapp://net?action=showNetWorkErrorView&params={'type':'profile'}
            // supportFrame hybrid,hasCB: false
            name: 'easyBrowser.showNetworkError',
            invoke: 'method',
            method: 'Bdbox_android_net.showNetWorkErrorView',
            env: {
                main: {light: '8.2'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // 落地页调用NA Dialog弹窗，支持文字icon，button可配置
        // baiduboxapp://easybrowse/showDialog?params=$params&callback=$callback
        {
            name: 'easyBrowser.showDialog',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/showDialog',
            authority: 'easybrowse',
            args: [{
                name: 'title',
                value: 'string'
            },
            {
                name: 'subtitle',
                value: 'string'
            },
            {
                name: 'sourceUrl',
                value: {
                    type: {
                        default: 'string',
                        night: 'string'
                    }
                }
            },
            {
                name: 'buttons',
                value: {
                    arrayOf: 'Object='
                }
            }
            ],
            nextCallback: 'callback',
            env: {
                main: {
                    light: '11.17'
                }
            }
        },
        // 保留图片
        // http://es.baidu-int.com/#/apidetail/1448
        {
            name: 'easyBrowser.image.save',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/saveImage',
            authority: 'utils',
            args: [
                {
                    name: 'url',
                    value: 'string'
                }
            ],
            env: {
                main: {
                    search: '8.2.0.0',
                    light: '8.2.0.0'
                }
            }
        },
        // 时政文章屏蔽部分活动弹窗
        // http://es.baidu-int.com/#/apidetail/1497
        {
            name: 'easyBrowser.sensitiveNews',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/sensitiveNews',
            authority: 'easybrowse',
            args: [
                {
                    name: 'msg',
                    value: 'string='
                }
            ],
            scheme: 'baiduboxapp',
            env: {
                main: {
                    light: '12.17.0.0'
                }
            }
        }
    ];


    var ios$1 = [
        // ios 特有
        // 'url': 'baiduboxapp://v34/easybrowse/thirdPartyPageInfo?params=$params&callback=$callback',
        // 'doc': 'http://ft.baidu.com/#/scheme/easybrowse_thirdpartypageinfo',
        {
            name: 'easyBrowser.getThirdPartyPageInfo',
            invoke: 'iframe.bdboxURL',
            authority: 'v34',
            path: '/easybrowse/thirdPartyPageInfo',
            args: [
                {name: 'type', value: 'string'},
                {name: 'info', value: 'Object'}
            ],
            env: {
                main: {light: '10.13'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 轻浏览框架showmenu
            // baiduboxapp://v9/easybrowse/showmenu?callback=_bdbox_js_526&upgrade=0
            name: 'easyBrowser.showMenu',
            scheme: 'baiduboxapp',
            invoke: 'iframe.bdboxURL',
            authority: 'v9',
            path: '/easybrowse/showmenu',
            args: [
                {name: 'menuItem', value: 'Object='}
            ],
            env: {
                main: {light: '9.0'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // 网络 imageCache
        // baiduboxapp://v28/net/imageCache?params=$params&callback=$callback
        {
            // data 和 header 字段，为了友好，对开发者要求的是对象。但低版本客户端要求的是字符串，boxx 对齐进行了兼容处理
            name: 'easyBrowser.imageCache',
            invoke: 'iframe.bdboxURL',
            path: '/imageCache',
            authority: 'net',
            args: [
                {
                    name: 'urls',
                    value: {
                        arrayOf: 'string'
                    }
                }
            ],
            env: {
                main: {light: '10.8'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 打开落地页
            // 'baiduboxapp://v1/easybrowse/open?url=http://cp01-renzhonghua.epc.baidu.com:8888/html/fjm/newfeed/demo.html?
            //  frame=h5&style={'toolbaricons':{'tids':['4','1','2','3']}}&newbrowser=1&callback=_bdbox_js_403&upgrade=0
            //  &oauthType=feed'
            name: 'easyBrowser.open',
            invoke: 'iframe.url',
            scheme: 'baiduboxapp',
            authority: 'v1',
            path: '/easybrowse/open',
            // todo 确认是否需要 query
            query: {
                upgrade: 0,
                oauthType: 'feed'
            },
            args: [
                {name: 'url', value: 'string'},
                {name: 'type', value: 'string='},
                {name: 'append', value: 'string='},
                {name: 'context', value: 'Object='},
                {name: 'style', value: 'Object='},
                {name: 'newbrowser', value: 'string='},
                {name: 'forbidautorotate', value: 'string='},
                {name: 'fullscreen', value: 'string='},
                {name: 'slog', value: 'Object='},
                {name: 'videoinfo', value: 'string='},
                {name: 'jssids', value: 'string='},
                {name: 'launchlogin', value: 'string='},
                {name: 'stay', value: 'string='},
                {name: 'next', value: 'string='},
                {name: 'commentInfo', value: 'Object='},
                {name: 'pagefrom', value: 'string='},
                {name: 'isSearchVideo', value: 'string='}
            ],
            env: {
                main: {
                    search: '7.3',
                    light: '7.3',
                    imagesearch: '12.20'
                },
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 进入落地页或者个人中心无网，展示大WiFi页面
            // baiduboxapp://net?action=showNetWorkErrorView&func=_bdbox_js_101
            // baiduboxapp://net?action=showNetWorkErrorView&params={'type':'profile'}
            name: 'easyBrowser.showNetworkError',
            invoke: 'iframe.bdboxURL',
            authority: 'net',
            query: {
                action: 'showNetWorkErrorView'
            },
            args: [
                {name: 'type', value: 'string='}
            ],
            env: {
                main: {light: '8.2.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            },
            // todo 端上没这个字段。实际上传了也不触发。
            callbackName: 'func'
        },
        {
            // http://bdbox.baidu.com/jssdk/api/cache.html#setstorage
            // scheme: 'baiduboxapp://utils?action=webStorage&params={'key':'name','data':'\'val\'','action':'set'}&func=_bdbox_js_357',
            name: 'easyBrowser.webStorage',
            invoke: 'iframe.bdboxURL',
            path: '/webStorage',
            authority: 'utils',
            args: [
                {name: 'key', value: 'string'},
                {name: 'data', value: 'string='},
                {name: 'action', value: 'string'}
            ],
            query: {
                action: 'webStorage',
                minver: '7.3'
            },
            callbackName: 'func',
            env: {
                main: {
                    search: '7.3',
                    light: '7.3'
                },
                lite: {
                    search: '2.6.0.0',
                    light: '2.6.0.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 网络请求
            // scheme: 'baiduboxapp://net?action=request&params={'url':'searchbox://?action=feed&nomust=searchbox&cmd=104','method':'post','data':{'nid':'news_17596451975550153468'},'cache':0,'refresh':1,'pu':1,'header':{}}&func=_bdbox_js_179',
            //     todo 确认以下参数是否需要
            // {name: 'cache', value: 'number='},
            // {name: 'expires', value: 'number='},
            // {name: 'refresh', value: 'number='},
            // {name: 'pu', value: 'number='},
            // {name: 'dataType', value: 'string'}
            name: 'easyBrowser.request',
            invoke: 'iframe.bdboxURL',
            authority: 'net',
            args: [
                {name: 'header', value: 'Object='},
                {name: 'pu', value: 'string'},
                {name: 'method', value: 'string'},
                {name: 'data', value: 'Object='},
                {name: 'url', value: 'string'}
            ],
            query: {
                action: 'request',
                minver: '7.3'
            },
            callbackName: 'func',
            env: {
                main: {light: '7.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 轻框架 通用运营任务
            // baiduboxapp://v46/easybrowse/universalOperation?params=%7B%22type%22%3A%221%22%2C%22data%22%3A%7B%22type%22%3A%221%22%2C%22actionId%22%3A%221234%22%2C%22tip%22%3A%22aaa%22%2C%22subTip%22%3A%22bbb%22%2C%22cmd%22%3A%22jjjj%22%2C%22imageURL%22%3A%22kkk%22%2C%22isGif%22%3A%220%22%2C%22canClose%22%3A%221%22%7D%7D&callback=_bdbox_js_214&upgrade=0&oauthType=feed
            name: 'easyBrowser.universalOperation',
            invoke: 'iframe.bdboxURL',
            authority: 'v46',
            path: '/easybrowse/universalOperation',
            args: [
                {name: 'type', value: 'string'},
                {name: 'data', value: 'Object'}
            ],
            // todo 确认是否需要
            query: {
                oauthType: 'feed'
            },
            env: {
                main: {light: '11.8'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 注入inject.js文件
            // baiduboxapp://feed?action=injectJS&params=%7B%22name%22:%22injectjs%22,%22files%22:%5B%22inject:boxjs/inject.js%22%5D%7D&func=_BoxJS_feed_injectJS_545
            // 'baiduboxapp://feed?action=injectJS&params={'name':'injectjs','files':['inject:boxjs/inject.js']}&func=_BoxJS_feed_injectJS_545'
            name: 'easyBrowser.injectJS',
            invoke: 'iframe.bdboxURL',
            authority: 'feed',
            query: {
                action: 'injectJS',
                minver: '6.3'
            },
            args: [
                {name: 'name', value: 'string'},
                {name: 'files', value: 'string[]'}
            ],
            callbackName: 'func',
            env: {
                main: {light: '6.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 打开 feed hybrid
            name: 'easyBrowser.openHybrid',
            invoke: 'iframe.url',
            scheme: 'baiduboxapp',
            authority: 'v1',
            path: '/easybrowse/hybrid',
            args: [
                {name: 'type', value: 'string='},
                {name: 'url', value: 'string='},
                {name: 'tpl_id', value: 'string='},
                {name: 'append', value: 'string='},
                {name: 'tplpath', value: 'string='},
                {name: 'context', value: 'Object'},
                {name: 'style', value: 'Object='},
                {name: 'newbrowser', value: 'string='},
                {name: 'forbidautorotate', value: 'string='},
                {name: 'slog', value: 'Object='},
                {name: 'videoinfo', value: 'string='},
                {name: 'jssids', value: 'string='},
                {name: 'launchlogin', value: 'string='},
                {name: 'stay', value: 'string='},
                {name: 'next', value: 'string='},
                {name: 'commentInfo', value: 'Object='},
                {name: 'pagefrom', value: 'string='},
                {name: 'isSearchVideo', value: 'string='},
                {name: 'ch_url', value: 'string='},
                {name: 'first_image', value: 'string='}
            ],
            env: {
                main: {light: '8.2'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 轻框架 通用性能打点接口
            // doc: http://ft.baidu.com/#/scheme/utils_onperformanceflowevent
            // baiduboxapp://v20/utils/onPerformanceFlowEvent?params=$params&callback=$callback
            name: 'easyBrowser.performanceFlowLog',
            invoke: 'iframe.bdboxURL',
            authority: 'v20',
            path: '/utils/onPerformanceFlowEvent',
            args: [
                {name: 'actionId', value: 'string'},
                {name: 'operation', value: 'string'},
                {name: 'value', value: 'string'},
                {name: 'frameSource', value: 'string='},
                {name: 'businessType', value: 'string='}
            ],
            env: {
                main: {light: '10.3.6'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // 落地页调用NA Dialog弹窗，支持文字icon，button可配置
        // baiduboxapp://easybrowse/showDialog?params=$params&callback=$callback
        {
            name: 'easyBrowser.showDialog',
            invoke: 'iframe.bdboxURL',
            path: '/showDialog',
            authority: 'easybrowse',
            args: [{
                name: 'title',
                value: 'string'
            },
            {
                name: 'subtitle',
                value: 'string'
            },
            {
                name: 'sourceUrl',
                value: {
                    isRequired: true,
                    type: {
                        default: 'string',
                        night: 'string'
                    }
                }
            },
            {
                name: 'buttons',
                value: {
                    arrayOf: 'Object'
                }
            }
            ],
            nextCallback: 'callback',
            env: {
                main: {
                    light: '11.17'
                }
            }
        },
        // 保留图片
        // http://es.baidu-int.com/#/apidetail/1448
        {
            name: 'easyBrowser.image.save',
            invoke: 'iframe.url',
            authority: 'utils',
            path: '/saveImage',
            args: [
                {
                    name: 'url',
                    value: 'string'
                }
            ],
            env: {
                main: {
                    search: '8.2.0.0',
                    light: '8.2.0.0'
                }
            }
        },
        // 时政文章屏蔽部分活动弹窗
        // http://es.baidu-int.com/#/apidetail/1497
        {
            name: 'easyBrowser.sensitiveNews',
            invoke: 'iframe.bdboxURL',
            path: '/sensitiveNews',
            authority: 'easybrowse',
            args: [
                {
                    name: 'msg', value: 'string='
                }
            ],
            scheme: 'baiduboxapp',
            env: {
                main: {
                    light: '12.17.0.0'
                }
            }
        }
    ];


    var androidCompatibility = [
        {
            // 轻框架 通用性能打点接口 旧版
            // baiduboxapp://v20/utils/onPerformanceFlowEvent?params=$params&callback=$callback
            name: 'easyBrowser.onPerformanceFlowEvent',
            invoke: 'method.args-json',
            method: 'Bdbox_android_utils.onPerformanceFlowEvent',
            args: [
                {name: 'actionId', value: 'string'},
                {name: 'operation', value: 'string'},
                {name: 'value', value: 'string'}
            ],
            env: {
                main: {light: '8.2'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 打开 feed hybrid 旧版
            // {"intent":"intent:#Intent;S.backup={"url":"http://cp01-renzhonghua.epc.baidu.com:8888/html/meijishuang/newfeed/demotest.html?frame=hybrid"};
            // component=com.baidu.searchbox/.home.feed.FeedDetailActivity;S.toolbaricons={"toolids":
            // ["4","1","2","3"]};S.menumode=2;end","min_v":"16783629","mode":"0"}
            name: 'easyBrowser.openHybridOld',
            args: [
                {name: 'intent', value: 'string='},
                {name: 'min_v', value: 'string='},
                {name: 'mode', value: 'string='}
            ],
            method: 'Bdbox_android_utils.command',
            authority: 'easybrowse',
            invoke: 'method.args-json+callback',
            env: {
                main: {light: '5.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        }
    ];

    var iosCompatibility = [
        {
            // "baiduboxapp://common?action=easybrowse&params={"menumode":2,"sfrom":"feed","context":{"nid":"news_17596451975550153468"},
            // "backup":{"url":"http://cp01-renzhonghua.epc.baidu.com:8888/html/meijishuang/newfeed/demotest.html?frame=hybrid","version":"0"},
            // "newbrowser":"0","toolbar":"{\"toolids\":[\"4\",\"1\",\"2\",\"3\"]}",
            // "openurl":"undefined","minver":"5.3.0.0","isla":0,"opentype":1,"append":0,"rbtnstyle":2,
            // "type":"Hybrid"}&minver=5.3.0.0"
            // 打开 feed hybrid 旧版
            name: 'easyBrowser.openHybridOld',
            invoke: 'iframe.bdboxURL',
            authority: 'common',
            args: [
                {name: 'menumode', value: 'number='},
                {name: 'sfrom', value: 'string='},
                {name: 'context', value: 'Object='},
                {name: 'backup', value: 'Object='},
                {name: 'newbrowser', value: 'string='},
                {name: 'toolbar', value: 'Object='},
                {name: 'openurl', value: 'string='},
                {name: 'isla', value: 'number='},
                {name: 'opentype', value: 'number='},
                {name: 'append', value: 'number='},
                {name: 'rbtnstyle', value: 'number='},
                {name: 'type', value: 'string='}
            ],
            query: {
                action: 'easybrowse',
                minver: '5.3.0.0'
            },
            env: {
                main: {light: '5.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        {
            // 轻框架 通用性能打点旧版
            // scheme: 'baiduboxapp://utils?action=onPerformanceFlowEvent&params={'actionId':'end','value':1558776748132,'operation':'end'}&minver=8.2.0.0',
            // baiduboxapp://utils?action=onPerformanceFlowEvent&params=$params&minver=8.2
            name: 'easyBrowser.onPerformanceFlowEvent',
            invoke: 'iframe.bdboxURL',
            authority: 'utils',
            args: [
                {name: 'actionId', value: 'string'},
                {name: 'operation', value: 'string'},
                {name: 'value', value: 'string'}
            ],
            query: {
                action: 'onPerformanceFlowEvent',
                minver: '8.2.0.0'
            },
            env: {
                main: {light: '8.2'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        }
    ];

    var easyBrowserDescrioption = /*#__PURE__*/Object.freeze({
        __proto__: null,
        android: android$1,
        ios: ios$1,
        androidCompatibility: androidCompatibility,
        iosCompatibility: iosCompatibility
    });

    /**
     * @file 预置描述表
     * @author wanwenqi
     * @date 2021/05/19
     */

    var android$2 = [
        {
            name: 'search.invokeSearchBarMaskView',
            path: '/invokeSearchBarMaskView',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'searchframe',
            invoke: 'method.bdboxURL',
            args: [{
                name: 'maskAlpah',
                value: 'string='
            }, {
                name: 'animateDuration',
                value: 'string='
            }, {
                name: 'clickMaskCallback',
                value: 'string='
            }, {
                name: 'type',
                value: 'string'
            }],
            env: {
                main: {
                    search: '11.25.0.0',
                    light: '11.25.0.0'
                }
            }
        },
        {
            name: 'search.showMobileNetPushVideoSettingPageToast',
            args: [
                {
                    name: 'pageId',
                    value: 'string'
                },
                {
                    name: 'bussiness',
                    value: 'string'
                }
            ],
            path: '/showMobileNetPushVideoSettingPageToast',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'searchframe',
            invoke: 'method.bdboxURL',
            env: {
                main: {
                    search: '12.3',
                    light: '12.3'
                }
            }
        },
        {
            name: 'search.getMobileNetVideoAutoplaySwitchType',
            args: [],
            path: '/getMobileNetVideoAutoplaySwitchType',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'searchVideo',
            invoke: 'method.bdboxURL',
            env: {
                main: {search: '12.3.0.0'}
            }
        },
        {
            name: 'search.getWifiVideoAutoplaySwitchType',
            args: [],
            path: '/getWifiVideoAutoplay',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'utils',
            invoke: 'method.bdboxURL',
            env: {
                main: {
                    search: '10.6.0.0',
                    light: '10.6.0.0'
                },
                lite: {
                    search: '3.3.0.0',
                    light: '3.3.0.0'
                }
            }
        },
        {
            name: 'search.videoPrefetch',
            path: '/videoPrefetch',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'search',
            invoke: 'method.bdboxURL',
            args: [
                {
                    name: 'action',
                    value: 'number'
                },
                {
                    name: 'data',
                    value: 'Array'
                }
            ],
            env: {
                main: {search: '11.7.0.0'},
                lite: {
                    search: '4.17.0.0'
                }
            }
        },
        {
            name: 'search.getSummary',
            args: [
                {
                    name: 'loadingpageurl',
                    value: 'string',
                    comment: '对应的落地页url'
                },
                {
                    name: 'type',
                    value: 'number',
                    comment: 'type 1：只锚点 2：只高亮 3：锚点和高亮'
                },
                {
                    name: 'keyword',
                    value: 'string',
                    comment: '高亮文本'
                }
            ],
            path: '/getSummary',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'searchframe',
            invoke: 'method.bdboxURL',
            action: 'getSummary',
            env: {
                main: {
                    search: '12.4',
                    light: '12.4'
                }
            }
        },
        {
            name: 'search.setQuery',
            invoke: 'method.args-json',
            path: '/setQuery',
            method: 'Bdbox_android_common.setQuery',
            args: [
                {
                    name: 'type',
                    value: 'string',
                    comment: '类型'
                },
                {
                    name: 'history',
                    value: 'boolean',
                    comment: '入历史'
                },
                {
                    name: 'query',
                    value: 'string',
                    comment: '上框搜索词'
                }
            ],
            env: {
                main: {
                    search: '5.5'
                },
                lite: {
                    search: '1.1'
                }
            },
            query: {
                minver: '5.5.0.0'
            }
        },

        // http://es.baidu-int.com/#/apidetail/799
        // baiduboxapp://v1/searchframe/jumpToTab
        {
            name: 'search.changeNATab',
            invoke: 'method.bdboxURL',
            path: '/jumpToTab',
            authority: 'searchframe',
            method: 'Bdbox_android_jsbridge.dispatch',
            action: 'jumpToTab',
            args: [
                {
                    name: 'seClickID',
                    value: 'string=',
                    comment: '串联打点的id'
                },
                {
                    name: 'total',
                    value: 'string=',
                    comment: '打点日志相关参数 详见（打点功能：jumpToTab端能力跳转六合tab）'
                },
                {
                    name: 'word',
                    value: 'string',
                    comment: '用户判断是否是同query 不同query切新页面 需base64解码'
                },
                {
                    name: 'extra',
                    value: 'string',
                    comment: '打开页面需要拼接的字符串参数'
                },
                {
                    name: 'url',
                    value: 'string',
                    comment: '打开页面的url'
                },
                {
                    name: 'pd',
                    value: 'string',
                    comment: '需要打开的垂搜栏目关键字'
                }
            ],
            env: {
                main: {
                    search: '11.13.0.0'
                },
                lite: {
                    search: '4.13.0.0'
                }
            }
        },
        {
            // http://es.baidu-int.com/#/apidetail/1056
            // baiduboxapp://v1/browser/search?upgrade=1&query=%e9%ab%98%e5%9c%86%e5%9c%86&simple=0&newwindow=0&append=1
            name: 'search.invokeSearch',
            invoke: 'method.url',
            authority: 'v1',
            method: 'Bdbox_android_jsbridge.dispatch',
            action: 'search',
            path: '/browser/search',
            args: [
                {
                    name: 'sa',
                    value: 'string=',
                    comment: '搜索来源，Android单端'
                },
                {
                    name: 'stay',
                    value: 'string=',
                    comment: '退出搜索是否停留在手百，Android单端'
                },
                {
                    name: 'source',
                    value: 'string=',
                    comment: '双端：调起方来源，iOS目前只有\'spotlight\''
                },
                {
                    name: 'from',
                    value: 'string=',
                    comment: '是否是talos调起，iOS未使用'
                },
                {
                    name: 'next',
                    value: 'string=',
                    comment: '下个scheme，iOS未使用'
                },
                {
                    name: 'config',
                    value: 'string=',
                    comment: '配置参数，iOS未使用'
                },
                {
                    name: 'fromwise',
                    value: 'string=',
                    comment: '是否来自wise，iOS未使用'
                },
                {
                    name: 'launchMode',
                    value: 'string=',
                    comment: '调起模式，Android单端'
                },
                {
                    name: 'query',
                    value: 'string',
                    comment: '搜索词'
                },
                {
                    name: 'simple',
                    value: 'string=',
                    comment: '是否简版浏览框架'
                },
                {
                    name: 'append',
                    value: 'string=',
                    comment: '是否加公参'
                },
                {
                    name: 'newwindow',
                    value: 'string=',
                    comment: '新窗口'
                }
            ],
            env: {
                main: {
                    search: '0.0.0.0',
                    light: '0.0.0.0'
                },
                lite: {
                    search: '2.2.0.0',
                    light: '2.2.0.0'
                }
            }
        },
        {
            // http://es.baidu-int.com/#/apidetail/33
            // baiduboxapp://v1/browser/open?upgrade=1&url=http%3a%2f%2fwww.163.com%2f&simple=0&newwindow=0&append=1
            name: 'search.open',
            invoke: 'method.url',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/open',
            authority: 'browser',
            action: 'open',
            args: [
                {
                    name: 'tplId',
                    value: 'string=',
                    comment: '历史模板'
                },
                {
                    name: 'logargs',
                    value: 'Object=',
                    comment: ''
                },
                {
                    name: 'loftSearchType',
                    value: 'string=',
                    comment: '负一楼业务方类型'
                },
                {
                    name: 'loftSearchUrl',
                    value: 'string=',
                    comment: '负一楼url'
                },
                {
                    name: 'loftSearchContainerState',
                    value: 'string=',
                    comment: '负一楼webview状态'
                },
                {
                    name: 'privateTips',
                    value: 'string=',
                    comment: '搜索框架隐私模式，首次进入在底bar弹出文案提示'
                },
                {
                    name: 'stay',
                    value: 'string=',
                    comment: 'Android单端：退出搜索框架是否返回首页'
                },
                {
                    name: 'noHistory',
                    value: 'string=',
                    comment: '0|1,表示当前端能力打开的页面是否要进入历史记录。默认是0，表示正常记录历史，1表示不进入历史。'
                },
                {
                    name: 'openParallelLoad',
                    value: 'string=',
                    comment: 'iOS单端：是否开启并行化加载'
                },
                {
                    name: 'from',
                    value: 'string=',
                    comment: 'Android单端：是否是talos调起'
                },
                {
                    name: 'sfUrl',
                    value: 'string=',
                    comment: 'Android单端：sf调用的Url'
                },
                {
                    name: 'launchMode',
                    value: 'string=',
                    comment: 'Android单端：调起模式'
                },
                {
                    name: 'isBee',
                    value: 'string=',
                    comment: 'Android单端：Feed hybird落地页内打开搜索bee框架标志 11.17添加'
                },
                {
                    name: 'next',
                    value: 'string=',
                    comment: '下个scheme'
                },
                {
                    name: 'fromwise',
                    value: 'string=',
                    comment: '是否来自wise'
                },
                {
                    name: 'isCollect',
                    value: 'number=',
                    comment: ''
                },
                {
                    name: 'isHistory',
                    value: 'number=',
                    comment: ''
                },
                {
                    name: 'logUrl',
                    value: 'string=',
                    comment: 'Android单端：V11.13版本新增打点日志的url\nAndroid该参数废弃，非必填'
                },
                {
                    name: 'url',
                    value: 'string',
                    comment: 'V8.2.0.0版本新增, url'
                },
                {
                    name: 'append',
                    value: 'string=',
                    comment: 'V8.2.0.0版本新增, 是否加公参。\'0\':默认不加。'
                },
                {
                    name: 'newwindow',
                    value: 'string=',
                    comment: 'V8.2.0.0版本新增, 是否从新窗口打开。 \'0\':默认从当前窗口打开'
                },
                {
                    name: 'simple',
                    value: 'string=',
                    comment: 'V8.2.0.0版本新增, 是否简版搜索浏览框架。\'0\':默认打开主浏览框架。'
                },
                {
                    name: 'origin',
                    value: 'string=',
                    comment: 'V10.13.0.0版本新增, url的origin信息，端走同步跳转时，用origin拼接url去跳转（可选）'
                },
                {
                    name: 'animation',
                    value: 'string=',
                    comment: 'V10.13.0.0版本新增, 入场动画类型（可选）\nAndroid该参数废弃'
                },
                {
                    name: 'extra',
                    value: 'string=',
                    comment: 'Android单端：V10.13.0.0版本新增, H5需要的扩展参数（可选）（用于透传）'
                },
                {
                    name: 'isContainer',
                    value: 'string=',
                    comment: 'V10.13.0.0版本新增, 判断是否是新容器打开（可选）。\'0\' 不是，\'1\' 是新容器打开。'
                },
                {
                    name: 'config',
                    value: 'string=',
                    comment: '配置参数，包含参数： immersed 、hideBack、feedbackUrl、background'
                },
                {
                    name: 'canAsync',
                    value: 'string=',
                    comment: '是否可以异步。0 不可以异步，走同步加载，\n\n1 可以异步，端上按需异步加载'
                }
            ],
            env: {
                main: {
                    search: '8.2.0.0',
                    light: '8.2.0.0',
                    imagesearch: '13.0'
                },
                lite: {
                    search: '1.3.0.0',
                    light: '1.3.0.0'
                }
            }
        },
        {
            name: 'search.getSSession',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/getSSession',
            authority: 'browser',
            args: [],
            scheme: 'baiduboxapp',
            env: {
                main: {
                    search: '12.16.0.0',
                    light: '12.16.0.0'
                }
            }
        },
        {
            name: 'search.invokePoiLandingPage',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/invokePoiLandingPage',
            authority: 'browser',
            args: [
                {
                    name: '_tplExt',
                    value: 'Object=',
                    comment: '11.23新增，模版扩展参数，透传到1013统计'
                },
                {
                    name: 'prefetch',
                    value: 'Object=',
                    comment: '预取'
                },
                {
                    name: 'query',
                    value: 'string',
                    comment: '检索关键词'
                },
                {
                    name: 'nwd',
                    value: 'string',
                    comment: '检索地图接口数据和请求gss'
                },
                {
                    name: 'title',
                    value: 'string',
                    comment: 'POI页面的标题'
                },
                {
                    name: 'poiType',
                    value: 'string=',
                    comment: 'POI类型'
                },
                {
                    name: 'pageType',
                    value: 'number',
                    comment: '1：单点，2：多点'
                },
                {
                    name: 'url',
                    value: 'string',
                    comment: '如果不进入地图NA页，从结果页进入地图h5页面的url'
                },
                {
                    name: 'ext',
                    value: {
                        type: {
                            clickTime: 'string=',
                            address: 'string',
                            userCityCode: 'string',
                            userLoc: 'string=',
                            loc: 'string=',
                            cityName: 'string',
                            cityCode: 'string'
                        },
                        isRequired: true
                    },
                    comment: '位置信息'
                },
                {
                    name: 'ubs',
                    value: {
                        type: {
                            srcid: 'string=',
                            qid: 'string=',
                            tempName: 'string=',
                            sid: 'string='
                        },
                        isRequired: true
                    },
                    comment: '打点相关字段'
                }
            ],
            env: {
                main: {
                    search: '11.10.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1095
        // 安卓单端
        {
            name: 'search.onShowDialogOrToast',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/onShowDialogOrToast',
            authority: 'searchframe',
            action: 'onShowDialogOrToast',
            args: [
                {
                    name: 'position',
                    value: 'Array'
                },
                {
                    name: 'isShow',
                    value: 'number'
                }
            ],
            scheme: 'baiduboxapp',
            env: {
                main: {
                    search: '11.23.0.0',
                    light: '11.23.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1614
        {
            name: 'search.advanceFilterClear',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/clear',
            authority: 'advanceFilter',
            args: [
                {
                    name: 'clear',
                    value: 'number'
                }
            ],
            env: {
                main: {
                    search: '12.22.0.0'
                }
            }
        },
        {
            // http://es.baidu-int.com/#/apidetail/987
            // 不支持 lite
            name: 'search.showBottomNaBar',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/showBottomNaBar',
            authority: 'searchframe',
            args: [
                {
                    name: 'fullScreenCallback',
                    value: 'string='
                },
                {
                    name: 'type',
                    value: 'string'
                },
                {
                    name: 'isShowDelay',
                    value: 'number='
                },
                {
                    name: 'result',
                    value: {
                        type: {
                            ext: 'Object=',
                            sid: 'string=',
                            recList: 'Array=',
                            word: 'string=',
                            emptyPage: 'Object=',
                            source: 'string=',
                            lid: 'string=',
                            pu: 'string=',
                            avatarList: 'Array=',
                            dispNum: 'string=',
                            url: 'string='
                        },
                        isRequired: true
                    }
                }
            ],
            env: {
                main: {
                    search: '11.23.0.0'
                },
                lite: false
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1374
        {
            name: 'search.invokeVoiceSettingPage',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/openVoiceSetting',
            authority: 'speech',
            args: [
                {
                    name: 'voiceSource',
                    value: 'string'
                }
            ],
            env: {
                main: {
                    search: '12.8.0.0'
                }
            }
        },
        // 无es平台描述，属于search.invokeVoiceSettingPage对应的低版本端能力
        // android版本，参数为3个，参数名和参数个数和iOS版本完全不一致，因此设计为两个独立的端能力描述
        {
            name: 'search.openSetting.android',
            scheme: 'baiduboxapp',
            invoke: 'method.args-json+callback',
            method: 'Bdbox_android_utils.startVoice',
            args: [
                {
                    name: 'voiceSource',
                    value: 'string='
                },
                {
                    name: 'method_name',
                    value: 'string='
                },
                {
                    name: 'source_app',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    search: '8.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1142
        {
            name: 'search.tts.stop',
            invoke: 'method.args-json',
            method: 'Bdbox_android_tts.stop',
            args: [
                {
                    name: 'option',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    search: '8.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1141
        {
            name: 'search.tts.speak',
            invoke: 'method.args-json',
            method: 'Bdbox_android_tts.speak',
            args: [
                {
                    name: 'speaker',
                    value: 'number='
                },
                {
                    name: 'isMuteCheck',
                    value: 'number='
                },
                {
                    name: 'voiceType',
                    value: 'number='
                },
                {
                    name: 'pitch',
                    value: 'string='
                },
                {
                    name: 'speed',
                    value: 'string='
                },
                {
                    name: 'auto',
                    value: 'number='
                },
                {
                    name: 'text',
                    value: 'string'
                },
                {
                    name: 'callback',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    search: '8.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1176
        // 同步歌曲播放状态
        {
            name: 'search.getPlayerStatus',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/getPlayerStatus',
            authority: 'searchmusic',
            args: [],
            env: {
                main: {
                    light: '11.26.0.0',
                    search: '11.26.0.0'
                }
            }
        },

        // 安卓单端 http://es.baidu-int.com/#/apidetail/1086
        {
            name: 'search.invokeImageSearch',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/imageSearch',
            authority: 'browser',
            args: [
                {
                    name: 'jsup',
                    value: 'string='
                },
                {
                    name: 'from',
                    value: 'string'
                },
                {
                    name: 'category',
                    value: 'string='
                },
                {
                    name: 'imageSearch_type',
                    value: 'string='
                },
                {
                    name: 'imageSearch_mode',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    search: '11.17.0.0'
                },
                lite: {
                    search: '4.13.0.0'
                }
            }
        },
        {
            // http://es.baidu-int.com/#/apidetail/1307
            // 不支持 lite
            name: 'search.transData',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/transData',
            authority: 'searchframe',
            args: [
                {
                    name: 'transData',
                    value: 'Object'
                }
            ],
            env: {
                main: {
                    search: '12.5.0.0'
                },
                lite: false
            }
        },
        // 安卓单端 http://es.baidu-int.com/index#/apidetail/1169
        {
            name: 'search.isNATipsShowing',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/isNATipsShowing',
            authority: 'searchframe',
            args: [],
            env: {
                main: {
                    search: '11.25.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1239
        {
            name: 'search.showVoiceOperation',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/showVoiceOperation',
            authority: 'searchframe',
            args: [
                {
                    name: 'ui_params',
                    value: {
                        type: {
                            delayMs: 'number=',
                            timeoutMs: 'number',
                            text: 'string',
                            /* eslint-disable camelcase */
                            bg_color: 'string=',
                            bg_color_night: 'string=',
                            showTips: 'string',
                            xiaodu_url: 'string',
                            xiaodu_url_night: 'string',
                            click_scheme: 'string='
                            /* eslint-enable camelcase */
                        },
                        isRequired: true
                    }
                },
                {
                    name: 'callback_url',
                    value: 'string'
                },
                {
                    name: 'click_url',
                    value: 'string'
                },
                {
                    name: 'cb_params',
                    value: {
                        type: {
                            key1: 'string',
                            key2: 'string'
                        },
                        isRequired: true
                    }
                },
                {
                    name: 'click_params',
                    value: {
                        type: {
                            key1: 'string',
                            key2: 'string'
                        },
                        isRequired: true
                    }
                },
                {
                    name: 'display',
                    value: 'number'
                }
            ],
            env: {
                main: {
                    search: '12.0.0.0'
                }
            }
        },
        // 安卓单端 http://es.baidu-int.com/#/apidetail/947
        {
            name: 'search.invokeVoiceSearchPage',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/startVoiceSearch',
            authority: 'speech',
            args: [
                {
                    name: 'voiceSource',
                    value: 'string'
                }
            ],
            env: {
                main: {
                    search: '11.16.0.0'
                },
                lite: {
                    search: '4.13.0.0'
                }
            }
        }
    ];

    var ios$2 = [
        {
            name: 'search.invokeSearchBarMaskView',
            path: '/invokeSearchBarMaskView',
            invoke: 'iframe.bdboxURL',
            args: [{
                name: 'maskAlpah',
                value: 'string='
            }, {
                name: 'animateDuration',
                value: 'string='
            }, {
                name: 'clickMaskCallback',
                value: 'string='
            }, {
                name: 'type',
                value: 'string'
            }],
            authority: 'searchframe',
            env: {
                main: {
                    search: '12.5.0.0',
                    light: '12.5.0.0'
                }
            }
        },
        {
            name: 'search.showMobileNetPushVideoSettingPageToast',
            invoke: 'iframe.bdboxURL',
            path: '/showMobileNetPushVideoSettingPageToast',
            authority: 'searchframe',
            args: [
                {
                    name: 'pageId',
                    value: 'string',
                    comment: '页面的id，端用来判断当前页面冷启动是否显示过toast，只显示一次'
                },
                {
                    name: 'bussiness',
                    value: 'string',
                    comment: '业务线，ubc打点区分不同业务minivideo_talos（小视频tab）、aminivideo（小视频卡）、avideo（短视频卡）、aNR（自然结果）'
                }
            ],
            env: {
                main: {
                    search: '12.3.0.0',
                    light: '12.3.0.0'
                }
            }
        },
        {
            name: 'search.getMobileNetVideoAutoplaySwitchType',
            path: '/getMobileNetVideoAutoplaySwitchType',
            invoke: 'iframe.bdboxURL',
            authority: 'searchVideo',
            env: {
                main: {
                    search: '12.3.0.0',
                    light: '12.3.0.0'
                }
            }
        },
        {
            name: 'search.getWifiVideoAutoplaySwitchType',
            path: '/getWifiVideoAutoplay',
            invoke: 'iframe.bdboxURL',
            authority: 'utils',
            env: {
                main: {
                    search: '10.6.0.0',
                    light: '10.6.0.0'
                },
                lite: {
                    search: '3.6.0.0',
                    light: '3.6.0.0'
                }
            }
        },
        {
            name: 'search.videoPrefetch',
            invoke: 'iframe.bdboxURL',
            path: '/videoPrefetch',
            authority: 'search',
            args: [
                {
                    name: 'action',
                    value: 'number',
                    comment: '【必传】0：进入页面; 1：滑停; 2：点击;'
                },
                {
                    name: 'data',
                    value: {
                        arrayOf: 'Object'
                    },
                    comment: '【必传】视频播放器数据,和外部不重复，是视频的数据，数据信息见基本信息介绍\n\'data\':[\n//第一条视频信息\n{\n \'nid\':\'123\',          /* 视频唯一标识*/\n \'prefetchType\':0,   /* 预取类型-视频 */\n\'businessType\': \'search\'  /* 业务方 */\n\'prefetchInfo\':\n{\n\'size\': 200*1024, /* 预取大小、字节 */\n\'url\':\'http://ft.baidu.com/\' /* 预取url */\n},\n}\n     //第二条视频信息\n{\n \'nid\':\'234\',          /* 视频唯一标识*/\n \'prefetchType\':0,   /* 预取类型-视频 */\n\'businessType\': \'search\'  /* 业务方 */\n\'prefetchInfo\':\n{\n\'size\': 200*1024, /* 预取大小、字节 */\n\'url\':\'http://ft.baidu.com/\'   /* 预取url */\n}\n}'
                }
            ],
            env: {
                main: {
                    search: '11.7.0.0',
                    light: '11.7.0.0'
                },
                lite: {
                    search: '4.3.0.0',
                    light: '4.3.0.0'
                }
            }
        },
        {
            name: 'search.getSummary',
            invoke: 'iframe.bdboxURL',
            path: '/getSummary',
            authority: 'searchframe',
            args: [
                {
                    name: 'loadingpageurl',
                    value: 'string',
                    comment: '对应的落地页url'
                },
                {
                    name: 'type',
                    value: 'number',
                    comment: 'type 1：只锚点 2：只高亮 3：锚点和高亮'
                },
                {
                    name: 'keyword',
                    value: 'string',
                    comment: '高亮文本'
                }
            ],
            env: {
                main: {
                    search: '12.4',
                    light: '12.4'
                }
            }
        },
        {
            name: 'search.setQuery',
            invoke: 'iframe.bdboxURL',
            path: '',
            authority: 'setquery',
            args: [
                {
                    name: 'type',
                    value: 'string',
                    comment: '类型'
                },
                {
                    name: 'history',
                    value: 'number',
                    comment: '入历史'
                },
                {
                    name: 'query',
                    value: 'string',
                    comment: '上框搜索词'
                }
            ],
            env: {
                main: {
                    search: '5.5'
                },
                lite: {
                    search: '1.1'
                }
            },
            query: {
                minver: '5.5.0.0'
            }
        },
        // http://es.baidu-int.com/#/apidetail/799
        // baiduboxapp://v1/searchframe/jumpToTab
        {
            name: 'search.changeNATab',
            invoke: 'iframe.bdboxURL',
            path: '/jumpToTab',
            authority: 'searchframe',
            args: [
                {
                    name: 'seClickID',
                    value: 'string=',
                    comment: '串联打点的id'
                },
                {
                    name: 'total',
                    value: 'string=',
                    comment: '打点日志相关参数 详见（打点功能：jumpToTab端能力跳转六合tab）'
                },
                {
                    name: 'word',
                    value: 'string',
                    comment: '用户判断是否是同query 不同query切新页面 需base64解码'
                },
                {
                    name: 'extra',
                    value: 'string',
                    comment: '打开页面需要拼接的字符串参数'
                },
                {
                    name: 'url',
                    value: 'string',
                    comment: '打开页面的url'
                },
                {
                    name: 'pd',
                    value: 'string',
                    comment: '需要打开的垂搜栏目关键字'
                }
            ],
            env: {
                main: {
                    search: '11.13.0.0'
                },
                lite: {
                    search: '4.7.0.0'
                }
            }
        },
        {
            // baiduboxapp://v1/browser/search?upgrade=1&query=%e9%ab%98%e5%9c%86%e5%9c%86&simple=0&newwindow=0&append=1
            name: 'search.invokeSearch',
            invoke: 'iframe.url',
            authority: 'v1',
            path: '/browser/search',
            args: [
                {
                    name: 'sa',
                    value: 'string=',
                    comment: '搜索来源，Android单端'
                },
                {
                    name: 'stay',
                    value: 'string=',
                    comment: '退出搜索是否停留在手百，Android单端'
                },
                {
                    name: 'source',
                    value: 'string=',
                    comment: '双端：调起方来源，iOS目前只有\'spotlight\''
                },
                {
                    name: 'from',
                    value: 'string=',
                    comment: '是否是talos调起，iOS未使用'
                },
                {
                    name: 'next',
                    value: 'string=',
                    comment: '下个scheme，iOS未使用'
                },
                {
                    name: 'config',
                    value: 'string=',
                    comment: '配置参数，iOS未使用'
                },
                {
                    name: 'fromwise',
                    value: 'string=',
                    comment: '是否来自wise，iOS未使用'
                },
                {
                    name: 'launchMode',
                    value: 'string=',
                    comment: '调起模式，Android单端'
                },
                {
                    name: 'query',
                    value: 'string',
                    comment: '搜索词'
                },
                {
                    name: 'simple',
                    value: 'string=',
                    comment: '是否简版浏览框架'
                },
                {
                    name: 'append',
                    value: 'string=',
                    comment: '是否加公参'
                },
                {
                    name: 'newwindow',
                    value: 'string=',
                    comment: '新窗口'
                }
            ],
            env: {
                main: {
                    search: '0.0.0.0',
                    light: '0.0.0.0'
                },
                lite: {
                    search: '1.3.0.0',
                    light: '1.3.0.0'
                }
            }
        },
        {
            // baiduboxapp://v1/browser/open?upgrade=1&url=http%3a%2f%2fwww.163.com%2f&simple=0&newwindow=0&append=1
            name: 'search.open',
            invoke: 'iframe.url',
            authority: 'v1',
            path: '/browser/open',
            args: [
                {
                    name: 'tplId',
                    value: 'string=',
                    comment: '历史模板'
                },
                {
                    name: 'logargs',
                    value: 'Object=',
                    comment: ''
                },
                {
                    name: 'loftSearchType',
                    value: 'string=',
                    comment: '负一楼业务方类型'
                },
                {
                    name: 'loftSearchUrl',
                    value: 'string=',
                    comment: '负一楼url'
                },
                {
                    name: 'loftSearchContainerState',
                    value: 'string=',
                    comment: '负一楼webview状态'
                },
                {
                    name: 'privateTips',
                    value: 'string=',
                    comment: '搜索框架隐私模式，首次进入在底bar弹出文案提示'
                },
                {
                    name: 'stay',
                    value: 'string=',
                    comment: 'Android单端：退出搜索框架是否返回首页'
                },
                {
                    name: 'noHistory',
                    value: 'string=',
                    comment: '0|1,表示当前端能力打开的页面是否要进入历史记录。默认是0，表示正常记录历史，1表示不进入历史。'
                },
                {
                    name: 'openParallelLoad',
                    value: 'string=',
                    comment: 'iOS单端：是否开启并行化加载'
                },
                {
                    name: 'from',
                    value: 'string=',
                    comment: 'Android单端：是否是talos调起'
                },
                {
                    name: 'sfUrl',
                    value: 'string=',
                    comment: 'Android单端：sf调用的Url'
                },
                {
                    name: 'launchMode',
                    value: 'string=',
                    comment: 'Android单端：调起模式'
                },
                {
                    name: 'isBee',
                    value: 'string=',
                    comment: 'Android单端：Feed hybird落地页内打开搜索bee框架标志 11.17添加'
                },
                {
                    name: 'next',
                    value: 'string=',
                    comment: '下个scheme'
                },
                {
                    name: 'fromwise',
                    value: 'string=',
                    comment: '是否来自wise'
                },
                {
                    name: 'isCollect',
                    value: 'number=',
                    comment: ''
                },
                {
                    name: 'isHistory',
                    value: 'number=',
                    comment: ''
                },
                {
                    name: 'logUrl',
                    value: 'string=',
                    comment: 'Android单端：V11.13版本新增打点日志的url\nAndroid该参数废弃，非必填'
                },
                {
                    name: 'url',
                    value: 'string',
                    comment: 'V8.2.0.0版本新增, url'
                },
                {
                    name: 'append',
                    value: 'string=',
                    comment: 'V8.2.0.0版本新增, 是否加公参。\'0\':默认不加。'
                },
                {
                    name: 'newwindow',
                    value: 'string=',
                    comment: 'V8.2.0.0版本新增, 是否从新窗口打开。 \'0\':默认从当前窗口打开'
                },
                {
                    name: 'simple',
                    value: 'string=',
                    comment: 'V8.2.0.0版本新增, 是否简版搜索浏览框架。\'0\':默认打开主浏览框架。'
                },
                {
                    name: 'origin',
                    value: 'string=',
                    comment: 'V10.13.0.0版本新增, url的origin信息，端走同步跳转时，用origin拼接url去跳转（可选）'
                },
                {
                    name: 'animation',
                    value: 'string=',
                    comment: 'V10.13.0.0版本新增, 入场动画类型（可选）\nAndroid该参数废弃'
                },
                {
                    name: 'extra',
                    value: 'string=',
                    comment: 'Android单端：V10.13.0.0版本新增, H5需要的扩展参数（可选）（用于透传）'
                },
                {
                    name: 'isContainer',
                    value: 'string=',
                    comment: 'V10.13.0.0版本新增, 判断是否是新容器打开（可选）。\'0\' 不是，\'1\' 是新容器打开。'
                },
                {
                    name: 'config',
                    value: 'string=',
                    comment: '配置参数，包含参数： immersed 、hideBack、feedbackUrl、background'
                },
                {
                    name: 'canAsync',
                    value: 'string=',
                    comment: '是否可以异步。0 不可以异步，走同步加载，\n\n1 可以异步，端上按需异步加载'
                }
            ],
            env: {
                main: {
                    search: '8.2.0.0',
                    light: '8.2.0.0',
                    imagesearch: '12.20'
                },
                lite: {
                    search: '2.2.0.0',
                    light: '2.2.0.0'
                }
            }
        },
        {
            name: 'search.getSSession',
            invoke: 'iframe.bdboxURL',
            path: '/getSSession',
            authority: 'browser',
            args: [],
            scheme: 'baiduboxapp',
            env: {
                main: {
                    search: '12.16.0.0',
                    light: '12.16.0.0'
                }
            }
        },
        {
            name: 'search.invokePoiLandingPage',
            invoke: 'iframe.bdboxURL',
            path: '/invokePoiLandingPage',
            authority: 'browser',
            args: [
                {
                    name: '_tplExt',
                    value: 'Object=',
                    comment: '11.23新增，模版扩展参数，透传到1013统计'
                },
                {
                    name: 'prefetch',
                    value: 'Object=',
                    comment: '预取'
                },
                {
                    name: 'query',
                    value: 'string',
                    comment: '检索关键词'
                },
                {
                    name: 'nwd',
                    value: 'string',
                    comment: '检索地图接口数据和请求gss'
                },
                {
                    name: 'title',
                    value: 'string',
                    comment: 'POI页面的标题'
                },
                {
                    name: 'poiType',
                    value: 'string=',
                    comment: 'POI类型'
                },
                {
                    name: 'pageType',
                    value: 'number',
                    comment: '1：单点，2：多点'
                },
                {
                    name: 'url',
                    value: 'string',
                    comment: '如果不进入地图NA页，从结果页进入地图h5页面的url'
                },
                {
                    name: 'ext',
                    value: {
                        type: {
                            clickTime: 'string=',
                            address: 'string=',
                            userCityCode: 'string',
                            userLoc: 'string=',
                            loc: 'string=',
                            cityName: 'string',
                            cityCode: 'string'
                        },
                        isRequired: true
                    },
                    comment: '位置信息'
                },
                {
                    name: 'ubs',
                    value: {
                        type: {
                            srcid: 'string=',
                            qid: 'string=',
                            tempName: 'string=',
                            sid: 'string='
                        },
                        isRequired: true
                    },
                    comment: '打点相关字段'
                }
            ],
            env: {
                main: {
                    search: '11.10.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1374
        {
            name: 'search.invokeVoiceSettingPage',
            invoke: 'iframe.bdboxURL',
            path: '/openVoiceSetting',
            authority: 'speech',
            args: [
                {
                    name: 'voiceSource',
                    value: 'string'
                }
            ],
            env: {
                main: {
                    search: '12.25.0.0'
                }
            }
        },
        // 无es平台描述，属于search.invokeVoiceSettingPage对应的低版本端能力
        // iOS版本，参数为2个，参数名和参数个数和android版本完全不一致，因此设计为两个独立的端能力描述
        {
            name: 'search.openSetting.ios',
            invoke: 'iframe.bdboxURL',
            authority: 'v6',
            path: '/vendor/speech/open',
            query: {
                upgrade: 0,
                oauthType: 'search'
            },
            args: [
                {
                    name: 'plugin_id',
                    value: 'string='
                },
                {
                    name: 'url',
                    value: 'string='
                }
            ],
            env: {
                main: {
                    search: '8.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1142
        {
            name: 'search.tts.stop',
            invoke: 'iframe.bdboxURL',
            path: '/stop',
            authority: 'tts',
            args: [],
            env: {
                main: {
                    search: '11.2.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1141
        {
            name: 'search.tts.speak',
            invoke: 'iframe.bdboxURL',
            path: '/speak',
            authority: 'tts',
            args: [
                {
                    name: 'speaker',
                    value: 'number='
                },
                {
                    name: 'isMuteCheck',
                    value: 'number='
                },
                {
                    name: 'voiceType',
                    value: 'number='
                },
                {
                    name: 'pitch',
                    value: 'string='
                },
                {
                    name: 'speed',
                    value: 'string='
                },
                {
                    name: 'auto',
                    value: 'number='
                },
                {
                    name: 'text',
                    value: 'string'
                },
                {
                    name: 'callback',
                    value: 'string'
                }
            ],
            env: {
                main: {
                    search: '11.2.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1176
        // 同步歌曲播放状态
        {
            name: 'search.getPlayerStatus',
            invoke: 'iframe.bdboxURL',
            path: '/getPlayerStatus',
            authority: 'searchmusic',
            args: [],
            env: {
                main: {
                    light: '11.26.0.0',
                    search: '11.26.0.0'
                }
            }
        },
        {
            // http://es.baidu-int.com/#/apidetail/1307
            // 不支持 lite
            name: 'search.transData',
            invoke: 'iframe.bdboxURL',
            path: '/transData',
            authority: 'searchframe',
            args: [
                {
                    name: 'transData',
                    value: 'Object'
                }
            ],
            env: {
                main: {
                    search: '12.5.0.0'
                },
                lite: false
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1614
        {
            name: 'search.advanceFilterClear',
            invoke: 'iframe.bdboxURL',
            path: '/clear',
            authority: 'advanceFilter',
            args: [
                {
                    name: 'clear',
                    value: 'number'
                }
            ],
            env: {
                main: {
                    search: '12.29.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1239
        {
            name: 'search.showVoiceOperation',
            invoke: 'iframe.bdboxURL',
            path: '/showVoiceOperation',
            authority: 'searchframe',
            args: [
                {
                    name: 'ui_params',
                    value: {
                        type: {
                            delayMs: 'number=',
                            timeoutMs: 'number',
                            text: 'string',
                            /* eslint-disable camelcase */
                            bg_color: 'string=',
                            bg_color_night: 'string=',
                            showTips: 'string',
                            xiaodu_url: 'string',
                            xiaodu_url_night: 'string',
                            click_scheme: 'string='
                            /* eslint-enable camelcase */
                        },
                        isRequired: true
                    }
                },
                {
                    name: 'callback_url',
                    value: 'string'
                },
                {
                    name: 'click_url',
                    value: 'string'
                },
                {
                    name: 'cb_params',
                    value: {
                        type: {
                            key1: 'string',
                            key2: 'string'
                        },
                        isRequired: true
                    }
                },
                {
                    name: 'click_params',
                    value: {
                        type: {
                            key1: 'string',
                            key2: 'string'
                        },
                        isRequired: true
                    }
                },
                {
                    name: 'display',
                    value: 'number'
                }
            ],
            env: {
                main: {
                    search: '12.0.0.0'
                }
            }
        }
    ];

    var searchDescription = /*#__PURE__*/Object.freeze({
        __proto__: null,
        android: android$2,
        ios: ios$2
    });

    /**
     * @file  预置描述表
     * @author chenqi14
     * @date 2019/09/26
     */

    var android$3 = [
        // http://es.baidu-int.com/index#/apidetail/1293
        {
            name: 'share.doShare',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/doShare',
            authority: 'share',
            args: [{
                    name: 'shareCallback',
                    value: 'string'
                },
                {
                    name: 'shareData',
                    value: 'Object'
                }
            ],
            env: {
                main: {
                    imagesearch: '13.0',
                    search: '12.24'
                }
            }
        }
    ];


    var ios$3 = [
        // http://es.baidu-int.com/index#/apidetail/1293
        {
            name: 'share.doShare',
            invoke: 'iframe.bdboxURL',
            path: '/doShare',
            authority: 'share',
            args: [{
                    name: 'shareCallback',
                    value: 'string'
                },
                {
                    name: 'shareData',
                    value: 'Object'
                }
            ],
            env: {
                main: {
                    imagesearch: '12.20',
                    search: '12.24'
                }
            }
        }
    ];

    var shareDescription = /*#__PURE__*/Object.freeze({
        __proto__: null,
        android: android$3,
        ios: ios$3
    });

    /**
     * @file  预置描述表
     * @author chenqi14
     * @date 2019/09/26
     */

    var android$4 = [
        // 判断scheme对应的应用是否在该设备安装
        // http://es.baidu-int.com/#/apidetail/501
        // http://ft.baidu.com/#/scheme/utils_queryscheme
        // baiduboxapp://v16/utils/queryScheme?params={'pkg':'wechat'}&callback=_bdbox_js
        {
            name: 'system.queryScheme',
            invoke: 'method.bdboxURL',
            authority: 'v16',
            path: '/utils/queryScheme',
            method: 'Bdbox_android_jsbridge.dispatch',
            args: [
                {name: 'pkg', value: 'string'}
            ],
            env: {
                main: {light: '10.0'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // 主题getNightMode
        // baiduboxapp://v16/theme/getNightMode?callback=$callback
        {
            name: 'system.getNightMode',
            scheme: 'baiduboxapp',
            invoke: 'method.url',
            authority: 'v16',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/theme/getNightMode',
            env: {
                main: {light: '10.0'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/471
        {
            name: 'system.getAppDuration',
            args: [],
            path: '/getAppDuration',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'utils',
            invoke: 'method.bdboxURL',
            scheme: 'baiduboxapp',
            env: {
                main: {light: '10.3.0.0'}
            }
        },
        // http://es.baidu-int.com/#/apidetail/470
        // baiduboxapp://v19/utils/getClipboardData?callback=\$callback
        {
            name: 'system.getClipboardData',
            path: '/getClipboardData',
            authority: 'utils',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            args: [],
            env: {
                main: {
                    search: '10.3.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/502
        {
            name: 'system.setClipboardData',
            args: [{
                name: 'data',
                value: 'string'
            }],
            path: '/setClipboardData',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'utils',
            invoke: 'method.bdboxURL',
            scheme: 'baiduboxapp',
            env: {
                main: {
                    light: '10.3.0.0'
                }
            }
        },
        // http://es.baidu-int.com/index#/apidetail/1681
        // Android 调起添加桌面Widget的弹窗
        {
            'name': 'system.addWidgetDialog',
            'invoke': 'method.bdboxURL',
            'method': 'Bdbox_android_jsbridge.dispatch',
            'path': '/addWidgetDialog',
            'authority': 'newFeature',
            'args': [
                {
                    /**
                     * 经典小部件 1
                     * 极简小部件 2
                     * 时钟小部件 3
                     * feed小部件 5
                     * feed小部件 中等尺寸 6
                     * feed小部件 小尺寸 7
                     * 拍照搜题 8
                     * MIUI适配 快捷搜索 9
                     * 1*1 widget,默认跳转首页 10
                     * */
                    'name': 'type',
                    'value': 'number'
                },
                {
                    'name': 'from',
                    'value': 'string'
                }
            ],
            'env': {
                'main': {
                    'search': '12.27.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/491
        {
            name: 'system.makePhoneCall',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            path: '/makePhoneCall',
            authority: 'utils',
            args: [{
                name: 'phoneNumber',
                value: 'string'
            }],
            env: {
                main: {
                    imagesearch: '13.0'
                }
            }
        }
    ];


    var ios$4 = [
        {
            // baiduboxapp://v16/utils/queryScheme?params={'scheme':'wechat'}&callback=_bdbox_js
            // http://es.baidu-int.com/#/apidetail/501
            // http://ft.baidu.com/#/scheme/utils_queryscheme
            // 判断scheme对应的应用是否在该设备安装
            name: 'system.queryScheme',
            invoke: 'iframe.bdboxURL',
            authority: 'v16',
            path: '/utils/queryScheme',
            args: [
                {name: 'scheme', value: 'string'}
            ],
            env: {
                main: {light: '10.0'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // baiduboxapp://theme/getTheme?callback=callback
        // http://es.baidu-int.com/#/apidetail/785
        {
            name: 'system.getTheme',
            invoke: 'iframe.bdboxURL',
            path: '/getTheme',
            authority: 'theme',
            args: [],
            env: {
                main: {
                    search: '11.13',
                    light: '11.13'
                },
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/471
        {
            name: 'system.getAppDuration',
            path: '/getAppDuration',
            invoke: 'iframe.bdboxURL',
            args: [],
            authority: 'utils',
            scheme: 'baiduboxapp',
            env: {
                main: {light: '10.3.0.0'}
            }
        },
        // http://es.baidu-int.com/#/apidetail/470
        // baiduboxapp://v19/utils/getClipboardData?callback=\$callback
        {
            name: 'system.getClipboardData',
            invoke: 'iframe.bdboxURL',
            path: '/getClipboardData',
            authority: 'utils',
            args: [],
            env: {
                main: {
                    search: '10.3.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/502
        {
            name: 'system.setClipboardData',
            path: '/setClipboardData',
            invoke: 'iframe.bdboxURL',
            args: [{
                name: 'data',
                value: 'string'
            }],
            authority: 'utils',
            scheme: 'baiduboxapp',
            env: {
                main: {
                    light: '10.3.0.0'
                }
            }
        }
    ];

    var systemDescription = /*#__PURE__*/Object.freeze({
        __proto__: null,
        android: android$4,
        ios: ios$4
    });

    /**
     * @file account预置描述表
     * @author chenqi14
     * @date 2019/09/26
     */

    var android$5 = [
        {
            name: 'account.login',
            invoke: 'method.bdboxURL',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'v26',
            path: '/account/login',
            args: [
                {name: 'displayName', value: 'string='},
                {name: 'app', value: 'string='},
                {name: 'loginType', value: 'string'},
                {name: 'showThirdLogin', value: 'string'},
                {name: 'loginSource', value: 'string'},
                {name: 'loginSubSource', value: 'string='},
                {name: 'normalizeAccount', value: 'string'},
                {name: 'normalizeTitle', value: 'string='},
                {name: 'normalizeSubTitle', value: 'string='},
                {name: 'ext', value: 'string='},
                {name: 'customCSS', value: 'string='}
            ],
            nextCallback: 'loginCallback',
            env: {
                main: {
                    light: '10.7.0.0',
                    search: '10.7.0.0'
                },
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/12
        // baiduboxapp://v14/account/getZid?params=$params&callback=$callback&upgrade=0
        {
            name: 'account.getZid',
            scheme: 'baiduboxapp',
            method: 'Bdbox_android_jsbridge.dispatch',
            invoke: 'method.bdboxURL',
            authority: 'v14',
            path: '/account/getZid',
            args: [
                {name: 'v', value: 'number'},
                {name: 'jsParams', value: 'string'}
            ],
            env: {
                main: {light: '9.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1156
        {
            name: 'account.getPhoneNumArgs',
            path: '/getPhoneNumArgs',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'account',
            invoke: 'method.url',
            scheme: 'baiduboxapp',
            args: [],
            env: {
                main: {
                    light: '11.24.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1173
        {
            path: '/launchLoginComponent',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'account',
            name: 'account.launchLoginComponent',
            invoke: 'method.url',
            scheme: 'baiduboxapp',
            args: [{
                comment: '是否支持游客（0：不支持；1：支持）',
                name: 'supportGuest',
                value: 'number='
            }, {
                comment: '登录按钮文案',
                name: 'buttonText',
                value: 'string='
            }, {
                comment: '登录来源',
                name: 'src',
                value: 'string'
            }, {
                comment: '弹窗业务对应的云控节点，用于控制半屏引导弹窗频次。不传代表不控制频次',
                name: 'cloudControlName',
                value: 'string='
            }, {
                comment: '弹窗实验，用于控制半屏弹窗实验。不传代表不需要实验',
                name: 'abTestName',
                value: 'string='
            }, {
                comment: '主标题',
                name: 'title',
                value: 'string='
            }, {
                comment: '登录组件类型',
                name: 'type',
                value: 'string'
            }],
            env: {
                main: {
                    light: '11.25.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1206
        {
            name: 'account.openPassPage',
            path: '/openPassPage',
            method: 'Bdbox_android_jsbridge.dispatch',
            authority: 'account',
            invoke: 'method.url',
            scheme: 'baiduboxapp',
            args: [{
                comment: 'accountCancel:账号注销；accountManage:账号管理；accountAppeal:账号申诉',
                name: 'page',
                value: 'string'
            }, {
                comment: '登录来源/调起pass页面的来源;\n帮助与反馈页面：personal_accountmanage',
                name: 'source',
                value: 'string'
            }],
            env: {
                main: {
                    light: '11.26.0.0'
                }
            }
        }
    ];


    var ios$5 = [
        // 客户端 在11.11计划废弃account.openLoginDialog。需要引导大家调用此端能力
        {
            name: 'account.login',
            invoke: 'iframe.bdboxURL',
            authority: 'v26',
            path: '/account/login',
            args: [
                {name: 'loginType', value: 'string'},
                {name: 'showThirdLogin', value: 'string'},
                {name: 'loginSource', value: 'string'},
                {name: 'ext', value: 'string='},
                {name: 'customCSS', value: 'string='},
                {name: 'normalizeAccount', value: 'string'},
                {name: 'normalizeTitle', value: 'string='},
                {name: 'normalizeSubTitle', value: 'string='},
                {name: 'loginSubSource', value: 'string='}
            ],
            nextCallback: 'loginCallback',
            env: {
                main: {
                    search: '10.7.0.0',
                    light: '10.7.0.0'
                },
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/12
        // baiduboxapp://v14/account/getZid?params=$params&callback=$callback&upgrade=0(正确)
        {
            name: 'account.getZid',
            invoke: 'iframe.bdboxURL',
            scheme: 'baiduboxapp',
            authority: 'v14',
            path: '/account/getZid',
            args: [
                {name: 'v', value: 'number'},
                {name: 'jsParams', value: 'string'}
            ],
            env: {
                main: {light: '9.3'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1156
        {
            name: 'account.getPhoneNumArgs',
            invoke: 'iframe.bdboxURL',
            path: '/getPhoneNumArgs',
            args: [],
            authority: 'account',
            scheme: 'baiduboxapp',
            env: {
                main: {
                    light: '11.24.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1173
        {
            name: 'account.launchLoginComponent',
            invoke: 'iframe.bdboxURL',
            path: '/launchLoginComponent',
            args: [{
                comment: '是否支持游客（0：不支持；1：支持）',
                name: 'supportGuest',
                value: 'number='
            }, {
                comment: '登录按钮文案',
                name: 'buttonText',
                value: 'string='
            }, {
                comment: '登录来源',
                name: 'src',
                value: 'string'
            }, {
                comment: '弹窗业务对应的云控节点，用于控制半屏引导弹窗频次。不传代表不控制频次',
                name: 'cloudControlName',
                value: 'string='
            }, {
                comment: '弹窗实验，用于控制半屏弹窗实验。不传代表不需要实验',
                name: 'abTestName',
                value: 'string='
            }, {
                comment: '主标题',
                name: 'title',
                value: 'string='
            }, {
                comment: '登录组件类型',
                name: 'type',
                value: 'string'
            }],
            authority: 'account',
            scheme: 'baiduboxapp',
            env: {
                main: {
                    light: '11.25.0.0'
                }
            }
        },
        // http://es.baidu-int.com/#/apidetail/1206
        {
            name: 'account.openPassPage',
            invoke: 'iframe.bdboxURL',
            path: '/openPassPage',
            authority: 'account',
            scheme: 'baiduboxapp',
            args: [{
                comment: 'accountCancel:账号注销；accountManage:账号管理；accountAppeal:账号申诉',
                name: 'page',
                value: 'string'
            }, {
                comment: '登录来源/调起pass页面的来源;\n帮助与反馈页面：personal_accountmanage',
                name: 'source',
                value: 'string'
            }],
            env: {
                main: {
                    light: '11.26.0.0'
                }
            }
        }
    ];


    var androidCompatibility$1 = [
        // http://cp01-invoker-3-mgt161657.epc.baidu.com:8007/invokers?id=546d74b92870cf906ddafc9b
        // 返回值 0为登录成功，1为失败
        {
            name: 'account.openLoginDialog',
            invoke: 'method.args-json+callback',
            method: 'Bdbox_android_account.loginDialog',
            args: [
                {name: 'tpl', value: 'string'},
                {name: 'third_login', value: 'string'},
                {name: 'login_type', value: 'string'}
            ],
            env: {
                main: {light: '5.5'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        }
    ];

    var iosCompatibility$1 = [
        // baiduboxapp://account?action=logindialog&params=$options&minver=5.5.0.0
        // http://cp01-invoker-3-mgt161657.epc.baidu.com:8007/invokers?id=546d74b92870cf906ddafc9b
        // 注意：此端能力的一级回调在 args 里传递（相当于二级回调了）。在外不接受一级回调了。
        {
            name: 'account.openLoginDialog',
            invoke: 'iframe.bdboxURL',
            authority: 'account',
            args: [
                {name: 'tpl', value: 'string'},
                {name: 'third_login', value: 'string'},
                {name: 'login_type', value: 'string'}
            ],
            query: {
                action: 'logindialog',
                minver: '5.5.0.0'
            },
            nextCallback: 'func',
            // 回调返回值：0为登录成功，1为失败  但是 ios切换账号失败也是0,
            env: {
                main: {light: '5.5'},
                lite: {
                    search: '2.2.2.0',
                    light: '2.2.2.0'
                },
                mission: {
                    search: '1.0.0.0',
                    light: '1.0.0.0'
                }
            }
        }
    ];

    var accountDescription = /*#__PURE__*/Object.freeze({
        __proto__: null,
        android: android$5,
        ios: ios$5,
        androidCompatibility: androidCompatibility$1,
        iosCompatibility: iosCompatibility$1
    });

    /**
     * @file 多模视觉搜索端能力描述表；
     * @author zhuxingyu01
     * @data 2021/06/25
     */

    // 输出全集描述表
    var allSet = [
        imageSearchDescription
    ];
    // 按需输出的描述表
    var customSet = [
        easyBrowserDescrioption,
        searchDescription,
        shareDescription,
        systemDescription,
        accountDescription
    ];
    // 需要输出的端能力名称
    var invokeNameList = [
        'search.open',
        'easyBrowser.open',
        'share.doShare',
        'system.makePhoneCall',
        'easyBrowser.image.save',
        'account.login'
    ];
    var moduleTypeList = ['android', 'ios', 'androidCompatibility', 'iosCompatibility'];
    var mergedDescription = {};
    allSet.forEach(function (module) {
        moduleTypeList.forEach(function (type) {
            if (module[type]) {
                !mergedDescription[type] && (mergedDescription[type] = []);
                mergedDescription[type] = mergedDescription[type].concat(module[type]);
            }
        });
    });

    customSet.forEach(function (module) {
        moduleTypeList.forEach(function (type) {
            if (module[type]) {
                !mergedDescription[type] && (mergedDescription[type] = []);
                mergedDescription[type] = mergedDescription[type]
                    .concat(module[type].filter(function (ivk) { return invokeNameList.indexOf(ivk.name) > -1; }));
            }
        });
    });

    return mergedDescription;

})));
