// @baidu/auc-base/boxLocalStorage.js
import boxx from '@baidu/auc-base/boxx';

/**
 * args
 * @param args object {data {string} key {string} action {string}}
 * @return {Promise.<resolveType, rejectType>}
 * boxLocalStorage({action: 'set', key: 'topicVideoTipsFlag' + networtType, data: '0'});
 */
export const boxLocalStorage = (args) => {
    return new Promise((resolove, rejetc) => {
        boxx.call('easyBrowser.webStorage', {
            ...args,
            success(res) {
                if (+res.status === 0) {
                    resolove(res.data);
                } else {
                    resolove('');
                }
            },
            fail(err) {
                resolove('');
            }
        });
    });
};

export const networkTips = {
    wifi: 'wifi环境自动播放视频，\n可到“我的-设置”关闭',
    mobileNet: '移动网络自动播放视频，\n可到“我的-设置”关闭',
};

export const netWorkToast = async (networtType) => {
    // 根据网络情况及是否出现过给出不同的提示
    const isFirst = await boxLocalStorage({action: 'get', key: 'topicVideoTipsFlag' + networtType, data: ''});
    if (!+isFirst) {
    // if (true) {
        boxx.call('commonUI.showToast', {
            type: '1',
            message: networkTips[networtType],
            time: '5',
        });
        boxLocalStorage({action: 'set', key: 'topicVideoTipsFlag' + networtType, data: '1'});
    }
    // 4g 调端能力给出提示框
    if (networtType !== 'wifi') {
        boxx.call('search.showMobileNetPushVideoSettingPageToast', {
            pageId: 'boxTopicVedio',
            bussiness: 'topic'
        });
    }
};
