// @baidu/auc-base/env.ts
/**
 * @file env.js
 * @author leizhihao@baidu.com
 */

import getBoxVersion from '@baidu/xbox/get-box-version';

const isSSR = typeof window !== 'object';

const ua = !isSSR ? window.navigator.userAgent : '';

export const env = {
    ua,
    isSSR,
    isMobile: !!ua.match(/mobile/i),
    // isPC: !isMobile,
    isIphone: ua.match(/(iPhone\sOS)\s([\d_]+)/),
    isIOS: ua.match(/iphone|ipod|ipad/i),
    isAndroid: ua.match(/android/i),
    isUC: ua.match(/UCBrowser/gi),
    isQQBrowser: ua.match(/MQQBrowser/gi),
    isBDBox: ua.match(/baiduboxapp/gi),
    isBaiduApp: / baiduboxapp\//i.test(ua) &&  !/ (lite|info|mission|pro) baiduboxapp/.test(ua),
    isBoxSeries: / baiduboxapp\//i.test(ua),
    isWeixin: ua.match(/micromessenger/gi), // 微信
    isMap: ua.match(/baidumap/i),
    isSafari: ua.match(/Safari/i),
    isBDbrowser: !!ua.match(/baidubrowser/i), // 百度浏览器app
    boxVersion: !isSSR && getBoxVersion(),
    isLightFrame: !!ua.match(/light/i),
    isHuawei: !!ua.match(/huawei|honor/i), // 华为浏览器
    isOppo: !!ua.match(/oppobrowser|heytapbrowser/i), // OPPO浏览器
    isXiaoMiBaiPai: !!ua.match(/knews/i), // 判断是不是小米白牌
    isMatrix: !!ua.match(/ bdapp/i),
    isTomas: !!ua.match(/ tomas/i)
};
