// @baidu/auc-base/jump.ts
import {env} from './env';
import invoke from '@baidu/xbox/na/invoke';
import {command} from '@baidu/xbox/na/android';
import {JumpProps} from './types/jumpProps';
import boxx from '@baidu/auc-base/boxx';

export const jump = (data:JumpProps) => {
    const {link, cmd, swan} = data;
    if (env.isBoxSeries || env.isXiaoMiBaiPai || env.isMatrix) {
        if (swan) {
            boxx.call('swan.open', {
                appKey: swan.appId,
                path: swan.url,
                _baiduboxapp: {
                    ext: swan.ext,
                    from: swan.from
                },
                query: swan.params
            });
        }
        else if (typeof cmd === 'string') {
            // invoke(cmd.replace(/%5Cu0026/g, '%26'));
            invoke(cmd);
        }
        else if (+cmd.mode === 2 || +cmd.mode === -1) {
            // invoke(cmd.url.replace(/%5Cu0026/g, '%26'));
            invoke(cmd.url);
        }
        else {
            if (/baiduboxapp/.test(cmd.intent)) {
                invoke(cmd.intent);
            } else {
                command(cmd.intent, cmd.mode, cmd.class, cmd.min_v);
            }
        }
    } else {
        // 小米白牌分享出去的点击事件去掉
        if (/_refluxos=a8/.test(window.location.search)) {
            return;
        }
        location.href = link;
    }
};
