// @baidu/auc-base/reflow.ts

/**
 * initReflow
 *
 * @param openBoxConfig openBox的选项
 */

/* eslint-disable fecs-camelcase */
export const initReflow = openBoxConfig => {
    return new window.OpenBox({...openBoxConfig});
};

export const initReflowByUrl = url => {
    const config = getOpenBoxConfigByUrl(url);
    return initReflow({
        ...config,
        showTip: true,
        waitTime: 300
    });
};

/* eslint-disable max-len */
export const getOpenBoxConfigByUrl = (url) => {
    let encodedUrl = encodeURIComponent(url);

    return {
        iosScheme: `baiduboxapp://easybrowse?opentype=1&sfrom=feed&newbrowser=1&openurl=${encodedUrl}&rbtnstyle=0&isBdboxShare=1&isla=0`,
        androidCommand: {
            'mode':'0',
            'intent':'intent:#Intent;S.toolbaricons=%7B%22toolids%22%3A%5B%223%22%5D%7D;S.menumode=2;S.create_menu_key=false;S.bdsb_light_start_url=' + encodedUrl + ';end',
            'class':'com.baidu.searchbox.xsearch.UserSubscribeCenterActivity',
            'min_v':'16787968'
        },
        channel: '1019151h',
        notUseIdm: true,
        from: 'feed_midbox', //调起来源,统计使用
        sfrom: 'feed_midbox', //调起来源,统计使用
    };
};

export const openReflow = config => {
    let openbox = initReflow(config);
    openbox.open(config);
};

export const openReflowByUrl = (url) => {
    const config = getOpenBoxConfigByUrl(url);
    openReflow(config);
};
