// @baidu/nano/es/tabs/panel.js
/**
 * @file 组件 panel
 * @author baozhixin <baozhixin@baidu.com>
 */

import {Component, DataTypes} from 'san';

const prefixCls = 's-tab-panel';

export default class Panel extends Component {
    // eslint-disable-next-line
    static template = `
        <div class="{{panelClass}}">
            <slot/>
        </div>
    `

    static dataTypes = {
        active: DataTypes.bool,
        disabled: DataTypes.bool,
        name: DataTypes.string
    };

    static computed = {
        panelClass() {
            const active = this.data.get('active');

            return [
                prefixCls,
                active && `${prefixCls}-active`
            ].filter(c => c);
        }
    };

    get isTabPanel() {
        return true;
    }

    initData() {
        return {
            active: false,
            disabled: false,
            name: ''
        };
    }

    attached() {
        this.dispatch('s-tabs:updateTabPanel');
    }

    // updated() {
    //     this.dispatch('s-tabs:updateTabPanel');
    // }

    detached() {
        this.dispatch('s-tabs:updateTabPanel');
    }
}
