// @baidu/nano/es/tabs/tab.js
/**
 * @file 组件 tab
 * @author baozhixin <baozhixin@baidu.com>
 */

import {Component, DataTypes} from 'san';

const prefixCls = 's-tab';

export default class Tab extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{tabClass}}">
            <span s-ref="node">{{name}}</span>
        </div>
    `;

    static dataTypes = {
        active: DataTypes.bool,
        disabled: DataTypes.bool,
        name: DataTypes.string,
        last: DataTypes.bool
    };

    static computed = {
        tabClass() {
            const active = this.data.get('active');
            const disabled = this.data.get('disabled');
            return [
                prefixCls,
                active && `${prefixCls}-active`,
                disabled && `${prefixCls}-disabled`
            ].filter(c => c);
        }
    };

    get offset() {
        let ret = {};
        const $node = this.ref('node');
        if ($node) {
            ret.width = $node.offsetWidth;
            ret.left = $node.offsetLeft;
        }

        return ret;
    }

    attached() {
        const {index, last} = this.data.get();
        const owner = this.owner;
        owner.$tabs[index] = this;

        // 最后一个tab挂载之后的下一个时钟处理
        if (last) {
            this.owner.updateCursorStyle();
            this.owner.scrollIntoView(true);
        }

        this.watch('active', active => {
            if (active) {
                this.owner.updateCursorStyle(this.offset);
            }
        });
    }

    detached() {
        const index = this.data.get('index');
        this.owner.$tabs[index] = null;
    }
}
