// @baidu/auc-image-header/index.ts
/**
 * @file index.ts
 * @author gaohong01
 */

import {Component} from 'san';
import styles from './style.module.less';
import {AucImageHeaderData} from './types';
import {WujiImage} from '@baidu/wuji-san';
import {hex2rgb} from '@baidu/auc-base/utils';

// <div 
//             class="{{styles.mask}}"
//             style="{{maskStyle}}">
//         </div>
// <div class="{{styles.defaultBlock}}" style="{{defaultBlock}}"></div>

export class AucImageHeader extends Component {
    static template:string = /* html */ `
    <div class="{{styles.aucImageHeaderWrapper}} {{imgType === 3 ? styles.aucImageHeader21 : ''}}">
        <wuji-image
            lazy="{{true}}"
            lazyOffset="{{400}}"
            ratio="{{imgType === 3 ? 'auto' : '16:9'}}"
            src="{{src}}"
            border-radius="0"
        ></wuji-image>
    </div>
    `;
    static components:object = {
        'wuji-image': WujiImage
    };

    static computed:object = {
        maskStyle() {
            const imgBg = this.data.get('headImgColor');
            let imgBgObj = hex2rgb(imgBg);
            return imgBg ? `background:linear-gradient(to bottom, ${imgBgObj.rgb0}, ${imgBgObj.rgb});` : '';
        },
        defaultBlock() {
            const imgBg = this.data.get('headImgColor');
            let imgBgObj = hex2rgb(imgBg);
            return imgBg ? `background-color: ${imgBgObj.rgb};` : '';
        }
    }
    initData(): AucImageHeaderData {
        return {
            styles,
            src: '',
            imgType: 2
        };
    }

    attached(): void {
    }
}

export default AucImageHeader;
