// @baidu/xbox-native/es5/hybrid/loadFile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _detect = require("../detect");

var _utils = require("../utils");

var _register = _interopRequireDefault(require("../register"));

var _index = _interopRequireDefault(require("./index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file loadFile
 * @module data
 * @author panming <panming0@baidu.com>
 */
(0, _register.default)('feed/loadFile', {
  v: 6,
  versionRange: ['8.5'],
  liteVersionRange: ['2.3'],
  missionVersionRange: ['1.0'],
  androidinfoVersionRange: ['1.1'],
  notSupportFrame: {
    android: ['search']
  },
  filter: function filter() {
    return {
      module: 'feed',
      action: 'loadFile'
    };
  },
  thunk: function thunk(res) {
    if ((0, _detect.version_compare)((0, _detect.getBoxVersion)(), '9.2.5') >= 0) {
      return res;
    } else {
      return {
        status: 0,
        data: res
      };
    }
  }
});
/**
 * loadFile
 * @function loadFile
 */

var _default = function _default(data) {
  return (0, _index.default)((0, _utils.getParams)('feed-loadFile', data));
};

exports.default = _default;
