// @baidu/nano/lib/scroll-list/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("./style/index.less");

var _san = require("san");

var _error = _interopRequireDefault(require("../error"));

var _flash = _interopRequireDefault(require("../flash"));

var _loader = _interopRequireDefault(require("../loader"));

require("../assets/svg/error-network.svg");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 's-scroll-list';

var ScrollList =
/*#__PURE__*/
function (_Component) {
  _inherits(ScrollList, _Component);

  function ScrollList() {
    _classCallCheck(this, ScrollList);

    return _possibleConstructorReturn(this, _getPrototypeOf(ScrollList).apply(this, arguments));
  }

  _createClass(ScrollList, [{
    key: "initData",
    // eslint-disable-next-line
    value: function initData() {
      return {
        isFirst: true,
        isLoading: false,
        isError: false,
        isEmpty: false,
        hasMore: true,
        showLoader: true,
        loadingConfig: {
          layout: 'full'
        },
        errorConfig: {
          layout: 'full',
          type: 'network',
          text: '网络不给力，请稍后重试',
          button: {
            text: '重新加载'
          }
        }
      };
    }
  }, {
    key: "attached",
    value: function attached() {
      var _this = this;

      this.data.get('isFirst') && this.fire('fetch'); // 首页加载模式时第一次进入页面要拉取数据
      // 下拉获取新的数据

      window.addEventListener('scroll', function () {
        if (_this.isBottom && _this.data.get('hasMore')) {
          _this.fire('fetch');
        }
      });
    } // 是否滚动到页面底部

  }, {
    key: "clickErrorLoader",
    value: function clickErrorLoader() {
      this.fire('error-click');
    }
  }, {
    key: "isBottom",
    get: function get() {
      var _document$documentEle = document.documentElement,
          clientHeight = _document$documentEle.clientHeight,
          scrollHeight = _document$documentEle.scrollHeight;
      var scrollTop = document.documentElement.scrollTop || document.body.scrollTop || 0;
      return clientHeight + scrollTop + 20 > scrollHeight;
    }
  }]);

  return ScrollList;
}(_san.Component);

exports.default = ScrollList;

_defineProperty(ScrollList, "template",
/*html*/
"\n        <div class=\"".concat(prefixCls, "\">\n            <!-- \u9996\u5C4F\u52A0\u8F7D\u6A21\u5F0F\u4E0B\uFF0C\u5B58\u5728 flash-loading\u3001error \u53CA empty \u6001 -->\n            <template s-if=\"isFirst\">\n                <s-flash\n                    s-if=\"isLoading\"\n                    s-bind=\"{{loadingConfig}}\"\n                />\n                <s-error\n                    s-elif=\"isError\"\n                    s-bind=\"{{errorConfig}}\"\n                    on-button-click=\"clickErrorLoader\"\n                />\n                <s-error\n                    s-elif=\"isEmpty\"\n                    s-bind=\"{{emptyConfig}}\"\n                />\n            </template>\n            <!-- \u6B63\u5E38\u6A21\u5F0F\u4E0B\uFF0C\u53EA\u6709 slot \u4F20\u5165\u7684\u5C55\u793A\u5217\u8868\u548C\u5E95\u90E8\u7684 loader bar -->\n            <template s-else>\n                <slot />\n                <s-loader\n                    s-if=\"{{showLoader}}\"\n                    state=\"{{loaderState}}\"\n                    s-bind=\"{{loaderConfig}}\"\n                    on-error-click=\"clickErrorLoader\"\n                />\n            </template>\n        </div>\n    "));

_defineProperty(ScrollList, "components", {
  's-error': _error.default,
  's-flash': _flash.default,
  's-loader': _loader.default
});

_defineProperty(ScrollList, "dataTypes", {
  isFirst: _san.DataTypes.bool,
  isLoading: _san.DataTypes.bool,
  loadingConfig: _san.DataTypes.object,
  isError: _san.DataTypes.bool,
  errorConfig: _san.DataTypes.object,
  isEmpty: _san.DataTypes.bool,
  emptyConfig: _san.DataTypes.object,
  showLoader: _san.DataTypes.bool,
  loaderConfig: _san.DataTypes.object,
  hasMore: _san.DataTypes.bool
});

_defineProperty(ScrollList, "computed", {
  loaderState: function loaderState() {
    if (this.data.get('isLoading')) {
      return 1;
    }

    if (!this.data.get('hasMore')) {
      return 2;
    }

    if (this.data.get('isError')) {
      return 3;
    }

    return 0;
  }
});
