// @baidu/nano/es/image/index.js
/**
 * @file 组件 image 图像
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';

const prefixCls = 's-image';
const sizeMap = {};

// 图片宽高比: http://bdbox.baidu.com/ceued-docs/design/component/image/
const sizeType = [
    '1:1',
    '2:3', // extra added
    '3:1',
    '3:2',
    '3:4',
    '4:3',
    '5:2',
    '16:9',
    '21:4',
    'full'
];

// 默认宽高比
const defaultSize = sizeType[0];
sizeType.forEach(size => {
    if ('full' === size) {
        sizeMap[size] = 0;
    }
    else {
        const sizeVal = size.split(':');
        sizeMap[size] = sizeVal[1] / sizeVal[0];
    }
});

export default class Image extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{wrapClass}}"
            on-touchstart="handleTouchEvent($event, cancelBubble)"
            on-touchend="handleTouchEvent($event, cancelBubble)"
            on-touchcancel="handleTouchEvent($event, cancelBubble)"
        >
            <div class="${prefixCls}-default"/>
            <div class="${prefixCls}" style="{{imageStyle}}"><slot/></div>
        </div>
    `;

    static dataTypes = {
        cancelBubble: DataTypes.bool,
        size: DataTypes.oneOf(sizeType),
        hairline: DataTypes.bool,
        url: DataTypes.string
    };

    static computed = {
        wrapClass() {
            const hairline = this.data.get('hairline');
            return [
                `${prefixCls}-wrapper`,
                hairline && `${prefixCls}-hairline`
            ].filter(c => c);
        },

        imageStyle() {
            const size = this.data.get('size');
            const url = this.data.get('url');
            const isFull = 'full' === size;
            return {
                'padding-top': `${sizeMap[size] * 100}%`,
                'height': isFull ? '100%' : 'auto',
                'background-image': url ? `url(${url})` : ''
            };
        }
    };

    initData() {
        return {
            hairline: true,
            size: defaultSize,
            url: '',
            cancelBubble: true
        };
    }

    handleTouchEvent(e, cancelBubble) {
        !cancelBubble && e.stopPropagation();
    }
}

Image.sizeType = sizeType;
