// @baidu/xbox-native/ui/closeLoading.js
/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {close} from './index';
import {version_compare as versionCompare, getBoxVersion, isMatrix, isIOS} from '../detect';
import {prompt as androidCmd} from '../android';
import iosInvoke from '../ios';

const boxVersion = getBoxVersion();

register('loading/close', {
    androidlite: {
        versionRange: ['2.2.2'],
        filter() {
            return {
                module: 'Bdbox_android_net',
                action: 'hideLoadingView'
            };
        }
    },
    androidmission: {
        versionRange: ['1.0'],
        filter() {
            return {
                module: 'Bdbox_android_net',
                action: 'hideLoadingView'
            };
        }
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidinfo: {
        versionRange: ['1.0'],
        filter() {
            return {
                module: 'bd_searchbox_interface',
                action: 'progressCompleted'
            };
        }
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        versionRange: ['5.1'],
        hasCB: false,
        filter() {
            return {
                module: 'speedProgressCompleted',
                action: '',
                extData: {
                    minver: '5.1.0.0'
                }
            };
        }
    },
    android: {
        versionRange: ['4.9'],
        supportFrame: ['hybrid'],
        hasCB: false,
        filter() {
            if (versionCompare(boxVersion, '8.2') >= 0) {
                return {
                    module: 'Bdbox_android_net',
                    action: 'hideLoadingView'
                };
            } else {
                return {
                    module: 'bd_searchbox_interface',
                    action: 'progressCompleted'
                };
            }
        }
    }
});
/**
 * @function closeLoading
 * @param {Object} [params={}] - 参数
 */
export default (params = {}) => {
    if (isMatrix('baiduboxvision')) {
        isIOS() ? iosInvoke('baiduboxapp://speedProgressCompleted?minver=1.0.0.0') : androidCmd('Bdbox_android_net', 'hideLoadingView', []);
        return Promise.resolve();
    }
    return close(getParams('loading', params));
};
