// @baidu/xbox-native/ui/openShare.js

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {open} from './index';
import {isMatrix, isIOS} from '../detect';
import stringify from '../stringify';
import {prompt as androidCmd} from '../android';
import iosInvoke from '../ios';

const dataFormat = data => {
    let d = {
        imageUrl: data.iconUrl,
        title: data.title,
        content: data.content,
        linkUrl: data.linkUrl,
        mediaType: data.mediaType,
        channel: data.pannel,
        shareSuccessCB: data.shareSuccessCB || 'console.log',
        shareErrorCB: data.shareErrorCB || 'console.log'
    };
    if (data.type === 'audio') {
        d.type = '2';
    } else if (data.type === 'image') {
        d.type = '3';
    } else if (data.type === 'video') {
        d.type = '4';
    } else {
        d.type = '1';
    }
    return d;
};
register('utils/openShareBox', {
    androidlite: {
        versionRange: ['2.2.2']
    },
    androidinfo: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    ios: {
        hasCB: false,
        versionRange: ['5.3.5'],
        filter(module, action, data) {
            return {
                module: 'callShare',
                action: '',
                paramName: 'options',
                extData: {
                    minver: '5.3.5.0'
                }
            };
        }
    },
    android: {
        versionRange: ['5.3.5'],
        hasCB: false,
        filter(module, action, data) {
            return {
                module: 'Bdbox_android_utils',
                action: 'callShare',
                data: dataFormat(data)
            };
        }
    }
});

/**
 * 分享面板
 * @function openShare
 * @param {Object} params 端能力参数
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    const defaultConf = {
        mediaType: 'all',
        type: 'url',
        shareSuccessCB: '',
        shareErrorCB: ''
    };
    /* eslint-disable max-len */
    if (isMatrix('baiduboxvision')) {
        return isIOS() ? iosInvoke('baiduboxapp://callShare?minver=1.0.0.0&options='
            + stringify(Object.assign({}, defaultConf, params), true))
            : androidCmd('Bdbox_android_utils', 'callShare', [stringify(Object.assign({}, defaultConf, dataFormat(params))), '', '']);
    }
    /* eslint-enable max-len */
    return open(
        getParams('utils-shareBox', params, {
            data: defaultConf
        })
    );
};
