// @baidu/xbox-native/ui/openToast.js
/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {open} from './index';
import stringify from '../stringify';
import {invokeP} from '../invoke';
import {isMatrix} from '../detect';

register(
    'toast/open',
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        ios: {
            versionRange: ['6.0'],
            hasCB: false,
            filter: (module, action, data) => {
                return {
                    module: 'utils',
                    action: 'toast',
                    extData: {
                        minver: '6.0.0.0'
                    },
                    data: data.message,
                    paramName: 'string'
                };
            }
        },
        android: {
            versionRange: ['4.0'],
            hasCB: false,
            filter: (module, action, data) => {
                return {
                    module: 'Bdbox_android_utils',
                    action: 'toast',
                    data: data.message,
                    paramName: 'string'
                };
            }
        }
    },
    {
        v: 11,
        hasCB: true,
        versionRange: ['9.1'],
        androidliteVersionRange: ['2.2.0.10'],
        iosinfoVersionRange: ['1.3.1.10'],
        androidmissionVersionRange: ['1.0'],
        iosmissionVersionRange: ['1.0'],
        filter: data => {
            if (data.style) {
                let style = data.style;
                for (let key in style) {
                    if (key !== 'style') {
                        data[key] = style[key];
                    }
                }
                if (style.style) {
                    data.style = style.style;
                }
            }
            return {
                action: 'showToast',
                module: 'utils',
                data: {
                    params: stringify(data, true)
                }
            };
        }
    }
);

/**
 * 展示toast
 * @function openToast
 * @param {Object} params 端能力参数，参数参看 http://ft.baidu.com/#/scheme/utils_showtoast
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    if (isMatrix('baiduboxvision')) {
        return invokeP('v11/utils/showToast', {params});
    }
    return open(
        getParams('toast', params, {
            data: {
                type: 1
            }
        })
    );
};
