// @baidu/xbox-native/ui/openCommentBox.js
/**
 * @file Created on 2018-12-5
 * @module ui
 * @author panming <panming@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {open} from './index';
import getBoxVersion from '@baidu/xbox/get-box-version';
import versionCompare from '@baidu/xbox/version_compare';
const boxVersion = getBoxVersion();

register(
    'comment/openCommentBox',
    {
        lite: {
            versionRange: ['2.2.2'],
            filter: function (module, action, data) {
                var params = data;
                if (versionCompare(boxVersion, '2.3') < 0) {
                    params.data && delete params.data.placeholder;
                }
                return {
                    module: 'Bdbox_android_comment',
                    action: 'showCommentBox',
                    data: params
                };
            }
        },
        androidinfo: {
            versionRange: ['1.0'],
            filter: function (module, action, data) {
                var params = data;
                if (versionCompare(boxVersion, '1.1') < 0) {
                    params.data && delete params.data.placeholder;
                }
                return {
                    module: 'Bdbox_android_comment',
                    action: 'showCommentBox',
                    data: params
                };
            }
        },
        ios: {
            versionRange: ['8.0'],
            supportFrame: ['tags', 'hybrid', 'video', 'h5'],
            filter: function (module, action, data) {
                var params = data;
                if (versionCompare(boxVersion, '8.2') < 0) {
                    params.data && delete params.data.rename;
                }
                if (versionCompare(boxVersion, '8.5') < 0) {
                    params.data && delete params.data.placeholder;
                }
                return {
                    module: 'comment',
                    action: 'showCommentBox',
                    extData: {
                        'minver': '8.0.0.0'
                    },
                    data: params
                };
            }
        },
        android: {
            supportFrame: ['tags', 'hybrid', 'video', 'h5'],
            versionRange: ['8.0'],
            filter: function (module, action, data) {
                var params = data;
                if (versionCompare(boxVersion, '8.2') < 0) {
                    params.data && delete params.data.rename;
                }
                if (versionCompare(boxVersion, '8.5') < 0) {
                    params.data && delete params.data.placeholder;
                }
                return {
                    module: 'Bdbox_android_comment',
                    action: 'showCommentBox',
                    data: params
                };
            }
        }
    }, {
        v: 4,
        hasCB: true,
        versionRange: ['8.3'],
        filter: function () {
            return {
                module: 'comment',
                action: 'showBox'
            };
        }
    }
);
/**
 * @function openCommentBox
 * @param {Object} [params={}] - 参数
 */
export default (params = {}) => {
    open(getParams('comment-commentBox', params ));
}
