// @baidu/xbox-native/ui/openNetworkErrorView.js
/**
 * @file Created on 2018-12-5
 * @module ui
 * @author panming <panming@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {open} from './index';
import {prompt as androidCmd} from '../android';
import iosInvoke from '../ios';
import {isMatrix, isIOS} from '../detect';

register('net/openNetworkErrorView', {
    lite: {
        versionRange: ['2.2.2'],
        supportFrame: ['tags']
    },
    androidinfo: {
        versionRange: ['1.0'],
        supportFrame: ['tags']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.3']
    },
    ios: {
        versionRange: ['8.2.3'],
        supportFrame: ['h5', 'hybrid', 'video', 'tags'],
        filter: function () {
            return {
                module: 'net',
                action: 'showNetWorkErrorView'
            };
        }
    },
    android: {
        versionRange: ['8.2'],
        hasCB: false,
        supportFrame: ['hybrid'],
        filter: function () {
            return {
                module: 'Bdbox_android_net',
                action: 'showNetWorkErrorView'
            };
        }
    }
});
/**
 * @function closeToast
 * @param {Object} [params={}] - 参数
 */
export default (params = {}) => {
    if (isMatrix('baiduboxvision')) {
        if (isIOS()) {
            return iosInvoke('baiduboxapp://net?action=showNetWorkErrorView');
        }
        return androidCmd('Bdbox_android_net', 'showNetWorkErrorView', []);
    }

    return open(getParams('net-networkErrorView', params));
};
