// @baidu/xbox-native/ui/openProfile.js
/**
 * @file 调起个人主页
 * @module ui
 * @author panming <panming@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {open} from './index';

register('profile/open', {
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidlite: {
        versionRange: ['2.2.2']
    },
    androidinfo: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        hasCB: false,
        versionRange: ['7.0'],
        filter(module, action, data) {
            return {
                module: 'account',
                action: 'profile',
                extData: {
                    minver: '7.0.0.0'
                },
                data
            };
        }
    },
    android: {
        hasCB: false,
        filter(module, action, data) {
            return {
                module: 'Bdbox_android_account',
                action: 'profile',
                data
            };
        }
    }
});
/**
 * @function openProfile
 * @param {object} [data={}] - openProfile 参数
 */

export default (params = {}) => open(getParams('profile', params));
