// @baidu/xbox-native/ui/openLoginDialog.js

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */


import register from '../register';
import {getParams} from '../utils';
import stringify from '../stringify';

import {open} from './index';

register(
    'account/openLoginDialog',
    {
        v: 26,
        versionRange: ['10.7'],
        iosinfoVersionRange: ['3.5'],
        androidinfoVersionRange: ['3.5'],
        iosmissionVersionRange: ['1.0'],
        androidmissionVersionRange: ['1.0'],
        iosliteVersionRange: ['3.5'],
        androidliteVersionRange: ['3.5'],
        filter(data) {
            if (data.third_login) {
                data.showThirdLogin = data.third_login;
                delete data.third_login;
            }
            if (data.login_type) {
                data.loginType = data.login_type;
                delete data.login_type;
            }
            if (data.func) {
                data.loginCallback = data.func;
                delete data.func;
            }
            if (data.tpl) {
                data.loginSource = data.tpl;
                delete data.tpl;
            }

            return {
                module: 'account',
                action: 'login',
                data: {
                    params: stringify(data, true)
                }
            };
        }
    },
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        ios: {
            versionRange: ['5.5'],
            filter() {
                return {
                    module: 'account',
                    action: 'logindialog',
                    extData: {
                        minver: '5.5.0.0'
                    }
                };
            },
            thunk(res) {
                return {
                    status: res,
                    message: ''
                };
            }
        },
        android: {
            versionRange: ['5.5'],
            filter() {
                return {
                    module: 'Bdbox_android_account',
                    action: 'loginDialog'
                };
            },
            thunk(res) {
                return {
                    status: res,
                    message: ''
                };
            }
        }
    }
);

/**
 * 展示登录面板
 * @function openLoginDialog
 * @param {Object} params 端能力参数，参数参看 http://ft.baidu.com/#/scheme/utils_showtoast
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    return open(getParams('account-loginDialog', params));
};
