// @baidu/xbox-native/ui/openPassAssist.js
/**
 * @file open pass sdk - created on 2019-4-1
 * @module ui
 * @author jiaowenhua <jiaowenhua@baidu.com>
 */

import register from '../register';
import {open} from './index';
import {getParams, decode} from '../utils';
import getGlobalFunc from '@baidu/xbox/get-global-func';

register('account/openPassAssist', {
    v: 31,
    versionRange: ['10.10'],
    filter: data => {
        return {
            module: 'account',
            action: 'passAssist'
        };
    }
});

/**
 * passNA版调用，目前只支持人脸
 *
 * @function openPassAssist
 * @param {Object} params 端能力参数，具体值参考 http://ft.baidu.com/#/scheme/account_passassist
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    return new Promise((resolve, reject) => {
        params.passCallback = getGlobalFunc(data => {
            data = JSON.parse(decode(data));
            let dataInfo = data.data || {};
            if (dataInfo.result) {
                dataInfo = dataInfo.result || {};
            }
            if (+data.status === 0 && +dataInfo.retCode === 0) {
                resolve({
                    errno: 0
                });
            } else {
                reject({
                    errno: dataInfo.retCode,
                    errmsg: dataInfo.retMsg
                });
            }
        });
        open(getParams('account-passAssist', params));
    });
};
