// @baidu/xbox-native/ui/openUGCPublish.js
/**
 * @file Created on Thu Nov 15 2018
 * @module ui/openUGCPublish
 * @author panming <panming@baidu.com>
 */

import register from '../register';
import {
    getParams,
    urlToSchema
} from '../utils';
import {open} from './index';

register('ugc/openPublish', {
    v: 6,
    hasCB: true,
    versionRange: ['8.5'],
    filter: data => {
        if (data['topic_pageurl']) {
            const style = encodeURIComponent(JSON.stringify({'menumode': '2', 'showtoolbar': '1'}));
            const url = encodeURIComponent(data['topic_pageurl']);
            data['topic_pageurl'] = 'baiduboxapp://v1/easybrowse/open?newbrowser=1&style=' + style + '&url=' + url;
        }
        // 兼容历史处理
        const urlToSchemaArr = ((data.conf && data.conf.urlToSchema) || []).filter(item => item !== 'topic_pageurl');
        urlToSchemaArr.forEach(function (item) {
            data[item] = urlToSchema(data[item]);
        });
        return {
            module: 'ugc',
            action: 'publish',
            data: {
                params: encodeURIComponent(JSON.stringify(data))
            }
        };
    }
});
/**
 * @function openProfile
 * @param {object} [data={}] - openProfile 参数
 */

export default (params = {}) => open(getParams('ugc-publish', params));
