// @baidu/xbox-native/ui/openChat.js
/**
 * @file Created on 2018-12-5
 * @module ui
 * @author panming <panming@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {open} from './index';
import stringify from '../stringify';
import {invokeP} from '../invoke';
import {isMatrix} from '../detect';

register(
    'chat/open',
    {
        ios: {
            versionRange: ['8.5'],
            filter(module, action, data) {
                /* eslint-disable*/
                var str = 'source_app=baiduboxapp&referer=myduzhan&paid=';
                var d = {
                    plugin_id: 'box.plugin.cambrian',
                    args: str + data.paid + '&title=' + data.title + '&from=' + data.from + '&method_name=invokeChat'
                };
                /* eslint-enable*/

                return {
                    module: 'invokePlug',
                    action: 'open',
                    data: JSON.stringify(d),
                    extData: {
                        minver: '8.5.0.0'
                    }
                };
            },
            thunk(res) {
                return {
                    status: res.result === 0 ? 0 : parseInt(res.result, 10),
                    message: ''
                };
            }
        },
        android: {
            versionRange: ['8.5'],
            filter(module, action, data) {
                /* eslint-disable*/
                var d = {
                    package_name: 'com.baidu.cambrian',
                    method_name: 'invokeChat',
                    by_user: 1,
                    params: {
                        paid: data.paid,
                        title: data.title,
                        from: data.from
                    }
                };
                /* eslint-enable*/

                return {
                    module: 'Bdbox_android_plugin',
                    action: 'invokePlugin',
                    data: JSON.stringify(d)
                };
            },
            thunk(res) {
                return {
                    status: res === 0 ? 0 : parseInt(res, 10),
                    message: ''
                };
            }
        }
    },
    {
        v: 9,
        versionRange: ['9.0.0'],
        filter: data => {
            return {
                module: 'account/profile',
                action: 'chat',

                data: {
                    params: stringify(data, true),
                    oauthType: 'h5api'
                }
            };
        },
        thunk(res) {
            return {
                status: res === 0 ? 0 : parseInt(res, 10),
                message: ''
            };
        }
    }
);
/**
 * @function openChat
 * @param {Object} [params={}] - 参数
 */
export default (params = {}) => {
    if (isMatrix('baiduboxvision')) {
        return invokeP('v9/account/profile/chat', {
            params: stringify(params, true),
            oauthType: 'h5api'
        });
    }
    return open(getParams('chat', params));
};
