// @baidu/xbox-native/ui/openImage.js
/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {version_compare as versionCompare, getBoxVersion, isMatrix, isIOS} from '../detect';
import {prompt as androidCmd} from '../android';
import iosInvoke from '../ios';

import {open} from './index';
import stringify from '@baidu/xbox/stringify';
const boxVersion = getBoxVersion();

register('utils/openImage', {
    androidlite: {
        versionRange: ['2.2.2'],
        filter(module, action, data) {
            data.type = '0';
            return {
                module: 'Bdbox_android_utils',
                action: 'lightImage',
                data: data
            };
        }
    },
    androidmission: {
        versionRange: ['1.0'],
        filter(module, action, data) {
            data.type = '0';
            return {
                module: 'Bdbox_android_utils',
                action: 'lightImage',
                data: data
            };
        }
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidinfo: {
        versionRange: ['1.0'],
        filter(module, action, data) {
            data.type = '0';
            return {
                module: 'Bdbox_android_utils',
                action: 'lightImage',
                data: data
            };
        }
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        versionRange: ['6.2'],
        filter(module, action, data) {
            data.type = '0';
            data.index = data.index || 0;
            return {
                module: 'utils',
                action: 'image',
                data: data
            };
        },
        thunk(res) {
            return {
                status: 0,
                message: '',
                data: ''
            };
        }
    },
    android: {
        versionRange: ['6.1'],
        filter(module, action, data) {
            data.type = '0';
            if (versionCompare(boxVersion, '7.4') >= 0) {
                return {
                    module: 'Bdbox_android_utils',
                    action: 'lightImage',
                    data: data
                };
            } else {
                return {
                    module: 'Bdbox_android_utils',
                    action: 'image',
                    data: data
                };
            }
        },
        thunk() {
            return {
                status: 0,
                message: ''
            };
        }
    }
});

/**
 * 大图浏览器
 * @function openImage
 * @param {Object} params 端能力参数
 * @return {Promise} 调用的返回
 */
export default (params = {}) => {
    if (isMatrix('baiduboxvision')) {
        params.type = '0';
        return isIOS() ? iosInvoke(`baiduboxapp://utils?action=image&params=${stringify(params, true)}`)
            : androidCmd('Bdbox_android_utils', 'lightImage', [stringify(params), 'syncCallback']);
    }
    return open(getParams('utils-image', params));
};
