// @baidu/xbox-native/log/ubc.js
/**
 * @file Created on Thu Nov 15 2018
 * @module log
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import log from './index';
register('utils/ubcReport', {
    androidlite: {
        versionRange: ['2.2.2']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidinfo: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        hasCB: false,
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'utils',
                action: 'ubcReport',
                extData: {
                    minver: '7.3.0.0'
                }
            };
        }
    },
    android: {
        hasCB: false,
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'Bdbox_android_utils',
                action: 'ubcEvent',
                extData: {
                    min_v: '16789504'
                }
            };
        }
    }
});
/**
 * ubcreport
 * @function ubc
 * @param {object} [data={}] - ubc 参数
 */
export default (data = {}) => {
    return log(
        getParams('ubcReport', {
            data,
            abtest: data.abtest
        })
    );
};
