// @baidu/xbox/es5/na/android/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prompt = prompt;
exports.command = command;
exports.syncInvoke = syncInvoke;
exports.default = void 0;

var _stringify = _interopRequireDefault(require("../../stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

/**
 * androidInvoke传入schema打开框架
 * @function default
 * @param {String} schema 协议
 * @return {Object} 发起schema请求
 */
var _default = function _default(schema) {
  if ((typeof window === "undefined" ? "undefined" : _typeof(window)) !== 'object') {
    return;
  }

  var androidJsBridge = window.Bdbox_android_jsbridge;

  if (androidJsBridge && androidJsBridge.dispatch) {
    return androidJsBridge.dispatch(schema);
  } else {
    return prompt('Bdbox_android_jsbridge', 'dispatch', schema);
  }
};
/**
 * 老版本的调起，prompt实现
 * @function prompt
 * @param {String} obj - 类似 Bdbox_android_utils
 * @param {String} func - 类似 command
 * @param {String} args 参数数组
 */


exports.default = _default;

function prompt(obj, func, args) {
  if ((typeof window === "undefined" ? "undefined" : _typeof(window)) === 'object' && typeof window.prompt === 'function') {
    if ({}.toString.call(args).slice(-6, -1) !== 'Array') {
      args = [args];
    }

    var data = (0, _stringify.default)({
      obj: obj,
      func: func,
      args: args
    });
    return window.prompt("BdboxApp:".concat(data));
  }
}
/**
 * 安卓command调起，使用 Bdbox_android_utils.command
 * @function command
 * @param  {String} intent 包含发送给客户端的参数
 * @param  {Number} [mode=0]   0代表调起框界面；1代表调起第三方界面
 * @param  {String} [comClass='']  调起客户端具体activity，缺省
 * @param  {String} [minV='']   调起的最小版本，可缺省
 * @return {Object}         返回对象格式的json
 */


function command(intent) {
  var mode = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  var comClass = arguments.length > 2 ? arguments[2] : undefined;
  var minV = arguments.length > 3 ? arguments[3] : undefined;
  var commParams = {
    mode: mode,
    intent: intent
  };

  if (typeof comClass !== 'undefined') {
    commParams['class'] = comClass;
  }

  if (typeof minV !== 'undefined') {
    // eslint-disable-next-line
    commParams.min_v = minV;
  }

  return prompt('Bdbox_android_utils', 'command', (0, _stringify.default)(commParams));
}
/**
 * 同步invoke， Bdbox_aiapps_jsbridge
 * @function syncInvoke
 * @param {String} scheme - scheme
 */


function syncInvoke(scheme) {
  if ((typeof window === "undefined" ? "undefined" : _typeof(window)) !== 'object') {
    return;
  }

  var androidJsBridge = window.Bdbox_aiapps_jsbridge;

  if (androidJsBridge && androidJsBridge.dispatch) {
    return androidJsBridge.dispatch(scheme);
  }

  return;
}
