// boxjs/dist/thirdParty/bdbox/android/invokeInfoApp.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /**
   * @file static/bdbox/android/invokeInfoApp.js
   * @author wangfupeng
  */

  /**
   * 调用资讯版端能力
   *
   * @param  {string} name android提供的命名空间
   * @param  {string} func android命名空间下的方法名
   * @param  {Array} args 接受参数
   * @return {Object}      返回对象格式的json
   * @version  $Id: invokeInfoApp.js 2017-08-17 wangfupeng
   */


  function invokeInfoApp(name, func, args) {
    if (args && !$.isArray(args)) {
      args = Array.prototype.slice.call(arguments, 0).slice(2);
    }

    if (!$.$isInfoBox() || !$.$isAndroid()) {
      return {
        error: 201
      };
    }

    var back = execCaller(name, func, args);
    back = back ? JSON.parse(back) : {};
    back.__from = 'infoApp';
    return back;
  }

  function execCaller(nameSpace, funcName, funcArgs) {
    // 保证要有[]
    var caller = {
      obj: nameSpace,
      func: funcName,
      args: funcArgs ? funcArgs : []
    };

    try {
      return window.prompt('BdboxApp:' + JSON.stringify(caller));
    } catch (e) {
      return {
        error: 201
      };
    }
  }

  return invokeInfoApp;
}();
