// @baidu/wuji-san/lib/wuji-quick-evaluate/star.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

require("./style/star.less");

var _san = require("san");

var _wujiIcon = _interopRequireDefault(require("../wuji-icon"));

require("@baidu/wuji-icons/svg/score-star-f-96.svg");

require("@baidu/wuji-icons/svg/score-star-o-96.svg");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 'w-qes'; // eslint-disable-next-line

var sparkHtml = '<div class="spark"><svg class="line" xmlns="http://www.w3.org/2000/svg"><rect width="100%" height="100%"></rect></svg></div>';

var WujiQuickEvaluate = /*#__PURE__*/function (_Component) {
  _inherits(WujiQuickEvaluate, _Component);

  var _super = _createSuper(WujiQuickEvaluate);

  function WujiQuickEvaluate() {
    _classCallCheck(this, WujiQuickEvaluate);

    return _super.apply(this, arguments);
  }

  _createClass(WujiQuickEvaluate, [{
    key: "initData",
    value: function initData() {
      return {
        animate: false,
        initTitle: '',
        succTitle: '',
        starDesc: [],
        score: 0,
        defaultIcon: 'score-star-o-96',
        activeIcon: 'score-star-f-96',
        spacing: true
      };
    } // 点击评分：1、设置分数；2、重置动效；

  }, {
    key: "handleSelect",
    value: function handleSelect(_ref, index) {
      var _this = this;

      var event = _ref.event;
      event.stopPropagation();
      this.data.set('score', index + 1);
      this.data.set('animate', false);
      this.fire('click', {
        event: event,
        score: index + 1
      });
      setTimeout(function () {
        _this.data.set('animate', true);
      }, 20);
    }
  }]);

  return WujiQuickEvaluate;
}(_san.Component);

exports["default"] = WujiQuickEvaluate;

_defineProperty(WujiQuickEvaluate, "template",
/* html */
"\n        <div class=\"".concat(prefixCls, "-wrap {{spacing ? '").concat(prefixCls, "-spacing' : ''}}\">\n            <div s-if=\"title\" class=\"").concat(prefixCls, "-title\">{{title | raw}}</div>\n            <div class=\"").concat(prefixCls, "-score {{animate ? 'animate' : ''}}\">\n                <wuji-icon\n                    s-for=\"item, index in itemList trackBy item\"\n                    class=\"").concat(prefixCls, "-score-item ipos{{index + 1}} {{score > index ? 'active' : ''}}\"\n                    type=\"{{score > index ? activeIcon : defaultIcon}}\"\n                    size=\"{{96}}\"\n                    on-click=\"handleSelect($event, index)\"\n                />\n                <div s-if=\"score > 3\" class=\"").concat(prefixCls, "-fireworks\">\n                    <div class=\"fire-bind for{{score}}\">{{html1 | raw}}</div>\n                    <div s-if=\"score === 5\">{{html2 | raw}}</div>\n                </div>\n            </div>\n            <div\n                s-if=\"stext\"\n                class=\"").concat(prefixCls, "-stext\"\n            >{{stext}}</div>\n        </div>\n    "));

_defineProperty(WujiQuickEvaluate, "components", {
  'wuji-icon': _wujiIcon["default"]
});

_defineProperty(WujiQuickEvaluate, "dataTypes", {
  animate: _san.DataTypes.bool,
  initTitle: _san.DataTypes.string,
  succTitle: _san.DataTypes.string,
  starDesc: _san.DataTypes.arrayOf(_san.DataTypes.string),
  score: _san.DataTypes.number,
  defaultIcon: _san.DataTypes.string,
  activeIcon: _san.DataTypes.string,
  spacing: _san.DataTypes.bool
});

_defineProperty(WujiQuickEvaluate, "computed", {
  title: function title() {
    var data = this.data;
    var initTitle = data.get('initTitle');
    var succTitle = data.get('succTitle');
    var score = data.get('score');
    return score > 0 && succTitle || initTitle;
  },
  stext: function stext() {
    var data = this.data;
    var starDesc = data.get('starDesc');
    var score = data.get('score');
    return starDesc[score] || '';
  },
  itemList: function itemList() {
    return new Array(5).fill('');
  },
  html1: function html1() {
    return sparkHtml.repeat(5);
  },
  html2: function html2() {
    var html = [];

    for (var index = 1; index < 9; index++) {
      var wraper = ["<div class=\"fire fire".concat(index, "\">"), '</div>'];
      var sparks = sparkHtml.repeat(8);
      html.push(wraper.join(sparks));
    }

    return html.join('');
  }
});
