// @baidu/feed-baselog/src/lib/interface.ts
/**
 * @file ts type
 * @author okaychen<chenglong13@baidu.com>
 */

export interface UbcType<T> {
    page?: string;
    value?: string | number;
    type?: string | number;
    ext?: T;
}

// init & send ubc
export interface ConfigType<T, U> {
    ubcId: number | string;
    params?: UbcType<T> | Record<string, unknown>;
    queryParams?: U;
}

// ubc 初始化参数
export interface UbcBase {
    // init ubc params
    page?: string;
    from?: string;
    source?: string;
    queryParams?: object;
    testMode?: boolean;
    testUrl?: string;
    idfrom?: string;
    ifHeartbeatTime?: boolean;
    useSendBeacon?: boolean;
}

// thunder 初始化参数
export interface ThunderBase {
    ct?: number | string;
    logFrom?: string;
}

// baselog 初始化
export interface InitArgs {
    // baselog mode
    mode?: number | string;
    params?: UbcBase | Record<string, unknown>;
    tParams?: ThunderBase | Record<string, unknown>;
}

export interface ArgsType {
    tid?: number| string;
    ubcId?: number | string;
}
