// @baidu/ug-invoke-app/src/config/rules/invokeTypeRules.ts
/**
 * @file invokeTypeRules
 * @author zhangsiyuan(zhangsiyuan@baidu.com)
 */

import {IOS_VERSION, IS_IOS, IS_IPAD} from '../../utils/detect/os';
import {AppName} from '../app/appConfigMap';
import {Browser} from '../app/browserMap';


// [browser,appName][]
export type BrowserAppNameTuples = Array<[Browser | '*', AppName | '*']>;
export const notUlinkRules: BrowserAppNameTuples = [
    ['weibo', 'baiduboxapp'],
    ['uc', 'baiduboxapp'],
    ['quark', 'baiduboxapp'],
    ['chrome', '*'],
];


// [browser,appName][]
export const notIframeRules: BrowserAppNameTuples = [
    ['uc', '*'],
    ['chrome', '*'],
    ['samsung', '*'],
];

// 工具函数，检测是否匹配某个规则
const matchRules = (
    rules: BrowserAppNameTuples, browser: Browser, appName: AppName) => rules.some(([browserRule, appNameRule]) => {
    return (browserRule === '*' || browserRule === browser)
        && (appNameRule === '*' || appNameRule === appName);
});

// ios系统，且版本》9.或者是个iapd  无所谓版本了
const matchIosWithVersion = IS_IOS && IOS_VERSION > 9 || IS_IPAD;

export function matchedUlink(browser: Browser, appName: AppName) {
    const notUlink = matchRules(notUlinkRules, browser, appName);
    // ios系统支持，且不在notUlink名单中
    return matchIosWithVersion && !notUlink;
}


export function notMatchedIframe(browser: Browser, appName: AppName) {
    return matchIosWithVersion || matchRules(notIframeRules, browser, appName);
}
