// @baidu/mo-danmaku-item/dist/index.amd.js
define(['exports', '@baidu/hd-core', '@baidu/mo-actions', '@baidu/mo-base', '@baidu/xbox-emoticon', './style.module.less'], (function (exports, hdCore, moActions, Base, xboxEmoticon, styles) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var Base__default = /*#__PURE__*/_interopDefaultLegacy(Base);
    var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * Copyright (C) 2021 ~ present baidu.com
     *
     * @author fangbinbin02 <fangbinbin02@baidu.com>
     */
    // 表情@baidu/xbox-emoticon的统计参数 http://wiki.baidu.com/pages/viewpage.action?pageId=1519332107
    var XBOX_EMOTICON_PROD = 'mo-danmaku-item';
    // @san/component
    var DanmakuItem = /** @class */ (function (_super) {
        __extends(DanmakuItem, _super);
        function DanmakuItem() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        DanmakuItem.prototype.initData = function () {
            return {
                styles: styles__default["default"],
                data: {
                    listUrl: '',
                    replyId: '',
                    uname: '',
                    avatar: '',
                    content: '',
                    likeCount: '',
                    isUped: false,
                    isSelf: false,
                },
            };
        };
        DanmakuItem.prototype.danmakuItemClickHandler = function (item) {
            this.eventDispatch(moActions.CLICK_DANMAKU_H_ITEM, item);
        };
        DanmakuItem.prototype.praise = function (e) {
            e.stopPropagation();
            e.preventDefault();
            var data = this.data.get('data');
            this.eventDispatch(moActions.CLICK_PRAISE, {
                replyId: data.replyId,
                isUped: data.isUped,
                likeCount: +data.likeCount,
            });
        };
        DanmakuItem.template = "\n        <div\n            class=\"{{styles.danmakuItem}} {{data.isSelf ? styles.isSelf : ''}}\"\n            on-click=\"danmakuItemClickHandler(data)\"\n        >\n            <div class=\"{{styles.avatar}}\">\n                <img\n                    src=\"{{data.avatar}}\"\n                    alt=\"{{data.uname}}\"\n                />\n            </div>\n            <!--bca-disable-->\n            <div\n                class=\"{{styles.content}}\"\n                title=\"{{contentXss}}\"\n            >\n                {{contentWithEmoji | raw}}\n            </div>\n            <!--bca-enable-->\n            <div\n                class=\"{{[\n                    styles.praise,\n                    data.isUped ? styles.liked : styles.like\n                ]}}\"\n                on-click=\"praise($event)\"\n            >\n                <i class=\"{{styles.icon}}\"></i>\n                <span\n                    class=\"{{styles.count}}\"\n                >{{likeCount}}</span>\n            </div>\n        </div>\n    ";
        DanmakuItem.computed = {
            // 处理表情后的content
            contentWithEmoji: function () {
                return xboxEmoticon.fuzzyReplace(this.data.get('contentXss'), {
                    prod: XBOX_EMOTICON_PROD,
                });
            },
            likeCount: function () {
                var data = this.data.get('data');
                return +data.likeCount > 0 ? data.likeCount : '赞';
            },
            // content预防xss攻击
            contentXss: function () {
                return hdCore.encodeHtml(this.data.get('data.content'));
            },
        };
        return DanmakuItem;
    }(Base__default["default"]));

    exports["default"] = DanmakuItem;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
