// @baidu/mo-danmaku-item/dist/index.js
/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author fangbinbin02 <fangbinbin02@baidu.com>
 */
import { encodeHtml } from '@baidu/hd-core';
import { CLICK_DANMAKU_H_ITEM, CLICK_PRAISE, } from '@baidu/mo-actions';
import Base from '@baidu/mo-base';
import { fuzzyReplace } from '@baidu/xbox-emoticon';
import styles from './style.module.less';
export * from './interface';
// 表情@baidu/xbox-emoticon的统计参数 http://wiki.baidu.com/pages/viewpage.action?pageId=1519332107
const XBOX_EMOTICON_PROD = 'mo-danmaku-item';
// @san/component
export default class DanmakuItem extends Base {
    static template = /* html */ `
        <div
            class="{{styles.danmakuItem}} {{data.isSelf ? styles.isSelf : ''}}"
            on-click="danmakuItemClickHandler(data)"
        >
            <div class="{{styles.avatar}}">
                <img
                    src="{{data.avatar}}"
                    alt="{{data.uname}}"
                />
            </div>
            <!--bca-disable-->
            <div
                class="{{styles.content}}"
                title="{{contentXss}}"
            >
                {{contentWithEmoji | raw}}
            </div>
            <!--bca-enable-->
            <div
                class="{{[
                    styles.praise,
                    data.isUped ? styles.liked : styles.like
                ]}}"
                on-click="praise($event)"
            >
                <i class="{{styles.icon}}"></i>
                <span
                    class="{{styles.count}}"
                >{{likeCount}}</span>
            </div>
        </div>
    `;
    static computed = {
        // 处理表情后的content
        contentWithEmoji() {
            return fuzzyReplace(this.data.get('contentXss'), {
                prod: XBOX_EMOTICON_PROD,
            });
        },
        likeCount() {
            const data = this.data.get('data');
            return +data.likeCount > 0 ? data.likeCount : '赞';
        },
        // content预防xss攻击
        contentXss() {
            return encodeHtml(this.data.get('data.content'));
        },
    };
    initData() {
        return {
            styles,
            data: {
                listUrl: '',
                replyId: '',
                uname: '',
                avatar: '',
                content: '',
                likeCount: '',
                isUped: false,
                isSelf: false,
            },
        };
    }
    danmakuItemClickHandler(item) {
        this.eventDispatch(CLICK_DANMAKU_H_ITEM, item);
    }
    praise(e) {
        e.stopPropagation();
        e.preventDefault();
        const data = this.data.get('data');
        this.eventDispatch(CLICK_PRAISE, {
            replyId: data.replyId,
            isUped: data.isUped,
            likeCount: +data.likeCount,
        });
    }
}
