// @baidu/nano/es/scroll/index.js
/**
 * @file 组件 scroll 横滑
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import '../assets/svg/new-right.svg';
import {Component, DataTypes} from 'san';
import Icon from '../icon';

const prefixCls = 's-scroll';

export default class Scroll extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="${prefixCls}-wrap">
            <div s-if="showHead || title" class="${prefixCls}-header">
                <slot name="header">
                    <div s-if="title" class="${prefixCls}-title">{{title}}</div>
                    <div s-if="showMore" class="${prefixCls}-more" on-click="handleMoreClick">
                        更多<s-icon type="new-right"/>
                    </div>
                </slot>
            </div>
            <div class="${prefixCls}">
                <div class="${prefixCls}-container"><slot/></div>
                <div class="${prefixCls}-placeholder"/>
            </div>
        </div>
    `;

    static components = {
        's-icon': Icon
    };

    static dataTypes = {
        showHead: DataTypes.bool,
        showMore: DataTypes.bool,
        title: DataTypes.string
    };

    initData() {
        return {
            title: '',
            showHeader: false,
            showMore: false
        };
    }

    handleMoreClick(e) {
        e.stopPropagation();
        this.fire('more-click', e);
    }
}
