// @baidu/feed-baselog/dist/lib/thunder.h5.js
/**
 * @file thunder h5 image
 * @author okaychen(chenglong13@baidu.com)
 */
var ThunderH5 = /** @class */ (function () {
    function ThunderH5(opt) {
        this.baseURL = '';
        this.baseParams = {};
        this.opt = opt || {};
        this.baseParams = this.opt.baseParams || {};
        this.config(this.opt);
    }
    // simpleMerge
    ThunderH5.prototype.simpleMerge = function (obj1, obj2) {
        if (!obj2 || !obj1) {
            return obj1;
        }
        for (var i in obj2) {
            if (obj2.hasOwnProperty(i)) {
                obj1[i] = obj2[i];
            }
        }
        return obj1;
    };
    // assembleUrl
    ThunderH5.prototype.assembleUrl = function (url, baseParams, assParams) {
        var params = '';
        /* istanbul ignore next */
        assParams = assParams || {};
        assParams.r = 'l' + Date.now();
        assParams = this.simpleMerge(this.simpleMerge({}, baseParams), assParams);
        // 拼接字符串
        for (var i in assParams) {
            if (assParams.hasOwnProperty(i)) {
                params += '&' + i + '=' + encodeURIComponent(assParams[i]);
            }
        }
        return url + '?' + params.slice(1);
    };
    ThunderH5.prototype.config = function (opt) {
        this.simpleMerge(this.opt, opt);
        // 日志打印基础URL
        this.baseURL = this.opt.baseURL || 'https://hpd.baidu.com/v.gif';
        // 需要混入日志打印的所有基础字段
        this.baseParams = this.simpleMerge(this.baseParams, opt.baseParams);
        return this;
    };
    ThunderH5.prototype.send = function (pobj, callback, url) {
        var req = new Image();
        var timer = null;
        url = url || this.baseURL;
        /* istanbul ignore next */
        req.onload = req.onerror = req.onabort = function () {
            if (timer) {
                clearTimeout(timer);
                timer = null;
                callback && callback();
            }
        };
        req.src = this.assembleUrl(url, this.baseParams, pobj);
        // 防止回调被hold住
        if (typeof callback === 'function') {
            /* istanbul ignore next */
            timer = setTimeout(function () {
                timer = null;
                callback();
            }, 500);
        }
    };
    return ThunderH5;
}());
export default ThunderH5;
;
