// @baidu/mo-topic-miaodong/dist/index.amd.js
define(['exports', '@baidu/mo-base', '@baidu/wuji-san', './style.module.less'], (function (exports, Base, wujiSan, styles) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var Base__default = /*#__PURE__*/_interopDefaultLegacy(Base);
    var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var MIAODONG_CLICK = 'miaodong:cilck';

    /**
     * Copyright (C) 2021 ~ present baidu.com
     *
     * @author gaohong01 <gaohong01@baidu.com>
     */
    // @san/component
    var TopicMiaodong = /** @class */ (function (_super) {
        __extends(TopicMiaodong, _super);
        function TopicMiaodong() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TopicMiaodong.prototype.initData = function () {
            return {
                styles: styles__default["default"],
                visited: false,
            };
        };
        TopicMiaodong.prototype.attached = function () {
            _super.prototype.attached.call(this);
        };
        TopicMiaodong.prototype.handleClick = function (e) {
            e.stopPropagation();
            this.data.set('visited', true);
            this.eventDispatch(MIAODONG_CLICK);
        };
        TopicMiaodong.template = "\n        <div\n            class=\"{{styles.miaodongWrapper}}\"\n            on-click=\"handleClick\"\n        >\n            <wuji-title\n                text=\"{{title}}\"\n                spacing=\"{{false}}\"\n                visited=\"{{visited}}\"\n                class=\"{{styles.title}}\"\n            />\n            <wuji-image\n                s-bind=\"{{imageProps}}\"\n            />\n        </div>\n    ";
        TopicMiaodong.components = {
            'wuji-image': wujiSan.WujiImage,
            'wuji-title': wujiSan.WujiTitle,
        };
        TopicMiaodong.computed = {};
        return TopicMiaodong;
    }(Base__default["default"]));

    exports.MIAODONG_CLICK = MIAODONG_CLICK;
    exports["default"] = TopicMiaodong;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
