// @baidu/wuji-uikit/src/text/text.js
/**
 * @file 文本编码 text.js
 *
 * DO NOT CHANGE, THIS IS GENERATED BY SCRIPT!
 */

module.exports.Txt_T = {
    Txt_X001: 'T_X01',
    Txt_X002: 'T_X06',
    Txt_X003: 'T_X08',
    Txt_X004: 'T_X01',
    Txt_X006: 'T_X04',
    Txt_X007: 'T_X06',
    Txt_X008: 'T_X08',
    Txt_X009: 'T_X09',
    Txt_X010: 'T_X03',
    Txt_X011: 'T_X06',
    Txt_X012: 'T_X08',
    Txt_X013: 'T_X08',
    Txt_X014: 'T_X08',
    Txt_X015: 'T_X08',
    Txt_X016: 'T_X01',
    Txt_X017: 'T_X04',
    Txt_X020: 'T_X06',
    Txt_X021: 'T_X04',
    Txt_X022: 'T_X09',
    Txt_X023: 'T_X09',
    Txt_X024: 'T_X08',
    Txt_X025: 'T_X13',
    Txt_X026: 'T_X13',
    Txt_X027: 'T_X13',
    Txt_X028: 'T_X04',
    Txt_X029: 'T_X09',
    Txt_X030: 'T_X10',
    Txt_X032: 'T_X09',
    Txt_X033: 'T_X09',
    Txt_X034: 'T_X09',
    Txt_X035: 'T_X03',
    Txt_X036: 'T_X04',
    Txt_X037: 'T_X04',
    Txt_X038: 'T_X04',
    Txt_X039: 'T_X10',
    Txt_X040: 'T_X06',
    Txt_X041: 'T_X09',
    Txt_X042: 'T_X04',
    Txt_X043: 'T_X03',
    Txt_X044: 'T_X06',
    Txt_X045: 'T_X04',
    Txt_X046: 'T_X04',
    Txt_X047: 'T_X04',
    Txt_X048: 'T_X10',
    Txt_X049: 'T_X01',
    Txt_X050: 'T_X01',
    Txt_X051: 'T_X08',
    Txt_X052: 'T_X01',
    Txt_X053: 'T_X01',
    Txt_X054: 'T_X01',
    Txt_X055: 'T_X14',
    Txt_X056: 'T_X14',
    Txt_X057: 'T_X14',
    Txt_X058: 'T_X14',
    Txt_X059: 'T_X14',
    Txt_X060: 'T_X14',
    Txt_X061: 'T_X14',
    Txt_X062: 'T_X14',
    Txt_X063: 'T_X14',
    Txt_X064: 'T_X14',
    Txt_X065: 'T_X14',
    Txt_X066: 'T_X14',
    Txt_X067: 'T_X10',
    Txt_X068: 'T_X10',
    Txt_X069: 'T_X10',
    Txt_X070: 'T_X01',
    Txt_X071: 'T_X01',
    Txt_X072: 'T_X01',
    Txt_X073: 'T_X01',
    Txt_X074: 'T_X01',
    Txt_X075: 'T_X01',
    Txt_X076: 'T_X01',
    Txt_X077: 'T_X04',
    Txt_X078: 'T_X04',
    Txt_X079: 'T_X06',
    Txt_X080: 'T_X06',
    Txt_X081: 'T_X06',
    Txt_X082: 'T_X06',
    Txt_X083: 'T_X08',
    Txt_X084: 'T_X08',
    Txt_X085: 'T_X08',
    Txt_X086: 'T_X08',
    Txt_X087: 'T_X09',
    Txt_X088: 'T_X09',
    Txt_X089: 'T_X09',
    Txt_X090: 'T_X09',
    Txt_X091: 'T_X15',
    Txt_X092: 'T_X02',
    Txt_X093: 'T_X14',
    Txt_X094: 'T_X14',
    Txt_X095: 'T_X09',
    Txt_X096: 'T_X08',
    Txt_X097: 'T_X08',
    Txt_X098: 'T_X09',
    Txt_X099: 'T_X04',
    Txt_X100: 'T_X06',
    Txt_X101: 'T_X06',
    Txt_X102: 'T_X06',
    Txt_X103: 'T_X04',
    Txt_X104: 'T_X08',
    Txt_X105: 'T_X09',
    Txt_X106: 'T_X04',
    Txt_X107: 'T_X09',
    Txt_X108: 'T_X17',
    Txt_X109: 'T_X08',
    Txt_X110: 'T_X08',
    Txt_X111: 'T_X04',
    Txt_X112: 'T_X03',
    Txt_X113: 'T_X03',
    Txt_X114: 'T_X01',
    Txt_X115: 'T_X07',
    Txt_X117: 'T_X06',
    Txt_X118: 'T_X06',
    Txt_X119: 'T_X03',
    Txt_X120: 'T_X03',
    Txt_X121: 'T_X03',
    Txt_X122: 'T_X03',
    Txt_X123: 'T_X03',
    Txt_X124: 'T_X07',
    Txt_X125: 'T_X07',
    Txt_X126: 'T_X07',
    Txt_X127: 'T_X03',
    Txt_X128: 'T_X03',
    Txt_X129: 'T_X03',
    Txt_X130: 'T_X07',
    Txt_X131: 'T_X07',
    Txt_X132: 'T_X07',
    Txt_X133: 'T_X07',
    Txt_X134: 'T_X04',
    Txt_X135: 'T_X06',
    Txt_X137: 'T_X06',
    Txt_X138: 'T_X06',
    Txt_X139: 'T_X17',
    Txt_X140: 'T_X03',
    Txt_X141: 'T_X08',
    Txt_X142: 'T_X06',
    Txt_X600: 'T_X04',
    Txt_X601: 'T_X08',
    Txt_X602: 'T_X06',
    Txt_X605: 'T_X04',
    Txt_X606: 'T_X06',
    Txt_X607: 'T_X06',
    Txt_X608: 'T_X08',
    Txt_X609: 'T_X05',
    Txt_X610: 'T_X07',
    Txt_X611: 'T_X08',
    Txt_X612: 'T_X13',
    Txt_X613: 'T_X08',
    Txt_X614: 'T_X07',
};

module.exports.Txt_F = {
    Txt_X001: 'F_X01',
    Txt_X002: 'F_X01',
    Txt_X003: 'F_X01',
    Txt_X004: 'F_X01',
    Txt_X006: 'F_X01',
    Txt_X007: 'F_X01',
    Txt_X008: 'F_X01',
    Txt_X009: 'F_X01',
    Txt_X010: 'F_X01',
    Txt_X011: 'F_X01',
    Txt_X012: 'F_X01',
    Txt_X013: 'F_X01',
    Txt_X014: 'F_X01',
    Txt_X015: 'F_X01',
    Txt_X016: 'F_X01',
    Txt_X017: 'F_X01',
    Txt_X020: 'F_X01',
    Txt_X021: 'F_X01',
    Txt_X022: 'F_X01',
    Txt_X023: 'F_X01',
    Txt_X024: 'F_X01',
    Txt_X025: 'F_X01',
    Txt_X026: 'F_X01',
    Txt_X027: 'F_X01',
    Txt_X028: 'F_X02',
    Txt_X029: 'F_X01',
    Txt_X030: 'F_X01',
    Txt_X032: 'F_X01',
    Txt_X033: 'F_X01',
    Txt_X034: 'F_X01',
    Txt_X035: 'F_X01',
    Txt_X036: 'F_X01',
    Txt_X037: 'F_X01',
    Txt_X038: 'F_X01',
    Txt_X039: 'F_X01',
    Txt_X040: 'F_X01',
    Txt_X041: 'F_X01',
    Txt_X042: 'F_X01',
    Txt_X043: 'F_X02',
    Txt_X044: 'F_X01',
    Txt_X045: 'F_X01',
    Txt_X046: 'F_X01',
    Txt_X047: 'F_X01',
    Txt_X048: 'F_X01',
    Txt_X049: 'F_X01',
    Txt_X050: 'F_X01',
    Txt_X051: 'F_X01',
    Txt_X052: 'F_X02',
    Txt_X053: 'F_X01',
    Txt_X054: 'F_X01',
    Txt_X055: 'F_X02',
    Txt_X056: 'F_X01',
    Txt_X057: 'F_X02',
    Txt_X058: 'F_X01',
    Txt_X059: 'F_X02',
    Txt_X060: 'F_X01',
    Txt_X061: 'F_X02',
    Txt_X062: 'F_X01',
    Txt_X063: 'F_X02',
    Txt_X064: 'F_X01',
    Txt_X065: 'F_X02',
    Txt_X066: 'F_X01',
    Txt_X067: 'F_X02',
    Txt_X068: 'F_X01',
    Txt_X069: 'F_X02',
    Txt_X070: 'F_X02',
    Txt_X071: 'F_X01',
    Txt_X072: 'F_X02',
    Txt_X073: 'F_X01',
    Txt_X074: 'F_X02',
    Txt_X075: 'F_X01',
    Txt_X076: 'F_X02',
    Txt_X077: 'F_X01',
    Txt_X078: 'F_X01',
    Txt_X079: 'F_X01',
    Txt_X080: 'F_X01',
    Txt_X081: 'F_X01',
    Txt_X082: 'F_X01',
    Txt_X083: 'F_X01',
    Txt_X084: 'F_X01',
    Txt_X085: 'F_X01',
    Txt_X086: 'F_X01',
    Txt_X087: 'F_X01',
    Txt_X088: 'F_X01',
    Txt_X089: 'F_X01',
    Txt_X090: 'F_X01',
    Txt_X091: 'F_X01',
    Txt_X092: 'F_X02',
    Txt_X093: 'F_X01',
    Txt_X094: 'F_X01',
    Txt_X095: 'F_X01',
    Txt_X096: 'F_X02',
    Txt_X097: 'F_X02',
    Txt_X098: 'F_X01',
    Txt_X099: 'F_X01',
    Txt_X100: 'F_X01',
    Txt_X101: 'F_X01',
    Txt_X102: 'F_X02',
    Txt_X103: 'F_X02',
    Txt_X104: 'F_X01',
    Txt_X105: 'F_X01',
    Txt_X106: 'F_X02',
    Txt_X107: 'F_X01',
    Txt_X108: 'F_X02',
    Txt_X109: 'F_X02',
    Txt_X110: 'F_X02',
    Txt_X111: 'F_X01',
    Txt_X112: 'F_X01',
    Txt_X113: 'F_X01',
    Txt_X114: 'F_X01',
    Txt_X115: 'F_X01',
    Txt_X117: 'F_X02',
    Txt_X118: 'F_X02',
    Txt_X119: 'F_X02',
    Txt_X120: 'F_X02',
    Txt_X121: 'F_X02',
    Txt_X122: 'F_X02',
    Txt_X123: 'F_X02',
    Txt_X124: 'F_X01',
    Txt_X125: 'F_X01',
    Txt_X126: 'F_X01',
    Txt_X127: 'F_X01',
    Txt_X128: 'F_X02',
    Txt_X129: 'F_X01',
    Txt_X130: 'F_X01',
    Txt_X131: 'F_X02',
    Txt_X132: 'F_X02',
    Txt_X133: 'F_X02',
    Txt_X134: 'F_X02',
    Txt_X135: 'F_X01',
    Txt_X137: 'F_X02',
    Txt_X138: 'F_X01',
    Txt_X139: 'F_X02',
    Txt_X140: 'F_X02',
    Txt_X141: 'F_X01',
    Txt_X142: 'F_X02',
    Txt_X600: 'F_X02',
    Txt_X601: 'F_X01',
    Txt_X602: 'F_X02',
    Txt_X605: 'F_X01',
    Txt_X606: 'F_X02',
    Txt_X607: 'F_X01',
    Txt_X608: 'F_X01',
    Txt_X609: 'F_X01',
    Txt_X610: 'F_X01',
    Txt_X611: 'F_X01',
    Txt_X612: 'F_X01',
    Txt_X613: 'F_X01',
    Txt_X614: 'F_X01',
};

module.exports.Txt_CAM = {
    Txt_X001: 'CAM_X350',
    Txt_X002: 'CAM_X273',
    Txt_X003: 'CAM_X351',
    Txt_X004: 'CAM_X370',
    Txt_X006: 'CAM_X352',
    Txt_X007: 'CAM_X352',
    Txt_X008: 'CAM_X352',
    Txt_X009: 'CAM_X352',
    Txt_X010: 'CAM_X350',
    Txt_X011: 'CAM_X351',
    Txt_X012: 'CAM_X428',
    Txt_X013: 'CAM_X370',
    Txt_X014: 'CAM_X273',
    Txt_X015: 'CAM_X350',
    Txt_X016: 'CAM_X293',
    Txt_X017: 'CAM_X350',
    Txt_X020: 'CAM_X350',
    Txt_X021: 'CAM_X370',
    Txt_X022: 'CAM_X350',
    Txt_X023: 'CAM_X368',
    Txt_X024: 'CAM_X368',
    Txt_X025: 'CAM_X368',
    Txt_X026: 'CAM_X350',
    Txt_X027: 'CAM_X352',
    Txt_X028: 'CAM_X350',
    Txt_X029: 'CAM_X446',
    Txt_X030: 'CAM_X370',
    Txt_X032: 'CAM_X450',
    Txt_X033: 'CAM_X452',
    Txt_X034: 'CAM_X369',
    Txt_X035: 'CAM_X357',
    Txt_X036: 'CAM_X458',
    Txt_X037: 'CAM_X357',
    Txt_X038: 'CAM_X456',
    Txt_X039: 'CAM_X350',
    Txt_X040: 'CAM_X370',
    Txt_X041: 'CAM_X448',
    Txt_X042: 'CAM_X293',
    Txt_X043: 'CAM_X370',
    Txt_X044: 'CAM_X293',
    Txt_X045: 'CAM_X262',
    Txt_X046: 'CAM_X428',
    Txt_X047: 'CAM_X448',
    Txt_X048: 'CAM_X352',
    Txt_X049: 'CAM_X368',
    Txt_X050: 'CAM_X607',
    Txt_X051: 'CAM_X607',
    Txt_X052: 'CAM_X350',
    Txt_X053: 'CAM_X458',
    Txt_X054: 'CAM_X351',
    Txt_X055: 'CAM_X625',
    Txt_X056: 'CAM_X625',
    Txt_X057: 'CAM_X626',
    Txt_X058: 'CAM_X626',
    Txt_X059: 'CAM_X620',
    Txt_X060: 'CAM_X620',
    Txt_X061: 'CAM_X628',
    Txt_X062: 'CAM_X628',
    Txt_X063: 'CAM_X629',
    Txt_X064: 'CAM_X629',
    Txt_X065: 'CAM_X631',
    Txt_X066: 'CAM_X631',
    Txt_X067: 'CAM_X625',
    Txt_X068: 'CAM_X625',
    Txt_X069: 'CAM_X631',
    Txt_X070: 'CAM_X625',
    Txt_X071: 'CAM_X625',
    Txt_X072: 'CAM_X631',
    Txt_X073: 'CAM_X631',
    Txt_X074: 'CAM_X627',
    Txt_X075: 'CAM_X627',
    Txt_X076: 'CAM_X501',
    Txt_X077: 'CAM_X625',
    Txt_X078: 'CAM_X627',
    Txt_X079: 'CAM_X625',
    Txt_X080: 'CAM_X627',
    Txt_X081: 'CAM_X626',
    Txt_X082: 'CAM_X624',
    Txt_X083: 'CAM_X625',
    Txt_X084: 'CAM_X627',
    Txt_X085: 'CAM_X626',
    Txt_X086: 'CAM_X624',
    Txt_X087: 'CAM_X629',
    Txt_X088: 'CAM_X620',
    Txt_X089: 'CAM_X624',
    Txt_X090: 'CAM_X626',
    Txt_X091: 'CAM_X350',
    Txt_X092: 'CAM_X350',
    Txt_X093: 'CAM_X368',
    Txt_X094: 'CAM_X607',
    Txt_X095: 'CAM_X628',
    Txt_X096: 'CAM_X625',
    Txt_X097: 'CAM_X631',
    Txt_X098: 'CAM_X631',
    Txt_X099: 'CAM_X626',
    Txt_X100: 'CAM_X640',
    Txt_X101: 'CAM_X628',
    Txt_X102: 'CAM_X631',
    Txt_X103: 'CAM_X501',
    Txt_X104: 'CAM_X293',
    Txt_X105: 'CAM_X293',
    Txt_X106: 'CAM_X352',
    Txt_X107: 'CAM_X273',
    Txt_X108: 'CAM_X352',
    Txt_X109: 'CAM_X428',
    Txt_X110: 'CAM_X350',
    Txt_X111: 'CAM_X568',
    Txt_X112: 'CAM_X428',
    Txt_X113: 'CAM_X458',
    Txt_X114: 'CAM_X428',
    Txt_X115: 'CAM_X351',
    Txt_X117: 'CAM_X350',
    Txt_X118: 'CAM_X604',
    Txt_X119: 'CAM_X770',
    Txt_X120: 'CAM_X605',
    Txt_X121: 'CAM_X606',
    Txt_X122: 'CAM_X603',
    Txt_X123: 'CAM_X771',
    Txt_X124: 'CAM_X350',
    Txt_X125: 'CAM_X353',
    Txt_X126: 'CAM_X428',
    Txt_X127: 'CAM_X351',
    Txt_X128: 'CAM_X350',
    Txt_X129: 'CAM_X791',
    Txt_X130: 'CAM_X626',
    Txt_X131: 'CAM_X370',
    Txt_X132: 'CAM_X352',
    Txt_X133: 'CAM_X793',
    Txt_X134: 'CAM_X626',
    Txt_X135: 'CAM_X793',
    Txt_X137: 'CAM_X626',
    Txt_X138: 'CAM_X1122',
    Txt_X139: 'CAM_X791',
    Txt_X140: 'CAM_X791',
    Txt_X141: 'CAM_X1196',
    Txt_X142: 'CAM_X791',
    Txt_X600: 'CAM_X625',
    Txt_X601: 'CAM_X628',
    Txt_X602: 'CAM_X627',
    Txt_X605: 'CAM_X656',
    Txt_X606: 'CAM_X625',
    Txt_X607: 'CAM_X622',
    Txt_X608: 'CAM_X792',
    Txt_X609: 'CAM_X625',
    Txt_X610: 'CAM_X989',
    Txt_X611: 'CAM_X614',
    Txt_X612: 'CAM_X244',
    Txt_X613: 'CAM_X1017',
    Txt_X614: 'CAM_X793',
};
