// @baidu/xbox-native/event/off.js
/**
 * @file off
 * @author panming
 * @date 2019-01-25 20:15:45
 */

import register from '../register';
import {off, paramsFilter} from './index';
// import {isPlainObject} from '@baidu/xbox/is';
import invoke from '../invoke';
import {isMatrix} from '../detect';
import {json2query} from '@baidu/xbox';

register('datachannel/unregister', {
    v: 5,
    hasCB: false,
    versionRange: ['8.4'],
    iosinfoVersionRange: ['1.3.1.10'],
    androidliteVersionRange: ['2.2.0.10'],
    iosmissionVersionRange: ['1.0'],
    androidmissionVersionRange: ['1.0'],
    filter: data => {
        return {
            data: paramsFilter(data)
        };
    }
});

export default params => {
    if (isMatrix('baiduboxvision')) {
        const invokeParams = paramsFilter(params, false);
        if (invokeParams.data) {
            Object.assign(invokeParams, invokeParams.data);
            delete invokeParams.data;
        }

        invoke(`baiduboxapp://v5/datachannel/unregister?${json2query({
            ...invokeParams
        })}`);
        return Promise.resolve();
    }
    return off(params);
};
