// @baidu/xbox-native/es5/view/imageCollection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _detect = require("../detect");

var _stringify = _interopRequireDefault(require("../stringify"));

var _index = _interopRequireDefault(require("../ios/index"));

var _index2 = require("../android/index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @module view
 * @file 打开图集落地页
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

/**
 * 打开图集落地页
 * @function imageCollection
 * @param {string} nid - nid
 * @param {string} [from="feed"] - 来源
 * @return {Promise}
 */
var _default = function _default(nid) {
  var from = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'feed';

  if ((0, _detect.isBox)() || (0, _detect.isMatrix)()) {
    var toolids = (0, _stringify.default)({
      toolids: ['4', '1', '2', '3']
    }, true);
    var cmd = '';

    if ((0, _detect.isAndroid)()) {
      var intent = ["intent:#Intent;S.toolbaricons=".concat(toolids), 'S.menumode=2', 'launchFlags=0x10000000', 'component=com.baidu.searchbox/.discovery.picture.PictureBrowseActivity', 'S.context=' + (0, _stringify.default)({
        nid: nid
      }, true), 'S.sfrom=' + from, 'end'].join(';');
      /* eslint-disable max-len */

      (0, _index2.command)(intent, 0, 'com.baidu.searchbox.discovery.picture.PictureBrowseActivity', (0, _detect.isMatrix)('baiduboxvision') ? '1' : '16789248');
      return Promise.resolve();
      /* eslint-enable max-len */
    } else {
      cmd = ['baiduboxapp://common?action=imagecollection&params=' + (0, _stringify.default)({
        context: {
          nid: nid
        },
        sfrom: from
      }, true), 'minver=7.2.5.0', 'toolbaricons=' + toolids, 'menumode=2'].join('&');
      (0, _index.default)(cmd);
      return Promise.resolve();
    }
  }

  return Promise.reject({
    errCode: 902,
    errMsg: '手百外无法调用端能力'
  });
};

exports.default = _default;
