// @baidu/xbox-native/es5/view/invokeHN.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _stringify = _interopRequireDefault(require("../stringify"));

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 打开HN页面
 * @author panming
 * @date 2019-02-20 18:32:00
 */
(0, _register.default)('HTMLNative/open', {
  v: 19,
  versionRange: ['10.3'],
  filter: function filter(data) {
    data.backupURL = data.backup;
    delete data.backup;
    return {
      module: 'utils',
      action: 'invokeHNPage',
      data: {
        params: (0, _stringify.default)(data, true)
      }
    };
  }
});

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('HTMLNative', params));
};

exports.default = _default;
