// @baidu/xbox-native/es5/view/swangame.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _stringify = _interopRequireDefault(require("../stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module view
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('swan/open', {
  v: 22,
  versionRange: ['10.13.5'],
  schemeNoV: true,
  filter: function filter(data) {
    /* istanbul ignore if */
    if ((0, _stringify.default)(data) === '{}') {
      return {
        module: 'swangame',
        action: 'open',
        data: {}
      };
    } // data.url中可能包含域名：
    // 1、域名中可能有 smartapps.cn ，并且后面紧跟 webmapp，需要去掉 webmapp，再摘出path+query
    // 2、域名中可能有 smartapps.cn，但没有紧跟 webmapp，那么域名后面的就是需要的path+query
    // 3、域名中可能没有 smartapps.cn，就是普通三方域名，需要摘出path+query
    //
    // data.url中可能没有域名：
    // 1、将前面的杠去掉，其他就是需要的path+query


    var url = '';
    var host = data.url.match(/^http(s)?:\/\/(.*?)\//);

    if (host) {
      if (data.url.match('smartapps.cn/webmapp/')) {
        url = data.url.split('smartapps.cn/webmapp/')[1];
      } else {
        url = data.url.split(host[0])[1];
      }
    } else {
      url = data.url.substring(0, 1) === '/' ? data.url.substring(1) : data.url;
    }

    var path = url.split('?')[0];
    var params = {};

    if (url.indexOf('?') > -1) {
      params = (0, _utils.noEmptyQueryToJson)(data.url);
    } else {
      params = data.params || {};
    }

    var baiduboxapp = {
      from: params.from || data.from || '',
      ext: params.ext || data.ext || {}
    };
    var result = {
      _baiduboxapp: (0, _stringify.default)(baiduboxapp, true)
    };
    var appId = data.appId || data.appKey;

    if (appId) {
      appId = appId.replace(/\w-/g, function (item) {
        return item.substring(0, 1).toUpperCase();
      });
    }

    if (path && path.length > 0) {
      for (var key in params) {
        params[key] = encodeURIComponent(params[key]);
      }

      Object.assign(result, params);
    }

    return {
      module: 'swangame/' + appId,
      action: path,
      data: result
    };
  }
});
/**
 * @function swangame
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('swangame', params));
};

exports.default = _default;
