// san-ssr/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileToRenderer = exports.compileToSource = exports.TypeGuards = void 0;
const san_project_1 = require("./models/san-project");
const TypeGuards = __importStar(require("./utils/type-guards"));
exports.TypeGuards = TypeGuards;
// util functions
var id_generator_1 = require("./utils/id-generator");
Object.defineProperty(exports, "IDGenerator", { enumerable: true, get: function () { return id_generator_1.IDGenerator; } });
var case_1 = require("./utils/case");
Object.defineProperty(exports, "parseSanHTML", { enumerable: true, get: function () { return case_1.parseSanHTML; } });
Object.defineProperty(exports, "compareSanHTML", { enumerable: true, get: function () { return case_1.compareSanHTML; } });
Object.defineProperty(exports, "assertDeepEqual", { enumerable: true, get: function () { return case_1.assertDeepEqual; } });
Object.defineProperty(exports, "assertSanHTMLEqual", { enumerable: true, get: function () { return case_1.assertSanHTMLEqual; } });
var dom_util_1 = require("./utils/dom-util");
Object.defineProperty(exports, "autoCloseTags", { enumerable: true, get: function () { return dom_util_1.autoCloseTags; } });
Object.defineProperty(exports, "booleanAttributes", { enumerable: true, get: function () { return dom_util_1.booleanAttributes; } });
var anode_util_1 = require("./utils/anode-util");
Object.defineProperty(exports, "getANodePropByName", { enumerable: true, get: function () { return anode_util_1.getANodePropByName; } });
var emitter_1 = require("./utils/emitter");
Object.defineProperty(exports, "Emitter", { enumerable: true, get: function () { return emitter_1.Emitter; } });
var underscore_1 = require("./runtime/underscore");
Object.defineProperty(exports, "_", { enumerable: true, get: function () { return underscore_1._; } });
// class types
var san_source_file_1 = require("./models/san-source-file");
Object.defineProperty(exports, "TypedSanSourceFile", { enumerable: true, get: function () { return san_source_file_1.TypedSanSourceFile; } });
Object.defineProperty(exports, "DynamicSanSourceFile", { enumerable: true, get: function () { return san_source_file_1.DynamicSanSourceFile; } });
Object.defineProperty(exports, "isTypedSanSourceFile", { enumerable: true, get: function () { return san_source_file_1.isTypedSanSourceFile; } });
var san_project_2 = require("./models/san-project");
Object.defineProperty(exports, "SanProject", { enumerable: true, get: function () { return san_project_2.SanProject; } });
var component_info_1 = require("./models/component-info");
Object.defineProperty(exports, "TypedComponentInfo", { enumerable: true, get: function () { return component_info_1.TypedComponentInfo; } });
Object.defineProperty(exports, "DynamicComponentInfo", { enumerable: true, get: function () { return component_info_1.DynamicComponentInfo; } });
var component_reference_1 = require("./models/component-reference");
Object.defineProperty(exports, "ComponentReference", { enumerable: true, get: function () { return component_reference_1.ComponentReference; } });
var component_1 = require("./models/component");
Object.defineProperty(exports, "COMPONENT_RESERVED_MEMBERS", { enumerable: true, get: function () { return component_1.COMPONENT_RESERVED_MEMBERS; } });
let defaultProject;
function compileToSource(ComponentClass) {
    const proj = new san_project_1.SanProject();
    const targetCode = proj.compileToSource(ComponentClass, 'js', {
        bareFunction: true
    });
    return targetCode;
}
exports.compileToSource = compileToSource;
function compileToRenderer(ComponentClass, options) {
    defaultProject = defaultProject || new san_project_1.SanProject();
    const renderer = defaultProject.compileToRenderer(ComponentClass, options);
    return renderer;
}
exports.compileToRenderer = compileToRenderer;
