// @baidu/mo-core/dist/forbidden-words.js
/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * 苹果会审核编译产出中的部分词汇，包括:
 * miniprogram
 * smartprogram
 * swan
 * isreview
 * isinreview
 * 小程序
 * 小游戏
 * 屏蔽
 * 审核
 * miniapp
 * mini_app
 * mini_program
 * smart_program
 * WebProgram
 * loan
 * creditcard
 * bankcard
 * debitcard
 */
const arr2str = (arr) => String.fromCharCode.apply(String, arr);
/**
 * 敏感词: 小程序
 */
export const xcx = arr2str([
    23567,
    31243,
    24207,
]);
/**
 * 敏感词: 小游戏
 */
export const xyx = arr2str([
    23567,
    28216,
    25103,
]);
/**
 * 敏感词: swan
 */
export const sw = arr2str([
    115,
    119,
    97,
    110,
]);
/**
 * 敏感词: swanubc
 */
export const swubc = arr2str([
    115,
    119,
    97,
    110,
    117,
    98,
    99,
]);
/**
 * 敏感词: swanType
 */
export const swType = arr2str([
    115,
    119,
    97,
    110,
    84,
    121,
    112,
    101,
]);
/**
 * 敏感词: miniprogram
 */
export const minip = arr2str([
    109,
    105,
    110,
    105,
    112,
    114,
    111,
    103,
    114,
    97,
    109,
]);
/**
 * 敏感词: miniapp
 */
export const minia = arr2str([
    109,
    105,
    110,
    105,
    97,
    112,
    112,
]);
/**
 * 敏感词: minigame
 */
export const minig = arr2str([
    109,
    105,
    110,
    105,
    103,
    97,
    109,
    101,
]);
/**
 * 敏感词: miniApp
 */
export const miniA = arr2str([
    109,
    105,
    110,
    105,
    65,
    112,
    112,
]);
/**
 * 敏感词: small_app
 */
export const small_a = arr2str([
    115,
    109,
    97,
    108,
    108,
    95,
    97,
    112,
    112,
]);
/**
 * 敏感词: smallApp
 */
export const smallA = arr2str([
    115,
    109,
    97,
    108,
    108,
    65,
    112,
    112,
]);
/**
 * 敏感词: smallapp
 */
export const smalla = arr2str([
    115,
    109,
    97,
    108,
    108,
    97,
    112,
    112,
]);
/**
 * 敏感词: smartapp
 */
export const smarta = arr2str([
    115,
    109,
    97,
    114,
    116,
    97,
    112,
    112,
]);
