// @baidu/auc-baike/index.ts
/**
 * @file index.ts
 * @author gaohong01
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucBaikeData} from './types';
import TopicBaike, {BAIKE_CLICK, BAIKE_IMAGE_CLICK} from '@baidu/mo-topic-baike';
import {componentLog} from '@baidu/auc-log';
import {env} from '@baidu/auc-base/env';

export class AucBaike extends AucComponent {
    static template:string = /* html */ `
    <div class="{{wrapperClass}}">
        <topic-baike
            title="{{title}}"
            description="{{description}}"
            imageProps="{{imageProps}}"
            index="{{index}}"
            logo="{{logo}}"
        ></topic-baike>
    </div>
    `;
    static components:object = {
        'topic-baike': TopicBaike
    };

    static computed = {
        wrapperClass() {
            return [
                styles.aucBaikeWrapper,
                this.data.get('visited') ? styles.visited : '',
                this.data.get('isBigEvent') ? styles.bigEvent : ''
            ]
        }
    };

    static messages = {
        /**
         * 百科点击
         */
        [BAIKE_CLICK]: function () {
            this.fire('jump', this.data.get('jumpProps'));
            componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
        },

        /**
         * 百科图片点击
         */
        [BAIKE_IMAGE_CLICK]: function () {
            let jumpProps = this.data.get('jumpProps');
            let logType = 'cilck'; 
            if (this.data.get('type') === 'video' && env.isBoxSeries) {
                jumpProps = {
                    cmd: this.data.get('videoCmd')
                };
                logType = 'video_click';
            }
            componentLog(this.data.get('logBaseType'), logType, this.data.get('logExtraParams'));
            this.fire('jump', jumpProps);
        },
    };

    initData(): AucBaikeData {
        return {
            styles,
            logBaseType: 'baike_',
            logExtraParams: {},
            jumpProps: {}
        };
    }

    attached(): void {
        this.registerScrollInView().then(() => {
            componentLog(this.data.get('logBaseType'), 'show', this.data.get('logExtraParams'));
        });
    }
}

export default AucBaike;
