// @baidu/nano/lib/image/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("./style/index.less");

var _san = require("san");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 's-image';
var sizeMap = {}; // 图片宽高比: http://bdbox.baidu.com/ceued-docs/design/component/image/

var sizeType = ['1:1', '2:3', // extra added
'3:1', '3:2', '3:4', '4:3', '5:2', '16:9', '21:4', 'full']; // 默认宽高比

var defaultSize = sizeType[0];
sizeType.forEach(function (size) {
  if ('full' === size) {
    sizeMap[size] = 0;
  } else {
    var sizeVal = size.split(':');
    sizeMap[size] = sizeVal[1] / sizeVal[0];
  }
});

var Image =
/*#__PURE__*/
function (_Component) {
  _inherits(Image, _Component);

  function Image() {
    _classCallCheck(this, Image);

    return _possibleConstructorReturn(this, _getPrototypeOf(Image).apply(this, arguments));
  }

  _createClass(Image, [{
    key: "initData",
    // eslint-disable-next-line
    value: function initData() {
      return {
        hairline: true,
        size: defaultSize,
        url: '',
        cancelBubble: true
      };
    }
  }, {
    key: "handleTouchEvent",
    value: function handleTouchEvent(e, cancelBubble) {
      !cancelBubble && e.stopPropagation();
    }
  }]);

  return Image;
}(_san.Component);

exports.default = Image;

_defineProperty(Image, "template",
/*html*/
"\n        <div class=\"{{wrapClass}}\"\n            on-touchstart=\"handleTouchEvent($event, cancelBubble)\"\n            on-touchend=\"handleTouchEvent($event, cancelBubble)\"\n            on-touchcancel=\"handleTouchEvent($event, cancelBubble)\"\n        >\n            <div class=\"".concat(prefixCls, "-default\"/>\n            <div class=\"").concat(prefixCls, "\" style=\"{{imageStyle}}\"><slot/></div>\n        </div>\n    "));

_defineProperty(Image, "dataTypes", {
  cancelBubble: _san.DataTypes.bool,
  size: _san.DataTypes.oneOf(sizeType),
  hairline: _san.DataTypes.bool,
  url: _san.DataTypes.string
});

_defineProperty(Image, "computed", {
  wrapClass: function wrapClass() {
    var hairline = this.data.get('hairline');
    return ["".concat(prefixCls, "-wrapper"), hairline && "".concat(prefixCls, "-hairline")].filter(function (c) {
      return c;
    });
  },
  imageStyle: function imageStyle() {
    var size = this.data.get('size');
    var url = this.data.get('url');
    var isFull = 'full' === size;
    return {
      'padding-top': "".concat(sizeMap[size] * 100, "%"),
      'height': isFull ? '100%' : 'auto',
      'background-image': url ? "url(".concat(url, ")") : ''
    };
  }
});

Image.sizeType = sizeType;
