// @baidu/wuji-san/es/wuji-stext/index.js
/**
 * @file 组件 WujiStext 无极-短文本
 * @author yanghongxue <yanghongxue@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';

const prefixCls = 'w-stext';

export default class WujiStext extends Component {
    static template = /* html */ `
        <div class="{{wrapClass}}">
            <span>{{firstText}}</span>
            <span>{{secondText}}</span>
        </div>
    `;

    static dataTypes = {
        line: DataTypes.bool,
    };

    static computed = {
        wrapClass() {
            const data = this.data;
            const layout = data.get('layout');
            const size = data.get('size');
            const line = data.get('line');

            return [
                `${prefixCls}-wrap`,
                line && `${prefixCls}-hasline`,
                layout && `${prefixCls}-${layout}`,
                size && `${prefixCls}-${size}`,
            ].filter(c => c);
        },
    };

    initData() {
        return {
            layout: 'lr',
            size: 'sm',
            firstText: '',
            secondText: '',
            line: false,
        };
    }
}
