// @baidu/feed-baselog/dist/index.devtools.js
/**
 * @file baselog devtools cdp
 * @author okaychen<chenglong13@baidu.com>
 */
import { __extends } from "tslib";
import BaseLog from './index';
import merge from 'lodash.merge';
import ThunderH5 from './lib/thunder.h5';
import { TBASEURL, UBASEURL } from './lib/const';
// @ts-ignore
var $devtools = global.$devtools;
var isDevtools = typeof $devtools !== 'undefined' && $devtools.sendCommand && $devtools.registerEvent;
if (isDevtools) {
    $devtools.baselogMonitor = [];
    /* istanbul ignore next */
    $devtools.registerEvent('baselog.getData', function () {
        return $devtools.baselogMonitor;
    });
}
var $BaseLog = /** @class */ (function (_super) {
    __extends($BaseLog, _super);
    function $BaseLog(InitArgs) {
        return _super.call(this, InitArgs) || this;
    }
    $BaseLog.prototype.sendThunder = function (params, callback, url) {
        if (url === void 0) { url = TBASEURL; }
        _super.prototype.sendThunder.call(this, params, callback, url);
        var thunder = new ThunderH5(this.baseParams);
        var thunderURL = thunder.assembleUrl(url, this.baseParams, params);
        var tFinnalParams = merge({}, this.baseParams, params);
        this.$handleParams('thunder', tFinnalParams, url, thunderURL);
    };
    $BaseLog.prototype.sendUbc = function (ubcList, callback, url) {
        var _a;
        if (url === void 0) { url = UBASEURL; }
        _super.prototype.sendUbc.call(this, ubcList, callback);
        /* istanbul ignore next */
        var uFinnalParams = merge({}, (_a = this.finalParams) === null || _a === void 0 ? void 0 : _a.params, ubcList);
        var ubcName = 'ubc';
        // 如果打点ID是13587，则为thunder迁移ubc点位
        if (+ubcList.ubcId === 13587) {
            ubcName = 'thunder to ubc';
        }
        this.$handleParams(ubcName, uFinnalParams, url);
    };
    // devtools.getData 参数
    $BaseLog.prototype.$handleParams = function (ubcName, args, url, requestURL) {
        /* istanbul ignore next */
        if (!isDevtools) {
            return;
        }
        var $params = {
            args: args,
            path: url,
            method: 'get',
            tid: args === null || args === void 0 ? void 0 : args.tid,
            ubcid: args === null || args === void 0 ? void 0 : args.ubcId,
            name: "".concat(ubcName, "-").concat((args === null || args === void 0 ? void 0 : args.ubcId) || (args === null || args === void 0 ? void 0 : args.tid)),
            request: {
                'Request Method': 'get',
                'Request URL': requestURL
            }
        };
        $devtools.baselogMonitor.push($params);
    };
    return $BaseLog;
}(BaseLog));
;
export default $BaseLog;
