// @baidu/auc-title/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucTitleData} from './types';
import {WujiTitle, WujiWidget} from '@baidu/wuji-san';
import {componentLog} from '@baidu/auc-log';

export class AucTitle extends AucComponent {
    static template: string = /* html */ `
        <div class="{{styles.aucTitleWrapper}} {{visited ? 'visited' : ''}} {{isBigEvent ? styles.bigEvent : ''}}" on-click="clickItem">
            <div class="{{styles.aucTitle}}">
                <div class="{{styles.title}}">
                    <wuji-title
                        text="{{title}}"
                        spacing="{{false}}"
                        label-props="{{labelProp}}"
                        visited="{{visited}}"
                        lineClamp="{{3}}"
                    ></wuji-title>
                </div>
                <div
                    class="{{styles.widget}}"
                    s-if="{{widgetItems && widgetItems.length > 0}}"
                >
                    <wuji-widget
                        has-margin-top="{{false}}"
                        has-margin-bottom="{{false}}"
                        items="{{widgetItems}}"
                    ></wuji-widget>
                </div>
            </div>
        </div>
    `;
    static components: object = {
        'wuji-title': WujiTitle,
        'wuji-widget': WujiWidget
    };

    static computed: object = {
    };

    initData(): AucTitleData {
        return {
            styles,
            title: '',
            labelProp: {},
            widgetItems: [],
            visited: false,
            spacing: false,
            jumpProps: {},
            logBaseType: 'text_only_',
            logExtraParams: {}
        };
    }

    attached(): void {
        this.registerScrollInView().then(() => {
            componentLog(this.data.get('logBaseType'), 'show', this.data.get('logExtraParams'));
        });
    }

    clickItem(): void {
        this.fire('jump', this.data.get('jumpProps'));
        this.data.set('visited', true);
        componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
    }
}

export default AucTitle;
