// @baidu/fs-actions/dist/index.js
/**
 * action: 登陆事件
 */
export const LOGIN = 'login';
/**
 * action: 弹幕点赞事件
 */
export const CLICK_LIKE_DANMAKU = 'click:like:danmaku';
/**
 * action: 弹幕中单条弹幕点赞成功事件
 */
export const LIKE_DANMAKU_SUCCESS = 'like:danmaku:success';
/**
 * action: 弹幕中单条弹幕点赞失败事件
 */
export const LIKE_DANMAKU_ERROR = 'like:danmaku:error';
/**
 * action: 弹幕发布事件
 */
export const CLICK_PUBLISH_DANMAKU = 'click:publish:danmaku';
/**
 * action: 弹幕发布成功事件
 */
export const PUBLISH_DANMAKU_SUCCESS = 'publish:danmaku:success';
/**
 * action: 弹幕发布失败事件
 */
export const PUBLISH_DANMAKU_ERROR = 'publish:danmaku:error';
/**
 * action: 表态点赞成功
 */
export const LIKE_OPINION_SUCCESS = 'like:opinion:success';
/**
 * action: 表态点赞失败
 */
export const LIKE_OPINION_ERROR = 'like:opinion:error';
/**
 * action: 移动端水平弹幕点赞成功
 */
export const LIKE_DANMAKU_H_MO_SUCCESS = 'like:danmakuHMo:success';
/**
 * action: 移动端水平弹幕点赞失败
 */
export const LIKE_DANMAKU_H_MO_ERROR = 'like:danmakuHMo:error';
/**
 * action: 移动端水平弹幕单条点击
 */
export const CLICK_DANMAKU_H_ITEM_MO = 'click:danmakuHItemMo';
/**
 * action: 移动端水平弹幕组件的发布 点击发布按钮
 */
export const PUBLISH_DANMAKU_H_MO = 'publish:danmakuHMo';
/**
 * action: 移动端水平弹幕发布成功
 */
export const PUBLISH_DANMAKU_H_MO_SUCCESS = 'publish:danmakuHMo:success';
/**
 * action: 移动端水平弹幕发布成功
 */
export const PUBLISH_DANMAKU_H_MO_ERROR = 'publish:danmakuHMo:error';
/**
 * action: 移动端水平弹幕发布框调起
 */
export const PULISHER_DANMAKU_H_MO_AROUSE = 'publisher:danmakuHMo:arouse';
/**
 * action: 移动端水平弹幕input框点击事件
 */
export const INPUT_DANMAKU_H_MO_AROUSE = 'input:danmakuHMo:arouse';
/**
 * action: 移动端水平弹幕展现打点
 */
export const SHOW_DANMAKU_H_MO = 'show:danmakuHMo';
