// @baidu/fs-actions/dist/index.amd.js
define(['exports'], (function (exports) { 'use strict';

    /**
     * action: 登陆事件
     */
    var LOGIN = 'login';
    /**
     * action: 弹幕点赞事件
     */
    var CLICK_LIKE_DANMAKU = 'click:like:danmaku';
    /**
     * action: 弹幕中单条弹幕点赞成功事件
     */
    var LIKE_DANMAKU_SUCCESS = 'like:danmaku:success';
    /**
     * action: 弹幕中单条弹幕点赞失败事件
     */
    var LIKE_DANMAKU_ERROR = 'like:danmaku:error';
    /**
     * action: 弹幕发布事件
     */
    var CLICK_PUBLISH_DANMAKU = 'click:publish:danmaku';
    /**
     * action: 弹幕发布成功事件
     */
    var PUBLISH_DANMAKU_SUCCESS = 'publish:danmaku:success';
    /**
     * action: 弹幕发布失败事件
     */
    var PUBLISH_DANMAKU_ERROR = 'publish:danmaku:error';
    /**
     * action: 表态点赞成功
     */
    var LIKE_OPINION_SUCCESS = 'like:opinion:success';
    /**
     * action: 表态点赞失败
     */
    var LIKE_OPINION_ERROR = 'like:opinion:error';
    /**
     * action: 移动端水平弹幕点赞成功
     */
    var LIKE_DANMAKU_H_MO_SUCCESS = 'like:danmakuHMo:success';
    /**
     * action: 移动端水平弹幕点赞失败
     */
    var LIKE_DANMAKU_H_MO_ERROR = 'like:danmakuHMo:error';
    /**
     * action: 移动端水平弹幕单条点击
     */
    var CLICK_DANMAKU_H_ITEM_MO = 'click:danmakuHItemMo';
    /**
     * action: 移动端水平弹幕组件的发布 点击发布按钮
     */
    var PUBLISH_DANMAKU_H_MO = 'publish:danmakuHMo';
    /**
     * action: 移动端水平弹幕发布成功
     */
    var PUBLISH_DANMAKU_H_MO_SUCCESS = 'publish:danmakuHMo:success';
    /**
     * action: 移动端水平弹幕发布成功
     */
    var PUBLISH_DANMAKU_H_MO_ERROR = 'publish:danmakuHMo:error';
    /**
     * action: 移动端水平弹幕发布框调起
     */
    var PULISHER_DANMAKU_H_MO_AROUSE = 'publisher:danmakuHMo:arouse';
    /**
     * action: 移动端水平弹幕input框点击事件
     */
    var INPUT_DANMAKU_H_MO_AROUSE = 'input:danmakuHMo:arouse';
    /**
     * action: 移动端水平弹幕展现打点
     */
    var SHOW_DANMAKU_H_MO = 'show:danmakuHMo';

    exports.CLICK_DANMAKU_H_ITEM_MO = CLICK_DANMAKU_H_ITEM_MO;
    exports.CLICK_LIKE_DANMAKU = CLICK_LIKE_DANMAKU;
    exports.CLICK_PUBLISH_DANMAKU = CLICK_PUBLISH_DANMAKU;
    exports.INPUT_DANMAKU_H_MO_AROUSE = INPUT_DANMAKU_H_MO_AROUSE;
    exports.LIKE_DANMAKU_ERROR = LIKE_DANMAKU_ERROR;
    exports.LIKE_DANMAKU_H_MO_ERROR = LIKE_DANMAKU_H_MO_ERROR;
    exports.LIKE_DANMAKU_H_MO_SUCCESS = LIKE_DANMAKU_H_MO_SUCCESS;
    exports.LIKE_DANMAKU_SUCCESS = LIKE_DANMAKU_SUCCESS;
    exports.LIKE_OPINION_ERROR = LIKE_OPINION_ERROR;
    exports.LIKE_OPINION_SUCCESS = LIKE_OPINION_SUCCESS;
    exports.LOGIN = LOGIN;
    exports.PUBLISH_DANMAKU_ERROR = PUBLISH_DANMAKU_ERROR;
    exports.PUBLISH_DANMAKU_H_MO = PUBLISH_DANMAKU_H_MO;
    exports.PUBLISH_DANMAKU_H_MO_ERROR = PUBLISH_DANMAKU_H_MO_ERROR;
    exports.PUBLISH_DANMAKU_H_MO_SUCCESS = PUBLISH_DANMAKU_H_MO_SUCCESS;
    exports.PUBLISH_DANMAKU_SUCCESS = PUBLISH_DANMAKU_SUCCESS;
    exports.PULISHER_DANMAKU_H_MO_AROUSE = PULISHER_DANMAKU_H_MO_AROUSE;
    exports.SHOW_DANMAKU_H_MO = SHOW_DANMAKU_H_MO;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
