// @baidu/auc-star/index.ts
import styles from './style.module.less';
import {aucStarTypes} from './types';
import aucStarHeader from '@baidu/auc-star-header';
import aucStarContent from '@baidu/auc-star-content';
import aucStarFooter from '@baidu/auc-star-footer';
import {componentLog} from '@baidu/auc-log';
import AucComponent from "@baidu/auc-base/AucComponent";
import {jump} from "@baidu/auc-base/jump";
import {env} from '@baidu/auc-base/env';
import {topicVideoThunderLog} from '@baidu/auc-base/topicVideoThunderLog';

export class AucStar extends AucComponent {
    static template:string = `
    <div class="{{styles.AucStarWrap}} {{isBigEvent ? styles.bigEvent : ''}}">
        <div class="{{styles.aucStarActiveArea}} {{starActive ? styles.aucStarActiveAreaActive : ''}}"
            on-click="handleClick"
        >
            <auc-star-header
                auhomepageurl="{{auhomepageurl}}"
                auhomepagecmd="{{auhomepagecmd}}"
                auimage="{{auimage}}"
                auimageflag="{{auimageflag}}"
                auname="{{auname}}"
                auabs="{{auabs}}"
                vip="{{vflag}}"
                aumrid="{{aumrid}}"
                followtype="{{followtype}}"
                isBigEvent="{{isBigEvent}}"
                landType="{{landType}}"
                class="{{styles.aucHeader}}"
                on-log="log"
            ></auc-star-header>
            <auc-star-content
                title="{{title}}"
                titleLink="{{titleLink}}"
                layout="{{layout}}"
                images="{{images}}"
                type="{{type}}"
                nid="{{nid}}"
                on-back-shoubai="backShoubai"
                on-jump-detail="jumpDetail"
                on-out-jump="outOfBoxJump"
                on-set-active="setActive"
                on-delete-active="deleteActive"
            ></auc-star-content>
        </div>
        <auc-star-footer
            class="{{styles.AucStarFooter}}"
            title="{{title}}"
            shareImageUrl="{{shareImageUrl}}"
            link="{{link}}"
            commentcmd="{{commentcmd}}"
            comment_num="{{comment_num}}"
            praisenum="{{praisenum}}"
            zanid="{{zanid}}"
            isliked="{{isliked}}"
            icon_type="{{icon_type}}"
            box-prase-icon-list="{{boxPraseIconList}}"
            praise-style="{{praise_style}}"
            page-mode="{{pageMode}}"
            is-immersive="{{isImmersive}}"
            status-bar-height="{{statusBarHeight}}"
            on-back-shoubai="backShoubai"
            on-log="log"
        ></auc-star-footer>
    </div>
    `;
    static components:object = {
        'auc-star-header': aucStarHeader,
        'auc-star-footer': aucStarFooter,
        'auc-star-content': aucStarContent
    };

    initData(): aucStarTypes {
        return {
            styles,
            layout: '',
            feedback: '',
            id: '',
            //
            auhomepageurl: '', // 跳转url
            auhomepagecmd: {}, // 头部跳转cmd
            auimage: '', // 头像imgurl
            auimageflag: '', // v标url
            auname: '', // 名称
            auabs: '', // 简介
            vflag: '0',
            aumrid: '',// 作者id
            followtype: '', // 关注类型
            //
            link: '',
            commentcmd: '',
            comment_num: 0,
            praisenum: 0,
            zanid: '', // 赞id
            isliked: 0, // 是否喜欢
            title: '',
            //
            liketype: '', //数据来源
            type: '', // 模板类型
            items: [], // 图片集合
            time: '',
            titleLink: {}, // 头部跳转信息 包括话题，@等
            source: '', // 来源信息
            ts: '', // 时间
            nid: '', // nid
            rid: '', // rid
            description: '', // 描述信息
            duration: '',
            cmd: {},
            images:[],
            starActive: false,
            shareImageUrl: '',
            logExtraParams: {
            },
            icon_type: ''
        };
    }
    static computed = {

    };
    inited():void {
    }
    attached(): void {
        this.registerScrollInView().then(_ => {
            this.log('dt_show');
            this.data.get('type') === 'video' && this.videoThunderLog('click');
        });
    }

    updated(): void {
    }

    backShoubai(): void {
        if (/_refluxos=a8/.test(window.location.search) || env.isSSR) {
            return;
        }
        this.fire('backShoubai');
    }
    jumpDetail(): void {
        if (/_refluxos=a8/.test(window.location.search)) {
            return;
        }
        this.log('dt_click');
        this.data.get('type') === 'video' && this.videoThunderLog('click');
        let cmd = this.data.get('cmd');
        const link = this.data.get('link');
        jump({
            cmd,
            link
        })
    }
    handleClick():void {
        this.jumpDetail();
    }
    log(type):void {
        componentLog(type, '', this.data.get('logExtraParams'));
    }
    outOfBoxJump(): void {
        if (/_refluxos=a8/.test(window.location.search)) {
            return;
        }
        window.location.href = this.data.get('link');
    }
    setActive(event) {
        this.data.set('starActive', true);
    }
    deleteActive(event) {
        this.data.set('starActive', false);
    }

    videoThunderLog(type) {
        // 视频切换短小融合需单独打点
        const videoThunderLog = this.data.get('videoThunderLog');
        if (videoThunderLog.resourceType) {
            topicVideoThunderLog(type === 'show' ? 'show' : 'click', {
                rid: this.data.get('rid'),
                ...videoThunderLog,
                topic_id: this.data.get('topicId')
            });
        }
    }
}

export default AucStar;
