// @baidu/auc-star/types.ts
import {JumpProps} from '@baidu/auc-base/types/jumpProps';
import {LogExtraParams} from "@baidu/auc-base/types/logExtraParams";

export interface aucStarTypes {
    styles: object;
    layout: string; // 类型
    feedback: string; // 打点参数
    id: string; // id
    starActive: boolean;
    //
    auhomepageurl: string; // 跳转url
    auhomepagecmd: object; // 头部跳转cmd
    auimage: string; // 头像imgurl
    auimageflag: string; // v标url
    auname: string; // 名称
    auabs: string; // 简介
    vflag: string; // vip等级
    aumrid: string; // 作者id
    followtype: string; // 关注类型
    //
    link: string; // 动态跳转url
    commentcmd: ''; // 评论cmd
    comment_num: number; // 评论数
    praisenum: number; // 点赞数
    zanid: string; // 赞id
    isliked: number; // 是否喜欢
    title: string;// 标题
    //
    liketype: string; //数据来源
    type: string; // 模板类型
    items: object[]; // 图片集合
    time: string;
    titleLink: object; // 头部跳转信息 包括话题，@等
    source: string; // 来源信息
    ts: string; // 时间
    nid: string; // nid
    rid: string; // rid
    description: string; // 描述信息
    cmd: JumpProps; // 跳转信息
    images: object[];
    duration: string;
    shareImageUrl: string;
    logExtraParams: LogExtraParams;
    icon_type: string; //点赞图标是否特殊显示 0普通 1 缅怀
}
