// @baidu/auc-topic-trans/componentLogExtra.ts
import {LogExtraParams} from "@baidu/auc-base/types/logExtraParams";

export interface ComponentLogExtra extends LogExtraParams {
    column_type: string;
    column_id: string;
    nid: string;
    res_type: string;
}

export interface videoThunderLog extends LogExtraParams {
    resourceType: string;
    landingType: string;
}

export const getBaseComponentLogExtra = (
    newsData:any,
    columnType:string,
    columnIndex:string|number,
    res_type?: string
):ComponentLogExtra => {
    const data = newsData.data;
    res_type = res_type ? res_type : data.item_type;
    if (newsData.layout.startsWith('star')) {
        res_type = 'dt';
    }
    return {
        column_type: columnType,
        column_id: columnIndex + '',
        nid: data.rid,
        res_type
    };
};

export const getBaseVideoThunderLog = (
    newsData:any,
    columnType:string,
    columnIndex:string|number,
    res_type?: string
) => {
    const data = newsData.data;
    let result:videoThunderLog = {
        resourceType: data.resource_type
    };
    data.landing_type ? result.landingType = data.landing_type : '';
    return result;
};
