// @baidu/auc-topic-trans/guardian.ts

import {transJumpProps, transitionBigImageIcon} from '@baidu/auc-base/dataTransition';


export default (newsData, columnType, columnIndex) => {
    const data = newsData.data;

    let isLive = false;
    let isPlayBack = false;
    let isPreload = false;

    if (newsData.layout === 'livevideo' || newsData.tplType === 'large') {
        if (data.title_prefix === 'live') {
            isLive = true;
        } else if (data.title_prefix === 'playback') {
            isPlayBack = true;
        } else if (data.title_prefix === 'preload') {
            isPreload = true;
        }
    } else if (newsData.tplType === 'large') {
        if (data.livetype === 'livevideo') {
            isLive = true;
        } else if (data.livetype === 'playback') {
            isPlayBack = true;
        }
    }

    return {
        time: data.time,
        timelineList: [{
            type: 'guardian',
            title: data.summary || data.title,
            kolName: data.kol_name,
            description: data.description,
            source: data.source,
            imageProps: transitionBigImageIcon(newsData),
            jumpProps: transJumpProps(newsData),
            logExtraParams: {
                column_type: columnType,
                column_id: columnIndex,
                nid: data.rid,
                res_type: data.item_type
            },
            widgetProps: {
                items: [{
                    text: `来自${data.source}`
                }]
            },
            isLive,
            isPlayBack,
            isPreload
        }]
    };
};
