// @baidu/auc-topic-trans/header.ts

import {AucHeaderData} from '@baidu/auc-header/types';
export const bingEventBgs = ['#444954', '#1167F1', '#B61D27', '#2A4B8C'];


export default (topicData):AucHeaderData => {
    const landType = +topicData.land_type;
    const headImgColor = landType !== 0 ? topicData.land_bg_color ? topicData.land_bg_color : bingEventBgs[landType - 1] : topicData.head_img_color
    let imgUrl = '';
    let imgType = 1;
    if (!topicData.img_url) {
        imgUrl = 'https://b.bdstatic.com/searchbox/icms/searchbox/img/default-head-1626935858.jpg';
        imgType = 1;
    } else {
        imgUrl = topicData.img_url;
        imgType = +topicData.img_type;
    }

    return {
        title: topicData.title,
        imgType,
        imgUrl,
        description: topicData.description,
        hotType: +topicData.hot_type,
        readingNum: +topicData.reading_num,
        authorNameList: topicData.bjh_author_info.author_namelist,
        authorImgList: topicData.bjh_author_info.author_imglist,
        authorNum: topicData.bjh_author_info.author_num,
        isNewBanner: topicData.modify_time > 1599567000,
        landType: +topicData.land_type || 0,
        isBigLogo: landType === 5,

        modifyTime: +topicData.modify_time,
        headImgLogo: topicData.head_img_logo,
        headImgColor: headImgColor,
        isBigEvent: topicData.land_type !== 0
    };
};
