// @baidu/hd-actions/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Copyright (C) 2021 ~ present baidu.com
 * @description 枚举常量
 * @author fangbinbin02 <fangbinbin02@baidu.com>
 */
// 操作状态
exports.ACTION_STATUS = void 0;
(function (ACTION_STATUS) {
    ACTION_STATUS["SUCCESS"] = "success";
    ACTION_STATUS["FAIL"] = "fail";
})(exports.ACTION_STATUS || (exports.ACTION_STATUS = {}));
exports.HOVER_STATUS = void 0;
(function (HOVER_STATUS) {
    HOVER_STATUS["ENTER"] = "enter";
    HOVER_STATUS["LEAVE"] = "leave";
})(exports.HOVER_STATUS || (exports.HOVER_STATUS = {}));
exports.DANMAKU_TYPE = void 0;
(function (DANMAKU_TYPE) {
    DANMAKU_TYPE["HORIZONTAL"] = "horizontal";
    DANMAKU_TYPE["VERTICAL"] = "vertical";
})(exports.DANMAKU_TYPE || (exports.DANMAKU_TYPE = {}));
exports.UI_MODE = void 0;
(function (UI_MODE) {
    UI_MODE["SMALL"] = "6N";
    UI_MODE["LARGE"] = "8N";
})(exports.UI_MODE || (exports.UI_MODE = {}));

/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @description 事件 action的类型定义文件
 * @author fangbinbin02 <fangbinbin02@baidu.com>
 */
/**
 * action: 赞点击事件
 */
var CLICK_PRAISE = 'click:praise';
/**
 * action: 发布器表情点击事件
 */
var CLICK_PUBLISHER_EMOJI = 'click:publisher:emoji';
/**
 * action: 发布器发表按钮点击事件
 */
var CLICK_PUBLISHER_PUBLISH = 'click:publisher:publish';
/**
 * action: 触发toast
 */
var INVOKE_TOAST = 'invoke:toast';
/**
 * action: 表情选择器单个表情点击事件
 */
var CLICK_EMOJIPICKER_ITEM = 'click:emojiPicker:item';
/**
 * action: 单条弹幕的鼠标hover事件
 */
var HOVER_DANMAKU_ITEM = 'hover:danmaku:item';
/**
 * action: 单条表态点击事件
 */
var CLICK_OPINION_ITEM = 'click:opinion:item';

exports.CLICK_EMOJIPICKER_ITEM = CLICK_EMOJIPICKER_ITEM;
exports.CLICK_OPINION_ITEM = CLICK_OPINION_ITEM;
exports.CLICK_PRAISE = CLICK_PRAISE;
exports.CLICK_PUBLISHER_EMOJI = CLICK_PUBLISHER_EMOJI;
exports.CLICK_PUBLISHER_PUBLISH = CLICK_PUBLISHER_PUBLISH;
exports.HOVER_DANMAKU_ITEM = HOVER_DANMAKU_ITEM;
exports.INVOKE_TOAST = INVOKE_TOAST;
