// @baidu/auc-heated-discussion/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import {Component} from 'san';
import styles from './style.module.less';
import {AucHeatedDiscussionData} from './types';
import {WujiAvatar} from '@baidu/wuji-san';
import AucPicTextSm from '@baidu/auc-pic-text-sm';
import {componentLog} from '@baidu/auc-log';

import defaultAvatar from './img/default-avatar.jpeg';

export class AucHeatedDiscussion extends Component {
    static template:string = /* html */ `
    <div class="{{styles.heatedDiscussion}} {{isBigEvent ? styles.isBigEvent : ''}}">
        <div class="{{styles.authorArea}}">
            <wuji-avatar size="84" src="{{avatarSrc || defaultAvatar}}" class="{{styles.authorAvatar}}"></wuji-avatar>
            <div class="{{styles.authorInfoArea}}">
                <div class="{{styles.authorName}}">{{authorName}}</div>
                <div class="{{styles.time}}">{{time}}</div>
            </div>
        </div>
        <div class="{{styles.contentArea}} {{styles.mar}}" s-if="{{description}}">
            <div class="{{styles.descriptionWrapper}}">
                <div class="{{styles.quot}}"></div>
                <div class="{{styles.description}}">{{description}}</div>
            </div>
        </div>
        <div class="{{styles.aucPicTextWrapper}}">
            <auc-pic-text-sm
                on-click-item="handleClick"
                title="{{title}}"
                imageProps="{{imageProps}}"
                s-if="!isUgc"
                class="{{styles.heatedPicText}} {{description ? styles.mar : ''}}"
            ></auc-pic-text-sm>
        </div>
    </div>
    `;
    static components:object = {
        'wuji-avatar': WujiAvatar,
        'auc-pic-text-sm': AucPicTextSm
    };

    initData(): AucHeatedDiscussionData {
        return {
            styles,
            isUgc: false,
            authorName: '',
            avatarSrc: '',
            description: '',
            imageProps: {},
            time: '',
            title: '',
            defaultAvatar,
            logBaseType: 'heated_disc_',
            jumpProps: {},
            logExtraParams: {}
        };
    }

    handleClick() {
        this.fire('jump', this.data.get('jumpProps'));
        componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
    }

    attached(): void {
    }
}

export default AucHeatedDiscussion;
