// @baidu/auc-three-image-text/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucThreeImageTextData} from './types';
import {WujiTitle, WujiWidget, WujiThreeImages} from '@baidu/wuji-san';
import '@baidu/wuji-icons/svg/d20-capsule-img-f-30.svg';
import {componentLog} from '@baidu/auc-log';

export class AucThreeImageText extends AucComponent {
    static template:string = /* html */ `
    <div
        class="{{styles.aucThreeImageTextWrapper}} {{visited ? styles.visited : ''}} {{isBigEvent ? styles.bigEvent : ''}}"
        on-click="handleClick"
    >
        <div class="{{styles.aucThreeImageText}}">
            <div class="{{styles.title}}">
                <wuji-title
                    text="{{title}}"
                    spacing="{{false}}"
                    label-props="{{labelProp}}"
                    visited="{{visited}}"
                ></wuji-title>
            </div>
            <div class="{{styles.imageWrapper}}">
                <wuji-three-image
                    s-if="{{showThreeImage}}"
                    img-list="{{imgList}}"
                    spacing="{{false}}"
                    margin="{{margin}}"
                ></wuji-three-image>
            </div>
            <div class="{{styles.widget}}"
                s-if="{{widgetItems && widgetItems.length > 0}}"
            >
                <wuji-widget
                    has-margin-top="{{false}}"
                    has-margin-bottom="{{false}}"
                    items="{{widgetItems}}"
                ></wuji-widget>
            </div>
        </div>
    </div>
    `;
    static components:object = {
        'wuji-title': WujiTitle,
        'wuji-widget': WujiWidget,
        'wuji-three-image': WujiThreeImages
    };
    static computed = {
        imgStr() {
            let imgList = this.data.get('imgList');
            return JSON.stringify(imgList);
        }
    };

    initData(): AucThreeImageTextData {
        return {
            styles,
            title: '',
            labelProp: {},
            widgetItems: [],
            imgList: [],
            visited: false,
            margin: 0,
            showThreeImage: false,
            logBaseType: 'three_image_',
            logExtraParams: {},
            jumpProps: {}
        };
    }

    attached(): void {
        // 兼容一个疑似ssr的BUG
        this.data.set('showThreeImage', true);

        this.registerScrollInView().then(() => {
            componentLog(this.data.get('logBaseType'), 'show', this.data.get('logExtraParams'));
        });
    }

    handleClick(): void {
        this.fire('jump', this.data.get('jumpProps'));
        this.data.set('visited', true);
        componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
    }
}

export default AucThreeImageText;
