// @baidu/auc-header/index.ts
/**
 * @file 专题头图组件
 * @author gaohong01 <gaohong01@baidu.com>
 */

import './style.less';
import {Component} from 'san';
import {AucHeaderData} from './types';
import HeaderTitle from './components/header-title';
import HeaderDesc from './components/header-description';
import {WujiImage, HeaderAuthorInfo} from '@baidu/wuji-san';
import {hex2rgb} from './hex2rgb';

const prefixCls = 'w-topic-header';

export default class AucHeader extends Component {
    static template = /* html */ `
    <div wnid="wn0" class="header {{noHeadDesc ? 'noHeadDesc' : ''}}">
        <div class="{{topicHeaderClass}}">
            <template s-if="{{!(!isThirdEdition && imageType === 2))}}">
                <div class="header-container">
                <div
                    s-if="isThirdEdition && headImgLogo && isBigLogo"
                    s-ref="bigLogo"
                    class="{{isBigLogo ? 'big-logo' : ''}}"
                    style="{{logoStyle}}"
                ></div>
                    <div class="image-wrapper">
                        <wuji-image
                            src="{{imgUrl}}"
                            ratio="{{imgRatio}}"
                            banner="{{true}}"
                        ></wuji-image>
                        <div 
                            class="{{maskCls}}"
                            style="{{maskStyle}}">
                        </div>
                        <div class="defaultBlock" style="{{defaultBlock}}"></div>
                    </div>
                    <div class="main-area">
                        <div
                            s-if="isThirdEdition && headImgLogo && !isBigLogo"
                            class="topic-label"
                            style="{{headImgLogo ? 'background-image:url(' + headImgLogo + ')': ''}}"
                        ></div>
                        <div class="title-area">
                            <header-title
                                title="{{title}}"
                                type="{{imageType}}"
                                is-third-edition="{{isThirdEdition}}"
                                class="header-title {{isShowHeader ? 'header-title-nomb' : ''}}">
                            </header-title>
                            <header-author-info
                                s-if="isShowHeader"
                                class="{{isThirdEdition && (imgType === 1 || imgType === 0)
                                    ? 'new-subsidiary-top' : 'subsidiary-top'}}"
                                desc-color="FC6"
                                author-desc="{{authorDesc}}"
                                author-img-list="{{authorImgList}}"
                                border-width="{{1}}"
                            ></header-author-info>
                        </div>
                    </div>
                </div>
                <header-desc
                    s-if="description"
                    class="header-round"
                    desc="{{description}}"
                    headerRound="{{headerRound}}"
                    landType="{{landType}}"
                ></header-desc>
            </template>
            <template s-else-if="{{!isThirdEdition && imageType === 2}}">
                <div class="image-wrapper {{!isThirdEdition && imageType === 2 ? 'old-image-wrapper' : ''}}">
                    <wuji-image
                        src="{{imgUrl}}"
                        ratio="{{isSecondEdition ? '3:1' : '5:1'}}"
                        banner="{{true}}"
                        backgroundSize="{{fill}}"
                    ></wuji-image>
                    <div class="image-mask"></div>
                </div>
                <div class="content header-round" headerRound="{{headerRound}}">
                    <div class="title-area">
                        <header-title
                            title="{{title}}"
                            type="{{imageType}}"
                        ></header-title>
                    </div>
                    <header-desc
                        desc="{{description}}"
                        class="description-wrapper"
                        s-if="description"
                        landType="{{landType}}"
                    >
                    </header-desc>
                    <header-author-info
                        class="auth-area"
                        s-if="isShowHeader"
                        desc-color="FC2"
                        author-desc="{{authorDesc}}"
                        author-img-list="{{authorImgList}}"
                        border-width="{{1}}"
                    ></header-author-info>
                </div>
            </template>
        </div>
    </div>
    
    `;

    static trimWhitespace = 'all';

    static components = {
        'header-title': HeaderTitle,
        'header-author-info': HeaderAuthorInfo,
        'header-desc': HeaderDesc,
        'wuji-image': WujiImage,
    };

    static computed = {
        topicHeaderClass() {
            return [
                prefixCls,
                !this.data.get('isThirdEdition') && this.data.get('imageType') === 2 ? 'old-w-topic-header' : '',
                this.data.get('noHeadDesc') ? 'noHeadDesc' : '',
            ];
        },
        isShowHeader() {
            const authorNameList = this.data.get('authorNameList') || [];
            return +this.data.get('hotType') === 1 && authorNameList.length > 0;
        },
        authorDesc() {
            const isShowHeader = this.data.get('isShowHeader');
            const authorNameList = this.data.get('authorNameList') || [];
            const authorNum = this.data.get('authorNum');
            let str = '';

            if (!isShowHeader) {
                return str;
            }

            if (authorNameList[0]) {
                str += authorNameList[0];
            }

            if (authorNameList[1]) {
                str += '、' + authorNameList[1];
            }

            return `${str}等${authorNum}个百家号参与`;
        },
        imageType() {
            const imgUrl = this.data.get('imgUrl');
            const imgType = this.data.get('imgType');
            const isThirdEdition = this.data.get('isThirdEdition');
            return !imgUrl ? (isThirdEdition ? 2 : 1) : +imgType;
        },
        maskCls() {
            const imgUrl = this.data.get('imgUrl');
            return !imgUrl && !this.data.get('isBigEvent') ? '' : 'image-mask';
        },
        // 头部区域一、二版：区别1头图尺寸改版
        isSecondEdition() {
            const modifyTime = this.data.get('modifyTime');
            return modifyTime > 1599567000 && modifyTime < 1626852919;
        },
        // 头部区域二、三版区分：区别1：头图尺寸改版，区别2：统一将标题、辅助信息放到图上
        isThirdEdition() {
            const modifyTime = this.data.get('modifyTime');
            return modifyTime > 1626852919; // 1625443200 - 2021-07-21 15:35:19 人机具体上线时间
        },
        // 一二三版大图&三版大图 ratio
        imgRatio() {
            const isThirdEdition = this.data.get('isThirdEdition');
            const isSecondEdition = this.data.get('isSecondEdition');
            const imageType = +this.data.get('imageType'); // 0 无图 1 大图 2 小图；无图按大图处理
            let ratio = '4:3'; // 默认值：第三版大图尺寸
            if (isThirdEdition) {
                if (imageType === 2) {
                    ratio = '16:9'; // 第三版小图尺寸
                }
            }
            else {
                // 第一、二版小图样式独立
                if (isSecondEdition) {
                    ratio = '16:9'; // 第二版大图
                }
                else {
                    ratio = '5:2'; // 第一版大图
                }
            }
            return ratio;
        },
        maskStyle() {
            const imgBg = this.data.get('headImgColor');
            let imgBgObj = hex2rgb(imgBg);
            return imgBg ? `background:linear-gradient(to bottom, ${imgBgObj.rgb0}, ${imgBgObj.rgb});` : '';
        },
        headerRound() {
            const imgBg = this.data.get('headImgColor');
            let imgBgObj = hex2rgb(imgBg);
            return this.data.get('isBigEvent') && imgBg ? `background-color: ${imgBgObj.rgb};` : '';
        },
        defaultBlock() {
            const imgBg = this.data.get('headImgColor');
            let imgBgObj = hex2rgb(imgBg);
            return imgBg ? `background-color: ${imgBgObj.rgb};` : '';
        },
        logoStyle() {
            let result;
            const headImgLogo = this.data.get('headImgLogo');
            if (headImgLogo) {
                result = `background-image: url(${headImgLogo})`;
            }
            const isBigLogo = this.data.get('isBigLogo');
            const bigLogoDom = this.data.get('bigLogo')
            const statusBarHeight = +this.data.get('statusBarHeight');
            const top = bigLogoDom && statusBarHeight + 15;

            if (isBigLogo) {
                result = `background-image: url(${headImgLogo}); top: ${top}px;`;
            }
            return result;
        }
    };

    attached() {
        this.data.set('bigLogo', this.ref('bigLogo'));
    }

    initData(): AucHeaderData {
        return {
            title: '',
            imgType: 0,
            imgUrl: '',
            description: '',
            hotType: 0,
            authorNameList: [],
            authorImgList: [],
            authorNum: 0,
            modifyTime: 0,
            headImgLogo: '',
            headImgColor: '',
            isBigEvent: false,
            landType: 0,
            readingNum: 0,
            isNewBanner: false,
            isBigLogo: false
        };
    }
}
