// @baidu/auc-operation-position/components/opetration-kingkong/item.ts
/**
 * @file index.ts
 * @author gaohong01
 */
import {Component} from 'san';
import styles from './style.module.less';
import {kingKongItemData} from './types';
import {machiningColor} from '@baidu/auc-base/utils';

export class kingKongItem extends Component {
    static template: string = /* html */ `
        <template>
            <div class="{{styles.aucKingkongIcon}}" style="background-image:url({{itemIcon}})"></div>
            <div class="{{styles.aucKingkongText}}" style="color:{{divColor}}">{{divTitle}}</div>
        </template>
     `;
    static components: object = {
    };

    static computed: object = {
        itemIcon() {
            let divIcon = this.data.get('divIcon');
            const pageMode = this.data.get('pageMode');
            return divIcon ? typeof divIcon === 'string' ? divIcon : divIcon[pageMode === 'dark' ? 'day' : pageMode] : '';
        },
        divColor() {
            let divColor = this.data.get('divWordColor');
            const pageMode = this.data.get('pageMode');
            return divColor ? machiningColor(typeof divColor === 'string' ? divColor : divColor[pageMode]) : '';
        }
    };

    initData(): kingKongItemData {
        return {
            styles,
            divIcon: '',
            divWordColor: '',
            pageMode: 'day',
            divTitle: ''
        };
    }

    inited(): void {
    }

    attached(): void {
    }
}

export default kingKongItem;
