// @baidu/wuji-san/es/wuji-live-status/index.js
/**
 * @file 组件 WujiLiveStatus 无极-直播状态
 * @author panwenshuang <panwenshuang@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';
import '@baidu/wuji-icons/svg/d20-search-o-66.svg';

const prefixCls = 'w-live-status';
const validColor = [
    'red',
    'blue',
    'gray',
];
const tenThousand = 10000;
const hundredMillon = 100000000;

export default class WujiLiveStatus extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap" on-click="handleClick">
            <div class="{{statusClass}}">
                <svg s-if="isLiving" viewBox="0 0 33 33" class="${prefixCls}-live-icon ${prefixCls}-icon">
                    <rect class="${prefixCls}-line ${prefixCls}-line1" x="0" y="0" width="5" height="14" rx="3"/>
                    <rect class="${prefixCls}-line ${prefixCls}-line2" x="10" y="0" width="5" height="14" rx="3"/>
                    <rect class="${prefixCls}-line ${prefixCls}-line3" x="20" y="0" width="5" height="14" rx="3"/>
                </svg>
                <wuji-icon s-else type="{{icon}}" size="33" class="${prefixCls}-icon" />
                <span>{{statusText}}</span>
            </div>
            <div s-if="num || numberText" class="${prefixCls}-number">
                <span>{{num}}{{numberText}}</span>
            </div>
        </div>
    `;

    static validColor = validColor;

    static components = {
        'wuji-icon': Icon,
    };

    static dataTypes = {
        isLiving: DataTypes.bool,
        statusText: DataTypes.string,
        number: DataTypes.oneOfType([
            DataTypes.number,
            DataTypes.string,
        ]),
        numberText: DataTypes.string,
        icon: DataTypes.string,
        color: DataTypes.oneOf(validColor),
    };

    static computed = {
        statusClass() {
            const color = this.data.get('color');
            return `${prefixCls} ${prefixCls}-${color}`;
        },
        num() {
            const number = this.data.get('number');
            const text = this.data.get('numberText');

            if (!number) {
                return '';
            }

            if (number < tenThousand) {
                return number + '' + text;
            }

            // 保留 1 位小数
            const decimalFactor = 10;

            if (number >= tenThousand && number < hundredMillon) {
                const newVal = number / tenThousand;
                return Math.round(newVal * decimalFactor) / decimalFactor + '万';
            }

            if (number >= hundredMillon) {
                const newVal = number / hundredMillon;
                return Math.round(newVal * decimalFactor) / decimalFactor + '亿';
            }
        },
    };

    initData() {
        return {
            isLiving: true,
            statusText: '直播中',
            color: 'red',
            number: 12345,
            numberText: '观看',
        };
    }

    handleClick(event) {
        this.fire('click', {event});
    }
}
