// @baidu/wuji-san/es/wuji-popup/index.js
/**
 * @file 组件 WujiPopup 无极-弹窗
 * @author yangmei03 <yangmei03@baidu.com>
 *
 * TODO: 代码风格和一些实现细节都需要优化
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';
import List from './list';
import '@baidu/wuji-icons/svg/popupmenu-close-o-48.svg';
import '@baidu/wuji-icons/svg/arrow-back-o-48.svg';

const prefixCls = 'w-popup';

export default class WujiPopup extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap">
            <div class="${prefixCls}-mask" on-click="close" s-if="isOpen"/>
            <div class="${prefixCls}-box ${prefixCls}-{{type}}" s-if="isOpen">
                <div s-if="showToolbar" class="${prefixCls}-title">
                    <div class="${prefixCls}-left" on-click="close">
                        <wuji-icon
                            s-if="type === 'answer'"
                            type="arrow-back-o-48"
                        />
                        <div s-if="type === 'calendar'">取消</div>
                    </div>
                    <div class="${prefixCls}-title-text">{{title}}</div>
                    <div class="${prefixCls}-right" on-click="close">
                        <div s-if="type === 'calendar'">完成</div>
                        <wuji-icon s-else type="popupmenu-close-o-48"/>
                    </div>
                </div>
                <div class="${prefixCls}-content" style="max-height: {{contentHeight}}px">
                    <span s-if="type === 'text'">{{content}}</span>
                    <img
                        s-else-if="type === 'image'"
                        src="{{imageUrl}}"
                        class="image-box"
                    />
                    <wuji-list
                        s-else
                        options="{{options}}"
                        value="{{value}}"
                        type="{{type}}"
                        buttonText="{{buttonText}}"
                        on-select="handleSelect"
                        on-button-click="handleButtonClick"
                    />
                </div>
            </div>
        </div>
    `;

    static components = {
        'wuji-icon': Icon,
        'wuji-list': List,
    };

    static dataTypes = {
        title: DataTypes.string,
        type: DataTypes.string,
    };

    initData() {
        return {
            value: '',
            content: '',
        };
    }

    attached() {
        this.watch('isOpen', value => {
            if (value) {
                this.nextTick(() => {
                    const windowHeight = document.documentElement.offsetHeight;
                    const boxHeight = windowHeight * 0.7;
                    const title = this.el.querySelector(`.${prefixCls}-title`);
                    const titleHeight = title ? title.offsetHeight : 0;
                    this.data.set('contentHeight', boxHeight - titleHeight);
                });
            }
        });
    }

    close(event) {
        this.data.set('isOpen', false);
        this.fire('close', {event});
    }

    handleSelect({event, value}) {
        this.data.set('value', value);
        this.fire('select', {event, value});
    }

    handleButtonClick({event}) {
        this.close(event);
    }
}
