// @baidu/wuji-san/es/wuji-popup/list.js
/** @file 组件 WujiPopup wuji弹窗带表格
 * @author yangmei03 <yangmei03@baidu.com>
 *
 * TODO: 代码风格和一些实现细节都需要优化
 */

import './style/list.less';
import {Component} from 'san';
import Icon from '../wuji-icon';
import Button from '../wuji-button';
import '@baidu/wuji-icons/svg/popupmenu-selected-o-48.svg';
import '@baidu/wuji-icons/svg/arrow-goto-o-48.svg';

const prefixCls = 'w-popup-options';

export default class WujiPopupList extends Component {
    static template = /* html */ `
        <div s-if="options" class="${prefixCls} ${prefixCls}-{{type}}">
            <div
                s-for="item in options"
                class="${prefixCls}-items {{item.value === value ? 'selected' : ''}}"
                on-click="handleSelect($event, item)"
            >
                <div class="options-left">
                    <template s-if="item.icon">
                        <wuji-icon
                            s-if="type === 'iconList'"
                            class="options-icon"
                            type="{{item.icon}}"
                        />
                        <span
                            s-else
                            class="options-avatar"
                            style="background-image: url({{item.icon}})"
                        />
                    </template>
                    <div class="options-text {{item.subTitle? 'options-text-title': ''}}">
                        <span>{{item.text}}</span>
                        <span s-if="item.subTitle" class="options-subtitle">
                            {{item.subTitle}}
                        </span>
                    </div>
                </div>
                <wuji-icon type="{{iconType}}" class="options-selected"/>
            </div>
            <div s-if="type === 'answer'" class="options-button">
                <wuji-button
                    text="{{buttonText}}"
                    theme="outline"
                    color="dark-gray"
                    height="114"
                    on-click="handleButtonClick"
                />
            </div>
        </div>
    `;

    static components = {
        'wuji-icon': Icon,
        'wuji-button': Button,
    };

    static computed = {
        iconType() {
            return this.data.get('type') === 'iconList' ? 'arrow-goto-o-48' : 'popupmenu-selected-o-48';
        },
    };

    initData() {
        return {
            type: 'list',
        };
    }

    handleSelect(event, item) {
        this.fire('select', {event, value: item.value});
    }

    handleButtonClick(event) {
        this.fire('button-click', {event});
    }
}
