// @baidu/xbox/src/browser/load-file.ts
/**
 * @file Created on Tue Nov 27 2018
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import {isFunction} from '../is';

const noop = () => {};

type LoadFileCallback = (err: any, path: string) => void;
interface LoadFileParams {
    async?: boolean;
    maxTries?: number;
    timeout?: number;
    before?: (path: string, elem: Element) => unknown;
}

/**
 * 加载单个文件，支持 css/img/js
 *
 * @param path url 地址
 * @param callback 回调函数
 * @param params 配置参数，包括 async，timeout，maxTries，before 等
 * @param numTries 不对外参数，用于重试次数
 * @example
 * loadFile('/path/to/foo.js', function(err, path) {
 *     // foo.js loaded
 * });
 * loadFile('css!/path/to/foo', function(err, path) {
 *     // foo css loaded
 * });
 * loadFile('img!/path/to/foo', function(err, path) {
 *     // foo img loaded
 * });
 */
export function loadFile(path: string, callback: LoadFileCallback, params: LoadFileParams, numTries = 0) {
    const doc = document;
    const pathStripped = path.replace(/^(css|img)!/, '');
    const isCSS = /(^css!|\.css$)/.test(path);
    const {async = true, maxTries = 1, timeout = 2e4, before} = params;
    const beforeCallback = before && isFunction(before) ? before : noop;

    // eslint-disable-next-line @typescript-eslint/init-declarations
    let elem: any;
    let timer = 0;

    const cb = (ev: Event | Error) => {
        let result = '';
        timer && clearTimeout(timer);

        if (ev instanceof Error) {
            result = 'e';
        }
        else {
            result = ev.type[0];
        }

        // Note: The following code isolates IE using `hideFocus` and treats empty
        // stylesheets as failures to get around lack of onerror support
        if (isCSS && 'hideFocus' in elem) {
            try {
                if (!elem.sheet.cssText.length) {
                    result = 'e';
                }
            }
            catch (x) {
                // sheets objects created from load errors don't allow access to
                // `cssText`
                result = 'e';
            }
        }

        // handle retries in case of load failure
        if (result === 'e') {
            // increment counter
            numTries += 1;

            // exit function and try again
            if (numTries < maxTries) {
                loadFile(path, callback, params, numTries);
                return;
            }
        }
        elem.onload = elem.onerror = elem.onbeforeload = null;
        elem = null;
        // error first
        callback(result === 'e' ? (ev ? ev : 'error') : null, path);
    };

    if (isCSS) {
        // css
        elem = doc.createElement('link');
        elem.rel = 'stylesheet';
        elem.href = pathStripped;
    }
    else if (/(^img!|\.(png|gif|jpg|svg)$)/.test(path)) {
        // image
        elem = doc.createElement('img');
        elem.src = pathStripped;
    }
    else {
        // javascript
        elem = doc.createElement('script');
        elem.src = path;
        elem.async = !!async;
    }

    elem.onload = elem.onerror = cb;

    if (beforeCallback(path, elem) !== false) {
        const head = doc.head || doc.getElementsByTagName('script')[0];
        head.appendChild(elem);
        // timeout
        if (timeout > 0) {
            timer = window.setTimeout(() => {
                cb(new Error('timeout'));
            }, timeout);
        }
    }
}
