// @baidu/xbox-native/es5/device/cuid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _detect = require("../detect");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module device
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
var boxVersion = (0, _detect.getBoxVersion)();
(0, _register.default)('utils/getCUID', {
  androidlite: {
    versionRange: ['2.2.2'],
    thunk: function thunk(res) {
      var cuid = res.result;
      return {
        status: 0,
        data: {
          data: cuid
        }
      };
    }
  },
  androidmission: {
    versionRange: ['1.0'],
    thunk: function thunk(res) {
      var cuid = res.result;
      return {
        status: 0,
        data: {
          data: cuid
        }
      };
    }
  },
  iosmission: {
    versionRange: ['1.0'],
    thunk: function thunk(res) {
      try {
        // 非加密
        var _cuid = res.unique_id || res.result;

        return {
          status: 0,
          data: {
            data: _cuid
          }
        };
      } catch (e) {
        console.log(e);
      }
    }
  },
  androidinfo: {
    versionRange: ['1.0'],
    thunk: function thunk(res) {
      var cuid = res.result;
      return {
        status: 0,
        data: {
          data: cuid
        }
      };
    }
  },
  iosinfo: {
    versionRange: ['1.1'],
    thunk: function thunk(res) {
      try {
        // 非加密
        var _cuid2 = res.unique_id;
        return {
          status: 0,
          data: {
            data: _cuid2
          }
        };
      } catch (e) {
        console.log(e);
      }
    }
  },
  ios: {
    versionRange: ['6.1'],
    filter: function filter() {
      return {
        module: 'utils',
        action: 'getCUID'
      };
    },
    thunk: function thunk(res) {
      if ((0, _detect.version_compare)(boxVersion, '8.4') >= 0) {
        try {
          // 非加密
          var _cuid3 = res.unique_id;
          return {
            status: 0,
            data: {
              data: _cuid3
            }
          };
        } catch (e) {
          console.log(e);
        }
      } else {
        var reg = /baiduboxapp\/.*\/(\w+)\/\d/;
        var ua = navigator.userAgent;
        var list = ua.match(reg); // 非加密

        var _cuid4 = list ? list[1] : '';

        return {
          status: 0,
          data: {
            data: _cuid4
          }
        };
      }
    }
  },
  android: {
    versionRange: ['6.5'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_utils',
        action: 'getcuid',
        async: false
      };
    },
    thunk: function thunk(res) {
      if ((0, _detect.version_compare)(boxVersion, '7.4') >= 0) {
        var _cuid5 = res.result;
        return {
          status: 0,
          data: {
            data: _cuid5
          }
        };
      } else {
        var match = document.cookie.match(/BAIDUCUID=(.+?);/);

        var _cuid6 = match ? match[1] : '';

        return {
          status: 0,
          data: {
            data: _cuid6
          }
        };
      }
    }
  }
});
/**
 * 获取 cuid
 * @function cuid
 */

var _default = _index.cuid;
exports.default = _default;
