// @baidu/xbox-native/es5/device/screenInfo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 获取 sreenInfo
 * @module device
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
function thunk(res) {
  return {
    status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
    data: res.data,
    message: res.errmsg
  };
}

(0, _register.default)('utils/getScreenInfo', {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['7.3'],
    filter: function filter() {
      return {
        module: 'utils',
        action: 'getDeviceInfo'
      };
    },
    thunk: thunk
  },
  android: {
    versionRange: ['7.3'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_utils',
        action: 'getDeviceInfo'
      };
    },
    thunk: thunk
  }
});
/**
 * 获取 screenInfo 网络环境
 * @function screenInfo
 * @param {any} abtest - abtest 参数
 */

var _default = function _default(abtest) {
  return (0, _index.info)(['screenInfo'], abtest);
};

exports.default = _default;
