// boxjs/dist/thirdParty/bdbox/monitor/multilog.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /**
   * @file multilog.js 改进pblog多id冲突
   *
   * @author Created by youliang on 17/2/4.
   */

  /* eslint-disable fecs-camelcase */


  var monitor = require('./monitor');

  var extend = require('../extend');

  var query2Json = require('../utils/queryToJson');

  var getVersion = require('../utils/getVersion');

  var clone = require('../clone'); // url中存在公共参数，则给tcbox拼上，透传


  var URL = query2Json(location.search);
  var ua = navigator.userAgent;
  var protocol = window.location.protocol;

  if (protocol !== 'http:') {
    protocol = 'https:';
  }

  var GIF_URI = protocol + '//m.baidu.com/tcbox';
  var urlencode = encodeURIComponent; // 开始改造

  var Multilog = function (cateid) {
    // 继承monitor
    var self = this;
    this.CUSTOM_ARG = {
      service: 'bdbox',
      action: 'pblog',
      // 表明参数协议第二个版本
      ctv: 2,
      // 表明参数加密
      cen: 'uid_ua_ut',
      data: {
        // 手百为1
        appid: '1',
        // 客户端：1、JS：2、SERVER：3
        dataid: '2',
        // 区分主动/被动行为，主动：1、被动：0
        actiontype: '1',
        // 写死, actionid：2=事件统计，1=pv统计
        actionid: '2',
        actiondata: {
          ref: URL.ref || '',
          gmv: URL.vmgdb || '',
          source: URL.from || URL.ref || '',
          boxVersion: getVersion(),
          boxPlatform: ua.match(/(iPad|iPhone|iPod)/igm) ? 'ios' : 'android'
        }
      }
    };

    if (URL.uid && URL.osname) {
      ['osname', 'ua', 'ut', 'from', 'cfrom', 'uid', 'pkgname'].forEach(function (v) {
        URL[v] && (self.CUSTOM_ARG[v] = URL[v]);
      });
    }

    if (cateid) {
      this.CUSTOM_ARG.data.cateid = cateid;
    }

    this.tmpArg = {};
    this.options = {};

    this.options.customHandler = function (url) {
      var arr = [];

      if (self.tmpArg) {
        for (var i in self.tmpArg) {
          if (self.tmpArg.hasOwnProperty(i)) {
            var val = self.tmpArg[i];

            if ($.isPlainObject(val)) {
              val = JSON.stringify(val);
            }

            arr.push(i + '=' + urlencode(val));
          }
        }
      }

      if (arr.length) {
        url += arr.join('&');
      }

      return url;
    };
  };

  Multilog.prototype = monitor(GIF_URI);
  /**
   * pv统计
   *
   * @param  {string} url 统计的pv url
   * @param  {string} u  来源url，可选
   * @return {Object} report 上报
   */

  Multilog.prototype.pv = function (url, u) {
    this.tmpArg = clone(this.CUSTOM_ARG);
    var _data = this.tmpArg.data; // 写死, actionid：2=事件统计，1=pv统计

    _data.actionid = '1';
    var data = {};
    data.url = url || location.href;

    if (u) {
      data.u = u;
    }

    _data.actiondata = extend(_data.actiondata, data);
    this.tmpArg.data = _data;
    return this.report();
  };
  /**
   * /searchbox?action=pblog&service=bdbox&pu=六大公共参数data={
      "dataid":"__DATA_ID__", //区分sdk客户端：1、JS：2、SERVER：3
      "cateid":"__CATE_ID__", // 区分业务 卡片、小说、视频…
      "actionid":"__ACTION_ID__", //区分行为
      "actiontype":"__ACTION_TYPE__", //区分主动/被动行为，主动：1、被动：0
      "actiondata":"__ACTION_DATA__", //业务具体数据，json格式
  }
  
   */

  /**
   * 事件打点统计
   * @param  {string} evtName 事件名称
   * @param  {string} evtType   事件类型
   * @param  {string} evtTag    事件tag标示
   * @return {Object} report 上报
   */


  Multilog.prototype.event = function (evtName, evtType, evtTag) {
    if (!evtName) {
      throw 'monitor.tc.event need a evtName';
    }

    if ($.isPlainObject(evtType) && !evtTag) {
      var data = {
        evtName: evtName
      };

      for (var i in evtType) {
        data[i] = evtType[i];
      }
    } else {
      var data = {
        evtName: evtName,
        evtType: evtType || '',
        evtTag: evtTag || ''
      };
    }

    this.tmpArg = clone(this.CUSTOM_ARG);
    var _data = this.tmpArg.data; // 写死, actionid：2=事件统计，1=pv统计

    _data.actionid = '2';
    _data.actiondata = extend(_data.actiondata, data);
    this.tmpArg.data = _data;
    return this.report();
  }; // 结束改造

  /**
   * Multilog, 事件跟踪 可适用于多个cateid在一个页面相互独立打点
   * @param  {number} cateid server分配id
   * @return {Object} Multilog 新的实例对象
   *
   * @memberOf Bdbox.monitor
   * @name  Multilog
   * @example
   * var Multilog = Bdbox.monitor.multilog;
   * var demoLog = Multilog(18); //如cateid是18
   * //统计PV
   * demoLog.pv(18,{name:123})
   * //统计事件
   * demoLog.event(evtName, evtType, evtTag)
   */


  return function (cateid) {
    return new Multilog(cateid);
  };
}();
