// boxjs/dist/thirdParty/bdbox/monitor/pblog.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /**
   * @file pblog.js
   */

  /* eslint-disable fecs-camelcase */


  var monitor = require('./monitor');

  var extend = require('../extend');

  var query2Json = require('../utils/queryToJson');

  var getVersion = require('../utils/getVersion');

  var clone = require('../clone'); //url中存在公共参数，则给tcbox拼上，透传


  var URL = query2Json(location.search);
  var ua = navigator.userAgent;
  var protocol = window.location.protocol;

  if (protocol !== 'http:') {
    protocol = 'https:';
  }

  var GIF_URI = protocol + '//m.baidu.com/tcbox';
  var CUSTOM_ARG = {
    service: 'bdbox',
    action: 'pblog',
    ctv: 2,
    //表明参数协议第二个版本
    cen: 'uid_ua_ut',
    //表明参数加密
    data: {
      appid: '1',
      //手百为1
      dataid: '2',
      //客户端：1、JS：2、SERVER：3
      actiontype: '1',
      //区分主动/被动行为，主动：1、被动：0
      actionid: '2',
      //写死, actionid：2=事件统计，1=pv统计
      actiondata: {
        ref: URL.ref || '',
        gmv: URL.vmgdb || '',
        source: URL.from || URL.ref || '',
        boxVersion: getVersion(),
        boxPlatform: ua.match(/(iPad|iPhone|iPod)/igm) ? 'ios' : 'android'
      }
    }
  };
  var urlencode = encodeURIComponent;

  if (URL.uid && URL.osname) {
    var arr = [];
    ['osname', 'ua', 'ut', 'from', 'cfrom', 'uid', 'pkgname'].forEach(function (v) {
      URL[v] && (CUSTOM_ARG[v] = URL[v]);
    });
  }

  var tmpArg;
  var PBLOG = monitor(GIF_URI, {
    customHandler: function (url) {
      var arr = [];

      if (tmpArg) {
        for (var i in tmpArg) {
          if (tmpArg.hasOwnProperty(i)) {
            var val = tmpArg[i];

            if ($.isPlainObject(val)) {
              val = JSON.stringify(val);
            }

            arr.push(i + '=' + urlencode(val));
          }
        }
      }

      if (arr.length) {
        url += arr.join('&');
      }

      return url;
    }
  });

  PBLOG.init = function (cateId, obj) {
    if ($.isPlainObject(obj)) {
      CUSTOM_ARG = extend(CUSTOM_ARG, obj);
    }

    CUSTOM_ARG.data.cateid = cateId;
  };
  /**
   * pv统计
   * @param  {string} url 统计的pv url
   * @param  {string} su  来源url，可选
   * @return {this}     PBLOG
   */


  PBLOG.pv = function (url, u) {
    tmpArg = clone(CUSTOM_ARG);
    var _data = tmpArg.data;
    _data.actionid = '1'; ////写死, actionid：2=事件统计，1=pv统计

    var data = {};
    data.url = url || location.href;

    if (u) {
      data.u = u;
    }

    _data.actiondata = extend(_data.actiondata, data);
    return PBLOG.report();
  };
  /**
   * /searchbox?action=pblog&service=bdbox&pu=六大公共参数data={
      "dataid":"__DATA_ID__", //区分sdk客户端：1、JS：2、SERVER：3
      "cateid":"__CATE_ID__", // 区分业务 卡片、小说、视频…
      "actionid":"__ACTION_ID__", //区分行为
      "actiontype":"__ACTION_TYPE__", //区分主动/被动行为，主动：1、被动：0
      "actiondata":"__ACTION_DATA__", //业务具体数据，json格式
  }
  
   */

  /**
   * 事件打点统计
   * @param  {string} evtName 事件名称
   * @param  {string} evtType   事件类型
   * @param  {string} evtTag    事件tag标示
   * @param  {?number} cateId    业务类型，可选用于指定打点业务类型
   * @return {object}  tc    tc函数
   */


  PBLOG.event = function (evtName, evtType, evtTag, cateId, callback) {
    // console.log(arguments);
    if (!evtName) {
      throw 'monitor.tc.event need a evtName';
    }

    if ($.isPlainObject(evtType) && !evtTag) {
      var data = {
        evtName: evtName
      };

      for (var i in evtType) {
        data[i] = evtType[i];
      }
    } else {
      var data = {
        evtName: evtName,
        evtType: evtType || '',
        evtTag: evtTag || ''
      };
    }

    tmpArg = clone(CUSTOM_ARG);
    var _data = tmpArg.data;
    _data.actionid = '2'; ////写死, actionid：2=事件统计，1=pv统计

    if (cateId) {
      tmpArg.data.cateid = cateId;
    }

    _data.actiondata = extend(_data.actiondata, data);
    return PBLOG.report(null, callback);
  };
  /**
   * 事件打点统计
   * @param  {Object} data 事件上报数据
   * @param  {?number} cateId    业务类型，可选用于指定打点业务类型
   * @param  {Function} callback 回调函数
   * @return {Object}  tc    tc函数
   */


  PBLOG.newEvent = function (data, cateId, callback) {
    tmpArg = clone(CUSTOM_ARG);
    var _data = tmpArg.data;
    _data.actionid = '2'; // 写死, actionid：2=事件统计，1=pv统计

    if (cateId) {
      tmpArg.data.cateid = cateId;
    }

    _data.actiondata = extend(_data.actiondata, data);
    return PBLOG.report(null, callback);
  };
  /**
   * pblog, 事件跟踪
   * @memberOf Bdbox.monitor
   * @name  pblog
   * @author wangyongqing01
   * @version $Id: pblog.js 293584 2016-05-06 07:44:21Z yupeng07 $
   * @example
   * init方法一般不使用
   * Bdbox.monitor.pblog('init',[cateId])
   * //统计PV
   * Bdbox.monitor.pblog('pv')
   * //统计事件
   * Bdbox.monitor.pblog('event',[evtName, evtType, evtTag])
   * onclick="Bdbox.monitor.pblog('event',[evtName, evtType, evtTag])"
   */


  return function () {
    PBLOG.main.apply(PBLOG, arguments); //return PBLOG.report('ep=' + urlencode(ep.join('*')) + '&et=4');
  };
}();
