// @baidu/auc-zhengming/conponents/viewComp.ts
/**
 * @file index.ts
 * @author gaohong01
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucViewCompData} from './types';
import {WujiImage, WujiAvatar, WujiTextImage, WujiTitle} from '@baidu/wuji-san';
import {env} from '@baidu/auc-base/env';
import {jump} from "@baidu/auc-base/jump";
import boxx from '@baidu/auc-base/boxx';
import {request} from '@baidu/auc-star-header/request';
import {formatNum} from '@baidu/auc-base/utils';


const PRAISEAPI = 'https://ext.baidu.com/api/like/v1/like/receive';

export class AucViewComp extends AucComponent {
    static template:string = /* html */ `
        <div class="{{camp === '1' ? styles.rightSideWrap : styles.oppositeSideWrap}}">
            <div class="{{styles.authorArea}}" on-click="handleAuthorClick">
                <wuji-avatar
                    size="90"
                    src="{{auimage || defaultAvatar}}"
                    class="{{styles.authorAvatar}}"
                    vip="{{vflag || '0'}}"
                ></wuji-avatar>
                <div class="{{styles.authorInfoArea}}">
                    <span class="{{styles.authorName}}">{{userName}}</span>
                    <span s-if="{{auabs !== ''}}" class="{{styles.divideCircle}}"></span>
                    <span class="{{styles.authorDesc}}">{{auabs}}</span>
                </div>
            </div>
            <div class="{{styles.content}}" on-click="handleViewClick">
                <div class="{{styles.contentLeft}}">
                    <wuji-title
                        text="{{title}}"
                        spacing="{{false}}"
                        lineClamp="{{3}}"
                        size="t2"
                        class="{{image ? styles.textMinHeight : ''}}"
                    />
                    <div class="{{styles.interaction}}">
                        <div class="{{styles.comment}}" on-click="handleCommentClick($event)">
                            <span class="{{styles.commentIcon}}"></span>
                            <span class="{{styles.iconDesc}}"> {{commemtNum}}</span>
                        </div>
                        <div class="{{styles.praiseBox}}" id="id" on-click="handlePraiseClick($event)">
                            <span class="{{styles.praiseIcon}} {{isliked === '1' ? styles.isPraisIcon : styles.notPraiseIcon}}"></span>
                            <span class=" {{isliked === '1'? styles.isPraisDesc : styles.iconDesc}}">{{praisNum}}</span>
                        </div>
                    </div>
                </div>
                <div s-if="image !== ''" class="{{styles.contentRight}}">
                    <wuji-image
                        src="{{image}}"
                        s-bind="imageProps"
                        user-select="{{true}}"
                    >
                        <slot name="imageMask" slot="mask"/>
                    </wuji-image>
                </div>
            </div>
        </div>
    `;

    static components:object = {
        'wuji-image': WujiImage,
        'wuji-avatar': WujiAvatar,
        'wuji-text-image': WujiTextImage,
        'wuji-title': WujiTitle
    };

    static computed:Object = {
        praisNum() {
            const num = +this.data.get('praisenum');
            return  num > 0 ? formatNum(num) : '赞';
        },
        commemtNum() {
            const num = +this.data.get('comment_num');
            return num > 0 ? formatNum(num) : '评论';
        }
    }
    initData(): AucViewCompData {
        return {
            styles,
            defaultAvatar: 'https://b.bdstatic.com/searchbox/image/gcp/20221221/2699453771.jpeg',
            praiseFlag: true
        };
    }
    handleAuthorClick(): void {
        this.fire('log', 'zm_author_click');
        if (env.isBoxSeries|| env.isXiaoMiBaiPai || env.isMatrix) {
            jump({
                cmd: this.data.get('auhomepagecmd'),
                link: this.data.get('auhomepageurl')
            });
        } else {
            this.fire('back-shoubai', '');
        }

    }
    handleViewClick(): void {
        const cmd = this.data.get('cmd');
        const link = this.data.get('link');
        jump({
            cmd,
            link
        });
        this.fire('log', 'zm_view_click');
    }
    handleCommentClick(e): void {
        e.stopPropagation();
        const isLogin = window.sortsData.isLogin;
        if (env.isBoxSeries|| env.isXiaoMiBaiPai || env.isMatrix) {
            if (!isLogin) {
                this.userLogin();
                return;
            }
            const cmd = this.data.get('commentcmd');
            const link = this.data.get('link');
            this.fire('log', 'zm_comment_click');
            jump({
                cmd,
                link
            })
        } else {
            this.fire('back-shoubai', '');
        }
    }
    handlePraiseClick(e): void {
        e.stopPropagation();
        const isliked = +this.data.get('isliked');

        if (env.isBoxSeries || env.isXiaoMiBaiPai || env.isMatrix) {
            const params = {
                type: 'feed',
                op_type: !isliked ? 'add' : 'cancel',
                sfrom: 'star',
                source: env.isBoxSeries ? 'star_feedlist_theme' : 'star_feedlist_theme_w',
                id: this.data.get('rid')
            };
            request(PRAISEAPI, params).then(res => {
                if (+res.errno === 0) {
                    if (!isliked) {
                        this.data.set('isliked', '1');
                        const praisenum = +this.data.get('praisenum') + 1;
                        this.data.set('praisenum', praisenum);
                        this.fire('log', 'zm_praise_click');
                    } else {
                        this.data.set('isliked', '0');
                        const praisenum = +this.data.get('praisenum') - 1;
                        this.data.set('praisenum', praisenum < 0 ? 0 : praisenum);
                    }
                } else if (res.errno === 202206) {
                    // 跳转到图文落地页/视频详情页/动态落地页点赞
                    this.data.set('praisenum', this.data.get('praisenum') + 1);
                    this.data.set('isliked', '1');
                } else {
                    boxx.call('commonUI.showToast', {
                        type: '1',
                        message: '请求超时，请稍后重试'
                    });
                }
            });
        } else {
            this.fire('back-shoubai', '');
        }
    }
    userLogin(): void {
        const that = this;
        boxx.call('account.login', {
            loginType: 'fast',
            showThirdLogin: '1',
            loginSource: 'hudong_trusted',
            normalizeAccount: '1',
            success() {
                const cmd = that.data.get('commentcmd');
                const link = that.data.get('link');
                jump({
                    cmd,
                    link
                });
            },
            fail() {
                boxx.call('commonUI.showToast', {
                    type: '1',
                    message: '登录失败'
                });
            }
        });
        
    }
}

export default AucViewComp;
