// @baidu/xbox-native/es5/log/disappear.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = _interopRequireDefault(require("./index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module log
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('ubc/duration/disappear', {
  v: 4,
  hasCB: false,
  versionRange: ['8.3'],
  androidliteVersionRange: ['2.2.2'],
  androidmissionVersionRange: ['1.0'],
  iosmissionVersionRange: ['1.0'],
  androidinfoVersionRange: ['1.0'],
  notSupportFrame: {
    android: ['search']
  }
});
/**
 * ubc/duration/disappear
 * @function disappear
 * @param {object} [data={}] - ubc 参数
 */

var _default = function _default() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.default)((0, _utils.getParams)('durationDisappear', {
    data: data,
    abtest: data.abtest
  }));
};

exports.default = _default;
