// @baidu/xbox-native/es5/log/feedReport.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = _interopRequireDefault(require("./index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2018-12-4
 * @module log
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('feed/report', {
  ios: {
    versionRange: ['8.3'],
    hasCB: false,
    filter: function filter() {
      return {
        module: 'feed',
        action: 'report'
      };
    }
  },
  android: {
    versionRange: ['8.3'],
    hasCB: false,
    filter: function filter() {
      return {
        module: 'Bdbox_android_utils',
        action: 'report'
      };
    }
  }
});
/**
 * ubcreport
 * @function ubc
 * @param {object} [data={}] - ubc 参数
 */

var _default = function _default() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.default)((0, _utils.getParams)('feedReport', {
    data: data,
    abtest: data.abtest
  }));
};

exports.default = _default;
