// @baidu/xbox-native/es5/log/performanceFlowEvent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = _interopRequireDefault(require("./index"));

var _stringify = _interopRequireDefault(require("../stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2018-12-03
 * @module log
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('utils/onPerformanceFlowEvent', {
  v: 20,
  hasCB: false,
  versionRange: ['7.2'],
  filter: function filter(data) {
    return {
      action: 'onPerformanceFlowEvent',
      module: 'utils',
      data: {
        params: (0, _stringify.default)(data, true)
      }
    };
  }
});
/**
 * performanceFlowEvent
 * @function disappear
 * @param {object} [data={}] - ubc 参数
 */

var _default = function _default() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.default)((0, _utils.getParams)('performanceFlowEvent', data));
};

exports.default = _default;
