// @baidu/xbox-native/es5/log/ubc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = _interopRequireDefault(require("./index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module log
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('utils/ubcReport', {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    hasCB: false,
    versionRange: ['7.3'],
    filter: function filter() {
      return {
        module: 'utils',
        action: 'ubcReport',
        extData: {
          minver: '7.3.0.0'
        }
      };
    }
  },
  android: {
    hasCB: false,
    versionRange: ['7.3'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_utils',
        action: 'ubcEvent',
        extData: {
          min_v: '16789504'
        }
      };
    }
  }
});
/**
 * ubcreport
 * @function ubc
 * @param {object} [data={}] - ubc 参数
 */

var _default = function _default() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.default)((0, _utils.getParams)('ubcReport', {
    data: data,
    abtest: data.abtest
  }));
};

exports.default = _default;
