// @baidu/xbox-native/es5/detect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isMaybeBox = isMaybeBox;
Object.defineProperty(exports, "version_compare", {
  enumerable: true,
  get: function get() {
    return _version_compare.default;
  }
});
Object.defineProperty(exports, "versionCompare", {
  enumerable: true,
  get: function get() {
    return _version_compare.default;
  }
});
Object.defineProperty(exports, "getBoxVersion", {
  enumerable: true,
  get: function get() {
    return _getBoxVersion.default;
  }
});
Object.defineProperty(exports, "isIOS", {
  enumerable: true,
  get: function get() {
    return _os.isIOS;
  }
});
Object.defineProperty(exports, "isAndroid", {
  enumerable: true,
  get: function get() {
    return _os.isAndroid;
  }
});
Object.defineProperty(exports, "isBox", {
  enumerable: true,
  get: function get() {
    return _isBox.default;
  }
});
Object.defineProperty(exports, "isMatrix", {
  enumerable: true,
  get: function get() {
    return _isMatrix.default;
  }
});

var _version_compare = _interopRequireDefault(require("@baidu/xbox/es5/version_compare"));

var _getBoxVersion = _interopRequireDefault(require("@baidu/xbox/es5/get-box-version"));

var _os = require("@baidu/xbox/es5/os");

var _isBox = _interopRequireDefault(require("@baidu/xbox/es5/is-box"));

var _isMatrix = _interopRequireDefault(require("@baidu/xbox/es5/is-matrix"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 根据 UserAgent 进行判断
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

/**
 * 或判断是否是某个 box
 * @param {string} subnames - lite/info组合，[,]隔开
 * @param {*} ua - ua
 */
function isMaybeBox(subnames, ua) {
  subnames = subnames.split(',');
  var len = subnames.length;

  while (len--) {
    if ((0, _isBox.default)(subnames[len], ua) || (0, _isMatrix.default)(subnames[len], ua)) {
      return true;
    }
  }

  return false;
}
