// @baidu/auc-pic-text-sm/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import styles from './style.module.less';
import {AucPicTextSmData} from './types';
import {WujiImage} from '@baidu/wuji-san';
import AucComponent from "@baidu/auc-base/AucComponent";

export class AucPicTextSm extends AucComponent {
    static template:string = /* html */ `
    <div class="{{styles.aucPicTextSmWrapper}}" on-click="handleClick">
        <div class="{{styles.imageWrapper}}" s-if="imageProps.src">
            <wuji-image
                ratio="3:2"
                s-bind="imageProps"
            ></wuji-image>
        </div>
        <div class="{{styles.picText}}">{{title}}</div>
    </div>
    `;
    static components:object = {
        'wuji-image': WujiImage
    };

    initData(): AucPicTextSmData {
        return {
            styles,
            title: '',
            image: '',
            labelProps: {}
        };
    }

    attached(): void {
        this.registerScrollInView().then(_ => {
            this.fire('show', '');
        });
    }

    handleClick() {
        this.fire('click-item', '');
    }
}

export default AucPicTextSm;
