// @baidu/xbox/es5/browser/detect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isBox", {
  enumerable: true,
  get: function get() {
    return _isBox.default;
  }
});
Object.defineProperty(exports, "isAndroid", {
  enumerable: true,
  get: function get() {
    return _os.isAndroid;
  }
});
Object.defineProperty(exports, "isIOS", {
  enumerable: true,
  get: function get() {
    return _os.isIOS;
  }
});
Object.defineProperty(exports, "getOSName", {
  enumerable: true,
  get: function get() {
    return _os.getOSName;
  }
});
Object.defineProperty(exports, "getOSVersion", {
  enumerable: true,
  get: function get() {
    return _os.getOSVersion;
  }
});
exports.getUa = exports.isBaidubrowser = exports.isUC = exports.isWP = exports.isWinphone = exports.isWinPhone = exports.isSafari = exports.isWebview = exports.isWebView = exports.isIE = exports.isChrome = exports.isOSX = exports.isMac = exports.isWebkit = exports.isQQ = exports.isMatrix = exports.isWechat = exports.isWX = exports.isWB = exports.isWeixin = exports.isWeibo = void 0;

var _ua = _interopRequireDefault(require("../ua"));

var _isBox = _interopRequireDefault(require("../is-box"));

var _isMatrix = _interopRequireDefault(require("../is-matrix"));

var _os = require("../os");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 判断浏览器类型
 * @module browser/detect
 * @file Created on Thu Nov 22 2018
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

/**
 * @function isWeibo
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */
var isWeibo = function isWeibo() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /\bweibo\b/i.test(ua);
};
/**
 * @function isWeixin
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */


exports.isWeibo = isWeibo;

var isWeixin = function isWeixin() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /micromessenger\//i.test(ua);
};

exports.isWeixin = isWeixin;
var isWB = isWeibo;
exports.isWB = isWB;
var isWX = isWeixin;
exports.isWX = isWX;
var isWechat = isWeixin;
exports.isWechat = isWechat;
var isMatrix = _isMatrix.default;
/**
 * @function isQQ
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */

exports.isMatrix = isMatrix;

var isQQ = function isQQ() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /QQ\//.test(ua);
};
/**
 * @function isWebkit
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */


exports.isQQ = isQQ;

var isWebkit = function isWebkit() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /Web[kK]it[/]{0,1}([\d.]+)/.test(ua);
};
/**
 * @function isMac
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */


exports.isWebkit = isWebkit;

var isMac = function isMac() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /\(Macintosh; Intel /.test(ua);
};

exports.isMac = isMac;
var isOSX = isMac;
/**
 * @function isChrome
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */

exports.isOSX = isOSX;

var isChrome = function isChrome() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /Chrome\/([\d.]+)/.test(ua) || /CriOS\/([\d.]+)/.test(ua);
};
/**
 * @function isIE
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */
// eslint-disable-next-line


exports.isChrome = isChrome;

var isIE = function isIE() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /MSIE\s([\d.]+)/.test(ua) || /Trident\/[\d](?=[^\?]+).*rv:([0-9.].)/.test(ua);
};

exports.isIE = isIE;

var isWebView = function isWebView() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return !isChrome(ua) && /(iPhone|iPod|iPad).*AppleWebKit(?!.*Safari)/.test(ua);
};
/**
 * @function isWebview
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */


exports.isWebView = isWebView;
var isWebview = isWebView;
/**
 * @function isSafari
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */

exports.isWebview = isWebview;

var isSafari = function isSafari() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return isWebView(ua) || /Version\/([\d.]+)([^S](Safari)|[^M]*(Mobile)[^S]*(Safari))/.test(ua);
};
/**
 * @function isWinPhone
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */


exports.isSafari = isSafari;

var isWinPhone = function isWinPhone() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /Windows Phone ([\d.]+)/.test(ua);
};

exports.isWinPhone = isWinPhone;
var isWinphone = isWinPhone;
exports.isWinphone = isWinphone;
var isWP = isWinPhone;
/**
 * @function isUC
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */

exports.isWP = isWP;

var isUC = function isUC() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return / UCBrowser\//.test(ua);
};
/**
 * @function isBaidubrowser
 * @param {string} ua - 待判断的 UA，默认当前 userAgent
 */


exports.isUC = isUC;

var isBaidubrowser = function isBaidubrowser() {
  var ua = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _ua.default;
  return /baidubrowser/i.test(ua);
};

exports.isBaidubrowser = isBaidubrowser;

var getUa = function getUa() {
  var browsers = null;

  switch (!0) {
    case (0, _isBox.default)():
      browsers = 'bdbox';
      break;

    case isWeixin():
      browsers = 'wechat';
      break;

    case isWeibo():
      browsers = 'webo';
      break;

    case isQQ():
      browsers = 'qq';
      break;

    case isBaidubrowser():
      browsers = 'baidubrowser';
      break;

    case isUC():
      browsers = 'ucbrowser';
      break;

    default:
      browsers = 'others';
      break;
  }

  return browsers;
};

exports.getUa = getUa;
