// @baidu/wuji-san/lib/wuji-avatar/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _san = require("san");

var _wujiIcon = _interopRequireDefault(require("../wuji-icon"));

var _isNumeric = _interopRequireDefault(require("../utils/is-numeric"));

var _pr2rem = _interopRequireDefault(require("../utils/pr2rem"));

require("@baidu/wuji-icons/svg/defaultpic-logo-87.svg");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 'w-avatar';

var Avatar = /*#__PURE__*/function (_Component) {
  _inherits(Avatar, _Component);

  var _super = _createSuper(Avatar);

  function Avatar() {
    _classCallCheck(this, Avatar);

    return _super.apply(this, arguments);
  }

  _createClass(Avatar, [{
    key: "initData",
    value: function initData() {
      return {
        hairline: true,
        size: 129,
        src: '',
        vip: '0',
        motion: false,
        borderColor: 'FC18'
      };
    }
  }, {
    key: "handleError",
    value: function handleError(error) {
      error.target.style.visibility = 'hidden';
      this.fire('error', {
        error: error
      });
    }
  }]);

  return Avatar;
}(_san.Component);

exports["default"] = Avatar;

_defineProperty(Avatar, "template",
/* html */
"\n        <div class=\"{{wrapClass}}\" style=\"{{wrapStyle}}\">\n            <div class=\"".concat(prefixCls, "-default\">\n                <wuji-icon\n                    type=\"{{defaultIcon || 'defaultpic-logo-87'}}\"\n                    size=\"{{defaultIconSize}}\"\n                />\n            </div>\n            <img s-if=\"src\" class=\"").concat(prefixCls, "-img\" src=\"{{src}}\" alt=\"{{alt}}\" on-error=\"handleError\"/>\n            <div s-if=\"showVip\" class=\"{{vipClass}}\"/>\n        </div>\n    "));

_defineProperty(Avatar, "components", {
  'wuji-icon': _wujiIcon["default"]
});

_defineProperty(Avatar, "dataTypes", {
  defaultIcon: _san.DataTypes.string,
  hairline: _san.DataTypes.bool,
  size: _san.DataTypes.oneOfType([_san.DataTypes.number, _san.DataTypes.string]),
  src: _san.DataTypes.string,
  vip: _san.DataTypes.oneOf(['0', '1', '2', '3', '10', 'swan']),
  motion: _san.DataTypes.bool,
  borderWidth: _san.DataTypes.oneOf([1, 3]),
  borderColor: _san.DataTypes.oneOf(['FC18', 'FC107'])
});

_defineProperty(Avatar, "computed", {
  wrapClass: function wrapClass() {
    var hairline = this.data.get('hairline');
    var borderWidth = this.data.get('borderWidth');
    var borderColor = this.data.get('borderColor').toLocaleLowerCase(); // 自定义描边的权重大于hairline

    if (borderWidth) {
      return [prefixCls, "".concat(prefixCls, "-hairline-").concat(borderWidth, "-").concat(borderColor)];
    }

    return [prefixCls, hairline && "".concat(prefixCls, "-hairline")].filter(function (c) {
      return c;
    });
  },
  wrapStyle: function wrapStyle() {
    var ret = {};
    var size = this.data.get('size');

    if ((0, _isNumeric["default"])(size)) {
      var width = "".concat((0, _pr2rem["default"])(size), "rem");
      ret = {
        width: width,
        height: width
      };
    }

    return ret;
  },
  showVip: function showVip() {
    var size = this.data.get('size');
    var vip = this.data.get('vip');

    if (vip && vip !== '0' && size >= 66) {
      return true;
    }
  },
  vipClass: function vipClass() {
    var size = this.data.get('size');
    var vip = this.data.get('vip');
    var motion = this.data.get('motion');
    var vipSize;

    switch (true) {
      case size > 170:
        vipSize = 'lg';
        break;

      case size > 156:
        vipSize = 'md';
        break;

      case size > 84:
        vipSize = 'sm';
        break;

      default:
        vipSize = 'xs';
        break;
    }

    return ["".concat(prefixCls, "-vip"), "vip-".concat(vip), motion && 'motion', vipSize && "vip-".concat(vipSize)].filter(function (c) {
      return c;
    });
  },
  defaultIconSize: function defaultIconSize() {
    var defaultIconSize = 'half';
    var size = this.data.get('size');

    if (size > 129) {
      defaultIconSize = 87;
    } else if (size === 129) {
      defaultIconSize = 66;
    }

    return defaultIconSize;
  }
});
