// @baidu/wuji-san/es/wuji-link/index.js
/**
 * @file 组件 WujiLink 无极-链接
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';

import '@baidu/wuji-icons/svg/link-img-o-51.svg';
import '@baidu/wuji-icons/svg/link-url-o-51.svg';
import '@baidu/wuji-icons/svg/link-video-o-51.svg';
import '@baidu/wuji-icons/svg/link-vote-o-51.svg';

const prefixCls = 'w-link';

export default class WujiLink extends Component {
    static template = /* html */ `
        <span
            class="${prefixCls}"
            on-click="handleClick"
            on-touchstart="handleTouchstart"
        >
            <wuji-icon s-if="icon" class="${prefixCls}-icon" type="{{icon}}"/>
            <span class="${prefixCls}-text">{{text}}</span>
        </span>
    `;

    static components = {
        'wuji-icon': Icon,
    };

    static dataTypes = {
        text: DataTypes.string,
        icon: DataTypes.oneOf([
            'link-img-o-51',
            'link-url-o-51',
            'link-video-o-51',
            'link-vote-o-51',
        ]),
    };

    initData() {
        return {
            text: '链接文字',
        };
    }

    handleClick(event) {
        this.fire('click', {event});
    }

    handleTouchstart(e) {
        e.stopPropagation();
    }
}
