// @baidu/nano/lib/follow/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("./style/index.less");

var _san = require("san");

var _button = _interopRequireDefault(require("../button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 's-follow';
var statusToText = {
  // 5: '被拉黑', // RELATION_BLACKED
  // 4: '拉黑', // RELATION_BLACK
  3: '互相关注',
  // RELATION_FRIEND
  2: '关注',
  // RELATION_FANS
  1: '已关注',
  // RELATION_FOLLOW
  0: '关注' // RELATION_NONE

};

var Follow =
/*#__PURE__*/
function (_Component) {
  _inherits(Follow, _Component);

  function Follow() {
    _classCallCheck(this, Follow);

    return _possibleConstructorReturn(this, _getPrototypeOf(Follow).apply(this, arguments));
  }

  _createClass(Follow, [{
    key: "initData",
    // eslint-disable-next-line
    value: function initData() {
      return {
        adapter: 'normal',
        followKey: '',
        followStatus: 0,
        followParams: {},
        loading: false,
        size: 'sm'
      };
    }
  }, {
    key: "handleClick",
    value: function handleClick() {
      var data = this.data.get();

      if (!data.loading) {
        this.fire('click', data);

        if (data.action) {
          this.dispatch(data.action, data);
        }
      }
    }
  }]);

  return Follow;
}(_san.Component);

exports.default = Follow;

_defineProperty(Follow, "template",
/*html*/
"\n        <div class=\"{{wrapClass}}\" on-click=\"handleClick\">\n            <s-button\n                s-bind=\"{{buttonProps}}\"\n                loading=\"{{loading}}\"\n                size=\"{{size}}\"\n                width=\"{{width}}\"\n                height=\"{{height}}\"\n                on-click=\"handleClick\"\n            />\n        </div>\n    ");

_defineProperty(Follow, "components", {
  's-button': _button.default
});

_defineProperty(Follow, "dataTypes", {
  adapter: _san.DataTypes.oneOf(['normal', 'outline', 'light']),
  followStatus: _san.DataTypes.oneOf([0, 1, 2, 3]),
  size: _san.DataTypes.oneOf(['sm', 'md', 'lg']),
  width: _san.DataTypes.oneOfType([_san.DataTypes.number, _san.DataTypes.string]),
  height: _san.DataTypes.oneOfType([_san.DataTypes.number, _san.DataTypes.string])
});

_defineProperty(Follow, "computed", {
  buttonProps: function buttonProps() {
    var data = this.data;
    var adapter = data.get('adapter');
    var loading = data.get('loading');
    var status = data.get('followStatus');
    var text = statusToText[status];
    var hairline = adapter !== 'normal' || text !== '关注';
    var type = 'primary';

    if (adapter === 'light') {
      type = 'default';
    } else if (text !== '关注' && type !== 'outline') {
      type = 'weak';
    }

    var props = {
      ghost: adapter === 'light',
      hairline: hairline,
      text: loading ? '' : text,
      type: type
    };
    return props;
  },
  wrapClass: function wrapClass() {
    var adapter = this.data.get('adapter');
    var status = this.data.get('followStatus');
    return [prefixCls, "".concat(prefixCls, "-").concat(adapter), "".concat(prefixCls, "-status-").concat(status)];
  }
});
