// @baidu/xbox-native/es5/media/tts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.play = exports.stop = exports.status = exports.append = exports.pause = exports.prefetch = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module media/tts
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('tts/prefetch', {
  v: 18,
  versionRange: ['10.2']
});
/**
 * @function prefetch
 * @param {object} [data={}] - prefetch 参数
 */

var prefetch = function prefetch() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.tts)((0, _utils.getParams)('prefetch', data));
};

exports.prefetch = prefetch;
(0, _register.default)('tts/pause', {
  v: 18,
  versionRange: ['10.2']
});
/**
 * @function pause
 * @param {object} [data={}] - pause 参数
 */

var pause = function pause() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.tts)((0, _utils.getParams)('pause', data));
};

exports.pause = pause;
(0, _register.default)('tts/append', {
  v: 18,
  versionRange: ['10.2']
});
/**
 * @function append
 * @param {object} [data={}] - append 参数
 */

var append = function append() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.tts)((0, _utils.getParams)('append', data));
};

exports.append = append;
(0, _register.default)('tts/status', {
  v: 18,
  versionRange: ['10.2']
});
/**
 * @function status
 * @param {object} [data={}] - status 参数
 */

var status = function status() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.tts)((0, _utils.getParams)('status', data));
};

exports.status = status;
(0, _register.default)('tts/stop', {
  v: 18,
  versionRange: ['10.2'],
  filter: function filter() {
    return {
      module: 'tts',
      action: 'end'
    };
  }
});
/**
 * @function stop
 * @param {object} [data={}] - stop 参数
 */

var stop = function stop() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.tts)((0, _utils.getParams)('stop', data));
};

exports.stop = stop;
(0, _register.default)('tts/play', {
  v: 18,
  versionRange: ['10.2']
});
/**
 * @function play
 * @param {object} [data={}] - play 参数
 */

var play = function play() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.tts)((0, _utils.getParams)('play', data));
};

exports.play = play;
