// @baidu/xbox-native/es5/media/video.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.play = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module media/video
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('video/play', {
  androidlite: {
    versionRange: ['2.3']
  },
  androidinfo: {
    versionRange: ['1.1']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  ios: {
    hasCB: false,
    versionRange: ['7.3'],
    filter: function filter(module, action, data) {
      return {
        module: 'video',
        action: 'invokePlayer',
        extData: {
          minver: '7.3.0.0'
        }
      };
    }
  },
  android: {
    versionRange: ['7.3'],
    supportFrame: ['video'],
    hasCB: false,
    filter: function filter(module, action, data) {
      return {
        module: 'Bdbox_android_capi_video',
        action: 'invokePlayer'
      };
    }
  }
});
/**
 * @function play
 * @param {object} [data={}] - play 参数
 */

var play = function play() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.video)((0, _utils.getParams)('play', data));
};

exports.play = play;
