// @baidu/wuji-san/es/wuji-image-content/index.js
/**
 * @file 组件 WujiImageContent 无极-左图右文小程序
 * @author raowenjuan <raowenjuan@baidu.com>
 */

import './style/index.less';
import {Component} from 'san';
import WujiImageText from '../wuji-image-text';
import WujiDetail from './components/detail';

const prefixCls = 'wuji-image-content';

export default class WujiImageContent extends Component {
    static template = /* html */ `
        <div class="${prefixCls}-wrap">
            <wuji-image-text
                title-props="{{titleProps}}"
                image-props="{{imageProps}}"
                widgets="{{!widgets.isRight ? widgetsFormat : []}}"
                img-col-span=4
                discount="{{discount}}"
            >
                <wuji-detail
                    slot="detail"
                    detail-info="{{detailInfo}}"
                    textButton="{{textButton}}"
                    widgets="{{widgets.isRight ? widgetsFormat : []}}"
                />
            </wuji-image-text>
        </div>
    `;

    static components = {
        'wuji-image-text': WujiImageText,
        'wuji-detail': WujiDetail,
    };

    static computed = {
        someClass() {
            const someprop = this.data.get('someprop');
            return [
                prefixCls,
                someprop && `prefixCls-${prefixCls}`,
            ].filter(c => c);
        },
        widgetsFormat() {
            const widgetsData = this.data.get('widgets.data') || [];
            return Object.keys(widgetsData).map(item => {
                if (item === 'image') {
                    return {
                        type: 'image',
                        ...widgetsData[item],
                    };
                }
                return {
                    type: 'normal',
                    text: widgetsData[item],
                };
            });
        },
    };

    initData() {
        return {
            title: '',
            imageProps: {},
        };
    }
}
