// @baidu/ug-invoke-app/src/config/app/appConfigMap.ts
/**
 * @file appconfig
 * @author zhangsiyuan(zhangsiyuan@baidu.com)
 * @see http://agroup.baidu.com/ceug-fe/md/article/2799135
 */
type GetUlink = (scheme: string, target: string) => string;

interface AppConfig {
    getUlink: GetUlink;
    appStoreUrl: string;
    scheme: string;
    pkgName: string;
}

export type AppConfigName = 'baiduboxapp'
    | 'baiduboxlite'
    | 'baiduhaokan'
    | 'bdminivideo'
    | 'tieba'
    | 'bdnetdisk'
    | 'iqiyi'
    | 'baidumap'
    | 'bddrama'
    | 'wenku'
    | 'lemon'
    | 'baiduinput'
    | 'baiduboxvision'
    | 'weibaapp'
    | 'youjia'
    | 'askmybaby'
    | 'yike'
    | 'baiduhanyu'
    | 'yoopu'
    | 'fortunecat'
    | 'tomas'
    | 'bdhealthapp'
    | 'bdwkst'
    | 'wepod'
    | 'rap'
    | 'yymobile'
    | 'baiduboxsenior'
    | 'knews'
    | 'yiju';

// 兼容一下第三方的appName名字
export type AppName = AppConfigName | 'tp';
type AppConfigMap = Record<AppConfigName, AppConfig>;

export const appConfigMap: AppConfigMap = {
    baiduboxapp: {
        getUlink(scheme, target) {
            return `https://boxer.baidu.com/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://itunes.apple.com/cn/app/id382201985?mt=8',
        scheme: 'baiduboxapp://v11/appTab/select?item=home&upgrade=0',
        pkgName: 'com.baidu.searchbox',
    },
    baiduboxlite: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/baiduboxlite/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://itunes.apple.com/cn/app/id1281873118?mt=8',
        scheme: 'baiduboxlite://v11/appTab/select?item=home&upgrade=0',
        pkgName: 'com.baidu.searchbox.lite',
    },
    baiduhaokan: {
        getUlink(scheme, target) {
            const path = scheme.replace('baiduhaokan://', '');
            const hasQuestionMark = path.includes('?');
            return 'https://hku.baidu.com/h5/share/s/'
                + path + (hasQuestionMark ? '&' : '?')
                + `target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://itunes.apple.com/cn/app/bai-du-hao-kan/id1092031003',
        scheme: 'baiduhaokan://home/index',
        pkgName: 'com.baidu.haokan',
    },
    bdminivideo: {
        getUlink(scheme, target) {
            return `https://vv.baidu.com/feedvideoui/ulink?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://itunes.apple.com/cn/app/%E5%85%A8%E6%B0%91%E5%B0%8F%E8%A7%86%E9%A2%91/id1329385145?mt=8',
        scheme: 'bdminivideo://home/index',
        pkgName: 'com.baidu.minivideo',
    },
    tieba: {
        appStoreUrl: 'https://itunes.apple.com/app/apple-store/id477927812?pt=328057&ct=bottom_layer&mt=8',
        pkgName: 'com.baidu.tieba',
        // 贴吧没有提供ulink
        getUlink() {
            return '';
        },
        scheme: 'com.baidu.tieba://unidispatch/homepage',
    },
    bdnetdisk: {
        getUlink(scheme, target) {
            return `https://snsyun.baidu.com/wap/snsdeeplink?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        pkgName: 'com.baidu.netdisk',
        appStoreUrl: 'https://itunes.apple.com/cn/app/%E7%99%BE%E5%BA%A6%E7%BD%91%E7%9B%98/id547166701?mt=8',
        scheme: 'bdnetdisk://n/action.EXTERNAL_ACTIVITY?m_n_v=10.0.20',
    },
    iqiyi: {
        getUlink() {
            return '';
        },
        pkgName: 'com.qiyi.video',
        appStoreUrl: 'https://apps.apple.com/cn/app/id393765873',
        scheme: 'iqiyi://',
    },
    baidumap: {
        getUlink(scheme, target) {
            return `https://clientmap.baidu.com/map/maplink.php?cburl=${
                encodeURIComponent(target)}&openapi=${encodeURIComponent(scheme)}`;
        },
        appStoreUrl: 'https://itunes.apple.com/cn/app/id452186370',
        scheme: 'baidumap://map?src=uginvoke',
        pkgName: 'com.baidu.BaiduMap',
    },
    bddrama: {
        getUlink(scheme, target) {
            return `https://ulinkmvideo.baidu.com/wisedrama/system/ulink?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/%E7%95%AA%E4%B9%90/id1484301936',
        pkgName: 'com.baidu.mv.drama',
        scheme: 'bddrama://home?source=',
    },
    wenku: {
        getUlink() {
            return '';
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id426340811',
        pkgName: 'com.baidu.wenku',
        scheme: 'bdwenku://wenku/operation?type=136&tab=find',
    },
    lemon: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/lemonapp/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'http://click.hm.baidu.com/app.gif?ap=1801081&ch=47556',
        scheme: 'lemon://share?type=13',
        pkgName: 'com.baidu.lemon',
    },
    baiduinput: {
        // 百度输入法不支持失败后跳转链接
        getUlink(scheme) {
            return `https://srf.baidu.com/?from=1024129m&c=apple&e=imehd&native_url=${
                encodeURIComponent(scheme)
            }`;
        },
        appStoreUrl: 'https://itunes.apple.com/app/apple-store/id916139408?pt=625805&ct=1024129m&mt=8',
        scheme: 'baiduimsettings2://superskin',
        pkgName: 'com.baidu.input',
    },
    baiduboxvision: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/baiduboxvision/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1526110789',
        scheme: 'baiduboxvision://appTab/select?item=home',
        pkgName: 'com.baidu.searchbox.vision',
    },
    weibaapp: {
        getUlink() {
            return '';
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/%E5%BE%AE%E5%8F%AD/id1410178720',
        scheme: 'lazybd://com.melon.lazymelon/home',
        pkgName: 'com.melon.lazymelon',
    },
    youjia: {
        getUlink() {
            return '';
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1490227077',
        scheme: 'youjia://app/homepage',
        pkgName: 'com.baidu.autocar',
    },
    askmybaby: {
        getUlink() {
            return '';
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/%E5%AE%9D%E5%AE%9D%E7%9F%A5%E9%81%93-%E7%A7%91%E5%AD%A6%E5%A4%87%E5%AD%95%E6%80%80%E5%AD%95%E8%82%B2%E5%84%BF%E5%8A%A9%E6%89%8B/id858401906',
        scheme: 'askmybaby://com.baidu.mbaby/',
        pkgName: 'com.baidu.mbaby',
    },
    yike: {
        // 一刻相册不支持失败后跳转链接
        getUlink(scheme) {
            return `https://photo.baidu.com/app/scheme?&scheme=${encodeURIComponent(scheme)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/%E4%B8%80%E5%88%BB%E7%9B%B8%E5%86%8C-%E7%99%BE%E5%BA%A6%E7%BD%91%E7%9B%98%E5%87%BA%E5%93%81/id1467852587',
        scheme: 'youa://youa.com/home',
        pkgName: 'com.baidu.youavideo',
    },
    baiduhanyu: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/baidudict/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/1/id1065829176',
        scheme: 'baidudict://com.baidu.dict',
        pkgName: 'com.baidu.dict',
    },
    yoopu: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/yoopu/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://itunes.apple.com/cn/app/id1533615786?mt=8',
        scheme: 'yoopu://',
        pkgName: 'com.baidu.yoopu',
    },
    tomas: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/tomas/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1546173008',
        scheme: 'tomas://',
        pkgName: 'com.baidu.searchbox.tomas',
    },
    fortunecat: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/fortunecat/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1523487452',
        scheme: 'fortunecat://invoke?p=%7B%22type%22%3A0%2C%22source%22%3A1%7D&ubc_ext=%7B%22invoke_channel%22%3A%221024634f%22%7D',
        pkgName: 'com.baidu.fortunecat',
    },
    yiju: {
        getUlink(scheme, target) {
            return `https://ulinkmvideo.baidu.com/yiju/system/ulink?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1548246673',
        scheme: 'bdyiju://home/index',
        pkgName: 'com.baidu.yiju',
    },
    bdhealthapp: {
        getUlink(scheme, target) {
            return `https://jiankang.baidu.com/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1555616850',
        scheme: 'bdhealthapp://app/home',
        pkgName: 'com.baidu.bdhealth',
    },
    bdwkst: {
        getUlink() {
            return '';
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1527030248',
        scheme: 'bdwkst://student/operation?type=136',
        pkgName: 'com.baidu.student',
    },
    rap: {
        getUlink(scheme, target) {
            return `https://ulinkmvideo.baidu.com/wisedrama/system/yinciulink?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1528455665',
        scheme: 'bdhiphop://',
        pkgName: 'com.baidu.rap',
    },
    wepod: {
        getUlink(scheme, target) {
            return `https://ulinkmvideo.baidu.com/wisedrama/system/wepodulink?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1541812473',
        scheme: 'bdwepod://',
        pkgName: 'com.baidu.wepod',
    },
    yymobile: {
        getUlink(scheme) {
            return `https://ulink.yy.com/urlscheme?type=scheme&action=${encodeURIComponent(scheme)}&autodownload=1`;
        },
        appStoreUrl: 'https://apps.apple.com/app/id427941017',
        scheme: 'yymobile://',
        pkgName: 'com.duowan.mobile',
    },
    baiduboxsenior: {
        getUlink(scheme, target) {
            return `https://wakeup.baidu.com/baiduboxsenior/scheme?scheme=${
                encodeURIComponent(scheme)
            }&target=${encodeURIComponent(target)}`;
        },
        appStoreUrl: 'https://apps.apple.com/cn/app/id1581796662',
        scheme: 'baiduboxsenior://v11/appTab/select?item=home&upgrade=0',
        pkgName: 'com.baidu.searchbox.senior',
    },
    knews: {
        getUlink() {
            return '';
        },
        appStoreUrl: '',
        scheme: 'knews://v11/appTab/select?item=home&upgrade=0',
        pkgName: 'com.miui.knews',
    },

};
