// @baidu/wuji-san/es/wuji-three-images/index.js
/**
 * @file 组件 WujiThreeImages 无极-三图
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Grid from '../wuji-grid';
import Image from '../wuji-image';

const prefixCls = 'w-three-images';

export default class WujiThreeImages extends Component {
    static template = /* html */ `
        <div class="${prefixCls}" on-click="handleClick">
            <wuji-row gutter="{{gutter}}" margin="{{spacing ? margin : 0}}">
                <wuji-col
                    s-for="img in imgList"
                >
                    <wuji-img
                        src="{{img.src}}"
                        border-radius="{{img.borderRadius}}"
                        radius="{{img.radius}}"
                        ratio="{{imgRatio}}"
                        label-icon="{{img.labelIcon}}"
                        label-text="{{img.labelText}}"
                    />
                </wuji-col>
            </wuji-row>
        </div>
    `;

    static components = {
        'wuji-img': Image,
        'wuji-col': Grid.Col,
        'wuji-row': Grid.Row,
    };

    static dataTypes = {
        gutter: DataTypes.number,
        margin: DataTypes.number,
        spacing: DataTypes.bool,
        imgList: DataTypes.array,
        imgRatio: DataTypes.oneOf([
            '1:1',
            '3:2',
        ]),
        labelIcon: DataTypes.string,
        labelText: DataTypes.string,
    };

    initData() {
        return {
            gutter: 9, // M_W_X001
            margin: 51, // M_W_X010
            spacing: true,
            imgList: [],
        };
    }

    inited() {
        const data = this.data;
        const labelIcon = data.get('labelIcon');
        const labelText = data.get('labelText');
        let imgList = data.get('imgList');

        if (imgList.length > 3) {
            imgList = imgList.slice(0, 3);
        }
        else if (imgList.length < 3) {
            const more = Array(3 - imgList.length).fill({});
            imgList.splice(imgList.length - 3, 0, ...more);
        }

        imgList[0].borderRadius = {
            tr: false,
            br: false,
        };

        imgList[1].borderRadius = {
            tl: false,
            tr: false,
            bl: false,
            br: false,
        };

        imgList[2].borderRadius = {
            tl: false,
            bl: false,
        };

        if (labelIcon || labelText) {
            imgList[2].labelIcon = labelIcon;
            imgList[2].labelText = labelText;
        }

        this.data.set('imgList', imgList);
    }

    handleClick(event) {
        this.fire('click', {event, data: this.data.get()});
    }
}
