// @baidu/xbox-native/device/cuid.js
/**
 * @file Created on Thu Nov 15 2018
 * @module device
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {version_compare as versionCompare, getBoxVersion} from '../detect';
import {cuid} from './index';
const boxVersion = getBoxVersion();

register('utils/getCUID', {
    androidlite: {
        versionRange: ['2.2.2'],
        thunk(res) {
            const cuid = res.result;
            return {
                status: 0,
                data: {
                    data: cuid
                }
            };
        }
    },
    androidmission: {
        versionRange: ['1.0'],
        thunk(res) {
            const cuid = res.result;
            return {
                status: 0,
                data: {
                    data: cuid
                }
            };
        }
    },
    iosmission: {
        versionRange: ['1.0'],
        thunk(res) {
            try {
                // 非加密
                let cuid = res.unique_id || res.result;
                return {
                    status: 0,
                    data: {
                        data: cuid
                    }
                };
            } catch (e) {
                console.log(e);
            }
        }
    },
    androidinfo: {
        versionRange: ['1.0'],
        thunk(res) {
            const cuid = res.result;
            return {
                status: 0,
                data: {
                    data: cuid
                }
            };
        }
    },
    iosinfo: {
        versionRange: ['1.1'],
        thunk(res) {
            try {
                // 非加密
                let cuid = res.unique_id;
                return {
                    status: 0,
                    data: {
                        data: cuid
                    }
                };
            } catch (e) {
                console.log(e);
            }
        }
    },

    ios: {
        versionRange: ['6.1'],
        filter() {
            return {
                module: 'utils',
                action: 'getCUID'
            };
        },
        thunk(res) {
            if (versionCompare(boxVersion, '8.4') >= 0) {
                try {
                    // 非加密
                    const cuid = res.unique_id;
                    return {
                        status: 0,
                        data: {
                            data: cuid
                        }
                    };
                } catch (e) {
                    console.log(e);
                }
            } else {
                let reg = /baiduboxapp\/.*\/(\w+)\/\d/;
                let ua = navigator.userAgent;
                let list = ua.match(reg);
                // 非加密
                let cuid = list ? list[1] : '';
                return {
                    status: 0,
                    data: {
                        data: cuid
                    }
                };
            }
        }
    },
    android: {
        versionRange: ['6.5'],
        filter() {
            return {
                module: 'Bdbox_android_utils',
                action: 'getcuid',
                async: false
            };
        },
        thunk(res) {
            if (versionCompare(boxVersion, '7.4') >= 0) {
                let cuid = res.result;
                return {
                    status: 0,
                    data: {
                        data: cuid
                    }
                };
            } else {
                let match = document.cookie.match(/BAIDUCUID=(.+?);/);
                let cuid = match ? match[1] : '';
                return {
                    status: 0,
                    data: {
                        data: cuid
                    }
                };
            }
        }
    }
});
/**
 * 获取 cuid
 * @function cuid
 */
export default cuid;
