// @baidu/xbox-native/device/netInfo.js
/**
 * @file 获取 info
 * @module device
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import getGlobalFunc from '@baidu/xbox/get-global-func';
import register from '../register';
import {info} from './index';
import {prompt as androidCmd} from '../android';
import iosInvoke from '../ios';
import {isMatrix, isIOS} from '../detect';
import stringify from '../stringify';

register('utils/getDeviceInfo', {
    androidlite: {
        versionRange: ['2.2.2']
    },
    androidinfo: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'utils',
                action: 'getDeviceInfo'
            };
        },
        thunk(res) {
            return {
                status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
                data: res.data,
                message: res.errmsg
            };
        }
    },
    android: {
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'Bdbox_android_utils',
                action: 'getDeviceInfo'
            };
        },
        thunk(res) {
            return {
                status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
                data: res.data,
                message: res.errmsg
            };
        }
    }
});
/**
 * 获取 netinfo 网络环境
 * @function netInfo
 * @param {any} abtest - abtest 参数
 */
export default abtest => {
    if (isMatrix('baiduboxvision')) {
        return new Promise((resolve, reject) => {
            const params = {
                keys: ['netInfo']
            };
            const callback = getGlobalFunc(res => {
                if (typeof res === 'string') {
                    res = JSON.parse(res);
                }
                resolve(res.data || {});
            });
            if (isIOS()) {
                iosInvoke(`baiduboxapp://utils?action=getDeviceInfo&params=${stringify(params, true)}&func=${callback}`);
            } else {
                androidCmd('Bdbox_android_utils', 'getDeviceInfo', [stringify(params), callback]);
            }
        });
    }
    return info(['netInfo'], abtest);
};
