// @baidu/xbox-native/device/screenInfo.js
/**
 * @file 获取 sreenInfo
 * @module device
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {info} from './index';
function thunk(res) {
    return {
        status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
        data: res.data,
        message: res.errmsg
    };
}
register('utils/getScreenInfo', {
    androidlite: {
        versionRange: ['2.2.2']
    },
    androidinfo: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosinfo: {
        versionRange: ['1.1']
    },
    ios: {
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'utils',
                action: 'getDeviceInfo'
            };
        },
        thunk
    },
    android: {
        versionRange: ['7.3'],
        filter() {
            return {
                module: 'Bdbox_android_utils',
                action: 'getDeviceInfo'
            };
        },
        thunk
    }
});
/**
 * 获取 screenInfo 网络环境
 * @function screenInfo
 * @param {any} abtest - abtest 参数
 */
export default abtest => {
    return info(['screenInfo'], abtest);
};
