// @baidu/xbox-native/device/vibrate.js
/**
 * @file 手机震动 swan 版本
 * @module device
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
// http://ft.baidu.com/#/scheme/device_vibrate
import {xRegister} from '../invoke';
/**
 * @function vibrate
 * @param [number] type - 震动类型，ios 用
 * - 震动类型：
 * 0.微弱短震【iPhone7，iPhone7 Plus以上可用】
 * 1.中等短震【iPhone7，iPhone7 Plus以上可用】
 * 2.明显短震【iPhone7，iPhone7 Plus以上可用】
 * 3.长震动 【所有机型】
 * @param [object] vibrationEffect - 震动效果（android专用）
 * @todo
 * 值为：{"timings":[100, 150, 100, 150],
 * "amplitudes":[0, 255, 0, 255]};
 * timings为震动控制里的震动时间数组，单位为毫秒，综合不超过阈值（暂定2000），
 * amplitudes代表震动强度数组，震动强度为SDK26以上可用，值为0~255，分别为不震到震动强度最大，
 * -1表示默认震动强度。
 */
export default (type = 3, vibrationEffect = {timings: [100, 150, 100, 150], amplitudes: [0, 255, 0, 255]}) => {
    return xRegister(
        {
            lite: {
                v: 34,
                versionRange: '3.9'
            },
            info: {
                v: 34,
                versionRange: '3.9'
            },
            main: {
                v: 30,
                versionRange: '10.9'
            }
        },
        'device/vibrate',
        {type, vibrationEffect}
    );
};
