// @baidu/fs-danmaku-h-mo/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var boxx = require('@baidu/boxx');
var feedLogic = require('@baidu/feed-logic');
var fsActions = require('@baidu/fs-actions');
var moActions = require('@baidu/mo-actions');
var Base = require('@baidu/mo-base');
var MoDanmakuH = require('@baidu/mo-danmaku-h');
var xbox = require('@baidu/xbox');
var styles = require('./style.module.less');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var boxx__default = /*#__PURE__*/_interopDefaultLegacy(boxx);
var Base__default = /*#__PURE__*/_interopDefaultLegacy(Base);
var MoDanmakuH__default = /*#__PURE__*/_interopDefaultLegacy(MoDanmakuH);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

// 评论服务的域名
var EXT_URL = 'https://ext.baidu.com';
// 点赞API http://app.baidu-int.com/docs/comment/#/yapi/20?id=%E8%AF%84%E8%AE%BA%E7%82%B9%E8%B5%9E
var PRAISE_API = "".concat(EXT_URL, "/api/comment/v1/comment/updown");

/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author fangbinbin02 <fangbinbin02@baidu.com>
 */
var _a;
// @san/component
var FsDanmakuHMo = /** @class */ (function (_super) {
    __extends(FsDanmakuHMo, _super);
    function FsDanmakuHMo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FsDanmakuHMo.prototype.initData = function () {
        return {
            styles: styles__default["default"],
            list: [],
            speed: 75,
            publisherStyle: '',
            publisherBtnStyle: '',
            sourceData: {
                source: '',
                appid: '',
                threadid: '',
            },
            commentTip: '说说你的看法',
        };
    };
    FsDanmakuHMo.prototype.attached = function () {
        this.eventDispatch(fsActions.SHOW_DANMAKU_H_MO);
    };
    FsDanmakuHMo.prototype.publish = function () {
        var _this = this;
        // const item = {
        //     'replyId': Math.random().toString(),
        //     'uname': 'fangbinbin',
        //     'isUped': true,
        //     'likeCount': '11',
        //     'dislikeCount': '0',
        //     'avatar': 'https://user-center.cdn.bcebos.com/head/raw/uc.101.9074d3.wfpQcSKFXItxEqCrmIm9xQ?x-bce-process=image/resize,m_lfit,w_200,h_200&autime=5789',
        //     'content': '我喜欢看书'
        // };
        // 展现事件
        this.eventDispatch(fsActions.INPUT_DANMAKU_H_MO_AROUSE);
        var publishEventCallback = function (res) {
            var parsedRes = typeof res === 'string' ? JSON.parse(res) : res;
            var resParams = parsedRes.params ? JSON.parse(decodeURIComponent(parsedRes.params)) : parsedRes;
            var isSuccess = parsedRes.status ? +parsedRes.status === 0 : +parsedRes.code === 0;
            if (isSuccess) {
                /* eslint-disable @typescript-eslint/naming-convention */
                var content = resParams.content, avatar = resParams.avatar, replyId = resParams.reply_id;
                /* eslint-enable @typescript-eslint/naming-convention */
                var item = {
                    replyId: replyId,
                    uname: '',
                    isUped: false,
                    likeCount: '0',
                    dislikeCount: '0',
                    avatar: avatar,
                    content: content,
                };
                _this.ref('moDanmakuH').insertItem(item);
                _this.eventDispatch(fsActions.PUBLISH_DANMAKU_H_MO_SUCCESS, {
                    sdk: 'fs-danmaku-h-mo',
                    data: parsedRes,
                });
            }
            else {
                _this.eventDispatch(fsActions.PUBLISH_DANMAKU_H_MO_ERROR, {
                    sdk: 'fs-danmaku-h-mo',
                    data: parsedRes,
                });
            }
            _this.eventDispatch(fsActions.PUBLISH_DANMAKU_H_MO, {
                sdk: 'fs-danmaku-h-mo',
                data: parsedRes,
            });
        };
        var publishHandler = function () {
            _this.eventDispatch(fsActions.PULISHER_DANMAKU_H_MO_AROUSE);
        };
        var sourceData = this.data.get('sourceData');
        /* eslint-disable @typescript-eslint/naming-convention,camelcase */
        var threadid = sourceData.threadid;
        if (boxx__default["default"].canIUse('comment.showBox')) {
            // http://es.baidu-int.com/index#/apidetail/657/params
            var params = {
                type: 0,
                source: sourceData.source,
                topic_id: threadid,
                comment_conf: {
                    img: {
                        switch: 0,
                    },
                },
                // 异步发布结束
                event_callback: xbox.getGlobalFunc(publishEventCallback),
                // 以下为必传内容否则报错，跟端沟通直接传空即可
                parent_id: '',
                rename: '',
                placeholder: '',
                logid: '',
                page: '',
                /* eslint-disable @typescript-eslint/no-empty-function */
                // 发布框调起
                success: publishHandler,
                fail: function () { },
                /* eslint-enable @typescript-eslint/no-empty-function */
            };
            /* eslint-enable @typescript-eslint/naming-convention,camelcase */
            boxx__default["default"].comment.showBox(__assign({}, params));
        }
        else if (boxx__default["default"].canIUse('commonUI.showToast')) {
            boxx__default["default"].commonUI.showToast({
                type: '1',
                message: 'APP版本太低，请进行升级',
            });
        }
    };
    FsDanmakuHMo.template = "\n        <div class='{{styles.fsDanmakuHMo}}'>\n            <div class='{{styles.danmakuHMo}}'>\n                <mo-danmaku-h\n                    s-ref='moDanmakuH'\n                    list='{{list}}'\n                    speed='{{speed}}'\n                    is-gray='{{isGray}}'\n                />\n            </div>\n            <div\n                class='{{styles.publisher}} {{isGray ? styles.gray : \"\"}}'\n                style='{{publisherStyle}}'\n                on-click='publish'\n            >\n                <span class='{{styles.text}}'>{{commentTip}}</span>\n                <div\n                    class='{{styles.btn}}'\n                    style='{{publisherBtnStyle}}'\n                >\n                    <span></span>\n                </div>\n            </div>\n        </div>\n    ";
    FsDanmakuHMo.components = {
        'mo-danmaku-h': MoDanmakuH__default["default"],
    };
    FsDanmakuHMo.messages = (_a = {},
        _a[moActions.CLICK_PRAISE] = function (_a) {
            var _this = this;
            var value = _a.value;
            var moDanmakuH = this.ref('moDanmakuH');
            var panelData = moDanmakuH.data.get('panelData');
            var channel = -1;
            var index = 0;
            var originItem = null;
            var replyId = value.replyId, isUped = value.isUped, likeCount = value.likeCount;
            for (var channelIndex in panelData) {
                if (channel === -1) {
                    for (var itemIndex in panelData[channelIndex]) {
                        if (panelData[channelIndex][itemIndex].replyId === replyId) {
                            channel = +channelIndex;
                            index = +itemIndex;
                            originItem = panelData[channelIndex][itemIndex];
                            break;
                        }
                    }
                }
            }
            // 假评论成功
            moDanmakuH.data.set("panelData[".concat(channel, "][").concat(index, "]"), __assign(__assign({}, originItem), { isUped: !isUped, likeCount: isUped ? likeCount - 1 : +likeCount + 1 }));
            var sourceData = this.data.get('sourceData');
            var data = {
                /* eslint-disable @typescript-eslint/naming-convention,camelcase */
                appid: sourceData.appid,
                thread_id: sourceData.threadid,
                client_type: 'app',
                reply_id: replyId,
                /* eslint-enable @typescript-eslint/naming-convention,camelcase */
                ts: (+new Date() / 1000).toString(),
            };
            if (isUped) {
                /* eslint-disable camelcase */
                data.undo_type = 1;
                /* eslint-enable camelcase */
            }
            else {
                data.type = 1;
            }
            feedLogic.post(PRAISE_API, data)
                .then(function () {
                _this.eventDispatch(fsActions.LIKE_DANMAKU_H_MO_SUCCESS, {
                    sdk: 'fs-danmaku-h-mo',
                    data: value,
                });
            })
                .catch(function () {
                _this.eventDispatch(fsActions.LIKE_DANMAKU_H_MO_ERROR, {
                    sdk: 'fs-danmaku-h-mo',
                    data: value,
                });
            });
        },
        _a[moActions.CLICK_DANMAKU_H_ITEM] = function (_a) {
            var value = _a.value;
            this.eventDispatch(fsActions.CLICK_DANMAKU_H_ITEM_MO, {
                sdk: 'fs-danmaku-h-mo',
                data: value,
            });
        },
        _a);
    return FsDanmakuHMo;
}(Base__default["default"]));

Object.defineProperty(exports, 'CLICK_DANMAKU_H_ITEM_MO', {
    enumerable: true,
    get: function () { return fsActions.CLICK_DANMAKU_H_ITEM_MO; }
});
Object.defineProperty(exports, 'LIKE_DANMAKU_H_MO_ERROR', {
    enumerable: true,
    get: function () { return fsActions.LIKE_DANMAKU_H_MO_ERROR; }
});
Object.defineProperty(exports, 'LIKE_DANMAKU_H_MO_SUCCESS', {
    enumerable: true,
    get: function () { return fsActions.LIKE_DANMAKU_H_MO_SUCCESS; }
});
Object.defineProperty(exports, 'PUBLISH_DANMAKU_H_MO', {
    enumerable: true,
    get: function () { return fsActions.PUBLISH_DANMAKU_H_MO; }
});
Object.defineProperty(exports, 'PUBLISH_DANMAKU_H_MO_ERROR', {
    enumerable: true,
    get: function () { return fsActions.PUBLISH_DANMAKU_H_MO_ERROR; }
});
Object.defineProperty(exports, 'PUBLISH_DANMAKU_H_MO_SUCCESS', {
    enumerable: true,
    get: function () { return fsActions.PUBLISH_DANMAKU_H_MO_SUCCESS; }
});
Object.defineProperty(exports, 'PULISHER_DANMAKU_H_MO_AROUSE', {
    enumerable: true,
    get: function () { return fsActions.PULISHER_DANMAKU_H_MO_AROUSE; }
});
Object.defineProperty(exports, 'SHOW_DANMAKU_H_MO', {
    enumerable: true,
    get: function () { return fsActions.SHOW_DANMAKU_H_MO; }
});
exports["default"] = FsDanmakuHMo;
