// @baidu/fs-danmaku-h-mo/dist/index.js
/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author fangbinbin02 <fangbinbin02@baidu.com>
 */
import boxx from '@baidu/boxx';
import { post } from '@baidu/feed-logic';
import { CLICK_DANMAKU_H_ITEM_MO, INPUT_DANMAKU_H_MO_AROUSE, LIKE_DANMAKU_H_MO_ERROR, LIKE_DANMAKU_H_MO_SUCCESS, PUBLISH_DANMAKU_H_MO, PUBLISH_DANMAKU_H_MO_ERROR, PUBLISH_DANMAKU_H_MO_SUCCESS, PULISHER_DANMAKU_H_MO_AROUSE, SHOW_DANMAKU_H_MO, } from '@baidu/fs-actions';
import { CLICK_DANMAKU_H_ITEM, CLICK_PRAISE, } from '@baidu/mo-actions';
import Base from '@baidu/mo-base';
import MoDanmakuH from '@baidu/mo-danmaku-h';
import { getGlobalFunc } from '@baidu/xbox';
import { PRAISE_API } from './api';
import styles from './style.module.less';
export { CLICK_DANMAKU_H_ITEM_MO, LIKE_DANMAKU_H_MO_ERROR, LIKE_DANMAKU_H_MO_SUCCESS, PUBLISH_DANMAKU_H_MO, PUBLISH_DANMAKU_H_MO_ERROR, PUBLISH_DANMAKU_H_MO_SUCCESS, PULISHER_DANMAKU_H_MO_AROUSE, SHOW_DANMAKU_H_MO, };
export * from './interface';
// @san/component
export default class FsDanmakuHMo extends Base {
    static template = /* html */ `
        <div class='{{styles.fsDanmakuHMo}}'>
            <div class='{{styles.danmakuHMo}}'>
                <mo-danmaku-h
                    s-ref='moDanmakuH'
                    list='{{list}}'
                    speed='{{speed}}'
                    is-gray='{{isGray}}'
                />
            </div>
            <div
                class='{{styles.publisher}} {{isGray ? styles.gray : ""}}'
                style='{{publisherStyle}}'
                on-click='publish'
            >
                <span class='{{styles.text}}'>{{commentTip}}</span>
                <div
                    class='{{styles.btn}}'
                    style='{{publisherBtnStyle}}'
                >
                    <span></span>
                </div>
            </div>
        </div>
    `;
    static components = {
        'mo-danmaku-h': MoDanmakuH,
    };
    static messages = {
        [CLICK_PRAISE]({ value }) {
            const moDanmakuH = this.ref('moDanmakuH');
            const panelData = moDanmakuH.data.get('panelData');
            let channel = -1;
            let index = 0;
            let originItem = null;
            const { replyId, isUped, likeCount } = value;
            for (const channelIndex in panelData) {
                if (channel === -1) {
                    for (const itemIndex in panelData[channelIndex]) {
                        if (panelData[channelIndex][itemIndex].replyId === replyId) {
                            channel = +channelIndex;
                            index = +itemIndex;
                            originItem = panelData[channelIndex][itemIndex];
                            break;
                        }
                    }
                }
            }
            // 假评论成功
            moDanmakuH.data.set(`panelData[${channel}][${index}]`, {
                ...originItem,
                isUped: !isUped,
                likeCount: isUped ? likeCount - 1 : +likeCount + 1,
            });
            const sourceData = this.data.get('sourceData');
            const data = {
                /* eslint-disable @typescript-eslint/naming-convention,camelcase */
                appid: sourceData.appid,
                thread_id: sourceData.threadid,
                client_type: 'app',
                reply_id: replyId,
                /* eslint-enable @typescript-eslint/naming-convention,camelcase */
                ts: (+new Date() / 1000).toString(),
            };
            if (isUped) {
                /* eslint-disable camelcase */
                data.undo_type = 1;
                /* eslint-enable camelcase */
            }
            else {
                data.type = 1;
            }
            post(PRAISE_API, data)
                .then(() => {
                this.eventDispatch(LIKE_DANMAKU_H_MO_SUCCESS, {
                    sdk: 'fs-danmaku-h-mo',
                    data: value,
                });
            })
                .catch(() => {
                this.eventDispatch(LIKE_DANMAKU_H_MO_ERROR, {
                    sdk: 'fs-danmaku-h-mo',
                    data: value,
                });
            });
        },
        [CLICK_DANMAKU_H_ITEM]({ value }) {
            this.eventDispatch(CLICK_DANMAKU_H_ITEM_MO, {
                sdk: 'fs-danmaku-h-mo',
                data: value,
            });
        },
    };
    initData() {
        return {
            styles,
            list: [],
            speed: 75,
            publisherStyle: '',
            publisherBtnStyle: '',
            sourceData: {
                source: '',
                appid: '',
                threadid: '',
            },
            commentTip: '说说你的看法',
        };
    }
    attached() {
        this.eventDispatch(SHOW_DANMAKU_H_MO);
    }
    publish() {
        // const item = {
        //     'replyId': Math.random().toString(),
        //     'uname': 'fangbinbin',
        //     'isUped': true,
        //     'likeCount': '11',
        //     'dislikeCount': '0',
        //     'avatar': 'https://user-center.cdn.bcebos.com/head/raw/uc.101.9074d3.wfpQcSKFXItxEqCrmIm9xQ?x-bce-process=image/resize,m_lfit,w_200,h_200&autime=5789',
        //     'content': '我喜欢看书'
        // };
        // 展现事件
        this.eventDispatch(INPUT_DANMAKU_H_MO_AROUSE);
        const publishEventCallback = (res) => {
            const parsedRes = typeof res === 'string' ? JSON.parse(res) : res;
            const resParams = parsedRes.params ? JSON.parse(decodeURIComponent(parsedRes.params)) : parsedRes;
            const isSuccess = parsedRes.status ? +parsedRes.status === 0 : +parsedRes.code === 0;
            if (isSuccess) {
                /* eslint-disable @typescript-eslint/naming-convention */
                const { content, avatar, reply_id: replyId } = resParams;
                /* eslint-enable @typescript-eslint/naming-convention */
                const item = {
                    replyId,
                    uname: '',
                    isUped: false,
                    likeCount: '0',
                    dislikeCount: '0',
                    avatar,
                    content,
                };
                this.ref('moDanmakuH').insertItem(item);
                this.eventDispatch(PUBLISH_DANMAKU_H_MO_SUCCESS, {
                    sdk: 'fs-danmaku-h-mo',
                    data: parsedRes,
                });
            }
            else {
                this.eventDispatch(PUBLISH_DANMAKU_H_MO_ERROR, {
                    sdk: 'fs-danmaku-h-mo',
                    data: parsedRes,
                });
            }
            this.eventDispatch(PUBLISH_DANMAKU_H_MO, {
                sdk: 'fs-danmaku-h-mo',
                data: parsedRes,
            });
        };
        const publishHandler = () => {
            this.eventDispatch(PULISHER_DANMAKU_H_MO_AROUSE);
        };
        const sourceData = this.data.get('sourceData');
        /* eslint-disable @typescript-eslint/naming-convention,camelcase */
        const threadid = sourceData.threadid;
        if (boxx.canIUse('comment.showBox')) {
            // http://es.baidu-int.com/index#/apidetail/657/params
            const params = {
                type: 0,
                source: sourceData.source,
                topic_id: threadid,
                comment_conf: {
                    img: {
                        switch: 0,
                    },
                },
                // 异步发布结束
                event_callback: getGlobalFunc(publishEventCallback),
                // 以下为必传内容否则报错，跟端沟通直接传空即可
                parent_id: '',
                rename: '',
                placeholder: '',
                logid: '',
                page: '',
                /* eslint-disable @typescript-eslint/no-empty-function */
                // 发布框调起
                success: publishHandler,
                fail() { },
                /* eslint-enable @typescript-eslint/no-empty-function */
            };
            /* eslint-enable @typescript-eslint/naming-convention,camelcase */
            boxx.comment.showBox({
                ...params,
            });
        }
        else if (boxx.canIUse('commonUI.showToast')) {
            boxx.commonUI.showToast({
                type: '1',
                message: 'APP版本太低，请进行升级',
            });
        }
    }
}
