// @baidu/xbox-native/storage/get.js

/**
 * @file Created on Thu Nov 15 2018
 * @module storage
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {get} from './index';
import {decode} from '../utils';

function thunk(res) {
    let result = {};
    try {
        result = JSON.parse(decode(res.data));
    } catch (e) {
        result = decode(res.data);
    }
    return {
        status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
        message: res.errmsg,
        data: {
            data: result
        }
    };
}
register(
    'storage/get',
    {
        v: 16,
        hasCB: true,
        versionRange: ['10.0'],
        filter() {
            return {
                module: 'dataStorage',
                action: 'getStorage'
            };
        },
        thunk(res) {
            let result = {};
            try {
                result = JSON.parse(decode(res.data.data));
            } catch (e) {
                result = decode(res.data.data);
            }
            if (res.data) {
                return {
                    status: res.status,
                    message: res.message,
                    data: {
                        data: result
                    }
                };
            } else {
                return res;
            }
        }
    },
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        ios: {
            versionRange: ['7.3'],
            filter(module, action, data) {
                if (data.domain) {
                    delete data.domain;
                }
                data.action = 'get';
                return {
                    module: 'utils',
                    action: 'webStorage',
                    data: data
                };
            },
            thunk
        },
        android: {
            versionRange: ['7.3'],
            filter(module, action, data) {
                if (data.domain) {
                    delete data.domain;
                }
                data.action = 'get';
                return {
                    module: 'Bdbox_android_utils',
                    action: 'webStorage',
                    data: data
                };
            },
            thunk
        }
    }
);
/**
 * 详细文档：http://bdbox.baidu.com/jssdk/api/cache.html#setstorage
 * @function get
 */
export default get;
