// @baidu/xbox-native/storage/remove.js
/**
 * @file Created on Thu Nov 15 2018
 * @module storage
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {del} from './index';
import stringify from '../stringify';

function thunk(res) {
    return {
        status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
        message: res.errmsg
    };
}
register(
    'storage/delete',
    {
        v: 16,
        hasCB: true,
        versionRange: ['10.0'],
        filter(data) {
            let d = data;
            if (d.data) {
                delete d.data;
            }
            return {
                module: 'dataStorage',
                action: 'removeStorage',
                data: {
                    params: stringify(data, true)
                }
            };
        }
    },
    {
        androidlite: {
            versionRange: ['2.2.2']
        },
        androidinfo: {
            versionRange: ['1.0']
        },
        iosinfo: {
            versionRange: ['1.1']
        },
        androidmission: {
            versionRange: ['1.0']
        },
        iosmission: {
            versionRange: ['1.0']
        },
        ios: {
            versionRange: ['7.3'],
            filter(module, action, data) {
                data.action = 'delete';
                return {
                    module: 'utils',
                    action: 'webStorage',
                    data: data
                };
            },
            thunk
        },
        android: {
            versionRange: ['7.3'],
            filter(module, action, data) {
                data.action = 'delete';
                return {
                    module: 'Bdbox_android_utils',
                    action: 'webStorage',
                    data: data
                };
            },
            thunk
        }
    }
);
/**
 * strore/delete
 * @function remove
 */
export default del;
