// @baidu/wuji-san/es/wuji-image-content/components/detail.js
import {Component} from 'san';
import IconText from './iconText';
import WujiProgress from '../../wuji-progress';
import WujiRate from '../../wuji-rate';
import Price from './price';
import Coupon from './coupon';
import Widget from '../../wuji-widget';
import Tags from './tags';
import TextButton from './textButton';

const COMPONENTLIST = [
    'iconText',
    'iconText2',
    'progress',
    'rate',
    'price',
    'coupon1',
    'coupon2',
    'tags',
    'abstract',
];
export default class WujiDetail extends Component {
    static template = /* html */ `
        <div class="detailContainer">
            <div s-for="item in formatList" class="item {{item.type}}">
                <icon-text s-if="item.type === 'iconText'" s-bind="item.data"></icon-text>
                <icon-text s-if="item.type === 'iconText2'" s-bind="item.data" isSpecial></icon-text>
                <wuji-progress s-if="item.type === 'progress'" s-bind="item.data"></wuji-progress>
                <wuji-rate s-if="item.type === 'rate'" s-bind="item.data"></wuji-rate>
                <price s-if="item.type === 'price'" s-bind="item.data"></price>
                <coupon s-if="item.type === 'coupon1'" data="{{item.data}}"></coupon>
                <coupon s-if="item.type === 'coupon2'" data="{{item.data}}"></coupon>
                <tags s-if="item.type === 'tags'" tags="{{item.data}}"></tags>
                <div s-if="item.type === 'abstract'" class="text text-{{item.data.lineClamp}}">{{item.data.text}}</div>
            </div>
            <div class="flexEnd">
                <text-button
                    s-if="textButton"
                    s-bind="textButton"
                ></text-button>
                <wuji-widget
                    s-if="widgets && widgets.length !== 0"
                    items="{{widgets}}"
                ></wuji-widget>
            </div>
        </div>
    `;
    static computed = {
        formatList() {
            const detailInfo = this.data.get('detailInfo');
            return COMPONENTLIST.filter(key => detailInfo[key]).map(item => ({
                type: item,
                data: detailInfo[item],
            }));
        },
    };
    initData() {
        return {
            detailInfo: {},
        };
    }
    static components = {
        'icon-text': IconText,
        'wuji-progress': WujiProgress,
        'wuji-rate': WujiRate,
        'price': Price,
        'coupon': Coupon,
        'wuji-widget': Widget,
        'tags': Tags,
        'text-button': TextButton,
    };
}
