// @baidu/xbox/es5/date-format.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _type = require("./type");

var _typeError = _interopRequireDefault(require("./type-error"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 日期格式化
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

/**
 * 日期格式化
 * @author wangyongqing01
 * @function dateFormat
 * @param {Date} d - date 对象
 * @param {string} [pattern = 'yyyy-MM-dd'] - 字符串
 * @return {string} 处理后的字符串
 * @example
 *	var d = new Date();
 *  dateFormat(d," yyyy年M月d日\n yyyy-MM-dd\n MM-dd-yy\n yyyy-MM-dd hh:mm:ss")
 *  //2018年11月10日\n 2018-01-01\n 01-01-18\n 2018-11-12 12:01:02
 */
var _default = function _default(d) {
  var pattern = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'yyyy-MM-dd';

  if (!(0, _type.isType)(d, 'date')) {
    throw (0, _typeError.default)('d', 'Date', d);
  }

  var y = d.getFullYear().toString();
  var o = {
    M: d.getMonth() + 1,
    // month
    d: d.getDate(),
    // day
    h: d.getHours(),
    // hour
    m: d.getMinutes(),
    // minute
    s: d.getSeconds() // second

  };
  pattern = pattern.replace(/(y+)/gi, function (a, b) {
    return y.substr(4 - Math.min(4, b.length));
  });
  Object.keys(o).forEach(function (i) {
    pattern = pattern.replace(new RegExp('(' + i + '+)', 'g'), function (a, b) {
      return o[i] < 10 && b.length > 1 ? '0' + o[i] : o[i];
    });
  });
  return pattern;
};

exports.default = _default;
