// @baidu/xbox/es5/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "guid", {
  enumerable: true,
  get: function get() {
    return _guid.default;
  }
});
Object.defineProperty(exports, "escapeHtml", {
  enumerable: true,
  get: function get() {
    return _escapeHtml.default;
  }
});
Object.defineProperty(exports, "dateFormat", {
  enumerable: true,
  get: function get() {
    return _dateFormat.default;
  }
});
Object.defineProperty(exports, "typeError", {
  enumerable: true,
  get: function get() {
    return _typeError.default;
  }
});
Object.defineProperty(exports, "ready", {
  enumerable: true,
  get: function get() {
    return _ready.default;
  }
});
Object.defineProperty(exports, "storage", {
  enumerable: true,
  get: function get() {
    return _storage.default;
  }
});
Object.defineProperty(exports, "Storage", {
  enumerable: true,
  get: function get() {
    return _storage.Storage;
  }
});
Object.defineProperty(exports, "getGlobal", {
  enumerable: true,
  get: function get() {
    return _getGlobal.default;
  }
});
Object.defineProperty(exports, "strictURIEncode", {
  enumerable: true,
  get: function get() {
    return _strictUriEncode.default;
  }
});
Object.defineProperty(exports, "decodeURIComponent", {
  enumerable: true,
  get: function get() {
    return _decodeUriComponent.default;
  }
});
Object.defineProperty(exports, "broadcast", {
  enumerable: true,
  get: function get() {
    return _emitter.default;
  }
});
Object.defineProperty(exports, "EventEmitter", {
  enumerable: true,
  get: function get() {
    return _emitter.EventEmitter;
  }
});
Object.defineProperty(exports, "emitter", {
  enumerable: true,
  get: function get() {
    return _emitter.default;
  }
});
Object.defineProperty(exports, "Monitor", {
  enumerable: true,
  get: function get() {
    return _monitor.default;
  }
});
Object.defineProperty(exports, "json2query", {
  enumerable: true,
  get: function get() {
    return _json2query.default;
  }
});
Object.defineProperty(exports, "query2json", {
  enumerable: true,
  get: function get() {
    return _query2json.default;
  }
});
Object.defineProperty(exports, "parseUrl", {
  enumerable: true,
  get: function get() {
    return _parseUrl.default;
  }
});
Object.defineProperty(exports, "each", {
  enumerable: true,
  get: function get() {
    return _each.default;
  }
});
Object.defineProperty(exports, "versionCompare", {
  enumerable: true,
  get: function get() {
    return _version_compare.default;
  }
});
Object.defineProperty(exports, "version_compare", {
  enumerable: true,
  get: function get() {
    return _version_compare.default;
  }
});
Object.defineProperty(exports, "substr", {
  enumerable: true,
  get: function get() {
    return _substr.default;
  }
});
Object.defineProperty(exports, "str2Array", {
  enumerable: true,
  get: function get() {
    return _substr.str2Array;
  }
});
Object.defineProperty(exports, "visibilityChange", {
  enumerable: true,
  get: function get() {
    return _visibilitychange.default;
  }
});
Object.defineProperty(exports, "clone", {
  enumerable: true,
  get: function get() {
    return _clone.clone;
  }
});
Object.defineProperty(exports, "o2o", {
  enumerable: true,
  get: function get() {
    return _o2o.default;
  }
});
Object.defineProperty(exports, "invoke", {
  enumerable: true,
  get: function get() {
    return _invoke.default;
  }
});
Object.defineProperty(exports, "invokeP", {
  enumerable: true,
  get: function get() {
    return _invoke.invokeP;
  }
});
Object.defineProperty(exports, "isMaybeBox", {
  enumerable: true,
  get: function get() {
    return _invoke.isMaybeBox;
  }
});
Object.defineProperty(exports, "xRegister", {
  enumerable: true,
  get: function get() {
    return _invoke.xRegister;
  }
});
Object.defineProperty(exports, "isJSON", {
  enumerable: true,
  get: function get() {
    return _is.isJSON;
  }
});
Object.defineProperty(exports, "isString", {
  enumerable: true,
  get: function get() {
    return _is.isString;
  }
});
Object.defineProperty(exports, "isArray", {
  enumerable: true,
  get: function get() {
    return _is.isArray;
  }
});
Object.defineProperty(exports, "isBoolean", {
  enumerable: true,
  get: function get() {
    return _is.isBoolean;
  }
});
Object.defineProperty(exports, "isBlank", {
  enumerable: true,
  get: function get() {
    return _is.isBlank;
  }
});
Object.defineProperty(exports, "isEmpty", {
  enumerable: true,
  get: function get() {
    return _is.isEmpty;
  }
});
Object.defineProperty(exports, "isFunction", {
  enumerable: true,
  get: function get() {
    return _is.isFunction;
  }
});
Object.defineProperty(exports, "isNil", {
  enumerable: true,
  get: function get() {
    return _is.isNil;
  }
});
Object.defineProperty(exports, "isNumber", {
  enumerable: true,
  get: function get() {
    return _is.isNumber;
  }
});
Object.defineProperty(exports, "isPlainObject", {
  enumerable: true,
  get: function get() {
    return _is.isPlainObject;
  }
});
Object.defineProperty(exports, "isUndefined", {
  enumerable: true,
  get: function get() {
    return _is.isUndefined;
  }
});
Object.defineProperty(exports, "isWindow", {
  enumerable: true,
  get: function get() {
    return _is.isWindow;
  }
});
Object.defineProperty(exports, "getType", {
  enumerable: true,
  get: function get() {
    return _type.default;
  }
});
Object.defineProperty(exports, "isType", {
  enumerable: true,
  get: function get() {
    return _type.isType;
  }
});
Object.defineProperty(exports, "toArray", {
  enumerable: true,
  get: function get() {
    return _to.toArray;
  }
});
Object.defineProperty(exports, "toBoolean", {
  enumerable: true,
  get: function get() {
    return _to.toBoolean;
  }
});
Object.defineProperty(exports, "toString", {
  enumerable: true,
  get: function get() {
    return _to.toString;
  }
});
Object.defineProperty(exports, "toNumber", {
  enumerable: true,
  get: function get() {
    return _to.toNumber;
  }
});
Object.defineProperty(exports, "isMatrix", {
  enumerable: true,
  get: function get() {
    return _isMatrix.default;
  }
});
Object.defineProperty(exports, "getMatrixSchema", {
  enumerable: true,
  get: function get() {
    return _getMatrixSchema.default;
  }
});
Object.defineProperty(exports, "isBox", {
  enumerable: true,
  get: function get() {
    return _isBox.default;
  }
});
Object.defineProperty(exports, "isAndroid", {
  enumerable: true,
  get: function get() {
    return _os.isAndroid;
  }
});
Object.defineProperty(exports, "isIOS", {
  enumerable: true,
  get: function get() {
    return _os.isIOS;
  }
});
Object.defineProperty(exports, "getOSName", {
  enumerable: true,
  get: function get() {
    return _os.getOSName;
  }
});
Object.defineProperty(exports, "getOSVersion", {
  enumerable: true,
  get: function get() {
    return _os.getOSVersion;
  }
});
Object.defineProperty(exports, "getBoxName", {
  enumerable: true,
  get: function get() {
    return _getBoxName.default;
  }
});
Object.defineProperty(exports, "getBoxVersion", {
  enumerable: true,
  get: function get() {
    return _getBoxVersion.default;
  }
});
Object.defineProperty(exports, "extend", {
  enumerable: true,
  get: function get() {
    return _extend.default;
  }
});
Object.defineProperty(exports, "decodeHtml", {
  enumerable: true,
  get: function get() {
    return _decodeHtml.default;
  }
});
Object.defineProperty(exports, "byteLen", {
  enumerable: true,
  get: function get() {
    return _byteLen.default;
  }
});
Object.defineProperty(exports, "stringify", {
  enumerable: true,
  get: function get() {
    return _stringify.default;
  }
});
Object.defineProperty(exports, "kebab2camel", {
  enumerable: true,
  get: function get() {
    return _kebab2camel.default;
  }
});
exports.delegate = exports.version = void 0;

var _guid = _interopRequireDefault(require("./guid"));

var _escapeHtml = _interopRequireDefault(require("./escape-html"));

var _dateFormat = _interopRequireDefault(require("./date-format"));

var _typeError = _interopRequireDefault(require("./type-error"));

var _ready = _interopRequireDefault(require("./browser/ready"));

var _storage = _interopRequireWildcard(require("./storage"));

var _getGlobal = _interopRequireDefault(require("./browser/get-global"));

var _strictUriEncode = _interopRequireDefault(require("./strict-uri-encode"));

var _decodeUriComponent = _interopRequireDefault(require("./decode-uri-component"));

var _emitter = _interopRequireWildcard(require("./emitter"));

var _monitor = _interopRequireDefault(require("./monitor/monitor"));

var _json2query = _interopRequireDefault(require("./json2query"));

var _query2json = _interopRequireDefault(require("./query2json"));

var _parseUrl = _interopRequireDefault(require("./parse-url"));

var _each = _interopRequireDefault(require("./each"));

var _version_compare = _interopRequireDefault(require("./version_compare"));

var _substr = _interopRequireWildcard(require("./substr"));

var _visibilitychange = _interopRequireDefault(require("./visibilitychange"));

var _clone = require("./clone");

var delegate = _interopRequireWildcard(require("./delegate"));

exports.delegate = delegate;

var _package = _interopRequireDefault(require("./package.json"));

var _o2o = _interopRequireDefault(require("./na/o2o"));

var _invoke = _interopRequireWildcard(require("./na/invoke"));

var _is = require("./is");

var _type = _interopRequireWildcard(require("./type"));

var _to = require("./to");

var _isMatrix = _interopRequireDefault(require("./is-matrix"));

var _getMatrixSchema = _interopRequireDefault(require("./get-matrix-schema"));

var _isBox = _interopRequireDefault(require("./is-box"));

var _os = require("./os");

var _getBoxName = _interopRequireDefault(require("./get-box-name"));

var _getBoxVersion = _interopRequireDefault(require("./get-box-version"));

var _extend = _interopRequireDefault(require("./extend"));

var _decodeHtml = _interopRequireDefault(require("./decode-html"));

var _byteLen = _interopRequireDefault(require("./byte-len"));

var _stringify = _interopRequireDefault(require("./stringify"));

var _kebab2camel = _interopRequireDefault(require("./kebab2camel"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 一般不要这样用，单独引用哦
 * @file Created on Thu Nov 15 2018
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
var version = _package.default.version;
exports.version = version;
