// @baidu/wuji-san/es/wuji-header/index.js
/**
 * @file 组件 WujiHeader 无极-头部
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../wuji-icon';

import '@baidu/wuji-icons/svg/arrow-o-42.svg';
import '@baidu/wuji-icons/svg/dislike-o-36.svg';

const prefixCls = 'w-header';

export default class WujiHeader extends Component {
    static template = /* html */ `
        <div class="${prefixCls}">
            <slot name="icon">
                <s-icon
                    s-if="icon"
                    s-bind="{{iconProps}}"
                    class="${prefixCls}-icon"
                    type="{{icon}}"
                    color="{{iconColor}}"
                />
            </slot>
            <div class="${prefixCls}-title">
                <div class="${prefixCls}-text"><span>{{title}}</span></div>
                <div s-if="title && subtitle" class="${prefixCls}-text ${prefixCls}-subtitle">
                    <div class="${prefixCls}-subtitle-line"/>
                    <span>{{subtitle}}</span>
                </div>
            </div>
            <slot name="helper">
                <div class="${prefixCls}-helper" on-click="helperClickHandler">
                    <span s-if="!dislike && helper">{{helper}}</span>
                    <s-icon
                        s-if="helperIcon"
                        class="${prefixCls}-helper-icon"
                        color="{{helperIcon.color}}"
                        size="{{helperIcon.size}}"
                        type="{{helperIcon.type}}"
                    />
                </div>
            </slot>
            <div s-if="hairline" class="${prefixCls}-hairline"/>
        </div>
    `;

    static trimWhitespace = 'all';

    static components = {
        's-icon': Icon,
    };

    static dataTypes = {
        arrow: DataTypes.bool,
        dislike: DataTypes.bool,
        hairline: DataTypes.bool,
        helper: DataTypes.string,
        icon: DataTypes.string,
        iconColor: DataTypes.string,
        title: DataTypes.string,
        subtitle: DataTypes.string,
        iconProps: DataTypes.object,
    };

    static computed = {
        helperIcon() {
            if (this.data.get('dislike')) {
                return {
                    size: 36,
                    type: 'dislike-o-36',
                };
            }

            if (this.data.get('arrow')) {
                return {
                    type: 'arrow-o-42',
                };
            }
        },
    };

    initData() {
        return {
            arrow: false,
            dislike: false,
            hairline: false,
            helper: '',
            icon: '',
            iconColor: 'red',
            title: '标题',
            subtitle: '',
            iconProps: {},
        };
    }

    helperClickHandler(event) {
        this.fire('helper-click', {event});
    }
}
