// @baidu/wuji-uikit/src/radius/radius.js
/**
 * @file 边角编码 radius.js
 *
 * DO NOT CHANGE, THIS IS GENERATED BY SCRIPT!
 */

module.exports.Android_Radius = {
    J_X01: null,
    J_X02: null,
    J_X03: null,
    J_X04: null,
    J_X05: null,
    J_X07: null,
    J_X08: null,
    J_X06: '0.5H',
    J_X09: null,
    J_X10: null,
    J_X11: null,
    J_X12: null,
    J_X13: null,
    J_X14: null,
    J_X15: null,
    J_X16: null,
    J_X17: null,
    J_X18: null,
    J_X19: null,
    J_X20: null,
    J_X21: null,
    J_X22: null,
    J_X23: null,
    J_X24: null,
    J_X25: null,
    J_X26: null,
    J_X28: '8',
    J_X29: '10',
    J_X30: null,
    J_X31: '12',
    J_X32: '16',
    J_X33: '2',
    J_X34: '11',
    J_X35: '7',
    J_X36: '15',
    J_X37: '18',
    J_X38: '17',
    J_X39: '5',
    J_X40: null,
    J_X42: '5',
    J_X43: '14',
    J_X44: '25',
    J_X45: '3',
    J_X46: '3',
    J_X47: '13',
    J_X48: '17',
};

module.exports.Ios_Radius = {
    J_X01: null,
    J_X02: null,
    J_X03: null,
    J_X04: null,
    J_X05: null,
    J_X07: null,
    J_X08: null,
    J_X06: '0.5H',
    J_X09: null,
    J_X10: null,
    J_X11: null,
    J_X12: null,
    J_X13: null,
    J_X14: null,
    J_X15: null,
    J_X16: null,
    J_X17: null,
    J_X18: null,
    J_X19: null,
    J_X20: null,
    J_X21: null,
    J_X22: null,
    J_X23: null,
    J_X24: null,
    J_X25: null,
    J_X26: null,
    J_X28: '9',
    J_X29: '12',
    J_X30: null,
    J_X31: '14',
    J_X32: '18',
    J_X33: '2',
    J_X34: '14',
    J_X35: '8',
    J_X36: '17',
    J_X37: '21',
    J_X38: '19',
    J_X39: '6',
    J_X40: null,
    J_X42: '6',
    J_X43: '16',
    J_X44: '29',
    J_X45: '3',
    J_X46: '4',
    J_X47: '15',
    J_X48: '20',
};

module.exports.H5_Radius = {
    J_X01: '6',
    J_X02: '9',
    J_X03: '15',
    J_X04: '24',
    J_X05: '39',
    J_X07: '0',
    J_X08: '0 0 9 9',
    J_X06: '0.5H',
    J_X09: '30',
    J_X10: '15 0 0 15',
    J_X11: '15 0 15 0',
    J_X12: '36',
    J_X13: '54',
    J_X14: '72',
    J_X15: '27',
    J_X16: '18',
    J_X17: '27 0 0 27',
    J_X18: '0 27 27 0',
    J_X19: '27 27 0 0',
    J_X20: '12',
    J_X21: '0 27 0 0',
    J_X22: '0 0 27 0',
    J_X23: '0 0 15 15',
    J_X24: '15 15 0 0',
    J_X25: '9 0 24 0',
    J_X26: '27 0 24 0',
    J_X28: '27',
    J_X29: '36',
    J_X30: '0',
    J_X31: '42',
    J_X32: '54',
    J_X33: '6',
    J_X34: '42',
    J_X35: '24',
    J_X36: '51',
    J_X37: '63',
    J_X38: '57',
    J_X39: '18',
    J_X40: '54 54 36 36',
    J_X42: '18',
    J_X43: '48',
    J_X44: '87',
    J_X45: '9',
    J_X46: '12',
    J_X47: '45',
    J_X48: '60',
};
