// @baidu/auc-miaodong/index.ts
/**
 * @file index.ts
 * @author gaohong01
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucMiaodongData} from './types';
import TopicMiaodong, {MIAODONG_CLICK} from '@baidu/mo-topic-miaodong';
import {componentLog} from '@baidu/auc-log';
import {env} from '@baidu/auc-base/env';
import {openImage} from '@baidu/auc-base/openImage';

export class AucMiaodong extends AucComponent {
    static template:string = /* html */ `
    <div class="{{wrapperClass}}">
        <topic-miaodong
            title="{{title}}"
            imageProps="{{imageProps}}"
        ></topic-miaodong>
    </div>
    `;
    static components:object = {
        'topic-miaodong': TopicMiaodong
    };

    static computed = {
        wrapperClass() {
            return [
                styles.aucMiaodongWrapper,
                this.data.get('visited') ? styles.visited : '',
                this.data.get('isBigEvent') ? styles.bigEvent : ''
            ]
        }
    };

    static messages = {
        /**
         * 百科点击
         */
        [MIAODONG_CLICK]: function () {
            // 继承原逻辑小米白牌内点击取消
            if (/_refluxos=a8/.test(window.location.search)) {
                return;
            }

            if (env.isBoxSeries) {
                openImage({
                    nid: this.data.get('fullPicture'),
                    imageList: [this.data.get('fullPicture')],
                    index: 0
                });
            } else {
                this.fire('backShoubai');
            }
            componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
        }
    };

    initData(): AucMiaodongData {
        return {
            styles,
            logBaseType: 'miaodong_',
            logExtraParams: {},
        };
    }

    attached(): void {
        this.registerScrollInView().then(() => {
            componentLog(this.data.get('logBaseType'), 'show', this.data.get('logExtraParams'));
        });
    }
}

export default AucMiaodong;
