// @baidu/xbox-native/es5/net/imageCache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module net
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
function thunk(res) {
  // errno为1是成功，0是失败
  return {
    status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
    message: res.errmsg,
    data: res.data
  };
}

(0, _register.default)('image/cache', {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    supportFrame: ['h5', 'hybrid', 'video', 'tags'],
    filter: function filter() {
      return {
        module: 'net',
        action: 'imageCache',
        extData: {
          minver: '7.3.0.0'
        }
      };
    },
    thunk: thunk
  },
  android: {
    supportFrame: ['hybrid', 'tags'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_net',
        action: 'imageCache'
      };
    },
    thunk: thunk
  }
});
/**
 * @function imageCache
 * @param {object} [data={}] - 参数
 */

var _default = function _default() {
  var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.download)((0, _utils.getParams)('image-cache', data));
};

exports.default = _default;
