// @baidu/xbox-native/es5/net/request.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file net requrest
 * @module net
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('net/request', {
  androidlite: {
    versionRange: ['2.2.2'],
    supportFrame: ['tags']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidinfo: {
    versionRange: ['1.0'],
    supportFrame: ['tags']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['7.3'],
    supportFrame: ['hybrid', 'h5', 'tags', 'video', 'search'],
    filter: function filter() {
      return {
        module: 'net',
        action: 'request'
      };
    },
    thunk: function thunk(res) {
      var data;

      if (res && res.data && res.errno === '1') {
        data = JSON.parse((0, _utils.decode)(res.data));
      } else {
        data = res.data;
      }

      return {
        status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
        message: res.message,
        data: data
      };
    }
  },
  android: {
    supportFrame: ['hybrid', 'tags'],
    versionRange: ['7.3'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_net',
        action: 'request'
      };
    },
    thunk: function thunk(res) {
      return {
        status: res.errno === '1' ? 0 : parseInt(res.errno, 10),
        message: res.errmsg,
        data: JSON.parse(res.data)
      };
    }
  }
});
/**
 * net/requst
 * @function request
 */

var _default = _index.request;
exports.default = _default;
