// boxjs/dist/thirdParty/bdbox/invoke/invoke.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /* eslint-disable */

  /**
   * 测试要点
   * 框外
   * 1、轻应用调起；ios，android
   * 2、卡片添加；ios，android
   * @author wangyongqing01
   * @version $Id: invoke.js 286585 2016-03-15 11:57:15Z wangyongqing01 $
   * 外观：
   * 弹层条件
   */

  /*jshint -W802,-W079*/


  require('../io/xDomain');

  require('../android/invokeApp');

  require('../ios/open');

  require('../utils/detect');

  var T = require('../utils/template');

  var getVersion = require('../utils/getVersion');

  var VERSION = getVersion();
  var IS_IN_BOX = $.$isBox(); // var HOST_REG = /(?:baidu|hao123).com/i;

  var B = require('../event/broadcast');

  var version_compare = require('../utils/version_compare');

  var APP_STORE_URL = 'http://itunes.apple.com/cn/app/id382201985?mt=8';
  /**
   * @typedef API_HASH_MAP
   * @type {Object}
   */

  var API_HASH_MAP = {
    /**
     * ios对象
     * @type {Object}
     */
    ios: {
      /**
       * 轻应用调起
       */
      'lightapp.invoke': {
        action: 'laopen'
      },

      /**
       * 卡片添加
       */
      'card.add': {
        action: 'cardMadd'
      },

      /**
       * 卡片状态查询
       */
      'card.query': {
        action: 'cardMquery'
      }
    },

    /**
     * android对象
     * @type {Object}
     */
    android: {
      /**
       * 轻应用添加
       */
      'lightapp.add': {
        namespace: 'Bdbox_android_xsearch',
        action: 'addWz',
        intent: ''
      },

      /**
       * 轻应用查询
       */
      'lightapp.query': {
        namespace: 'Bdbox_android_xsearch',
        action: 'queryWzStatus',
        intent: ''
      },
      //moplsu只有微站调起接口

      /**
       * 轻应用调起
       */
      'lightapp.invoke': {
        namespace: 'Bdbox_android_xsearch',
        action: 'invokeWz',
        intent: '#Intent;action=com.baidu.searchbox.action.aloader.VIEW;category=android.intent.category.DEFAULT;launchFlags=0x18000000;S.packageName=com.baidu.searchbox.plugins.xsearch;B.showMenu=false;S.pageId=<%=args%>;S.src=h5_export;end'
      },
      // 'lightapp.pay': {
      //     namespace: 'Bdbox_android_xsearch',
      //     action: 'invokeWz',
      //     intent: '#Intent;action=com.baidu.searchbox.action.aloader.VIEW;category=android.intent.category.DEFAULT;launchFlags=0x18000000;S.packageName=com.baidu.searchbox;B.showMenu=false;S.pageId=<%=args%>;S.src=h5_export;end'
      // },
      // moplus只有卡片添加接口

      /**
       * 卡片添加
       */
      'card.add': {
        namespace: 'Bdbox_android_card',
        action: 'madd',
        intent: '#Intent;action=com.baidu.searchbox.intent.action.CARD_MADD;category=android.intent.category.DEFAULT;launchFlags=0x10000000;S.cardMadd=<%=args%>;S.launcher_from=aladdin;end'
      },

      /**
       * 卡片状态查询
       */
      'card.query': {
        namespace: 'Bdbox_android_card',
        action: 'mquery',
        intent: ''
      }
    }
  };
  var apis = API_HASH_MAP[$.os.name];
  /**
   * Invoke构造器，封装了所有的invoke调起，解决了moplus百度域名问题
   * 请使用`Bdbox.invoke`
   * @constructor Invoke
   * @inner
   * @param {string} url 父窗口使用需要传入代理页面url，子窗口不传入
   * @param {string} minVersion 功能支持的最小版本，小于此版本则弹层
   * @author wangyongqing01
   * @version $Id: invoke.js 286585 2016-03-15 11:57:15Z wangyongqing01 $
   */

  function Invoke(url, minVersion) {
    this.proxy_url = url;
    this.minVersion = minVersion; //是否已经将浮层插入到页面判断

    this.$layer = null;
    this.$mask = null;
    this.$node = null;
    this.init();
  }

  Invoke.prototype = {
    constructor: Invoke,

    /**
     * 绑定xDomain事件
     */
    bindEvent: function () {
      this.xDomain.on('showUpdate', this.showUpdate);
    },

    /**
     * 显示升级弹层
     */
    showUpdate: function () {
      //显示升级
      if ($.os.ios) {
        if (window.confirm('此功能需要最新版手机百度支持，继续前往下载')) {
          location.href = APP_STORE_URL;
        }
      } else {
        var self = this;

        if (this.$node) {
          _center(this.$layer, this.$mask);

          this.$node.style.display = 'block';
          return this;
        }

        var template = Template;

        if (template) {
          //显示浮层
          var $node = build(template);
          this.$layer = $node.list.layer;
          this.$node = $node.node;
          this.$mask = $node.list.mask;
          var layer = $node.list.layer;

          _center(layer, this.$mask);

          this.$node.style.display = 'block';
          $node.list.close.addEventListener('click', function () {
            self.$node.style.display = 'none'; //触发close事件
            // self.fire('close');
          }, false); // $node.list.download.addEventListener('click', function() {
          //     //触发download事件
          //     self.fire('download');
          // }, false);

          getBody().appendChild($node.node);
        }
      }

      return this;
    },

    /**
     * 初始化
     */
    init: function () {
      var self = this;

      if (IS_IN_BOX || $.os.ios) {//在框内，或者是$.os.ios
      } else {
        this.xDomain = $.io.xDomain(this.proxy_url);
        this.xDomain.send('minversion', this.minVersion);
        this.bindEvent();
      }
    },

    /**
     * 主要功能，调起的功能
     * @param  {String}   action   字符串，传入的功能，包括：lightapp.add/query，card.add/query等，ios和android功能不同，详见{@link API_HASH_MAP}
     * @param  {object}   args     参数列表
     * @param  {Function} callback 回调函数，第一个参数是：是否有错误
     * @return {Invoke}
     */
    cmd: function (action, args, callback) {
      if (!(action in apis) && apis[action]) {
        throw 'not support ' + action;
      }

      var aObj = apis[action];
      var ns, ac;

      if (IS_IN_BOX || $.os.ios) {
        if (IS_IN_BOX && version_compare(VERSION, this.minVersion) < 0) {
          //框内 && 版本低提示升级
          this.showUpdate();
          return this;
        } //恶心实现了……
        //接口定的太乱


        switch ($.os.name) {
          case 'android':
            ns = aObj.namespace;
            ac = aObj.action;

            switch (action) {
              case 'lightapp.invoke':
                args = [$.isString(args) ? args : JSON.stringify(args), '{}'];
                break;

              default:
                args = $.isString(args) ? args : JSON.stringify(args);
            }

            var result = $.android.invokeApp(ns, ac, args);

            if ($.isFunction(callback)) {
              callback(result);
            }

            break;

          case 'ios':
            args = $.isString(args) ? args : JSON.stringify(args);
            ac = aObj.action;

            switch (action) {
              case 'card.query':
                this.invokeIos(ac, {
                  params: encodeURIComponent(args)
                }, callback);
                break;

              default:
                this.invokeIos(ac, {
                  params: encodeURIComponent(args)
                });
            }

            break;
        }
      } else if (aObj && aObj.intent) {
        //认为是框外，并且是Android手机
        var intent = apis[action].intent;
        args = $.isString(args) ? args : JSON.stringify(args); //这里真恶心！因为moplus乱七八糟的接口！！传入给moplus数据有的需要encode，有的不要！
        //没有规律，只能单个处理 %>_<%

        switch (action) {
          case 'card.add':
            args = encodeURIComponent(args);
            break;
        }

        intent = T(intent, {
          args: args
        });
        this.sendIntent(intent, $.isFunction(callback) ? callback : $.emptyFn);
      } else {
        this.showUpdate();
      }

      return this;
    },

    /**
     * 调起ios功能
     * @param  {String}   action   功能
     * @param  {object}   params   参数
     * @param  {Function} callback 接口本身回调函数
     * @param  {Function} cb       接口调用成功回调函数
     * @return {Invoke}
     */
    invokeIos: function (action, params, callback, cb) {
      var self = this;
      var url = [];

      if ($.isFunction(params)) {
        callback = params;
      } else {
        for (var i in params) {
          url.push(i + '=' + params[i]);
        }
      }

      if ($.isFunction(callback)) {
        var funcName = '_bdbox_js_' + $.getId();

        window[funcName] = function () {
          callback.apply(window, [].slice.call(arguments, 0));
          delete window[funcName];
        };

        url.push('func=' + funcName);
      }

      if (IS_IN_BOX) {
        url = 'baiduboxapp://' + action + '?' + url.join('&');
      } else {
        //框外使用v5协议，保证最小调起版本号，不满足则升级
        url.push('minver=' + this.minVersion);
        url = 'baiduboxappv5://' + action + '?' + url.join('&');
      }

      $.ios.open(url, IS_IN_BOX ? null : function (isTimeout) {
        if (isTimeout) {
          self.showUpdate();
        } else {
          $.isFunction(cb) && cb();
        }
      });
      return this;
    },

    /**
     * 获取moplus信息
     * @param  {Function} callback 回调函数
     * @return {Invoke}
     */
    getMoplusInfo: function (callback) {
      return this.xAjax('getInfo', null, callback);
    },

    /**
     * 发送intent
     * @param  {String}   intent   intent字符串
     * @param  {Function} callback 回调函数
     * @return {Invoke}
     */
    sendIntent: function (intent, callback) {
      return this.xAjax('sendIntent', {
        intent: intent
      }, callback);
    },

    /**
     * 跨域通信
     * @param  {string}   name     唯一标示名称
     * @param  {*}   data     传送数据
     * @param  {Function} callback 回调函数
     * @return {Invoke}
     */
    xAjax: function (name, data, callback) {
      var id = $.getId();
      var nname = name + '_' + id;

      var cb = function (data) {
        B.off(nname, cb);
        callback(data);
      };

      B.on(nname, cb);
      data = data || {};
      data._cbname = nname + '_cb';
      this.xDomain.once(data._cbname, function (data) {
        B.fire(nname, data);
      }).send(name, data);
      return this;
    }
  };
  /**
   * 框调起功能，实现框内框外统一接口，
   * 通过xbox代理文件，私下通信，支持回调和callback，返回{@link Invoke}
   *
   *
   * @example
   * var invoke = Bdbox.invoke(5.2);
   * invoke.cmd('card.add', [{"card_id":"009_%E6%9C%80%E7%BE%8E%E7%9A%84%E6%97%B6%E5%85%89","word":"最美的时光","fetchkey":"最美的时光","resource_id":"3100","resource_name":"tv","season":"1","src":"爱奇艺","csrc":"app_mainbox_txt","dsp":"iphone","tn":"wisexmlnew","type":"2"}], function(data){
   *       alert(JSON.stringify(data));
   *   });
   *  invoke.cmd('card.add', [{
   *      'mdsign': '8f0e6554517d2e3f7c9d635d0e382ba5',
   *      'resource_id': '14520',
   *      'tn': 'wisexmlnew',
   *      'csrc': 'app_guanwang',
   *      'pfid': 'a012',
   *      'resource_name': 'sports',
   *      'query': 'worldcup',
   *      'type': '2',
   *      'card_id': '012_worldcup',
   *      'dsp': 'iphone'
   *  }], function(data){
   *      alert(JSON.stringify(data));
   *  });
   *
   * @namespace Bdbox.invoke
   * @function
   * @param {string} minVersion 功能支持的最小版本，不满足则弹层分发
   * @return {Invoke}
   */

  return function (minVersion) {
    minVersion = minVersion || 5;
    return new Invoke('http://xbox.m.baidu.com/app/invoke/main', minVersion);
  }; //a({"download":"http://mo.baidu.com/baidusearch/download/baidusearch_AndroidPhone_1000715v.apk","version":"4.5.1","size":"4.1M","date":"2013-07-26"})

  var Template = '<div data-node="box-openapp" style="margin:0;padding:0;-webkit-text-size-adjust:100%;">' + '<div data-node="mask" style="z-index:99999990;opacity:.5;background-color:#000;position: absolute;top:0;left:0;height:100%;width:100%;font-size:15px;"></div>' + '<div data-node="layer" style="z-index:99999999;width:280px;height:154px;overflow:hidden;position:absolute;top:10px;left:50%;margin-left:-140px;-webkit-border-radius:2px;border-radius:2px;background:#fff;">' + '<a data-node="close" href="javascript:void(0);" style="position: absolute; display: block; width: 48px; height: 48px; top: 0px; right: 0px;background-image: url(http://m.baidu.com/static/search/inapp/lib/popup_sprite_png8.png?v=20130424); background-size:14px 80px; -webkit-background-size:14px 80px; background-position:26px -58px; background-repeat:no-repeat no-repeat;"></a>' + '<p style="font-size:16px;color:#262626;margin:22px 16px 10px 16px;text-align:left;min-height:1px;">使用手机百度</p><p style="color:#2a8de8;font-size:16px;margin:10px 16px;text-align:left;">添加后下次访问更方便</p><p style="margin:10px 16px;color:#999;text-align:left;">请下载最新版</p><div style="margin:16px 16px 21px 16px;line-height:30px;text-align:left;font-size:13px;"><a href="http://dl.ops.baidu.com/baidusearch_AndroidPhone_1001608s.apk" style="display:inline-block;-webkit-box-sizing:border-box;height:30px;width:150px;text-decoration:none;text-align:center;border:1px solid #7da9f2;background:#9ac4ff;color:#fff;"><span style="padding-left:20px;background:url(http://m.baidu.com/static/search/inapp/lib/popup_sprite_png8.png?v=20130424) left -15px no-repeat;background-size:14px 80px;-webkit-background-size:14px 80px;">立即下载</span></a>' + '</div>' + '</div>' + '</div>';

  function build(html) {
    var $node = document.createElement('div');
    $node.innerHTML = html;
    var $nodes = $node.querySelectorAll('[data-node]'),
        attr;
    var back = {
      node: $node,
      list: {}
    }; // evts = evts || {};

    for (var i = 0, len = $nodes.length; i < len; i++) {
      attr = $nodes[i].getAttribute('data-node');
      back.list[attr] = $nodes[i]; // if (evts[attr]) {
      //     evts[attr].addEventListener('click', );
      // }
    }

    return back;
  }

  var getComputedStyle = document.defaultView.getComputedStyle;

  function getStyle(node, property) {
    return node.style[property] || getComputedStyle(node, '').getPropertyValue(property);
  }

  function getBody() {
    return document.body || document.getElementsByTagName('body')[0];
  }

  function _center(layer, mask) {
    var iHeight = window.innerHeight;
    var body = getBody();
    var maxH = Math.max(iHeight, body.scrollHeight);
    mask.style.height = maxH + 'px';
    var h = parseInt(getStyle(layer, 'height'));
    var w = parseInt(getStyle(layer, 'width'));
    var scrollTop = !$.isUndefined(window.pageYOffset) ? window.pageYOffset : (document.documentElement || body.parentNode || body).scrollTop;
    layer.style.marginLeft = -w / 2 + 'px';
    layer.style.top = scrollTop + (iHeight - h) / 2 + 'px';
  }
}();
