// boxjs/dist/thirdParty/bdbox/invoke/openBox.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /* eslint-disable */

  /**
   * @file
   * Created by yupeng on 15/10/8.
   *
   * require('common:bdbox/invoke/openBox')({
   *     iosScheme: 'bdboxapp://***', //IOS scheme
   *     androidCommand: {}，//Android command
   *     sfrom: 'feed' //调起来源,统计使用,
   *     channel:''//调起官网是统计渠道号
   *     waitTime: 1500,//默认等待时间，超时则认为未调起,默认1500ms。
   *     failCallback: ''//如果不传则跳转官网,
   *     overwrite:1//是使用中间页跳转，比如applink和ulink，微信下展示面板
   * });
   */


  var _tipCss = '.openBox-weixinTip .options .link {' + 'overflow: hidden;' + 'text-overflow: ellipsis;' + 'white-space: nowrap;' + '}' + '.openBox-weixinTip *,' + '.openBox-weixinTip *:before,' + '.openBox-weixinTip *:after {' + '-webkit-box-sizing: inherit;' + 'box-sizing: inherit;' + '}' + '.openBox-weixinTip {' + '-webkit-box-sizing: border-box;' + 'box-sizing: border-box;' + 'position: fixed;' + 'top: 0;' + 'width: 100%;' + 'z-index: 999999;' + 'display: -webkit-box;' + '-webkit-box-orient: vertical;' + 'box-orient: vertical;' + 'min-height: 100%;' + 'background-color: #FFFFFF;' + '}' + '.openBox-weixinTip .main {' + '-webkit-box-flex: 1;' + 'box-flex: 1;' + 'padding: 0 35px;' + '}' + '.openBox-weixinTip .tips {' + 'padding-top: 38px;' + 'font-size: 15px;' + 'color: #aaa;' + '}' + '.openBox-weixinTip .options dt {' + 'padding: 13px 0px 9px;' + '}' + '.openBox-weixinTip .options dt .num {' + 'float: left;' + '-webkit-border-radius: 100%;' + 'border-radius: 100%;' + 'width: 16px;' + 'height: 16px;' + 'line-height: 16px;' + 'font-size: 13px;' + 'text-align: center;' + 'color: #fff;' + 'background-color: #5d646f;' + '}' + '.openBox-weixinTip .options dt .case {' + 'font-size: 17px;' + 'padding-left: 6px;' + '}' + '.openBox-weixinTip .options dd {' + 'padding-bottom: 24px;' + 'margin-bottom: 8px;' + 'line-height: 22.5px;' + '}' + '.openBox-weixinTip .options dd.bb {' + 'border-bottom: 1px solid #efefef;' + '}' + '.openBox-weixinTip .options .wizard {' + 'position: relative;' + '}' + '.openBox-weixinTip .options .wizard .arr {' + 'position: absolute;' + 'right: -10px;' + 'top: -50px;' + 'width: 40px;' + 'height: 40px;' + 'background: url(//s.bdstatic.com/common/openjs/openBox/wechatPop-arrow.png) no-repeat;' + '-webkit-background-size: cover;' + 'background-size: cover;' + '}' + '.openBox-weixinTip .options .wizard .img {' + 'padding-top: 18.033%;' + 'width: 100%;' + 'background: url(//s.bdstatic.com/common/openjs/openBox/wechatPop-wizard.png) no-repeat;' + '-webkit-background-size: cover;' + 'background-size: cover;' + '}' + '.openBox-weixinTip .options .wizard p {' + 'display: none;' + '}' + '.openBox-weixinTip .options .text {' + 'font-size: 17px;' + 'color: #777;' + '}' + '.openBox-weixinTip .options .link {' + 'display: block;' + 'text-decoration: underline;' + 'font-size: 17px;' + 'width: 100%;' + 'color: #3c76ff;' + '}' + '.openBox-weixinTip .options .btn {' + 'display: block;' + 'border: 1px solid #ccc;' + '-webkit-border-radius: 6px;' + 'border-radius: 6px;' + 'height: 40px;' + 'line-height: 40px;' + 'font-size: 16px;' + 'text-align: center;' + 'color: #000;' + 'background-color: #fff;' + '}' + '.openBox-weixinTip .options .btn:active {' + 'background-color: rgba(0, 0, 0, 0.1);' + '}' + '.openBox-weixinTip footer {' + 'height: 115px;' + '}' + '.openBox-weixinTip footer .logo {' + 'overflow: hidden;' + 'margin: 0 auto;' + 'width: 216px;' + 'height: 63px;' + 'text-indent: -9999px;' + 'background: url(//s.bdstatic.com/common/openjs/openBox/wechatPop-logo.png) no-repeat;' + '-webkit-background-size: cover;' + 'background-size: cover;' + '}' + '.openBox-weixinTip footer .copy {' + 'margin-top: 15px;' + 'text-align: center;' + 'font-size: 10px;' + 'color: #999;' + '};';

  var _applinkBase = 'http://a.app.qq.com/o/simple.jsp?pkgname=com.baidu.searchbox&android_schema=';
  var _ulinkBase = 'https://boxer.baidu.com/scheme?scheme=';
  var _guanwangUrl = 'https://mo.baidu.com/mo/home';

  function invoke(scheme, fail, waitTime) {
    if (_os.ios && versionCompare(_os.version, '9.0') >= 0) {
      window.location.href = scheme;
    } else {
      var $node = document.createElement('iframe');
      $node.style.display = 'none';
      $node.src = scheme;
      var body = document.body || document.getElementsByTagName('body')[0];
      body.appendChild($node);
      setTimeout(function () {
        body.removeChild($node);
        $node = null;
      }, 0);
    } // 记录起始时间


    var last = +Date.now();
    setTimeout(function () {
      if (Date.now() - last < waitTime + 200) {
        if (Bdbox.isFunction(fail)) {
          fail();
        } else {
          window.location.href = fail;
        }
      }
    }, waitTime);
  }

  ;

  var _html;

  var _tip;

  function showWeiXinTip() {
    if (!_tip) {
      _html = document.getElementsByTagName('html')[0];

      require('common:bdbox/utils/addStyle')(_tipCss);

      _tip = document.createElement('div');
    }

    _tip.setAttribute('class', 'openBox-weixinTip');

    _tip.innerHTML = '<section class="main">' + '<div class="tips"></div>' + '<dl class="options">' + '<dt>' + '<span class="num">1</span> <span class="case">若您已安装手机百度</span>' + '</dt>' + '<dd class="bb">' + '<div class="wizard">' + '<div class="arr"></div>' + '<div class="img"></div>' + '<p>第1步 点击该页右上角的“更多”</p>' + '<p>第2步 选择在浏览器中打开</p>' + '</div>' + '</dd>' + '<dt>' + '<span class="num">2</span>' + '<span class="case">若您尚未安装手机百度</span>' + '</dt>' + '<dd>' + '<a href="http://a.app.qq.com/o/simple.jsp?pkgname=com.baidu.searchbox" class="btn">去Appstore下载</a>' + '</dd>' + '</dl>' + '</section>' + '<footer>' + '<div class="logo"><strong>手机百度</strong>' + '<p>百度一下 你就得到</p></div>' + '<div class="copy">Copyright © 2016 BAIDU Corporation. All rights reserved.</div>' + '</footer>';
    document.body.appendChild(_tip);
    var htmlOverFlow = _html && _html.style.overflow;
    _html && (_html.style.overflow = "hidden");

    _tip.addEventListener('click', function (e) {
      if (e.target.tagName !== 'A') {
        _html && (_html.style.overflow = htmlOverFlow);
        document.body.removeChild(_tip);
      }
    });
  }

  var appRegs = {
    sinaweibo: /weibo/i,
    weixin: /micromessenger\//i,
    qqfriend: /QQ\//,
    qqdenglu: /Qzone\//,
    qqbrowser: /MQQBrowser\//i,
    uc: /UCBrowser\//i,
    miuiBrowser: /MiuiBrowser\//i,
    baidubrowser: /baidubrowser\//,
    chrome: /(Chrome|CriOS)\/([\d.]+)/,
    safari: /Safari\//i
  };
  /**
   * 获取打开页面的APP
   * @return {String} App名称
   */

  function getAppName(ua) {
    for (var i in appRegs) {
      if (appRegs.hasOwnProperty(i) && appRegs[i].test(ua)) {
        if (_os.ios || i != 'safari') {
          return i;
        }
      }
    }

    return 'other';
  }
  /**
   * 判断是否是空对象
   * @param  {Object}  e
   * @return {Boolean}
   */


  function isEmptyObject(obj) {
    var t;

    for (t in obj) return !1;

    return !0;
  }

  var Bdbox = $;

  var _os = require('../utils/detect')();

  var versionCompare = require('../utils/version_compare'); //判断是否是框内分享出来，需要框内分享的URL上加isBdboxShare=1.


  var _queryToJson = require('../utils/queryToJson');

  var _urlParames = _queryToJson(window.location.href);

  var _appName = getAppName(window.navigator.userAgent);

  var _pblog = require("../monitor/pblog");

  _pblog('init', [14]);

  return function (opts) {
    if (Bdbox.$isBox() || !Bdbox.isObject(opts)) {
      return;
    }

    var overwrite = opts.overwrite == 1 ? 1 : 0;
    var sfrom = opts.from || opts.sfrom || 'other';
    var waitTime = opts.waitTime || 1500;
    var failUrl = opts.failUrl || '';
    var logargs = {
      source: sfrom,
      from: 'openbox',
      page: _appName,
      type: _urlParames && _urlParames.isBdboxShare == 1 ? 'share' : '',
      value: '',
      channel: opts.channel || '',
      extlog: opts.extLog || ''
    };
    var fail;

    if (Bdbox.isFunction(opts.failCallback)) {
      fail = opts.failCallback;
    } else {
      // if(failUrl != ''){
      //     fail = failUrl;
      // }else{
      fail = opts.channel ? _guanwangUrl + '?from=' + opts.channel : _guanwangUrl; // }
    }

    if (_os.android && opts.androidCommand) {
      var androidScheme;

      if (Bdbox.isObject(opts.androidCommand)) {
        var minver = '7.4';

        if (opts.androidCommand.minver) {
          minver = opts.androidCommand.minver;
          delete opts.androidCommand.minver;
        }

        var params = '{"intent":"intent:#Intent;action=com.baidu.searchbox.action.HOME;component=com.baidu.searchbox\/.MainActivity';

        if (isEmptyObject(opts.androidCommand)) {
          params = params + ';end"}';
        } else {
          params = params + ';S.targetCommand=' + encodeURIComponent(JSON.stringify(opts.androidCommand)) + ';end"}';
        }

        androidScheme = 'baiduboxapp://utils?action=sendIntent&minver=' + minver + '&params=' + encodeURIComponent(params);
      } else {
        androidScheme = opts.androidCommand;
      }

      androidScheme = androidScheme + '&needlog=1&logargs=' + encodeURIComponent(JSON.stringify(logargs));

      _pblog('event', ['openBox', sfrom, logargs]);

      if (!overwrite && (_os.isWechat || _os.isQQ)) {
        window.location.href = _applinkBase + encodeURIComponent(androidScheme);
      } else {
        invoke(androidScheme, fail, waitTime);
      }
    } else if (_os.ios && opts.iosScheme) {
      var iosScheme = opts.iosScheme + '&needlog=1&logargs=' + encodeURIComponent(JSON.stringify(logargs));
      logargs.isUL = versionCompare(_os.version, '9.0') >= 0 && !overwrite ? 1 : 0;

      _pblog('event', ['openBox', sfrom, logargs]);

      if (versionCompare(_os.version, '9.0') < 0 || overwrite) {
        if (!overwrite && _os.isWechat) {
          showWeiXinTip();
        } else {
          invoke(iosScheme, fail, waitTime);
        }
      } else {
        var link = _ulinkBase + encodeURIComponent(iosScheme);

        if (failUrl != '') {
          link += '&target=' + encodeURIComponent(failUrl);
        }

        window.location.href = link;
      }
    } else {
      if (Bdbox.isFunction(fail)) {
        fail();
      } else {
        window.location.href = fail;
      }
    }

    ;
  };
}();
