// @baidu/auc-vedio-player/types.ts
/**
 * @file types.ts
 * @author gaohong01
 */
export interface AucVedioPlayerData {
    styles: object;
    /*vedioData */
    nid: string;
    vid: string;// 视频id，用于和落地页播放进度同步，二级标识
    src: string;// 视频资源路径
    pageUrl: string;//视频落地页url，用于和落地页播放进度同步，一级标识
    title: string;
    poster: string;// 封面图
    duration: string;// 总时间
    /*vedioConfig */
    loop: boolean;// 循环播放
    needReadLocalProgress: boolean;// 是否读取本地进度（从本地存储的进度开始播放）默认不读取，若需读取置为true 双端12.9开始支持
    isIgnoreVolumeChange: boolean; // 是否忽略物理声音变化，默认为false不忽略，若为true忽略，则系统声音变化不会影响静音状态
    ignore4gToast: boolean; // 忽略4g提示
    isIgnoreDeviceOrientation: boolean; // 忽略随设备转屏事件
    isPlayerCanEnterEndPage: number; // 屏蔽重新播放蒙层
    posterRatio: string; //默认展位图比例
    customizedVideo: boolean;
    /*vedioControlConfig */
    playerControlType: number; // ios播放器控件层类型0：只有一个播放画面; 不显示任何控件，1：仅显示底部沉浸式进度条 2：显示默认的控件层，默认0
    /*globalConfig */
    autoplay: boolean;// 1代表自动播放 0代表不自动播放
    muted: boolean; // 静音播放
    /*status */
    vedioStatus: number;// 自定义播放器状态，0资源未加载成功，1资源加载成功
    playTime: number;
    ticking: boolean;
    player: null;
};
