// @baidu/auc-star-title/index.ts
import {Component} from 'san';
import styles from './style.module.less';
import {aucStarTitleTypes} from './types';
import invoke from '@baidu/xbox/na/invoke';
import {command} from '@baidu/xbox/na/android';
import {env} from '@baidu/auc-base/env';

export class aucStarTitle extends Component {
    static template:string = `
    <div>
        <div class="{{styles.aucStarTitle}} {{hideMore ? styles.aucStarTitleHide : ''}}" on-click="handleClick($event)" s-ref="aucStarTitle">
            <div s-ref="aucStarTitleInner">
                {{formatTitle | raw}}
<!--                <div class="{{styles.more}} {{starActive ? styles.moreActive : ''}}" s-if="showMoreBtn" on-click="moreClick($event)">-->
<!--                    <span class="{{styles.moresl}}">...</span>全文-->
<!--                </div>-->
            </div>
        </div>
    </div>
    `;
    static components:object = {
    };

    initData(): aucStarTitleTypes {
        return {
            styles,
            titleLink: {}, // 头部跳转信息 包括话题，@等
            title: '',
            links: [],
            showMoreBtn: false,
            hideMore: true
        };
    }
    static computed = {
        formatTitle() {
            let title = this.data.get('title');
            this.data.get('links').forEach((item,index) => {
                title = title.replace(item.title, `<span class="titleLink" data-index="${index}">${item.title}</span>`)
            });
            return title;
        }
    };
    inited(): void {
    }
    attached(): void {
        let titleLink = this.data.get('titleLink');
        for (let i in titleLink) {
            titleLink[i] && titleLink[i].forEach(item => {
                this.data.push('links',item);
            })
        }
        if (this.ref('aucStarTitle').clientHeight < this.ref('aucStarTitleInner').clientHeight){
            this.data.set('showMoreBtn', true);
        }
    }
    handleClick(e): void {
        e.stopPropagation();
        e.preventDefault();
        if (/_refluxos=a8/.test(window.location.search)) {
            return;
        }
        let link = this.data.get('links')[+e.target.getAttribute('data-index')];
        if (!link) {
            this.fire('jump-detail');
            return;
        }
        if (/titleLink/.test(e.target.className)){
            if (env.isBoxSeries|| env.isXiaoMiBaiPai || env.isMatrix) {
                if (+link.cmd.mode === 2 || +link.cmd.mode === -1) {
                    invoke(link.cmd.url);
                } else {
                    if (/baiduboxapp/.test(link.cmd.intent)) {
                        invoke(link.cmd.intent);
                        return;
                    }
                    command(link.cmd.intent, link.cmd.mode, link.cmd.class, link.cmd.min_v);
                }
            } else {
                this.fire('back-shoubai');
            }
        } else {
            if (!/more/.test(e.target.className)) {
                this.fire('jump-detail');
            }
        }
    }
    moreClick(e) {
        e.stopPropagation();
        this.data.set('showMoreBtn', false);
        this.data.set('hideMore', false);
    }
}

export default aucStarTitle;
