// @baidu/xbox-player-mini/plugins/ExitFullscreen.js
export default class ExitFullscreen {
    constructor(api, pluginOptions) {
        this.api = api;
        this.pluginOptions = pluginOptions;
    }
    /**
     * 退出全屏
     * run为本插件对应的方法实际执行的函数
     */
    run() {
        if (this.api.getMode() === 'na') {
            return this.naRun();
        }
        this.h5Run();
    }
    naRun() {
        const {isIOS, isAndroid, invokeP} = this.api.getBoxTools();
        if (isIOS()) {
            const {fullScreenType, containerId} = this.api.getPlayerOption();
            const scHead = fullScreenType === 'windowFullScreen'
                ? 'video/slrExitFullWindowScreen' : 'video/slrExitFullCustomScreen';
            invokeP(scHead, {params: {containerId}});
        }
        else if (isAndroid()) {
            this.h5Run();
        }
    }
    h5Run() {
        let cancelFullscreen = document.exitFullScreen || document.webkitCancelFullScreen
            || document.mozCancelFullScreen || document.msExitFullscreen;
        cancelFullscreen.call(document);
    }
}
