// @baidu/xbox-player-mini/plugins/index.js
import Autoplay from './Autoplay';
import Muted from './Muted';
import Play from './Play';
import Volume from './Volume';
import Duration from './Duration';
import PlayerError from './PlayerError';
import PlaybackRate from './PlaybackRate';
import Src from './Src';
import Create from './Create';
import Ended from './Ended';
import Pause from './Pause';
// import Preload from './Preload';
import CurrentTime from './CurrentTime';
import Loop from './Loop';
import Paused from './Paused';
import Poster from './Poster';
import ReadyState from './ReadyState';
import NetworkState from './NetworkState';
import Replay from './Replay';
/**
 * 播放器内部属性
* @inner
 */
export default {
    // 属性
    autoplay: Autoplay, // 设置/返回 自动播放属性
    paused: Paused, // 是否处于暂停状态
    currentTime: CurrentTime, // 设置/返回视频当前播放时间
    duration: Duration, // 返回视频时长，单位：秒
    ended: Ended, // 返回视频是否播放结束
    loop: Loop, // 是否开启了循环播放
    muted: Muted, // 静音
    networkState: NetworkState, // 返回视频的当前网络状态
    playbackRate: PlaybackRate, // 设置或返回音频/视频的当前播放速度。
    // preload: Preload, // h5 option初始化参数配置，不可修改 规定是否在页面加载后载入视频。
    readyState: ReadyState, // 返回视频的就绪状态
    src: Src, // 设置/返回当前视频的地址
    poster: Poster, // 设置/返回当前视频的封面
    volume: Volume, // 设置/返回视频的音量
    error: PlayerError,
    // 'played', // h5 video属性  played 属性返回 TimeRanges 对象。TimeRanges 对象表示用户已经播放或看到的音频/视频范围。
    // 'seekable', //   h5 video属性 seekable 属性返回 TimeRanges 对象。TimeRanges 对象表示音频/视频中用户可寻址的范围。可寻址范围指的是用户在音频/视频中可寻址（移动播放位置）的时间范围。
    // 'seeking', // h5 video属性 属性返回用户目前是否在音频/视频中寻址。寻址中（Seeking）指的是用户在音频/视频中移动/跳跃到新的位置。
    // 方法
    create: Create,
    play: Play,
    replay: Replay,
    pause: Pause
};
