// @baidu/xbox-player-mini/plugins/IsExistSuspense.js
export default class IsExistSuspense {
    constructor(api, pluginOptions) {
        this.api = api;
        this.pluginOptions = pluginOptions;
    }
    /**
    * 是否存在小窗
    */
    get() {
        const {isIOS, isAndroid, invokeP} = this.api.getBoxTools();
        if (this.api.getMode() === 'na') {
            const containerId = this.api.getPlayerOption('containerId');
            if (isIOS()) {
                return new Promise(resolve => {
                    invokeP('video/slrIsexistSuspensePlayer', {params: {containerId}})
                        .then(res => {
                            this.api.setPlayerOption('suspenseType', res.type);
                            resolve(!!res.exist);
                        });
                });
            }
            else if (isAndroid()) {
                const video = this.api.getVideoInstance();
                video.zeusVideoExt = JSON.stringify({
                    id: containerId, // 唯一标识，必选
                    name: 'h5samelayer', // 播放器类型，默认h5samelayer
                    action: 'hasFloatingPlayer' // 进入小窗模式，必选
                });
                return new Promise(resolve => {
                    this.api.setPlayerController('androidIsExistSuspense', resolve);
                });
            }
        }
    }
}
