// @baidu/xbox-player-mini/plugins/Play.js
export default class Play {
    constructor(api, pluginOptions) {
        this.api = api;
        this.pluginOptions = pluginOptions;
    }
    /**
     * 播放{
     * run为本插件对应的方法实际执行的函数
     */
    run() {
        if (this.api.getMode() === 'na') {
            return this.naRun();
        }
        this.h5Run();
    }
    h5Run() {
        let video = this.api.getVideoInstance();
        return video && video.play();
    }
    naRun() {
        const {isIOS, isAndroid, invokeP} = this.api.getBoxTools();
        if (isIOS()) {
            const containerId = this.api.getPlayerOption('containerId');
            const player = this.api.getPlayerInstance();
            const paused = player.paused;
            invokeP('video/slrPlayPlayer', {params: {containerId}})
                .then(() => {
                    // 主要解决12.13之前的版本暂停后play事件的触发
                    if (paused && this.api.getPlayerController('playerKernelInit')) {
                        this.api.emit('play');
                        this.api.setPlayerOption('paused', false);
                        this.api.emit('playing');
                    }
                });
        }
        else if (isAndroid()) {
            this.h5Run();
        }
    }
}
