// @baidu/wuji-san/es/wuji-scroll/index.js
/**
 * @file 组件 WujiScroll 无极-横滑
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import pr2rem from '../utils/pr2rem';

const prefixCls = 'w-scroll';

export default class WujiScroll extends Component {
    // eslint-disable-next-line
    static template = /* html */ `
        <div class="${prefixCls}-wrap">
            <slot name="header"/>
            <div
                class="${prefixCls} {{spacing > 0 ? 'spacing' : ''}}"
                style="{{scrollerStyle(spacing)}}"
                on-scroll="handleScroll"
            >
                <div
                    class="${prefixCls}-container"
                    style="{{containerStyle}}"
                ><slot/></div>
                <div
                    s-if="spacing > 0"
                    class="${prefixCls}-placeholder"
                    style="{{placeHolderStyle(spacing)}}"
                />
            </div>
        </div>
    `;

    static dataTypes = {
        containerStyle: DataTypes.oneOfType([
            DataTypes.string,
            DataTypes.object,
        ]),
        spacing: DataTypes.oneOfType([
            DataTypes.bool,
            DataTypes.number,
            DataTypes.string,
        ]),
    };

    initData() {
        return {
            spacing: true,
            containerStyle: '',
        };
    }

    scrollerStyle(spacing) {
        if (typeof spacing !== 'boolean' && spacing > 0) {
            const value = pr2rem(spacing) + 'rem';
            return {
                'padding-left': value,
                'padding-right': value,
            };
        }
    }

    placeHolderStyle(spacing) {
        if (typeof spacing !== 'boolean' && spacing > 0) {
            const value = pr2rem(spacing) + 'rem';
            return {
                width: value,
            };
        }
    }

    handleScroll(event) {
        this.fire('scroll', {event});
    }
}
