// @baidu/ubc-report-hn-sdk/src/config.js
/**
 * @file ubcLog的conf配置
 * @author songzimei<songzimei@baidu.com>
 */

export const CONFIG = {
    UBC_URL: 'https://mbd.baidu.com/ztbox',
    TEST_UBC_URL: 'http://bjyz-mco-searchbox201609-m12xi3-044.bjyz.baidu.com:8080/ztbox',
    DEFAULT_LOG_DATA: {
        cateid: '99',
        actiondata: {
            // 外部填写（ubcID）
            id: '0',
            // 日志类型：0—事件类（web类）1—流类型（客户端)
            type: '0',
            timestamp: 0,
            content: {
                // 页面名称
                page: '-',
                // 区块位置
                type: '-',
                // 页面来源
                value: '-',
                // 资源来源
                source: 'default',
                // 日志类型
                from: 'act',
                // 扩展字段
                ext: {
                    apptype: '-'
                }
            }
        }
    },
    DEFAULT_QUERY: {
        action: 'zpblog',
        v: '2.0',
        appname: 'baiduboxapp',
        from: '',
        cfrom: '',
        smfw: ''
    }
};

export const CONFIG_MERGE = {
    UBC_URL: 'https://mbd.baidu.com/ztbox',
    TEST_UBC_URL: 'http://bjyz-mco-searchbox201609-m12xi3-044.bjyz.baidu.com:8080/ztbox',
    DEFAULT_LOG_DATA: {
        cateid: '101',
        data: []
    },
    DEFAULT_QUERY: {
        action: 'mpblog',
        v: '2.0',
        appname: 'baiduboxapp',
        from: '',
        cfrom: '',
        smfw: '',
        sid: '',
        sn: ''
    }
};
