// @baidu/mo-card/dist/index.amd.js
define(['exports', '@baidu/mo-actions', '@baidu/mo-base', './style.module.less'], (function (exports, moActions, Base, styles) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var Base__default = /*#__PURE__*/_interopDefaultLegacy(Base);
    var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * Copyright (C) 2021 ~ present baidu.com
     *
     * @author lvkunjie <lvkunjie@baidu.com>
     */
    // @san/component
    var Card = /** @class */ (function (_super) {
        __extends(Card, _super);
        function Card() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Card.prototype.initData = function () {
            return {
                styles: styles__default["default"],
                fontColor: '',
            };
        };
        Card.prototype.handleMoreLinkClick = function () {
            var moreUrl = this.data.get('moreUrl');
            this.eventDispatch(moActions.CLICK_MORE_LINK, {
                moreUrl: moreUrl,
            });
        };
        Card.template = "\n        <div\n            class=\"{{[\n                styles.wrap\n            ]}}\"\n        >\n            <div\n                class=\"{{[\n                    styles.main,\n                    styles.titleOneLine\n                ]}}\"\n                on-click=\"handleMoreLinkClick\"\n            >\n                <h1\n                    s-if=\"{{titleImg || title}}\"\n                    class=\"{{[\n                        styles.title,\n                        titleImg ? styles.titleImg : ''\n                    ]}}\"\n                    style=\"{{specSt}}\"\n                >{{titleImg ? '' : title}}</h1>\n                <p\n                    s-if=\"{{hasDesc && desc}}\"\n                    class=\"{{styles.rightHead}}\"\n                >\n                    <span\n                        class=\"{{styles.desc}}\"\n                        style=\"{{descStyle}}\"\n                    >{{desc}}</span>\n                    <span\n                        s-if=\"{{moreUrl}}\"\n                        class=\"{{styles.moreIcon}}\"\n                    ></span>\n                </p>\n            </div>\n            <div class=\"{{styles.content}}\">\n                <slot name=\"content\">slot:content</slot>\n            </div>\n        </div>\n    ";
        Card.computed = {
            specSt: function () {
                var str = '';
                var titleImg = this.data.get('titleImg');
                var color = this.data.get('color');
                if (titleImg) {
                    str = "background-image: url(".concat(titleImg, ")");
                }
                else if (color) {
                    str = "color: ".concat(color);
                }
                return str;
            },
            descStyle: function () {
                var descColor = this.data.get('descColor');
                return descColor ? "color: ".concat(descColor) : '';
            },
        };
        return Card;
    }(Base__default["default"]));

    exports["default"] = Card;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
