// @baidu/mo-card/dist/index.js
/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author lvkunjie <lvkunjie@baidu.com>
 */
import { CLICK_MORE_LINK, } from '@baidu/mo-actions';
import Base from '@baidu/mo-base';
import styles from './style.module.less';
export * from './interface';
// @san/component
export default class Card extends Base {
    static template = /* html */ `
        <div
            class="{{[
                styles.wrap
            ]}}"
        >
            <div
                class="{{[
                    styles.main,
                    styles.titleOneLine
                ]}}"
                on-click="handleMoreLinkClick"
            >
                <h1
                    s-if="{{titleImg || title}}"
                    class="{{[
                        styles.title,
                        titleImg ? styles.titleImg : ''
                    ]}}"
                    style="{{specSt}}"
                >{{titleImg ? '' : title}}</h1>
                <p
                    s-if="{{hasDesc && desc}}"
                    class="{{styles.rightHead}}"
                >
                    <span
                        class="{{styles.desc}}"
                        style="{{descStyle}}"
                    >{{desc}}</span>
                    <span
                        s-if="{{moreUrl}}"
                        class="{{styles.moreIcon}}"
                    ></span>
                </p>
            </div>
            <div class="{{styles.content}}">
                <slot name="content">slot:content</slot>
            </div>
        </div>
    `;
    static computed = {
        specSt() {
            let str = '';
            const titleImg = this.data.get('titleImg');
            const color = this.data.get('color');
            if (titleImg) {
                str = `background-image: url(${titleImg})`;
            }
            else if (color) {
                str = `color: ${color}`;
            }
            return str;
        },
        descStyle() {
            const descColor = this.data.get('descColor');
            return descColor ? `color: ${descColor}` : '';
        },
    };
    initData() {
        return {
            styles,
            fontColor: '',
        };
    }
    handleMoreLinkClick() {
        const moreUrl = this.data.get('moreUrl');
        this.eventDispatch(CLICK_MORE_LINK, {
            moreUrl,
        });
    }
}
