// @baidu/nano/es/cell-group/index.js
/**
 * @file 组件 cell-group 单元格组
 * @author baozhixin <baozhixin@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import Icon from '../icon';

const prefixCls = 's-cell-group';

export default class CellGroup extends Component {
    // eslint-disable-next-line
    static template = /*html*/ `
        <div class="{{wrapClass}}">
            <slot name="title">
                <div
                    s-if="title || icon"
                    class="${prefixCls}-title"
                    style="{{titleStyle}}"
                    on-click="handleTitleClick"
                >
                    <s-icon
                        s-if="icon"
                        class="${prefixCls}-icon"
                        size="xs"
                        type="{{icon}}"
                    />
                    <span>{{title}}</span>
                    <div class="hairline"/>
                </div>
            </slot>
            <slot/>
        </div>
    `;

    static components = {
        's-icon': Icon
    };

    static dataTypes = {
        icon: DataTypes.string,
        title: DataTypes.string,
        titleStyle: DataTypes.object,
        gapTop: DataTypes.bool,
        gapBottom: DataTypes.bool
    };

    static computed = {
        wrapClass() {
            const gapTop = this.data.get('gapTop');
            const gapBottom = this.data.get('gapBottom');
            return [
                prefixCls,
                gapTop && `${prefixCls}-gap-top`,
                gapBottom && `${prefixCls}-gap-bottom`
            ].filter(c => c);
        }
    };

    initData() {
        return {
            icon: '',
            title: '',
            titleStyle: {},
            gapTop: false,
            gapBottom: false
        };
    }

    handleTitleClick(e) {
        this.fire('title-click', e);
    }
}
