// @baidu/xbox-native/es5/ui/cancelPraise.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2019-3-14 取消点赞动画
 * @module ui/cancelPraise
 * @author sunxiaoyu01 <sunxiaoyu01@baidu.com>
 */
(0, _register.default)('praise/openCancelPraise', {
  v: 42,
  androidVersionRange: ['11.3.6'],
  iosVersionRange: ['11.3.6'],
  filter: function filter() {
    return {
      module: 'praise',
      action: 'cancelPraise'
    };
  }
});
/**
 * @function openCancelPraise
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('praise-cancelPraise', params));
};

exports.default = _default;
