// @baidu/xbox-native/es5/ui/closeLoading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _detect = require("../detect");

var _android = require("../android");

var _ios = _interopRequireDefault(require("../ios"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
var boxVersion = (0, _detect.getBoxVersion)();
(0, _register.default)('loading/close', {
  androidlite: {
    versionRange: ['2.2.2'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_net',
        action: 'hideLoadingView'
      };
    }
  },
  androidmission: {
    versionRange: ['1.0'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_net',
        action: 'hideLoadingView'
      };
    }
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidinfo: {
    versionRange: ['1.0'],
    filter: function filter() {
      return {
        module: 'bd_searchbox_interface',
        action: 'progressCompleted'
      };
    }
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['5.1'],
    hasCB: false,
    filter: function filter() {
      return {
        module: 'speedProgressCompleted',
        action: '',
        extData: {
          minver: '5.1.0.0'
        }
      };
    }
  },
  android: {
    versionRange: ['4.9'],
    supportFrame: ['hybrid'],
    hasCB: false,
    filter: function filter() {
      if ((0, _detect.version_compare)(boxVersion, '8.2') >= 0) {
        return {
          module: 'Bdbox_android_net',
          action: 'hideLoadingView'
        };
      } else {
        return {
          module: 'bd_searchbox_interface',
          action: 'progressCompleted'
        };
      }
    }
  }
});
/**
 * @function closeLoading
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if ((0, _detect.isMatrix)('baiduboxvision')) {
    (0, _detect.isIOS)() ? (0, _ios.default)('baiduboxapp://speedProgressCompleted?minver=1.0.0.0') : (0, _android.prompt)('Bdbox_android_net', 'hideLoadingView', []);
    return Promise.resolve();
  }

  return (0, _index.close)((0, _utils.getParams)('loading', params));
};

exports.default = _default;
