// @baidu/xbox-native/es5/ui/openImage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _detect = require("../detect");

var _android = require("../android");

var _ios = _interopRequireDefault(require("../ios"));

var _index = require("./index");

var _stringify = _interopRequireDefault(require("@baidu/xbox/es5/stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
var boxVersion = (0, _detect.getBoxVersion)();
(0, _register.default)('utils/openImage', {
  androidlite: {
    versionRange: ['2.2.2'],
    filter: function filter(module, action, data) {
      data.type = '0';
      return {
        module: 'Bdbox_android_utils',
        action: 'lightImage',
        data: data
      };
    }
  },
  androidmission: {
    versionRange: ['1.0'],
    filter: function filter(module, action, data) {
      data.type = '0';
      return {
        module: 'Bdbox_android_utils',
        action: 'lightImage',
        data: data
      };
    }
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidinfo: {
    versionRange: ['1.0'],
    filter: function filter(module, action, data) {
      data.type = '0';
      return {
        module: 'Bdbox_android_utils',
        action: 'lightImage',
        data: data
      };
    }
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['6.2'],
    filter: function filter(module, action, data) {
      data.type = '0';
      data.index = data.index || 0;
      return {
        module: 'utils',
        action: 'image',
        data: data
      };
    },
    thunk: function thunk(res) {
      return {
        status: 0,
        message: '',
        data: ''
      };
    }
  },
  android: {
    versionRange: ['6.1'],
    filter: function filter(module, action, data) {
      data.type = '0';

      if ((0, _detect.version_compare)(boxVersion, '7.4') >= 0) {
        return {
          module: 'Bdbox_android_utils',
          action: 'lightImage',
          data: data
        };
      } else {
        return {
          module: 'Bdbox_android_utils',
          action: 'image',
          data: data
        };
      }
    },
    thunk: function thunk() {
      return {
        status: 0,
        message: ''
      };
    }
  }
});
/**
 * 大图浏览器
 * @function openImage
 * @param {Object} params 端能力参数
 * @return {Promise} 调用的返回
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if ((0, _detect.isMatrix)('baiduboxvision')) {
    params.type = '0';
    return (0, _detect.isIOS)() ? (0, _ios.default)("baiduboxapp://utils?action=image&params=".concat((0, _stringify.default)(params, true))) : (0, _android.prompt)('Bdbox_android_utils', 'lightImage', [(0, _stringify.default)(params), 'syncCallback']);
  }

  return (0, _index.open)((0, _utils.getParams)('utils-image', params));
};

exports.default = _default;
