// @baidu/xbox-native/es5/ui/openFeedback.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _index = require("./index");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file open feedback sdk - created on 2019-3-4
 * @module ui
 * @author jiaowenhua <jiaowenhua@baidu.com>
 */
(0, _register.default)('feedback/open', {
  v: 17,
  versionRange: ['10.1'],
  iosinfoVersionRange: ['2.8'],
  androidliteVersionRange: ['3.3'],
  iosmissionVersionRange: ['1.0'],
  androidmissionVersionRange: ['1.0'],
  filter: function filter(data) {
    return {
      module: 'utils',
      action: 'feedback'
    };
  }
});
/**
 * 打开举报反馈页面
 *
 * @function openFeedback
 * @param {Object} params 端能力参数，具体值端同学还在整理中，@：liushaohua02@baidu.com
 * @return {Promise} 调用的返回
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('feedback', params));
};

exports.default = _default;
