// @baidu/xbox-native/es5/ui/openRedPacket.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _stringify = _interopRequireDefault(require("../stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('redPacket/open', {
  v: 17,
  versionRange: ['10.1'],
  filter: function filter(data) {
    data.ext = (0, _stringify.default)(data.ext);
    return {
      module: 'redPacket',
      action: 'getRedPacket',
      data: {
        params: (0, _stringify.default)(data, true)
      }
    };
  }
});
/**
 * @function openRedPacket
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('redPacket', params));
};

exports.default = _default;
