// @baidu/xbox-native/es5/ui/openPassAssist.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _index = require("./index");

var _utils = require("../utils");

var _getGlobalFunc = _interopRequireDefault(require("@baidu/xbox/es5/get-global-func"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file open pass sdk - created on 2019-4-1
 * @module ui
 * @author jiaowenhua <jiaowenhua@baidu.com>
 */
(0, _register.default)('account/openPassAssist', {
  v: 31,
  versionRange: ['10.10'],
  filter: function filter(data) {
    return {
      module: 'account',
      action: 'passAssist'
    };
  }
});
/**
 * passNA版调用，目前只支持人脸
 *
 * @function openPassAssist
 * @param {Object} params 端能力参数，具体值参考 http://ft.baidu.com/#/scheme/account_passassist
 * @return {Promise} 调用的返回
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return new Promise(function (resolve, reject) {
    params.passCallback = (0, _getGlobalFunc.default)(function (data) {
      data = JSON.parse((0, _utils.decode)(data));
      var dataInfo = data.data || {};

      if (dataInfo.result) {
        dataInfo = dataInfo.result || {};
      }

      if (+data.status === 0 && +dataInfo.retCode === 0) {
        resolve({
          errno: 0
        });
      } else {
        reject({
          errno: dataInfo.retCode,
          errmsg: dataInfo.retMsg
        });
      }
    });
    (0, _index.open)((0, _utils.getParams)('account-passAssist', params));
  });
};

exports.default = _default;
