// @baidu/xbox-native/es5/ui/openShare.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _detect = require("../detect");

var _stringify = _interopRequireDefault(require("../stringify"));

var _android = require("../android");

var _ios = _interopRequireDefault(require("../ios"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
var dataFormat = function dataFormat(data) {
  var d = {
    imageUrl: data.iconUrl,
    title: data.title,
    content: data.content,
    linkUrl: data.linkUrl,
    mediaType: data.mediaType,
    channel: data.pannel,
    shareSuccessCB: data.shareSuccessCB || 'console.log',
    shareErrorCB: data.shareErrorCB || 'console.log'
  };

  if (data.type === 'audio') {
    d.type = '2';
  } else if (data.type === 'image') {
    d.type = '3';
  } else if (data.type === 'video') {
    d.type = '4';
  } else {
    d.type = '1';
  }

  return d;
};

(0, _register.default)('utils/openShareBox', {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  ios: {
    hasCB: false,
    versionRange: ['5.3.5'],
    filter: function filter(module, action, data) {
      return {
        module: 'callShare',
        action: '',
        paramName: 'options',
        extData: {
          minver: '5.3.5.0'
        }
      };
    }
  },
  android: {
    versionRange: ['5.3.5'],
    hasCB: false,
    filter: function filter(module, action, data) {
      return {
        module: 'Bdbox_android_utils',
        action: 'callShare',
        data: dataFormat(data)
      };
    }
  }
});
/**
 * 分享面板
 * @function openShare
 * @param {Object} params 端能力参数
 * @return {Promise} 调用的返回
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var defaultConf = {
    mediaType: 'all',
    type: 'url',
    shareSuccessCB: '',
    shareErrorCB: ''
  };
  /* eslint-disable max-len */

  if ((0, _detect.isMatrix)('baiduboxvision')) {
    return (0, _detect.isIOS)() ? (0, _ios.default)('baiduboxapp://callShare?minver=1.0.0.0&options=' + (0, _stringify.default)(Object.assign({}, defaultConf, params), true)) : (0, _android.prompt)('Bdbox_android_utils', 'callShare', [(0, _stringify.default)(Object.assign({}, defaultConf, dataFormat(params))), '', '']);
  }
  /* eslint-enable max-len */


  return (0, _index.open)((0, _utils.getParams)('utils-shareBox', params, {
    data: defaultConf
  }));
};

exports.default = _default;
