// @baidu/xbox-native/es5/ui/openProfile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 调起个人主页
 * @module ui
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('profile/open', {
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    hasCB: false,
    versionRange: ['7.0'],
    filter: function filter(module, action, data) {
      return {
        module: 'account',
        action: 'profile',
        extData: {
          minver: '7.0.0.0'
        },
        data: data
      };
    }
  },
  android: {
    hasCB: false,
    filter: function filter(module, action, data) {
      return {
        module: 'Bdbox_android_account',
        action: 'profile',
        data: data
      };
    }
  }
});
/**
 * @function openProfile
 * @param {object} [data={}] - openProfile 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('profile', params));
};

exports.default = _default;
