// @baidu/xbox-native/es5/ui/openUGCPublish.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module ui/openUGCPublish
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('ugc/openPublish', {
  v: 6,
  hasCB: true,
  versionRange: ['8.5'],
  filter: function filter(data) {
    if (data['topic_pageurl']) {
      var style = encodeURIComponent(JSON.stringify({
        'menumode': '2',
        'showtoolbar': '1'
      }));
      var url = encodeURIComponent(data['topic_pageurl']);
      data['topic_pageurl'] = 'baiduboxapp://v1/easybrowse/open?newbrowser=1&style=' + style + '&url=' + url;
    } // 兼容历史处理


    var urlToSchemaArr = (data.conf && data.conf.urlToSchema || []).filter(function (item) {
      return item !== 'topic_pageurl';
    });
    urlToSchemaArr.forEach(function (item) {
      data[item] = (0, _utils.urlToSchema)(data[item]);
    });
    return {
      module: 'ugc',
      action: 'publish',
      data: {
        params: encodeURIComponent(JSON.stringify(data))
      }
    };
  }
});
/**
 * @function openProfile
 * @param {object} [data={}] - openProfile 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('ugc-publish', params));
};

exports.default = _default;
