// @baidu/xbox-native/es5/ui/openNetworkErrorView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _android = require("../android");

var _ios = _interopRequireDefault(require("../ios"));

var _detect = require("../detect");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2018-12-5
 * @module ui
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('net/openNetworkErrorView', {
  lite: {
    versionRange: ['2.2.2'],
    supportFrame: ['tags']
  },
  androidinfo: {
    versionRange: ['1.0'],
    supportFrame: ['tags']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.3']
  },
  ios: {
    versionRange: ['8.2.3'],
    supportFrame: ['h5', 'hybrid', 'video', 'tags'],
    filter: function filter() {
      return {
        module: 'net',
        action: 'showNetWorkErrorView'
      };
    }
  },
  android: {
    versionRange: ['8.2'],
    hasCB: false,
    supportFrame: ['hybrid'],
    filter: function filter() {
      return {
        module: 'Bdbox_android_net',
        action: 'showNetWorkErrorView'
      };
    }
  }
});
/**
 * @function closeToast
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if ((0, _detect.isMatrix)('baiduboxvision')) {
    if ((0, _detect.isIOS)()) {
      return (0, _ios.default)('baiduboxapp://net?action=showNetWorkErrorView');
    }

    return (0, _android.prompt)('Bdbox_android_net', 'showNetWorkErrorView', []);
  }

  return (0, _index.open)((0, _utils.getParams)('net-networkErrorView', params));
};

exports.default = _default;
