// @baidu/xbox-native/es5/ui/openChat.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _stringify = _interopRequireDefault(require("../stringify"));

var _invoke = require("../invoke");

var _detect = require("../detect");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2018-12-5
 * @module ui
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('chat/open', {
  ios: {
    versionRange: ['8.5'],
    filter: function filter(module, action, data) {
      /* eslint-disable*/
      var str = 'source_app=baiduboxapp&referer=myduzhan&paid=';
      var d = {
        plugin_id: 'box.plugin.cambrian',
        args: str + data.paid + '&title=' + data.title + '&from=' + data.from + '&method_name=invokeChat'
      };
      /* eslint-enable*/

      return {
        module: 'invokePlug',
        action: 'open',
        data: JSON.stringify(d),
        extData: {
          minver: '8.5.0.0'
        }
      };
    },
    thunk: function thunk(res) {
      return {
        status: res.result === 0 ? 0 : parseInt(res.result, 10),
        message: ''
      };
    }
  },
  android: {
    versionRange: ['8.5'],
    filter: function filter(module, action, data) {
      /* eslint-disable*/
      var d = {
        package_name: 'com.baidu.cambrian',
        method_name: 'invokeChat',
        by_user: 1,
        params: {
          paid: data.paid,
          title: data.title,
          from: data.from
        }
      };
      /* eslint-enable*/

      return {
        module: 'Bdbox_android_plugin',
        action: 'invokePlugin',
        data: JSON.stringify(d)
      };
    },
    thunk: function thunk(res) {
      return {
        status: res === 0 ? 0 : parseInt(res, 10),
        message: ''
      };
    }
  }
}, {
  v: 9,
  versionRange: ['9.0.0'],
  filter: function filter(data) {
    return {
      module: 'account/profile',
      action: 'chat',
      data: {
        params: (0, _stringify.default)(data, true),
        oauthType: 'h5api'
      }
    };
  },
  thunk: function thunk(res) {
    return {
      status: res === 0 ? 0 : parseInt(res, 10),
      message: ''
    };
  }
});
/**
 * @function openChat
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if ((0, _detect.isMatrix)('baiduboxvision')) {
    return (0, _invoke.invokeP)('v9/account/profile/chat', {
      params: (0, _stringify.default)(params, true),
      oauthType: 'h5api'
    });
  }

  return (0, _index.open)((0, _utils.getParams)('chat', params));
};

exports.default = _default;
