// @baidu/xbox-native/es5/ui/openToast.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _stringify = _interopRequireDefault(require("../stringify"));

var _invoke = require("../invoke");

var _detect = require("../detect");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu Nov 15 2018
 * @module ui
 * @author wangyongqing <wangyongqing01@baidu.com>
 */
(0, _register.default)('toast/open', {
  androidlite: {
    versionRange: ['2.2.2']
  },
  androidinfo: {
    versionRange: ['1.0']
  },
  androidmission: {
    versionRange: ['1.0']
  },
  iosmission: {
    versionRange: ['1.0']
  },
  iosinfo: {
    versionRange: ['1.1']
  },
  ios: {
    versionRange: ['6.0'],
    hasCB: false,
    filter: function filter(module, action, data) {
      return {
        module: 'utils',
        action: 'toast',
        extData: {
          minver: '6.0.0.0'
        },
        data: data.message,
        paramName: 'string'
      };
    }
  },
  android: {
    versionRange: ['4.0'],
    hasCB: false,
    filter: function filter(module, action, data) {
      return {
        module: 'Bdbox_android_utils',
        action: 'toast',
        data: data.message,
        paramName: 'string'
      };
    }
  }
}, {
  v: 11,
  hasCB: true,
  versionRange: ['9.1'],
  androidliteVersionRange: ['2.2.0.10'],
  iosinfoVersionRange: ['1.3.1.10'],
  androidmissionVersionRange: ['1.0'],
  iosmissionVersionRange: ['1.0'],
  filter: function filter(data) {
    if (data.style) {
      var style = data.style;

      for (var key in style) {
        if (key !== 'style') {
          data[key] = style[key];
        }
      }

      if (style.style) {
        data.style = style.style;
      }
    }

    return {
      action: 'showToast',
      module: 'utils',
      data: {
        params: (0, _stringify.default)(data, true)
      }
    };
  }
});
/**
 * 展示toast
 * @function openToast
 * @param {Object} params 端能力参数，参数参看 http://ft.baidu.com/#/scheme/utils_showtoast
 * @return {Promise} 调用的返回
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if ((0, _detect.isMatrix)('baiduboxvision')) {
    return (0, _invoke.invokeP)('v11/utils/showToast', {
      params: params
    });
  }

  return (0, _index.open)((0, _utils.getParams)('toast', params, {
    data: {
      type: 1
    }
  }));
};

exports.default = _default;
