// @baidu/xbox-native/es5/ui/openCommentBox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

var _getBoxVersion = _interopRequireDefault(require("@baidu/xbox/es5/get-box-version"));

var _version_compare = _interopRequireDefault(require("@baidu/xbox/es5/version_compare"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2018-12-5
 * @module ui
 * @author panming <panming@baidu.com>
 */
var boxVersion = (0, _getBoxVersion.default)();
(0, _register.default)('comment/openCommentBox', {
  lite: {
    versionRange: ['2.2.2'],
    filter: function filter(module, action, data) {
      var params = data;

      if ((0, _version_compare.default)(boxVersion, '2.3') < 0) {
        params.data && delete params.data.placeholder;
      }

      return {
        module: 'Bdbox_android_comment',
        action: 'showCommentBox',
        data: params
      };
    }
  },
  androidinfo: {
    versionRange: ['1.0'],
    filter: function filter(module, action, data) {
      var params = data;

      if ((0, _version_compare.default)(boxVersion, '1.1') < 0) {
        params.data && delete params.data.placeholder;
      }

      return {
        module: 'Bdbox_android_comment',
        action: 'showCommentBox',
        data: params
      };
    }
  },
  ios: {
    versionRange: ['8.0'],
    supportFrame: ['tags', 'hybrid', 'video', 'h5'],
    filter: function filter(module, action, data) {
      var params = data;

      if ((0, _version_compare.default)(boxVersion, '8.2') < 0) {
        params.data && delete params.data.rename;
      }

      if ((0, _version_compare.default)(boxVersion, '8.5') < 0) {
        params.data && delete params.data.placeholder;
      }

      return {
        module: 'comment',
        action: 'showCommentBox',
        extData: {
          'minver': '8.0.0.0'
        },
        data: params
      };
    }
  },
  android: {
    supportFrame: ['tags', 'hybrid', 'video', 'h5'],
    versionRange: ['8.0'],
    filter: function filter(module, action, data) {
      var params = data;

      if ((0, _version_compare.default)(boxVersion, '8.2') < 0) {
        params.data && delete params.data.rename;
      }

      if ((0, _version_compare.default)(boxVersion, '8.5') < 0) {
        params.data && delete params.data.placeholder;
      }

      return {
        module: 'Bdbox_android_comment',
        action: 'showCommentBox',
        data: params
      };
    }
  }
}, {
  v: 4,
  hasCB: true,
  versionRange: ['8.3'],
  filter: function filter() {
    return {
      module: 'comment',
      action: 'showBox'
    };
  }
});
/**
 * @function openCommentBox
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  (0, _index.open)((0, _utils.getParams)('comment-commentBox', params));
};

exports.default = _default;
