// @baidu/xbox-native/es5/ui/setActionBarRightMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2019-4-10 轻浏览框架设置顶bar,右侧功能区菜单点击
 * @module ui/setActionBarRightMenu
 * @author sunxiaoyu01 <sunxiaoyu01@baidu.com>
 */
(0, _register.default)('easybrowse/openSetActionBarRightMenu', {
  v: 45,
  androidVersionRange: ['11.7.1'],
  filter: function filter() {
    return {
      module: 'easybrowse',
      action: 'setActionBarRightMenu'
    };
  }
});
/**
 * @function openSetActionBarRightMenu
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('easybrowse-setActionBarRightMenu', params));
};

exports.default = _default;
