// @baidu/xbox-native/es5/ui/setFloatingBallImg.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on Thu July 31 2019
 * @module ui
 * @author luoxiao09 <luoxiao09@baidu.com>
 */
(0, _register.default)('easybrowse/openImageUrlAddress', {
  v: 50,
  hasCB: true,
  versionRange: ['11.13'],
  supportFrame: ['h5', 'hybird'],
  filter: function filter() {
    return {
      module: 'easybrowse',
      action: 'imageUrlAddress'
    };
  }
});
/**
 * @function openImageUrlAddress
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('easybrowse-imageUrlAddress', params));
};

exports.default = _default;
