// @baidu/xbox-native/es5/ui/showPraiseMotion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _register = _interopRequireDefault(require("../register"));

var _utils = require("../utils");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Created on 2018-12-10 触发点赞动画
 * @module ui/showPraiseMotion
 * @author panming <panming@baidu.com>
 */
(0, _register.default)('praise/openMotionEvent', {
  v: 25,
  androidVersionRange: ['10.6.5'],
  iosVersionRange: ['10.10'],
  androidmissionVersionRange: ['1.0'],
  androidliteVersionRange: ['3.3'],
  filter: function filter() {
    return {
      module: 'praise',
      action: 'motionEvent'
    };
  }
});
/**
 * @function openPraiseMotionEvent
 * @param {Object} [params={}] - 参数
 */

var _default = function _default() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _index.open)((0, _utils.getParams)('praise-motionEvent', params));
};

exports.default = _default;
