// @baidu/wuji-san/lib/wuji-topic-header/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _san = require("san");

var _headerTitle = _interopRequireDefault(require("./components/header-title"));

var _headerDescription = _interopRequireDefault(require("./components/header-description"));

var _wujiImage = _interopRequireDefault(require("../wuji-image"));

var _wujiSubsidiaryAvatar = _interopRequireDefault(require("../wuji-subsidiary-avatar"));

var _hex2rgb = require("./hex2rgb");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var prefixCls = 'w-topic-header';
var defaultHeadSrc = 'https://b.bdstatic.com/searchbox/icms/searchbox/img/default-head-1626935858.jpg';

var WujiTopicHeader = /*#__PURE__*/function (_Component) {
  _inherits(WujiTopicHeader, _Component);

  var _super = _createSuper(WujiTopicHeader);

  function WujiTopicHeader() {
    _classCallCheck(this, WujiTopicHeader);

    return _super.apply(this, arguments);
  }

  _createClass(WujiTopicHeader, [{
    key: "initData",
    value: function initData() {
      return {
        title: '',
        imgUrl: '',
        description: '',
        authorNameList: [],
        authorImgList: [],
        headImgLogo: '',
        headImgColor: ''
      };
    }
  }]);

  return WujiTopicHeader;
}(_san.Component);

exports["default"] = WujiTopicHeader;

_defineProperty(WujiTopicHeader, "template",
/* html */
"\n        <div class=\"".concat(prefixCls, "\">\n            <template s-if=\"{{!(!isThirdEdition && imageType === 2))}}\">\n                <div class=\"header-container\">\n                    <div class=\"image-wrapper\">\n                        <wuji-image\n                            src=\"{{imageUrl}}\"\n                            ratio=\"{{imgRatio}}\"\n                            banner=\"{{true}}\"\n                        ></wuji-image>\n                        <div \n                            class=\"{{maskCls}}\"\n                            style=\"{{maskStyle}}\">\n                        </div>\n                    </div>\n                    <div class=\"main-area\">\n                        <div\n                            s-if=\"isThirdEdition && headImgLogo\"\n                            class=\"topic-label\"\n                            style=\"{{headImgLogo ? 'background-image:url(' + headImgLogo + ')': ''}}\"\n                        ></div>\n                        <div class=\"title-area\">\n                            <header-title\n                                title=\"{{title}}\"\n                                type=\"{{imageType}}\"\n                                is-third-edition=\"{{isThirdEdition}}\"\n                                class=\"header-title {{isShowHeader ? 'header-title-nomb' : ''}}\">\n                            </header-title>\n                            <header-author-info\n                                s-if=\"isShowHeader\"\n                                class=\"{{isThirdEdition && (imgType === 1 || imgType === 0)\n                                    ? 'new-subsidiary-top' : 'subsidiary-top'}}\"\n                                desc-color=\"FC6\"\n                                author-desc=\"{{authorDesc}}\"\n                                author-img-list=\"{{authorImgList}}\"\n                                border-width=\"{{1}}\"\n                            ></header-author-info>\n                        </div>\n                    </div>\n                </div>\n                <header-desc class=\"header-round\" desc=\"{{description}}\"></header-desc>\n            </template>\n            <template s-else-if=\"{{!isThirdEdition && imageType === 2}}\">\n                <div class=\"image-wrapper\">\n                    <wuji-image\n                        src=\"{{imgUrl}}\"\n                        ratio=\"{{isSecondEdition ? '3:1' : '5:1'}}\"\n                        banner=\"{{true}}\"\n                    ></wuji-image>\n                    <div class=\"image-mask\"></div>\n                </div>\n                <div class=\"content header-round\">\n                    <div class=\"title-area\">\n                        <header-title\n                            title=\"{{title}}\"\n                            type=\"{{imageType}}\"\n                        ></header-title>\n                    </div>\n                    <header-desc\n                        desc=\"{{description}}\"\n                        class=\"description-wrapper\"\n                        s-if=\"description\">\n                    </header-desc>\n                    <header-author-info\n                        class=\"auth-area\"\n                        s-if=\"isShowHeader\"\n                        desc-color=\"FC2\"\n                        author-desc=\"{{authorDesc}}\"\n                        author-img-list=\"{{authorImgList}}\"\n                        border-width=\"{{1}}\"\n                    ></header-author-info>\n                </div>\n            </template>\n        </div>\n    "));

_defineProperty(WujiTopicHeader, "trimWhitespace", 'all');

_defineProperty(WujiTopicHeader, "components", {
  'header-title': _headerTitle["default"],
  'header-author-info': _wujiSubsidiaryAvatar["default"],
  'header-desc': _headerDescription["default"],
  'wuji-image': _wujiImage["default"]
});

_defineProperty(WujiTopicHeader, "computed", {
  isShowHeader: function isShowHeader() {
    var authorNameList = this.data.get('authorNameList') || [];
    return +this.data.get('hotType') === 1 && authorNameList.length > 0;
  },
  authorDesc: function authorDesc() {
    var isShowHeader = this.data.get('isShowHeader');
    var authorNameList = this.data.get('authorNameList') || [];
    var authorNum = this.data.get('authorNum');
    var str = '';

    if (!isShowHeader) {
      return str;
    }

    if (authorNameList[0]) {
      str += authorNameList[0];
    }

    if (authorNameList[1]) {
      str += '、' + authorNameList[1];
    }

    return "".concat(str, "\u7B49").concat(authorNum, "\u4E2A\u767E\u5BB6\u53F7\u53C2\u4E0E");
  },
  imageType: function imageType() {
    var imgUrl = this.data.get('imgUrl');
    var imgType = this.data.get('imgType');
    var isThirdEdition = this.data.get('isThirdEdition');
    return !imgUrl ? isThirdEdition ? 2 : 1 : +imgType;
  },
  imageUrl: function imageUrl() {
    var imgUrl = this.data.get('imgUrl');
    return !imgUrl ? defaultHeadSrc : imgUrl;
  },
  maskCls: function maskCls() {
    var imgUrl = this.data.get('imgUrl');
    return !imgUrl ? '' : 'image-mask';
  },
  // 头部区域一、二版：区别1头图尺寸改版
  isSecondEdition: function isSecondEdition() {
    var modifyTime = this.data.get('modifyTime');
    return modifyTime > 1599567000 && modifyTime < 1626852919;
  },
  // 头部区域二、三版区分：区别1：头图尺寸改版，区别2：统一将标题、辅助信息放到图上
  isThirdEdition: function isThirdEdition() {
    var modifyTime = this.data.get('modifyTime');
    return modifyTime > 1626852919; // 1625443200 - 2021-07-21 15:35:19 人机具体上线时间
  },
  // 一二三版大图&三版大图 ratio
  imgRatio: function imgRatio() {
    var isThirdEdition = this.data.get('isThirdEdition');
    var isSecondEdition = this.data.get('isSecondEdition');
    var imageType = +this.data.get('imageType'); // 0 无图 1 大图 2 小图；无图按大图处理

    var ratio = '4:3'; // 默认值：第三版大图尺寸

    if (isThirdEdition) {
      if (imageType === 2) {
        ratio = '16:9'; // 第三版小图尺寸
      }
    } else {
      // 第一、二版小图样式独立
      if (isSecondEdition) {
        ratio = '16:9'; // 第二版大图
      } else {
        ratio = '5:2'; // 第一版大图
      }
    }

    return ratio;
  },
  maskStyle: function maskStyle() {
    var imgBg = this.data.get('headImgColor');
    var imgBgObj = (0, _hex2rgb.hex2rgb)(imgBg);
    return imgBg ? "background:linear-gradient(to bottom, ".concat(imgBgObj.rgb0, ", ").concat(imgBgObj.rgb, ");") : '';
  }
});

_defineProperty(WujiTopicHeader, "dataTypes", {
  title: _san.DataTypes.string,
  imgType: _san.DataTypes.oneOf([0, 1, 2]),
  imgUrl: _san.DataTypes.string,
  description: _san.DataTypes.string,
  hotType: _san.DataTypes.oneOf([0, 1]),
  authorNameList: _san.DataTypes.array,
  authorImgList: _san.DataTypes.array,
  authorNum: _san.DataTypes.number,
  modifyTime: _san.DataTypes.number,
  headImgLogo: _san.DataTypes.string,
  headImgColor: _san.DataTypes.string
});
