// boxjs/dist/thirdParty/bdbox/distribute.js
module.exports = function () {
  var $ = require('boxjs/dist/utils/_bdboxWrapper');
  /* eslint-disable */


  var schema = require('./schema');

  require('./utils/detect');

  var each = require('./each');

  var Moplus = require('./moplus');

  var emptyFn = $.emptyFn;
  var defaultOptions = {
    qqDownloadUrl: 'http://a.app.qq.com/o/simple.jsp?pkgname=com.baidu.searchbox',
    androidDownloadUrl: '',
    iosDownloadUrl: '',
    // wechatCallback: emptyFn,
    iosFailCallback: emptyFn,
    androidFailCallback: emptyFn,
    iosSchema: '',
    androidSchema: '',
    androidIntent: '',
    inBoxCallback: emptyFn
  };

  var Distribute = function (opts) {
    var that = this;
    opts && each(defaultOptions, function (value, key) {
      opts[key] = opts[key] || value;
    });
    var url = this.url = opts.url;

    if (url && url !== '') {
      this.androidIntent = 'intent://' + url.replace(/^http[s]?:\/\//, '') + '#Intent;scheme=http;action=com.baidu.searchbox.action.VIEW;category=android.intent.category.DEFAULT;end'; // this.androidSchema = '';

      this.iosSchema = 'baiduboxapp://easybrowse?openurl=' + encodeURIComponent(url) + '&opentype=0&isla=0&append=0&minver=5.3.0.0';
    } else {
      ['androidIntent', 'iosSchema', 'androidSchema'].forEach(function (v) {
        that[v] = opts[v];
      });
    }

    this.options = opts; //框外，并且非微信

    this.fail = function () {
      console.log($.os.name + ' fail');

      if ($.isFunction(opts[$.os.name + 'FailCallback'])) {
        var back = opts[$.os.name + 'FailCallback'](); //如果返回是false,则不往下执行，即不跳转下载

        if ($.isBoolean(back) && !back) {
          return;
        }
      }

      var url = opts[$.os.name + 'DownloadUrl'];

      if (url && url !== '') {
        location.href = opts[$.os.name + 'DownloadUrl'];
      }
    };

    this.success = function () {
      console.log($.os.name + ' success');

      if ($.isFunction(opts[$.os.name + 'SuccessCallback'])) {
        return opts[$.os.name + 'SuccessCallback']();
      }
    };
  };

  Distribute.prototype.wechat = function () {
    var opts = this.options;

    if ($.isFunction(opts.wechatCallback)) {
      var back = opts.wechatCallback(); //如果返回是false,则不往下执行，即不跳转下载

      if ($.isBoolean(back) && !back) {
        return;
      }
    }

    if (opts.qqDownloadUrl && opts.qqDownloadUrl !== '') {
      location.href = opts.qqDownloadUrl;
    }
  };

  Distribute.prototype.run = function () {
    var that = this;
    var url = that.url;
    var opts = that.options;

    if ($.os.isWechat) {
      return that.wechat();
    }

    if ($.$isBox() && $.isFunction(opts.inBoxCallback)) {
      //框内
      return opts.inBoxCallback();
    }

    this.invoke();
  };

  Distribute.prototype.invoke = function () {
    var that = this;

    if ($.os.android) {
      var moplus = Moplus();
      moplus.getHVersion(function (info) {
        if (info) {
          moplus.sendIntent(that.androidIntent, function (j) {
            if (j.error == 0) {
              that.success();
            } else {
              that.fail();
            }
          }, 1e3);
        } else {
          if (that.androidSchema) {
            schema(that.androidSchema, function (isTimeout) {
              if (isTimeout) {
                //跳转到下载地址或者提示
                that.fail();
              } else {
                that.success();
              }
            });
          } else {
            that.fail();
          }
        }
      });
    } else {
      schema(that.iosSchema, function (isFail) {
        if (isFail) {
          //调起失败的逻辑，比如打开官网下载
          that.fail();
        } else {
          that.success();
        }
      });
    }
  };
  /**
   * 分发逻辑
   * @param  {Object} opts 配置对象
   * @memberOf Bdbox
   * @function
   * @name distribute
   * @version $Id: distribute.js 286319 2016-03-14 08:25:23Z wangyongqing01 $
   * @example
   * var dis = Bdbox.distribute({
   *     url: '活动url',
   *     wechatCallback: '微信回调',
   *     androidFailCallback: '安卓失败回调',
   *     iosFailCallback: 'ios失败回调，如果返回为false,则不往下执行，否则跳转下载地址',
   *     iosSuccessCallback: 'ios成功回调',
   *     androidDownloadUrl: '下载地址'
   * });
   * //执行默认流程：微信→invoke→失败
   * dis.run();
   * //执行调起
   * dis.invoke();
   * //使用intent
   * var a = Box.distribute({
   *      androidIntent: '#Intent;action=com.baidu.searchbox.action.HOME;package=com.baidu.searchbox;category=android.intent.category.DEFAULT;end',
   *       iosSchema: 'baiduboxapp://',
   *       androidDownloadUrl: 'http://mo.baidu.com/baidusearch/',
   *       iosDownloadUrl: 'http://mo.baidu.com/baidusearch/',
   *       wechatCallback: function(){
   *           alert('微信');
   *       }
   *   });
   *   a.run();
   */


  return function (opts) {
    return new Distribute(opts);
  };
}();
