// @baidu/xbox-native/media/tts.js
/**
 * @file Created on Thu Nov 15 2018
 * @module media/tts
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {tts} from './index';
register('tts/prefetch', {
    v: 18,
    versionRange: ['10.2']
});
/**
 * @function prefetch
 * @param {object} [data={}] - prefetch 参数
 */
export const prefetch = (data = {}) => tts(getParams('prefetch', data));

register('tts/pause', {
    v: 18,
    versionRange: ['10.2']
});
/**
 * @function pause
 * @param {object} [data={}] - pause 参数
 */
export const pause = (data = {}) => tts(getParams('pause', data));

register('tts/append', {
    v: 18,
    versionRange: ['10.2']
});
/**
 * @function append
 * @param {object} [data={}] - append 参数
 */
export const append = (data = {}) => tts(getParams('append', data));

register('tts/status', {
    v: 18,
    versionRange: ['10.2']
});
/**
 * @function status
 * @param {object} [data={}] - status 参数
 */
export const status = (data = {}) => tts(getParams('status', data));

register('tts/stop', {
    v: 18,
    versionRange: ['10.2'],
    filter() {
        return {
            module: 'tts',
            action: 'end'
        };
    }
});
/**
 * @function stop
 * @param {object} [data={}] - stop 参数
 */
export const stop = (data = {}) => tts(getParams('stop', data));

register('tts/play', {
    v: 18,
    versionRange: ['10.2']
});

/**
 * @function play
 * @param {object} [data={}] - play 参数
 */
export const play = (data = {}) => tts(getParams('play', data));
