// @baidu/xbox-native/media/video.js
/**
 * @file Created on Thu Nov 15 2018
 * @module media/video
 * @author wangyongqing <wangyongqing01@baidu.com>
 */

import register from '../register';
import {getParams} from '../utils';
import {video} from './index';
register('video/play', {
    androidlite: {
        versionRange: ['2.3']
    },
    androidinfo: {
        versionRange: ['1.1']
    },
    androidmission: {
        versionRange: ['1.0']
    },
    iosmission: {
        versionRange: ['1.0']
    },
    ios: {
        hasCB: false,
        versionRange: ['7.3'],
        filter(module, action, data) {
            return {
                module: 'video',
                action: 'invokePlayer',
                extData: {
                    minver: '7.3.0.0'
                }
            };
        }
    },
    android: {
        versionRange: ['7.3'],
        supportFrame: ['video'],
        hasCB: false,
        filter(module, action, data) {
            return {
                module: 'Bdbox_android_capi_video',
                action: 'invokePlayer'
            };
        }
    }
});
/**
 * @function play
 * @param {object} [data={}] - play 参数
 */
export const play = (data = {}) => video(getParams('play', data));
