// @baidu/mo-topic-baike/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Base = require('@baidu/mo-base');
var wujiSan = require('@baidu/wuji-san');
var styles = require('./style.module.less');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var Base__default = /*#__PURE__*/_interopDefaultLegacy(Base);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var ATTACHED_SHOW = 'show:attached';
var BAIKE_CLICK = 'baike:cilck';
var BAIKE_IMAGE_CLICK = 'baike_image:click';

/**
 * Copyright (C) 2021 ~ present baidu.com
 *
 * @author gaohong01 <gaohong01@baidu.com>
 */
// @san/component
var TopicBaike = /** @class */ (function (_super) {
    __extends(TopicBaike, _super);
    function TopicBaike() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TopicBaike.prototype.initData = function () {
        return {
            styles: styles__default["default"],
            visited: false,
            showAction: ATTACHED_SHOW,
        };
    };
    TopicBaike.prototype.attached = function () {
        _super.prototype.attached.call(this);
        this.eventDispatch(ATTACHED_SHOW);
    };
    TopicBaike.prototype.infoClick = function (e) {
        e.stopPropagation();
        this.data.set('visited', true);
        this.eventDispatch(BAIKE_CLICK);
    };
    TopicBaike.prototype.imageClick = function (e) {
        e.stopPropagation();
        this.eventDispatch(BAIKE_IMAGE_CLICK);
    };
    TopicBaike.template = "\n        <div\n            class=\"{{wrapperClass}}\"\n        >\n            <wuji-row\n                gutter=\"12\"\n                margin=\"0\"\n            >\n                <wuji-col\n                    span=\"{{imageProps.src ? 8 : 12}}\"\n                    class=\"{{styles.info}}\"\n                >\n                    <div\n                        on-click=\"infoClick\"\n                    >\n                        <div class=\"{{styles.title}}\">\n                            <div class=\"{{styles.text}}\">{{title}}</div>\n                        </div>\n                        <div class=\"{{styles.desc}}\">\n                            {{description}}\n                        </div>\n                    </div>\n                </wuji-col>\n                <wuji-col\n                    s-if=\"imageProps.src\"\n                    span=\"4\"\n                >   \n                    <div\n                        class=\"{{styles.image}}\"\n                        on-click=\"imageClick\"\n                    >\n                        <wuji-image\n                            s-bind=\"{{imageProps}}\"\n                        />\n                        <div\n                            s-if=\"logo\"\n                            class=\"{{styles.mask}}\"\n                        ></div>\n                        <img\n                            s-if=\"logo\"\n                            src=\"{{logo}}\"\n                            class=\"{{styles.baikeLogo}}\"\n                        />\n                    </div>\n                </wuji-col>\n            </wuji-row>\n        </div>\n    ";
    TopicBaike.components = {
        'wuji-image': wujiSan.WujiImage,
        'wuji-row': wujiSan.WujiGrid.Row,
        'wuji-col': wujiSan.WujiGrid.Col,
    };
    TopicBaike.computed = {
        wrapperClass: function () {
            return [
                styles__default["default"].baikeWrapper,
                !this.data.get('imageProps.src') ? styles__default["default"].onlyTitle : '',
            ];
        },
    };
    return TopicBaike;
}(Base__default["default"]));

exports.ATTACHED_SHOW = ATTACHED_SHOW;
exports.BAIKE_CLICK = BAIKE_CLICK;
exports.BAIKE_IMAGE_CLICK = BAIKE_IMAGE_CLICK;
exports["default"] = TopicBaike;
