// @baidu/wuji-san/es/wuji-topic-header/index.js
/**
 * @file 组件 WujiTopicHeader San Component for Wuji
 * @author xiewenbo01 <xiewenbo01@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';
import HeaderTitle from './components/header-title';
import HeaderDesc from './components/header-description';
import WujiImage from '../wuji-image';
import HeaderAuthorInfo from '../wuji-subsidiary-avatar';
import {hex2rgb} from './hex2rgb';

const prefixCls = 'w-topic-header';
const defaultHeadSrc = 'https://b.bdstatic.com/searchbox/icms/searchbox/img/default-head-1626935858.jpg';

export default class WujiTopicHeader extends Component {
    static template = /* html */ `
        <div class="${prefixCls}">
            <template s-if="{{!(!isThirdEdition && imageType === 2))}}">
                <div class="header-container">
                    <div class="image-wrapper">
                        <wuji-image
                            src="{{imageUrl}}"
                            ratio="{{imgRatio}}"
                            banner="{{true}}"
                        ></wuji-image>
                        <div 
                            class="{{maskCls}}"
                            style="{{maskStyle}}">
                        </div>
                    </div>
                    <div class="main-area">
                        <div
                            s-if="isThirdEdition && headImgLogo"
                            class="topic-label"
                            style="{{headImgLogo ? 'background-image:url(' + headImgLogo + ')': ''}}"
                        ></div>
                        <div class="title-area">
                            <header-title
                                title="{{title}}"
                                type="{{imageType}}"
                                is-third-edition="{{isThirdEdition}}"
                                class="header-title {{isShowHeader ? 'header-title-nomb' : ''}}">
                            </header-title>
                            <header-author-info
                                s-if="isShowHeader"
                                class="{{isThirdEdition && (imgType === 1 || imgType === 0)
                                    ? 'new-subsidiary-top' : 'subsidiary-top'}}"
                                desc-color="FC6"
                                author-desc="{{authorDesc}}"
                                author-img-list="{{authorImgList}}"
                                border-width="{{1}}"
                            ></header-author-info>
                        </div>
                    </div>
                </div>
                <header-desc class="header-round" desc="{{description}}"></header-desc>
            </template>
            <template s-else-if="{{!isThirdEdition && imageType === 2}}">
                <div class="image-wrapper">
                    <wuji-image
                        src="{{imgUrl}}"
                        ratio="{{isSecondEdition ? '3:1' : '5:1'}}"
                        banner="{{true}}"
                    ></wuji-image>
                    <div class="image-mask"></div>
                </div>
                <div class="content header-round">
                    <div class="title-area">
                        <header-title
                            title="{{title}}"
                            type="{{imageType}}"
                        ></header-title>
                    </div>
                    <header-desc
                        desc="{{description}}"
                        class="description-wrapper"
                        s-if="description">
                    </header-desc>
                    <header-author-info
                        class="auth-area"
                        s-if="isShowHeader"
                        desc-color="FC2"
                        author-desc="{{authorDesc}}"
                        author-img-list="{{authorImgList}}"
                        border-width="{{1}}"
                    ></header-author-info>
                </div>
            </template>
        </div>
    `;

    static trimWhitespace = 'all';

    static components = {
        'header-title': HeaderTitle,
        'header-author-info': HeaderAuthorInfo,
        'header-desc': HeaderDesc,
        'wuji-image': WujiImage,
    };

    static computed = {
        isShowHeader() {
            const authorNameList = this.data.get('authorNameList') || [];
            return +this.data.get('hotType') === 1 && authorNameList.length > 0;
        },
        authorDesc() {
            const isShowHeader = this.data.get('isShowHeader');
            const authorNameList = this.data.get('authorNameList') || [];
            const authorNum = this.data.get('authorNum');
            let str = '';

            if (!isShowHeader) {
                return str;
            }

            if (authorNameList[0]) {
                str += authorNameList[0];
            }

            if (authorNameList[1]) {
                str += '、' + authorNameList[1];
            }

            return `${str}等${authorNum}个百家号参与`;
        },
        imageType() {
            const imgUrl = this.data.get('imgUrl');
            const imgType = this.data.get('imgType');
            const isThirdEdition = this.data.get('isThirdEdition');
            return !imgUrl ? (isThirdEdition ? 2 : 1) : +imgType;
        },
        imageUrl() {
            const imgUrl = this.data.get('imgUrl');
            return !imgUrl ? defaultHeadSrc : imgUrl;
        },
        maskCls() {
            const imgUrl = this.data.get('imgUrl');
            return !imgUrl ? '' : 'image-mask';
        },
        // 头部区域一、二版：区别1头图尺寸改版
        isSecondEdition() {
            const modifyTime = this.data.get('modifyTime');
            return modifyTime > 1599567000 && modifyTime < 1626852919;
        },
        // 头部区域二、三版区分：区别1：头图尺寸改版，区别2：统一将标题、辅助信息放到图上
        isThirdEdition() {
            const modifyTime = this.data.get('modifyTime');
            return modifyTime > 1626852919; // 1625443200 - 2021-07-21 15:35:19 人机具体上线时间
        },
        // 一二三版大图&三版大图 ratio
        imgRatio() {
            const isThirdEdition = this.data.get('isThirdEdition');
            const isSecondEdition = this.data.get('isSecondEdition');
            const imageType = +this.data.get('imageType');// 0 无图 1 大图 2 小图；无图按大图处理
            let ratio = '4:3';// 默认值：第三版大图尺寸
            if (isThirdEdition) {
                if (imageType === 2) {
                    ratio = '16:9';// 第三版小图尺寸
                }
            } else {
                // 第一、二版小图样式独立
                if (isSecondEdition) {
                    ratio = '16:9';// 第二版大图
                } else {
                    ratio = '5:2';// 第一版大图
                }
            }
            return ratio;
        },
        maskStyle() {
            const imgBg = this.data.get('headImgColor');
            let imgBgObj = hex2rgb(imgBg);
            return imgBg ? `background:linear-gradient(to bottom, ${imgBgObj.rgb0}, ${imgBgObj.rgb});` : '';
        },
    };

    static dataTypes = {
        title: DataTypes.string,
        imgType: DataTypes.oneOf([
            0,
            1,
            2,
        ]),
        imgUrl: DataTypes.string,
        description: DataTypes.string,
        hotType: DataTypes.oneOf([
            0,
            1,
        ]),
        authorNameList: DataTypes.array,
        authorImgList: DataTypes.array,
        authorNum: DataTypes.number,
        modifyTime: DataTypes.number,
        headImgLogo: DataTypes.string,
        headImgColor: DataTypes.string,
    };

    initData() {
        return {
            title: '',
            imgUrl: '',
            description: '',
            authorNameList: [],
            authorImgList: [],
            headImgLogo: '',
            headImgColor: '',
        };
    }
}
