// @baidu/wuji-san/es/wuji-table-covid/index.js
/**
 * @file 组件 WujiTableCovid 无极-疫情数据
 * @author jiangxinge <jiangxinge@baidu.com>
 */

import './style/index.less';
import {Component, DataTypes} from 'san';

const prefixCls = 'w-table-covid';

export default class WujiTableCovid extends Component {
    static template = /* html */ `
        <div class="${prefixCls} table-type-{{type}}" on-click="handleClick">
            <div s-if="table.showTitle" class="${prefixCls}-titles">
                <div s-for="title in table.titles" class="${prefixCls}-title">{{title}}</div>
            </div>
            <div class="${prefixCls}-lists">
                <div s-for="list in table.lists" class="${prefixCls}-list">
                    <div s-for="item, index in list" class="${prefixCls}-item">
                        <span s-if="showColName(item, type, index)" class="name">{{item.colName}}</span>
                        <span class="text text-{{getTextClass(item.wujiText)}}">{{item.colText}}</span>
                    </div>
                </div>
            </div>
        </div>
    `;

    static dataTypes = {
        data: DataTypes.array,
        type: DataTypes.oneOf([
            '1',
            '2',
            '3',
        ]),
    };

    static computed = {
        table() {
            const data = this.data.get('data');
            const table = {
                titles: [],
                lists: [],
                showTitle: true,
            };
            data.forEach(side => {
                table.titles.push(side.rowName);
                side.colData.forEach((item, index) => {
                    if (table.lists[index]) {
                        table.lists[index].push(item);
                    }
                    else {
                        table.lists.push([
                            item,
                        ]);
                    }
                });
            });
            if (!table.titles.filter(c => c).length) {
                table.showTitle = false;
            }
            return table;
        },
    };

    initData() {
        return {
            data: [],
            type: '1',
        };
    }

    getTextClass(code) {
        if (!code) {
            return 'x123';
        }

        return code.replace('Txt_', '').toLowerCase();
    }

    showColName(item, type, index) {
        if (!item.colName || type === '3') {
            return false;
        }
        if (type === '2' && index > 0) {
            return false;
        }

        return true;
    }

    handleClick(event) {
        this.fire('click', {event, data: this.data.get()});
    }
}
