// @baidu/auc-one-image-text/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import AucComponent from '@baidu/auc-base/AucComponent';
import styles from './style.module.less';
import {AucOneImageTextData} from './types';
import {WujiTextImage} from '@baidu/wuji-san';
import {componentLog} from '@baidu/auc-log';
import {topicVideoThunderLog} from '@baidu/auc-base/topicVideoThunderLog';

export class AucOneImageText extends AucComponent {
    static template:string = /* html */ `
    <div
        class="{{styles.aucOneImageTextWrapper}} {{visited ? styles.visited : ''}} {{isBigEvent ? styles.bigEvent : ''}}"
        on-click="handleClick"
    >
        <div class="{{styles.oneImageTextInner}}">
            <wuji-text-image
                title="{{title}}"
                justify="space-between"
                spacing="{{true}}"
                hairline="{{true}}"
                margin="{{margin}}"
                label-props="{{labelProp}}"
                widgets="{{widgetItems}}"
                image="{{image}}"
                image-props="{{imageProps}}"
                title-props="{{titleProps}}"
            >
            </wuji-text-image>
        </div>
    </div>
    `;
    static components:object = {
        'wuji-text-image': WujiTextImage
    };
    scrollView = null;

    initData(): AucOneImageTextData {
        return {
            styles,
            title: '',
            labelProp: {},
            widgetItems: [],
            imageProps: {},
            image: '',
            titleProps: {
                visited: false
            },
            logBaseType: 'image_text_',
            logExtraParams: {
            },
            jumpProps: {},
            // todo 换成编码
            margin: 51
        };
    }

    attached(): void {
        this.registerScrollInView().then(() => {
            componentLog(this.data.get('logBaseType'), 'show', this.data.get('logExtraParams'));
            this.data.get('type') === 'video' && this.videoThunderLog('show');
        });
    }

    handleClick(): void {
        this.fire('jump', this.data.get('jumpProps'));
        this.data.set('titleProps', {
            visited: true
        });
        this.data.get('type') === 'video' && this.videoThunderLog('click');
        componentLog(this.data.get('logBaseType'), 'click', this.data.get('logExtraParams'));
    }

    videoThunderLog(type) {
        // 视频切换短小融合需单独打点
        const videoThunderLog = this.data.get('videoThunderLog');
        if (videoThunderLog.resourceType) {
            topicVideoThunderLog(type === 'show' ? 'show' : 'click', {
                rid: this.data.get('rid'),
                videoThunderLog,
                topic_id: this.data.get('topicId')
            });
        }
    }
}

export default AucOneImageText;
