// @baidu/auc-kol/index.ts
/**
 * @file index.ts
 * @author leizhihao
 */

import styles from './style.module.less';
import {AucKolData} from './types';
import {WujiImage, WujiWidget} from '@baidu/wuji-san';
import AucComponent from "@baidu/auc-base/AucComponent";

export class AucKol extends AucComponent {
    static template:string = /* html */ `
    <div class="{{styles.aucKolWrapper}}">
        <div class="{{styles.contentWrapper}}" on-click="handleClick">
            <div class="{{styles.imageWrapper}}" s-if="imageProps.src">
                <div class="{{styles.liveIconWrapper}} {{styles.playBack}}" s-if="isPlayBack"><wuji-icon class="{{styles.liveIcon}}" type="live-newreview-f-30" size="30"></wuji-icon>直播回放</div>
                <div class="{{styles.liveIconWrapper}} {{styles.preload}}" s-if="isPreload"><wuji-icon class="{{styles.liveIcon}}" type="live-newtrailer-f-30" size="30"></wuji-icon>直播预告</div>
                <div class="{{styles.liveIconWrapper}} {{styles.living}}" s-if="isLive"><div class="{{styles.dotWrapper}}"><i class="{{styles.dot}} {{styles.dot1}}"></i><i class="{{styles.dot}} {{styles.dot2}}"></i><i class="{{styles.dot}} {{styles.dot3}}"></i></div>直播中</div>
                <wuji-image
                    s-bind="{{imageProps}}"
                ></wuji-image>
            </div>
            <div class="{{styles.textWrapper}}">
                <div class="{{styles.title}}">{{title}}</div>
                <div class="{{styles.kolName}}" s-if="kolName">—{{kolName}}</div>
            </div>
        </div>
        <wuji-widget s-bind="widgetProps" class="{{styles.source}}"></wuji-widget>
    </div>
    `;
    static components:object = {
        'wuji-image': WujiImage,
        'wuji-widget': WujiWidget
    };

    initData(): AucKolData {
        return {
            styles,
            jumpProps: {},
            isPlayBack: false,
            isLive: false,
            isPreload: false
        };
    }

    attached(): void {
        this.registerScrollInView().then(() => {
            this.fire('show', '');
        });
    }

    handleClick(): void {
        this.fire('click-kol', this.data.get('jumpProps'));
    }
}

export default AucKol;
